<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lieven Clement" />


<title>Introduction to proteomics data analysis: maxLFQ summarization</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./docs/intro.pdf">Position</a>
    </li>
    <li>
      <a href="recapGeneralLinearModel.html">Recap linear model</a>
    </li>
    <li>
      <a href="./docs/martens_proteomics_bioinformatics.pdf">PDA - identification</a>
    </li>
    <li>
      <a href="pda_quantification_preprocessing.html">Quantification - Preprocessing</a>
    </li>
    <li>
      <a href="pda_robustSummarisation_peptideModels.html">Wrap-up Preprocessing - Peptide-Level-Models</a>
    </li>
    <li>
      <a href="pda_quantification_inference.html">Quantification - Differential Abundance Analysis</a>
    </li>
    <li>
      <a href="pda_blocking_wrapup.html">Wrap-up DA - Blocking</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multipleRegression_KPNA2.html">Intro KPNA2</a>
    </li>
    <li>
      <a href="https://www.compomics.com/bioinformatics-for-proteomics/">Identification</a>
    </li>
    <li>
      <a href="pda_tutorialPreprocessing.html">DA - Preprocessing</a>
    </li>
    <li>
      <a href="pda_tutorialDesign.html">DA - Design</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Introduction to proteomics data analysis: maxLFQ summarization</h1>
<h4 class="author">Lieven Clement</h4>
<h4 class="date">statOmics, Ghent University (<a href="https://statomics.github.io" class="uri">https://statomics.github.io</a>)</h4>

</div>


<p><a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a></p>
<p>This is part of the online course <a href="https://statomics.github.io/PDA21/">Proteomics Data Analysis 2021 (PDA21)</a></p>
<div id="background" class="section level1">
<h1><span class="header-section-number">1</span> Background</h1>
<p>This case-study is a subset of the data of the 6th study of the Clinical Proteomic Technology Assessment for Cancer (CPTAC). In this experiment, the authors spiked the Sigma Universal Protein Standard mixture 1 (UPS1) containing 48 different human proteins in a protein background of 60 ng/<span class="math inline">\(\mu\)</span>L Saccharomyces cerevisiae strain BY4741. Two different spike-in concentrations were used: 6A (0.25 fmol UPS1 proteins/<span class="math inline">\(\mu\)</span>L) and 6B (0.74 fmol UPS1 proteins/<span class="math inline">\(\mu\)</span>L) [5]. We limited ourselves to the data of LTQ-Orbitrap W at site 56. The data were searched with MaxQuant version 1.5.2.8, and detailed search settings were described in Goeminne et al. (2016) [1]. Three replicates are available for each concentration.</p>
<ul>
<li>NOTE THAT maxLFQ SUMMARISATION IS SUBOPTIMAL!</li>
<li>THIS IS FOR DIDACTICAL PURPOSES ONLY.</li>
</ul>
</div>
<div id="data" class="section level1">
<h1><span class="header-section-number">2</span> Data</h1>
<p>We first import the data from proteinGroups.txt file. This is the file containing maxLFQ summarized protein-level intensities. For a MaxQuant search [6], this proteinGroups.txt file can be found by default in the “path_to_raw_files/combined/txt/” folder from the MaxQuant output, with “path_to_raw_files” the folder where the raw files were saved. In this vignette, we use a MaxQuant proteinRaws file which is a subset of the cptac study. To import the data we use the <code>QFeatures</code> package.</p>
<p>We generate the object proteinRawFile with the path to the proteinGroups.txt file. Using the <code>grepEcols</code> function, we find the columns that contain the LFQ expression data of the proteinRaws in the proteinGroups.txt file.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">library</span>(limma)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">library</span>(QFeatures)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw">library</span>(msqrob2)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="kw">library</span>(plotly)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="kw">library</span>(gridExtra)</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a>proteinsFile &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/statOmics/PDA21/data/quantification/cptacAvsB_lab3/proteinGroups.txt&quot;</span></span>
<span id="cb1-9"><a href="#cb1-9"></a></span>
<span id="cb1-10"><a href="#cb1-10"></a>ecols &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;LFQ</span><span class="ch">\\</span><span class="st">.intensity</span><span class="ch">\\</span><span class="st">.&quot;</span>, <span class="kw">names</span>(<span class="kw">read.delim</span>(proteinsFile)))</span>
<span id="cb1-11"><a href="#cb1-11"></a></span>
<span id="cb1-12"><a href="#cb1-12"></a>pe &lt;-<span class="st"> </span><span class="kw">readQFeatures</span>(</span>
<span id="cb1-13"><a href="#cb1-13"></a>    <span class="dt">table =</span> proteinsFile, <span class="dt">fnames =</span> <span class="dv">1</span>, <span class="dt">ecol =</span> ecols,</span>
<span id="cb1-14"><a href="#cb1-14"></a>    <span class="dt">name =</span> <span class="st">&quot;proteinRaw&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span></span>
<span id="cb1-15"><a href="#cb1-15"></a>)</span></code></pre></div>
<p>In the following code chunk, we can extract the spikein condition from the raw file name.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>cond &lt;-<span class="st"> </span><span class="kw">which</span>(</span>
<span id="cb2-2"><a href="#cb2-2"></a>  <span class="kw">strsplit</span>(<span class="kw">colnames</span>(pe)[[<span class="dv">1</span>]][<span class="dv">1</span>], <span class="dt">split =</span> <span class="st">&quot;&quot;</span>)[[<span class="dv">1</span>]] <span class="op">==</span><span class="st"> &quot;A&quot;</span>) <span class="co"># find where condition is stored</span></span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="kw">colData</span>(pe)<span class="op">$</span>condition &lt;-<span class="st"> </span><span class="kw">substr</span>(<span class="kw">colnames</span>(pe), cond, cond) <span class="op">%&gt;%</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="st">  </span>unlist <span class="op">%&gt;%</span><span class="st">  </span></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="st">  </span>as.factor</span></code></pre></div>
<p>We calculate how many non zero intensities we have per protein and this will be useful for filtering.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">rowData</span>(pe[[<span class="st">&quot;proteinRaw&quot;</span>]])<span class="op">$</span>nNonZero &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assay</span>(pe[[<span class="st">&quot;proteinRaw&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</span></code></pre></div>
<p>Proteins with zero intensities are missing and should be represent with a <code>NA</code> value rather than <code>0</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>pe &lt;-<span class="st"> </span><span class="kw">zeroIsNA</span>(pe, <span class="st">&quot;proteinRaw&quot;</span>) <span class="co"># convert 0 to NA</span></span></code></pre></div>
<div id="data-exploration" class="section level2">
<h2><span class="header-section-number">2.1</span> Data exploration</h2>
<p>45% of all peptide intensities are missing and for some proteins we do not even measure a signal in any sample.</p>
</div>
</div>
<div id="preprocessing" class="section level1">
<h1><span class="header-section-number">3</span> Preprocessing</h1>
<p>This section preforms preprocessing for the peptide data. This include</p>
<ul>
<li>log transformation,</li>
<li>filtering</li>
</ul>
<div id="log-transform-the-data" class="section level2">
<h2><span class="header-section-number">3.1</span> Log transform the data</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>pe &lt;-<span class="st"> </span><span class="kw">logTransform</span>(pe, <span class="dt">base =</span> <span class="dv">2</span>, <span class="dt">i =</span> <span class="st">&quot;proteinRaw&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;proteinLog&quot;</span>)</span></code></pre></div>
</div>
<div id="filtering" class="section level2">
<h2><span class="header-section-number">3.2</span> Filtering</h2>
<ol style="list-style-type: decimal">
<li>Remove reverse sequences (decoys) and contaminants</li>
</ol>
<p>We now remove the contaminants and proteins that map to decoys.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>pe[[<span class="st">&quot;proteinLog&quot;</span>]] &lt;-<span class="st"> </span>pe[[<span class="st">&quot;proteinLog&quot;</span>]][<span class="kw">rowData</span>(pe[[<span class="st">&quot;proteinLog&quot;</span>]])<span class="op">$</span>Reverse <span class="op">!=</span><span class="st"> &quot;+&quot;</span>, ]</span>
<span id="cb6-2"><a href="#cb6-2"></a>pe[[<span class="st">&quot;proteinLog&quot;</span>]] &lt;-<span class="st"> </span>pe[[<span class="st">&quot;proteinLog&quot;</span>]][<span class="kw">rowData</span>(pe[[<span class="st">&quot;proteinLog&quot;</span>]])<span class="op">$</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="st">    </span>Potential.contaminant <span class="op">!=</span><span class="st"> &quot;+&quot;</span>, ]</span></code></pre></div>
<p>We keep 1537 peptides upon filtering.</p>
</div>
<div id="normalize-the-data-using-median-centering" class="section level2">
<h2><span class="header-section-number">3.3</span> Normalize the data using median centering</h2>
<p>We normalize the data by substracting the sample median from every intensity for peptide <span class="math inline">\(p\)</span> in a sample <span class="math inline">\(i\)</span>:</p>
<p><span class="math display">\[y_{ip}^\text{norm} = y_{ip} - \hat\mu_i\]</span></p>
<p>with <span class="math inline">\(\hat\mu_i\)</span> the median intensity over all observed peptides in sample <span class="math inline">\(i\)</span>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>pe &lt;-<span class="st"> </span><span class="kw">normalize</span>(pe, </span>
<span id="cb7-2"><a href="#cb7-2"></a>                <span class="dt">i =</span> <span class="st">&quot;proteinLog&quot;</span>, </span>
<span id="cb7-3"><a href="#cb7-3"></a>                <span class="dt">name =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb7-4"><a href="#cb7-4"></a>                <span class="dt">method =</span> <span class="st">&quot;center.median&quot;</span>)</span></code></pre></div>
</div>
<div id="explore-normalized-data" class="section level2">
<h2><span class="header-section-number">3.4</span> Explore normalized data</h2>
<p>Upon the normalisation the density curves are nicely registered</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>pe[[<span class="st">&quot;protein&quot;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">condition =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;condition&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> intensity,<span class="dt">group =</span> sample,<span class="dt">color =</span> condition)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="st">    </span><span class="kw">geom_density</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 4052 rows containing non-finite values (stat_density).</code></pre>
<p><img src="cptac_maxLFQ_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>We can visualize our data using a Multi Dimensional Scaling plot, eg. as provided by the <code>limma</code> package.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>pe[[<span class="st">&quot;protein&quot;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="st">  </span>limma<span class="op">::</span><span class="kw">plotMDS</span>(<span class="dt">col =</span> <span class="kw">as.numeric</span>(<span class="kw">colData</span>(pe)<span class="op">$</span>condition))</span></code></pre></div>
<p><img src="cptac_maxLFQ_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>Note that the samples show a clear separation according to the spike-in condition in the second dimension of the MDS plot.</p>
</div>
</div>
<div id="data-analysis" class="section level1">
<h1><span class="header-section-number">4</span> Data Analysis</h1>
<div id="estimation" class="section level2">
<h2><span class="header-section-number">4.1</span> Estimation</h2>
<p>We model the protein level expression values using <code>msqrob</code>. By default <code>msqrob2</code> estimates the model parameters using robust regression.</p>
<p>We will model the data with a different group mean. The group is incoded in the variable <code>condition</code> of the colData. We can specify this model by using a formula with the factor condition as its predictor: <code>formula = ~condition</code>.</p>
<p>Note, that a formula always starts with a symbol ‘~’.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>pe &lt;-<span class="st"> </span><span class="kw">msqrob</span>(<span class="dt">object =</span> pe, <span class="dt">i =</span> <span class="st">&quot;protein&quot;</span>, <span class="dt">formula =</span> <span class="op">~</span>condition)</span></code></pre></div>
</div>
<div id="inference" class="section level2">
<h2><span class="header-section-number">4.2</span> Inference</h2>
<p>First, we extract the parameter names of the model by looking at the first model. The models are stored in the row data of the assay under the default name msqrobModels.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">getCoef</span>(<span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>msqrobModels[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>## [1] NA</code></pre>
<p>We can also explore the design of the model that we specified using the the package <code>ExploreModelMatrix</code></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">library</span>(ExploreModelMatrix)</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="kw">VisualizeDesign</span>(<span class="kw">colData</span>(pe),<span class="op">~</span>condition)<span class="op">$</span>plotlist[[<span class="dv">1</span>]]</span></code></pre></div>
<p><img src="cptac_maxLFQ_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>Spike-in condition <code>A</code> is the reference class. So the mean log2 expression for samples from condition A is ‘(Intercept). The mean log2 expression for samples from condition B is’(Intercept)+conditionB’. Hence, the average log2 fold change between condition b and condition a is modelled using the parameter ‘conditionB’. Thus, we assess the contrast ‘conditionB = 0’ with our statistical test.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>L &lt;-<span class="st"> </span><span class="kw">makeContrast</span>(<span class="st">&quot;conditionB=0&quot;</span>, <span class="dt">parameterNames =</span> <span class="kw">c</span>(<span class="st">&quot;conditionB&quot;</span>))</span>
<span id="cb15-2"><a href="#cb15-2"></a>pe &lt;-<span class="st"> </span><span class="kw">hypothesisTest</span>(<span class="dt">object =</span> pe, <span class="dt">i =</span> <span class="st">&quot;protein&quot;</span>, <span class="dt">contrast =</span> L)</span></code></pre></div>
</div>
<div id="plots" class="section level2">
<h2><span class="header-section-number">4.3</span> Plots</h2>
<div id="volcano-plot" class="section level3">
<h3><span class="header-section-number">4.3.1</span> Volcano-plot</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>volcano &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>conditionB,</span>
<span id="cb16-2"><a href="#cb16-2"></a>                  <span class="kw">aes</span>(<span class="dt">x =</span> logFC, <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(pval), <span class="dt">color =</span> adjPval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)) <span class="op">+</span></span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">cex =</span> <span class="fl">2.5</span>) <span class="op">+</span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">alpha</span>(<span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="fl">0.5</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">theme_minimal</span>()</span>
<span id="cb16-5"><a href="#cb16-5"></a>volcano</span></code></pre></div>
<p><img src="cptac_maxLFQ_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>Note, that only 7 proteins are found to be differentially abundant.</p>
</div>
<div id="heatmap" class="section level3">
<h3><span class="header-section-number">4.3.2</span> Heatmap</h3>
<p>We first select the names of the proteins that were declared signficant.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>sigNames &lt;-<span class="st"> </span><span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>conditionB <span class="op">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="st">  </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;protein&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="st">  </span><span class="kw">filter</span>(adjPval<span class="op">&lt;</span><span class="fl">0.05</span>) <span class="op">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="st">  </span><span class="kw">pull</span>(protein)</span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="kw">heatmap</span>(<span class="kw">assay</span>(pe[[<span class="st">&quot;protein&quot;</span>]])[sigNames, ])</span></code></pre></div>
<p><img src="cptac_maxLFQ_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>The majority of the proteins are indeed UPS proteins. 1 yeast protein is returned. Note, that the yeast protein indeed shows evidence for differential abundance.</p>
</div>
<div id="boxplots" class="section level3">
<h3><span class="header-section-number">4.3.3</span> Boxplots</h3>
<p>We make boxplot of the log2 FC and stratify according to the whether a protein is spiked or not.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>conditionB <span class="op">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="st">  </span><span class="kw">rownames_to_column</span>(<span class="dt">var =</span> <span class="st">&quot;protein&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span><span class="kw">grepl</span>(<span class="st">&quot;UPS&quot;</span>,protein),<span class="dt">y=</span>logFC)) <span class="op">+</span></span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;UPS&quot;</span>) <span class="op">+</span></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="st">  </span><span class="kw">geom_segment</span>(</span>
<span id="cb18-7"><a href="#cb18-7"></a>    <span class="dt">x =</span> <span class="fl">1.5</span>,</span>
<span id="cb18-8"><a href="#cb18-8"></a>    <span class="dt">xend =</span> <span class="fl">2.5</span>,</span>
<span id="cb18-9"><a href="#cb18-9"></a>    <span class="dt">y =</span> <span class="kw">log2</span>(<span class="fl">0.74</span><span class="op">/</span><span class="fl">0.25</span>),</span>
<span id="cb18-10"><a href="#cb18-10"></a>    <span class="dt">yend =</span> <span class="kw">log2</span>(<span class="fl">0.74</span><span class="op">/</span><span class="fl">0.25</span>),</span>
<span id="cb18-11"><a href="#cb18-11"></a>    <span class="dt">colour=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="st">  </span><span class="kw">geom_segment</span>(</span>
<span id="cb18-13"><a href="#cb18-13"></a>    <span class="dt">x =</span> <span class="fl">0.5</span>,</span>
<span id="cb18-14"><a href="#cb18-14"></a>    <span class="dt">xend =</span> <span class="fl">1.5</span>,</span>
<span id="cb18-15"><a href="#cb18-15"></a>    <span class="dt">y =</span> <span class="dv">0</span>,</span>
<span id="cb18-16"><a href="#cb18-16"></a>    <span class="dt">yend =</span> <span class="dv">0</span>,</span>
<span id="cb18-17"><a href="#cb18-17"></a>    <span class="dt">colour=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span></span>
<span id="cb18-18"><a href="#cb18-18"></a><span class="st">  </span><span class="kw">annotate</span>(</span>
<span id="cb18-19"><a href="#cb18-19"></a>    <span class="st">&quot;text&quot;</span>,</span>
<span id="cb18-20"><a href="#cb18-20"></a>    <span class="dt">x =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>),</span>
<span id="cb18-21"><a href="#cb18-21"></a>    <span class="dt">y =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="kw">log2</span>(<span class="fl">0.74</span><span class="op">/</span><span class="fl">0.25</span>))<span class="op">+</span>.<span class="dv">1</span>,</span>
<span id="cb18-22"><a href="#cb18-22"></a>    <span class="dt">label =</span> <span class="kw">c</span>(</span>
<span id="cb18-23"><a href="#cb18-23"></a>      <span class="st">&quot;log2 FC Ecoli = 0&quot;</span>,</span>
<span id="cb18-24"><a href="#cb18-24"></a>      <span class="kw">paste0</span>(<span class="st">&quot;log2 FC UPS = &quot;</span>,<span class="kw">round</span>(<span class="kw">log2</span>(<span class="fl">0.74</span><span class="op">/</span><span class="fl">0.25</span>),<span class="dv">2</span>))</span>
<span id="cb18-25"><a href="#cb18-25"></a>      ),</span>
<span id="cb18-26"><a href="#cb18-26"></a>    <span class="dt">colour =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 746 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="cptac_maxLFQ_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>What do you observe?</p>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">5</span> Session Info</h1>
<p>With respect to reproducibility, it is highly recommended to include a session info in your script so that readers of your output can see your particular setup of R.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.1.1 (2021-08-10)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Catalina 10.15.7
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ExploreModelMatrix_1.4.0    gridExtra_2.3              
##  [3] plotly_4.10.0               msqrob2_1.0.0              
##  [5] QFeatures_1.2.0             MultiAssayExperiment_1.18.0
##  [7] SummarizedExperiment_1.22.0 Biobase_2.52.0             
##  [9] GenomicRanges_1.44.0        GenomeInfoDb_1.28.4        
## [11] IRanges_2.26.0              S4Vectors_0.30.2           
## [13] BiocGenerics_0.38.0         MatrixGenerics_1.4.3       
## [15] matrixStats_0.61.0          limma_3.48.3               
## [17] forcats_0.5.1               stringr_1.4.0              
## [19] dplyr_1.0.7                 purrr_0.3.4                
## [21] readr_2.0.2                 tidyr_1.1.4                
## [23] tibble_3.1.5                ggplot2_3.3.5              
## [25] tidyverse_1.3.1            
## 
## loaded via a namespace (and not attached):
##  [1] minqa_1.2.4             colorspace_2.0-2        ellipsis_0.3.2         
##  [4] XVector_0.32.0          fs_1.5.0                clue_0.3-60            
##  [7] rstudioapi_0.13         farver_2.1.0            DT_0.19                
## [10] fansi_0.5.0             lubridate_1.8.0         xml2_1.3.2             
## [13] codetools_0.2-18        splines_4.1.1           knitr_1.36             
## [16] jsonlite_1.7.2          nloptr_1.2.2.2          broom_0.7.9            
## [19] cluster_2.1.2           dbplyr_2.1.1            shinydashboard_0.7.2   
## [22] shiny_1.7.1             compiler_4.1.1          httr_1.4.2             
## [25] backports_1.2.1         assertthat_0.2.1        Matrix_1.3-4           
## [28] fastmap_1.1.0           lazyeval_0.2.2          cli_3.0.1              
## [31] later_1.3.0             htmltools_0.5.2         tools_4.1.1            
## [34] gtable_0.3.0            glue_1.4.2              GenomeInfoDbData_1.2.6 
## [37] Rcpp_1.0.7              cellranger_1.1.0        jquerylib_0.1.4        
## [40] vctrs_0.3.8             nlme_3.1-152            rintrojs_0.3.0         
## [43] xfun_0.26               lme4_1.1-27.1           rvest_1.0.1            
## [46] mime_0.12               lifecycle_1.0.1         zlibbioc_1.38.0        
## [49] MASS_7.3-54             scales_1.1.1            promises_1.2.0.1       
## [52] hms_1.1.1               ProtGenerics_1.24.0     AnnotationFilter_1.16.0
## [55] yaml_2.2.1              sass_0.4.0              stringi_1.7.5          
## [58] highr_0.9               boot_1.3-28             BiocParallel_1.26.2    
## [61] rlang_0.4.11            pkgconfig_2.0.3         bitops_1.0-7           
## [64] evaluate_0.14           lattice_0.20-44         htmlwidgets_1.5.4      
## [67] labeling_0.4.2          cowplot_1.1.1           tidyselect_1.1.1       
## [70] magrittr_2.0.1          R6_2.5.1                generics_0.1.0         
## [73] DelayedArray_0.18.0     DBI_1.1.1               pillar_1.6.3           
## [76] haven_2.4.3             withr_2.4.2             MsCoreUtils_1.4.0      
## [79] RCurl_1.98-1.5          modelr_0.1.8            crayon_1.4.1           
## [82] utf8_1.2.2              tzdb_0.1.2              rmarkdown_2.11         
## [85] grid_4.1.1              readxl_1.3.1            data.table_1.14.2      
## [88] reprex_2.0.1            digest_0.6.28           xtable_1.8-4           
## [91] httpuv_1.6.3            munsell_0.5.0           viridisLite_0.4.0      
## [94] bslib_0.3.1             shinyjs_2.0.0</code></pre>
</div>

<div id="rmd-source-code">---
title: "Introduction to proteomics data analysis: maxLFQ summarization"
author: "Lieven Clement"
date: "statOmics, Ghent University (https://statomics.github.io)"
output:
    html_document:
      code_download: true
      theme: cosmo
      toc: true
      toc_float: true
      highlight: tango
      number_sections: true
    pdf_document:
      toc: true
      number_sections: true
linkcolor: blue
urlcolor: blue
citecolor: blue

bibliography: msqrob2.bib

---

<a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>

This is part of the online course [Proteomics Data Analysis 2021 (PDA21)](https://statomics.github.io/PDA21/)

# Background
This case-study is a subset of the data of the 6th study of the Clinical
Proteomic Technology Assessment for Cancer (CPTAC).
In this experiment, the authors spiked the Sigma Universal Protein Standard
mixture 1 (UPS1) containing 48 different human proteins in a protein background
of 60 ng/$\mu$L Saccharomyces cerevisiae strain BY4741.
Two different spike-in concentrations were used:
6A (0.25 fmol UPS1 proteins/$\mu$L) and 6B (0.74 fmol UPS1 proteins/$\mu$L) [5].
We limited ourselves to the data of LTQ-Orbitrap W at site 56.
The data were searched with MaxQuant version 1.5.2.8, and
detailed search settings were described in Goeminne et al. (2016) [1].
Three replicates are available for each concentration.


- NOTE THAT maxLFQ SUMMARISATION IS SUBOPTIMAL! 
- THIS IS FOR DIDACTICAL PURPOSES ONLY.

# Data

We first import the data from proteinGroups.txt file. This is the file containing
maxLFQ summarized protein-level intensities. For a MaxQuant search [6],
this proteinGroups.txt file can be found by default in the
"path_to_raw_files/combined/txt/" folder from the MaxQuant output,
with "path_to_raw_files" the folder where the raw files were saved.
In this vignette, we use a MaxQuant proteinRaws file which is a subset
of the cptac study. 
To import the data we use the `QFeatures` package.

We generate the object proteinRawFile with the path to the proteinGroups.txt file.
Using the `grepEcols` function, we find the columns that contain the LFQ expression
data of the proteinRaws in the proteinGroups.txt file.



```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(limma)
library(QFeatures)
library(msqrob2)
library(plotly)
library(gridExtra)

proteinsFile <- "https://raw.githubusercontent.com/statOmics/PDA21/data/quantification/cptacAvsB_lab3/proteinGroups.txt"

ecols <- grep("LFQ\\.intensity\\.", names(read.delim(proteinsFile)))

pe <- readQFeatures(
    table = proteinsFile, fnames = 1, ecol = ecols,
    name = "proteinRaw", sep = "\t"
)
```

In the following code chunk, we can extract the spikein condition from the raw file name.

```{r}
cond <- which(
  strsplit(colnames(pe)[[1]][1], split = "")[[1]] == "A") # find where condition is stored

colData(pe)$condition <- substr(colnames(pe), cond, cond) %>%
  unlist %>%  
  as.factor
```


We calculate how many non zero intensities we have per protein and this
will be useful for filtering.

```{r}
rowData(pe[["proteinRaw"]])$nNonZero <- rowSums(assay(pe[["proteinRaw"]]) > 0)
```


Proteins with zero intensities are missing and should be represent
with a `NA` value rather than `0`.
```{r}
pe <- zeroIsNA(pe, "proteinRaw") # convert 0 to NA
```


## Data exploration

`r format(mean(is.na(assay(pe[["proteinRaw"]])))*100,digits=2)`% of all peptide
intensities are missing and for some proteins we do not even measure a signal
in any sample.


# Preprocessing

This section preforms preprocessing for the peptide data. 
This include 

- log transformation, 
- filtering 

## Log transform the data

```{r}
pe <- logTransform(pe, base = 2, i = "proteinRaw", name = "proteinLog")
```

## Filtering

1. Remove reverse sequences (decoys) and contaminants

We now remove the contaminants and proteins that map to decoys.

```{r}
pe[["proteinLog"]] <- pe[["proteinLog"]][rowData(pe[["proteinLog"]])$Reverse != "+", ]
pe[["proteinLog"]] <- pe[["proteinLog"]][rowData(pe[["proteinLog"]])$
    Potential.contaminant != "+", ]
```
We keep `r nrow(pe[["proteinLog"]])` peptides upon filtering.


## Normalize the data using median centering 

We normalize the data by substracting the sample median from every intensity for peptide $p$  in a sample $i$: 

$$y_{ip}^\text{norm} = y_{ip} - \hat\mu_i$$ 

with $\hat\mu_i$ the median intensity over all observed peptides in sample $i$.

```{r}
pe <- normalize(pe, 
                i = "proteinLog", 
                name = "protein", 
                method = "center.median")
```


## Explore  normalized data

Upon the normalisation the density curves are nicely registered

```{r}
pe[["protein"]] %>% 
  assay %>%
  as.data.frame() %>%
  gather(sample, intensity) %>% 
  mutate(condition = colData(pe)[sample,"condition"]) %>%
  ggplot(aes(x = intensity,group = sample,color = condition)) + 
    geom_density()
```

We can visualize our data using a Multi Dimensional Scaling plot,
eg. as provided by the `limma` package.

```{r}
pe[["protein"]] %>% 
  assay %>%
  limma::plotMDS(col = as.numeric(colData(pe)$condition))
```

Note that the samples show a clear separation according to the spike-in condition in the second dimension of the MDS plot.

# Data Analysis

## Estimation

We model the protein level expression values using `msqrob`.
By default `msqrob2` estimates the model parameters using robust regression.

We will model the data with a different group mean. 
The group is incoded in the variable `condition` of the colData. 
We can specify this model by using a formula with the factor condition as its predictor: 
`formula = ~condition`.

Note, that a formula always starts with a symbol '~'.

```{r, warning=FALSE}
pe <- msqrob(object = pe, i = "protein", formula = ~condition)
```

## Inference

First, we extract the parameter names of the model by looking at the first model. 
The models are stored in the row data of the assay under the default name msqrobModels. 

```{r}
getCoef(rowData(pe[["protein"]])$msqrobModels[[1]])
```

We can also explore the design of the model that we specified using the the package `ExploreModelMatrix` 

```{r}
library(ExploreModelMatrix)
VisualizeDesign(colData(pe),~condition)$plotlist[[1]]
```

Spike-in condition `A` is the reference class. So the mean log2 expression
for samples from condition A is '(Intercept).
The mean log2 expression for samples from condition B is '(Intercept)+conditionB'.
Hence, the average log2 fold change between condition b and
condition a is modelled using the parameter 'conditionB'.
Thus, we assess the contrast 'conditionB = 0' with our statistical test.

```{r}
L <- makeContrast("conditionB=0", parameterNames = c("conditionB"))
pe <- hypothesisTest(object = pe, i = "protein", contrast = L)
```


## Plots

### Volcano-plot


```{r,warning=FALSE}
volcano <- ggplot(rowData(pe[["protein"]])$conditionB,
                  aes(x = logFC, y = -log10(pval), color = adjPval < 0.05)) +
  geom_point(cex = 2.5) +
  scale_color_manual(values = alpha(c("black", "red"), 0.5)) + theme_minimal()
volcano
```

Note, that only `r sum(rowData(pe[["protein"]])$conditionB$adjPval < 0.05, na.rm = TRUE)` proteins are found to be differentially abundant.

### Heatmap

We first select the names of the proteins that were declared signficant.

```{r}
sigNames <- rowData(pe[["protein"]])$conditionB %>%
  rownames_to_column("protein") %>%
  filter(adjPval<0.05) %>%
  pull(protein)
heatmap(assay(pe[["protein"]])[sigNames, ])
```

The majority of the proteins are indeed UPS proteins. 
1 yeast protein is returned. 
Note, that the yeast protein indeed shows evidence for differential abundance. 

### Boxplots

We make boxplot of the log2 FC and stratify according to the whether a protein is spiked or not.

```{r}
rowData(pe[["protein"]])$conditionB %>%
  rownames_to_column(var = "protein") %>%
  ggplot(aes(x=grepl("UPS",protein),y=logFC)) +
  geom_boxplot() +
  xlab("UPS") +
  geom_segment(
    x = 1.5,
    xend = 2.5,
    y = log2(0.74/0.25),
    yend = log2(0.74/0.25),
    colour="red") +
  geom_segment(
    x = 0.5,
    xend = 1.5,
    y = 0,
    yend = 0,
    colour="red") +
  annotate(
    "text",
    x = c(1,2),
    y = c(0,log2(0.74/0.25))+.1,
    label = c(
      "log2 FC Ecoli = 0",
      paste0("log2 FC UPS = ",round(log2(0.74/0.25),2))
      ),
    colour = "red")
```

What do you observe?

# Session Info

With respect to reproducibility, it is highly recommended to include a session info in your script so that readers of your output can see your particular setup of R. 

```{r}
sessionInfo()
```
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("cptac_maxLFQ.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
