<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Statistical Genomics Analysis 2021-2022</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./docs/intro.pdf">1.1 Position</a>
    </li>
    <li>
      <a href="recapGeneralLinearModel.html">1.2 Recap linear model</a>
    </li>
    <li>
      <a href="./docs/martens_proteomics_bioinformatics.pdf">2.1 PDA - identification</a>
    </li>
    <li>
      <a href="pda_quantification_preprocessing.html">2.2 PDA- Quantification - Preprocessing</a>
    </li>
    <li>
      <a href="pda_robustSummarisation_peptideModels.html">2.2 PDA - Wrap-up Preprocessing - Peptide-Level-Models</a>
    </li>
    <li>
      <a href="pda_quantification_inference.html">2.3 PDA - Quantification - Differential Abundance Analysis</a>
    </li>
    <li>
      <a href="pda_blocking_wrapup.html">2.3 PDA - Wrap-up DA - Blocking</a>
    </li>
    <li>
      <a href="technicalDetailsProteomics.html">2.4. PDA - DA Technical Details</a>
    </li>
    <li>
      <a href="./docs/stagewiseTesting.pdf">2.5 PDA - Stage-wise testing</a>
    </li>
    <li>
      <a href="sequencing_intro.html">3.1 SDA - Intro to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">3.2 SDA - Working with count data</a>
    </li>
    <li>
      <a href="sequencing_rnaseqIntro.html">3.3 SDA - Intro to RNA-seq analysis</a>
    </li>
    <li>
      <a href="sequencing_technicalDE.html">3.4 SDA - Technical topics in DE analysis</a>
    </li>
    <li>
      <a href="https://docs.google.com/presentation/d/1sfbtw52qWgA7TcDIrGB3fWksLM0Mu9O7gBqYgGrt9fM/edit?usp=sharing">4.1 scRNA-seq - General concepts and workflow</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multipleRegression_KPNA2.html">1 Intro KPNA2</a>
    </li>
    <li>
      <a href="https://www.compomics.com/bioinformatics-for-proteomics/">2.1 Identification</a>
    </li>
    <li>
      <a href="pda_tutorialPreprocessing.html">2.2 DA - Preprocessing</a>
    </li>
    <li>
      <a href="pda_tutorialDesign.html">2.3 DA - Design</a>
    </li>
    <li>
      <a href="sequencing_preprocessing.sh">3.1 Preprocessing</a>
    </li>
    <li>
      <a href="sequencing_lab_oneGene.html">3.2 GLM for one gene</a>
    </li>
    <li>
      <a href="sequencing_hw.html">3.3 Bulk RNA-seq homework</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Statistical Genomics Analysis 2021-2022</h1>

</div>


<p><img src="figures/wpGraph.jpeg" /><!-- --></p>
<div id="course-description" class="section level1">
<h1><span class="header-section-number">1</span> Course Description</h1>
<p>High-throughput ’omics studies generate ever larger datasets and, as a consequence, complex data interpretation challenges. This course focusses on statistical concepts involved in preprocessing, quantification and differential analysis of high-throughput ’omics data. The core focus will be on shotgun proteomics and (bulk and single-cell) RNA-sequencing. Experimental design is essential to allow for correct interpretation in all ’omics studies, and we will cover how to design a statistically efficient experiment, as well as discuss the impact experimental design has on how we model ’omics data, introducing concepts such as blocking. The course will rely exclusively on free and user-friendly open-source tools in R/Bioconductor. We hope that this will provide a solid basis for beginners, but will also bring new perspectives to those already familiar with standard data analysis workflows for proteomics and next-generation sequencing applications.</p>
</div>
<div id="target-audience" class="section level1">
<h1><span class="header-section-number">2</span> Target Audience</h1>
<p>This course is oriented towards biologists and bioinformaticians with a particular interest in differential analysis for quantitative ’omics data.</p>
</div>
<div id="github-repository" class="section level1">
<h1><span class="header-section-number">3</span> GitHub repository</h1>
<p>All source and data files for this course are available on the accompanying <a href="https://github.com/statOmics/SGA21">GitHub repository</a>.</p>
</div>
<div id="prerequisites" class="section level1">
<h1><span class="header-section-number">4</span> Prerequisites</h1>
<p>The prerequisites for the Statistical Genomics Analysis course are the successful completion of a basic course of statistics that covers topics on data exploration and descriptive statistics, statistical modeling, and inference: linear models, confidence intervals, t-tests, F-tests, anova, chi-squared test. The basis concepts may be revisited in the online course at <a href="https://gtpb.github.io/PSLS20/" class="uri">https://gtpb.github.io/PSLS20/</a> (English) and in <a href="https://statomics.github.io/statistiekCursusNotas/" class="uri">https://statomics.github.io/statistiekCursusNotas/</a> (Dutch).</p>
<p>In addition, knowledge of programming in <code>R</code> is preferred. A primer to <code>R</code> and Data visualization in <code>R</code> can be found at:</p>
<ul>
<li><code>R</code> Basics: <a href="https://dodona.ugent.be/nl/courses/335/" class="uri">https://dodona.ugent.be/nl/courses/335/</a></li>
<li><code>R</code> Data Exploration: <a href="https://dodona.ugent.be/nl/courses/345/" class="uri">https://dodona.ugent.be/nl/courses/345/</a></li>
</ul>
</div>
<div id="software" class="section level1">
<h1><span class="header-section-number">5</span> Software</h1>
<ul>
<li><p>Participants are required to bring their own laptop with <a href="https://www.r-project.org/">R</a> version 4.1.1 or greater.</p></li>
<li><p>We also recommend to also install the latest version of <a href="https://www.rstudio.com/products/rstudio/download/">RStudio</a>.</p></li>
<li><p>Installation script: to install all required packages, please copy and paste this line of code in your R console.</p></li>
</ul>
<pre><code>source(&quot;https://raw.githubusercontent.com/statOmics/SGA21/master/install.R&quot;)</code></pre>
<ul>
<li>Participants who have issues with the installation of the R/Rstudio can use an Rstudio instance in the cloud with all packages installed for the course in the mean time. Note, that this is instance is not for routine use.</li>
</ul>
<p><a href="https://mybinder.org/v2/gh/statOmics/SGA21/binder?urlpath=rstudio"><img src="http://mybinder.org/badge.svg" alt="Binder" /></a></p>
</div>
<div id="detailed-program" class="section level1">
<h1><span class="header-section-number">6</span> Detailed Program</h1>
<ol style="list-style-type: decimal">
<li><p>Position of the course: <a href="./docs/intro.pdf">PDF</a></p></li>
<li><p>Recap Linear Models (Week 1)</p>
<ul>
<li>Lecture: <a href="./recapGeneralLinearModel.html">HTML</a>, <a href="./recapGeneralLinearModel.pdf">PDF</a></li>
<li>Tutorial KPNA2: <a href="./multipleRegression_KPNA2.html">HTML</a>, <a href="./multipleRegression_KPNA2.pdf">PDF</a></li>
</ul></li>
</ol>
<div id="module-i-proteomics-data-analysis-week-1-5" class="section level2">
<h2><span class="header-section-number">6.1</span> Module I: Proteomics Data Analysis (Week 1-5)</h2>
<ol style="list-style-type: decimal">
<li><p>Bioinformatics for proteomics</p>
<ul>
<li>Lecture: <a href="./docs/martens_proteomics_bioinformatics.pdf">PDF</a>, <a href="https://www.youtube.com/watch?v=ZgwNWRul98o">youtube</a></li>
<li>Tutorials: <a href="https://www.compomics.com/bioinformatics-for-proteomics/">identification</a></li>
</ul></li>
</ol>

<ol start="2" style="list-style-type: decimal">
<li><p>Preprocessing &amp; Analysis of Simple Designs (week 3)</p>
<ul>
<li>Lecture: <a href="./pda_quantification_preprocessing.html">Preprocessing</a> [<a href="./pda_quantification_preprocessing.pdf">PDF</a>]</li>
<li>Tutorial: <a href="./pda_tutorialPreprocessing.html">Preprocessing</a></li>
<li>Wrap-up: <a href="./pda_robustSummarisation_peptideModels.html">Peptide-based Models</a> [<a href="./pda_robustSummarisation_peptideModels.pdf">PDF</a>]</li>
</ul></li>
</ol>

<ol start="3" style="list-style-type: decimal">
<li><p>Statistical Inference &amp; Analysis of Factorial Designs (Week 3-4)</p>
<ul>
<li>Lecture: <a href="./pda_quantification_inference.html">Differential Abundance Analysis</a> [<a href="./pda_quantification_inference.pdf">PDF</a>]</li>
<li>Tutorial: <a href="pda_tutorialDesign.html">Design</a></li>
<li>Wrap-up: <a href="./pda_blocking_wrapup.html">Blocking</a> [<a href="./pda_blocking_wrapup.pdf">PDF</a>]</li>
</ul></li>
<li><p>Advanced materials and reading materials</p>
<ul>
<li><p><a href="./technicalDetailsProteomics.html">Technical details: Inference upon summarization</a></p></li>
<li><p><a href="./docs/stagewiseTesting.pdf">Stage-wise testing</a></p></li>
<li><p>Papers:</p>
<ul>
<li><a href="https://www.biorxiv.org/content/10.1101/668863v1">Sticker et al. (2020) Robust summarization and inference in proteome-wide label-free quantification</a></li>
<li><a href="./docs/backgroundProteomicsDataAnalysis.pdf">Ludger Goeminne: Extensive Background on proteomics and proteomics data analysis (Introduction of PhD)</a></li>
<li><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1277-0">Van den Berge et al. (2017) Stage-wise testing: stageR paper</a></li>
</ul></li>
</ul></li>
<li><p>Solutions</p>
<ul>
<li>CPTAC study: See <a href="./pda_robustSummarisation_peptideModels.html">wrap-up preprocessing</a></li>
<li><a href="cancer2_6x6.html">cancer 6x6</a></li>
<li>mouse: See <a href="./pda_blocking_wrapup.html">wrap-up blocking</a></li>
<li><a href="heartMainInteractionStageR.html">Heart study</a></li>
</ul></li>
</ol>
</div>
<div id="module-ii-bulk-rna-sequencing" class="section level2">
<h2><span class="header-section-number">6.2</span> Module II: Bulk RNA-sequencing</h2>
<ol style="list-style-type: decimal">
<li><p>Introduction to sequencing technology, raw data and preprocessing.</p>
<ul>
<li>Lecture: <a href="./sequencing_intro.html">Introduction to sequencing</a> [<a href="./sequencing_intro.pdf">PDF</a>].</li>
<li>Tutorial: <a href="./sequencing_preprocessing.sh">Preprocessing RNA-seq data</a></li>
<li>Recommended reading: <a href="https://www.annualreviews.org/doi/abs/10.1146/annurev-biodatasci-072018-021255">RNA Sequencing Data: Hitchhiker’s Guide to Expression Analysis</a></li>
</ul></li>
</ol>

<ol start="2" style="list-style-type: decimal">
<li><p>Working with count data and generalized linear models.</p>
<ul>
<li>Lecture: <a href="./sequencing_countData.html">Working with count data and GLMs</a> [<a href="./sequencing_countData.pdf">PDF</a>].</li>
<li>Tutorial I: Embedded within the lecture: Analysis of bike-sharing data.</li>
<li>Tutorial II: <a href="sequencing_lab_oneGene.html">GLM for one gene</a>.</li>
</ul></li>
</ol>

<ol start="3" style="list-style-type: decimal">
<li><p>Analysis of RNA-seq data.</p>
<ul>
<li>Lecture + Tutorial: <a href="./sequencing_rnaseqIntro.html">Four major challenges in the analysis of RNA-seq datasets</a> [<a href="./sequencing_rnaseqIntro.pdf">PDF</a>].</li>
<li><a href="sequencing_scalingNormalization.html">Why we use offsets rather than count scaling</a>.</li>
</ul></li>
</ol>

<ol start="4" style="list-style-type: decimal">
<li><p>Technical topics in bulk RNA-seq differential expression analysis (DEA).</p>
<ul>
<li>Lecture: <a href="./sequencing_technicalDE.html">Technical topics in DEA</a> [<a href="./sequencing_technicalDE.pdf">PDF</a>].</li>
</ul></li>
</ol>
</div>
<div id="module-iii-single-cell-rna-sequencing" class="section level2">
<h2><span class="header-section-number">6.3</span> Module III: Single-cell RNA-sequencing</h2>
<ol style="list-style-type: decimal">
<li><p>General concepts and analysis workflow of single-cell RNA-seq data.</p>
<ul>
<li>Lecture: <a href="https://docs.google.com/presentation/d/1sfbtw52qWgA7TcDIrGB3fWksLM0Mu9O7gBqYgGrt9fM/edit?usp=sharing">Slides</a>.
<ul>
<li><a href="./singleCell_varStabilization.html">Variance stabilizing transformation for a Poisson random variable</a> [<a href="./singleCell_varStabilization.pdf">PDF</a>].</li>
</ul></li>
</ul></li>
</ol>
</div>
</div>
<div id="instructors" class="section level1">
<h1><span class="header-section-number">7</span> Instructors</h1>
<ul>
<li><a href="https://koenvandenberge.github.io/">Koen Van den Berge</a></li>
<li><a href="./instructors.html">Lieven Clement</a></li>
<li><a href="https://www.compomics.com/people/lennart-martens/">Lennart Martens</a></li>
</ul>
</div>
<div id="license" class="section level1">
<h1><span class="header-section-number">8</span> License</h1>
<p><a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a></p>
<p>This project is licensed under the <a href="https://creativecommons.org/licenses/by-nc-sa/4.0">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0)</a></p>
</div>

<div id="rmd-source-code">---
title: "Statistical Genomics Analysis 2021-2022"
---


```{r echo=FALSE, warning=FALSE, message=FALSE}
knitr::include_graphics("./figures/wpGraph.jpeg")
```

# Course Description

High-throughput 'omics studies generate ever larger datasets and, as a consequence, complex data interpretation challenges. This course focusses on statistical concepts involved in preprocessing, quantification and differential analysis of high-throughput 'omics data. The core focus will be on shotgun proteomics and (bulk and single-cell) RNA-sequencing. Experimental design is essential to allow for correct interpretation in all 'omics studies, and we will cover how to design a statistically efficient experiment, as well as discuss the impact experimental design has on how we model 'omics data, introducing concepts such as blocking. The course will rely exclusively on free and user-friendly open-source tools in R/Bioconductor. We hope that this will provide a solid basis for beginners, but will also bring new perspectives to those already familiar with standard data analysis workflows for proteomics and next-generation sequencing applications.

# Target Audience

This course is oriented towards biologists and bioinformaticians with a particular interest in differential analysis for quantitative 'omics data.

# GitHub repository

All source and data files for this course are available on the accompanying [GitHub repository](https://github.com/statOmics/SGA21).

# Prerequisites

The prerequisites for the Statistical Genomics Analysis course are the successful completion of a basic course of statistics that covers topics on data exploration and descriptive statistics, statistical modeling, and inference: linear models, confidence intervals, t-tests, F-tests, anova, chi-squared test.
The basis concepts may be revisited in the online course at https://gtpb.github.io/PSLS20/ (English) and in https://statomics.github.io/statistiekCursusNotas/ (Dutch).

In addition, knowledge of programming in `R` is preferred. A primer to `R` and Data visualization in `R` can be found at:

 - `R` Basics: https://dodona.ugent.be/nl/courses/335/
 - `R` Data Exploration: https://dodona.ugent.be/nl/courses/345/


# Software 

- Participants are required to bring their own laptop with [R](https://www.r-project.org/) version 4.1.1 or greater. 

- We also recommend to also install the latest version of [RStudio](https://www.rstudio.com/products/rstudio/download/).

- Installation script: to install all required packages, please copy and paste this line of code in your R console.

```
source("https://raw.githubusercontent.com/statOmics/SGA21/master/install.R")
```

- Participants who have issues with the installation of the R/Rstudio can use an Rstudio instance in the cloud with all packages installed for the course in the mean time. Note, that this is instance is not for routine use. 

[![Binder](http://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/statOmics/SGA21/binder?urlpath=rstudio)

# Detailed Program


1. Position of the course: [PDF](./docs/intro.pdf) 

2. Recap Linear Models (Week 1)

   - Lecture: [HTML](./recapGeneralLinearModel.html), [PDF](./recapGeneralLinearModel.pdf)
   - Tutorial KPNA2: [HTML](./multipleRegression_KPNA2.html),  [PDF](./multipleRegression_KPNA2.pdf)


## Module I: Proteomics Data Analysis (Week 1-5)


1. Bioinformatics for proteomics
   
   - Lecture: [PDF](./docs/martens_proteomics_bioinformatics.pdf), [youtube](https://www.youtube.com/watch?v=ZgwNWRul98o)
   - Tutorials: [identification](https://www.compomics.com/bioinformatics-for-proteomics/)

\newline

2. Preprocessing & Analysis of Simple Designs (week 3)
   
   - Lecture: [Preprocessing](./pda_quantification_preprocessing.html) [[PDF](./pda_quantification_preprocessing.pdf)]
   - Tutorial: [Preprocessing](./pda_tutorialPreprocessing.html)
   - Wrap-up: [Peptide-based Models](./pda_robustSummarisation_peptideModels.html) [[PDF](./pda_robustSummarisation_peptideModels.pdf)]

\newline

3. Statistical Inference & Analysis of Factorial Designs (Week 3-4)
   
   - Lecture: [Differential Abundance Analysis](./pda_quantification_inference.html) [[PDF](./pda_quantification_inference.pdf)]
   - Tutorial: [Design](pda_tutorialDesign.html)
   - Wrap-up: [Blocking](./pda_blocking_wrapup.html) [[PDF](./pda_blocking_wrapup.pdf)]
   
4. Advanced materials and reading materials

   - [Technical details: Inference upon summarization](./technicalDetailsProteomics.html)
   - [Stage-wise testing](./docs/stagewiseTesting.pdf)
   
   - Papers: 

      - [Sticker et al. (2020) Robust summarization and inference in proteome-wide label-free quantification](https://www.biorxiv.org/content/10.1101/668863v1)
      - [Ludger Goeminne: Extensive Background on proteomics and proteomics data analysis (Introduction of PhD)](./docs/backgroundProteomicsDataAnalysis.pdf)
      - [Van den Berge et al. (2017) Stage-wise testing: stageR paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1277-0)

5. Solutions

   - CPTAC study: See [wrap-up preprocessing](./pda_robustSummarisation_peptideModels.html)
   - [cancer 6x6](cancer2_6x6.html)
   - mouse: See [wrap-up blocking](./pda_blocking_wrapup.html)
   - [Heart study](heartMainInteractionStageR.html)
   
## Module II: Bulk RNA-sequencing


1. Introduction to sequencing technology, raw data and preprocessing.
   
   - Lecture: [Introduction to sequencing](./sequencing_intro.html) [[PDF](./sequencing_intro.pdf)].
   - Tutorial: [Preprocessing RNA-seq data](./sequencing_preprocessing.sh)
   - Recommended reading: [RNA Sequencing Data: Hitchhiker's Guide to Expression Analysis](https://www.annualreviews.org/doi/abs/10.1146/annurev-biodatasci-072018-021255)

\newline

2. Working with count data and generalized linear models.
   
   - Lecture: [Working with count data and GLMs](./sequencing_countData.html) [[PDF](./sequencing_countData.pdf)].
   - Tutorial I: Embedded within the lecture: Analysis of bike-sharing data.
   - Tutorial II: [GLM for one gene](sequencing_lab_oneGene.html).

\newline

3. Analysis of RNA-seq data.

   - Lecture + Tutorial: [Four major challenges in the analysis of RNA-seq datasets](./sequencing_rnaseqIntro.html) [[PDF](./sequencing_rnaseqIntro.pdf)].
   - [Why we use offsets rather than count scaling](sequencing_scalingNormalization.html).

\newline
   
4. Technical topics in bulk RNA-seq differential expression analysis (DEA).

   - Lecture: [Technical topics in DEA](./sequencing_technicalDE.html) [[PDF](./sequencing_technicalDE.pdf)].
   
## Module III: Single-cell RNA-sequencing

1. General concepts and analysis workflow of single-cell RNA-seq data.

   - Lecture: [Slides](https://docs.google.com/presentation/d/1sfbtw52qWgA7TcDIrGB3fWksLM0Mu9O7gBqYgGrt9fM/edit?usp=sharing).
      - [Variance stabilizing transformation for a Poisson random variable](./singleCell_varStabilization.html) [[PDF](./singleCell_varStabilization.pdf)].


# Instructors

 - [Koen Van den Berge](https://koenvandenberge.github.io/)
 - [Lieven Clement](./instructors.html)
 - [Lennart Martens](https://www.compomics.com/people/lennart-martens/)



# License

<a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>

This project is licensed under the [Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0)](https://creativecommons.org/licenses/by-nc-sa/4.0)

</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("index.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
