---
title: "Statistical Genomics Analysis 2021-2022"
---


**Picture**

# Course Description

High throughput omics studies generate ever larger datasets and, as a consequence, complex data interpretation challenges. This course focuses on the statistical concepts involved in preprocessing, quantification and differential analysis of high-throughput omics data. Experimental design is essential to allow for correct interpretation, and concepts such as blocking in advanced experimental designs will also be introduced. The core focus will be on shotgun proteomics and (bulk and single-cell) next generation sequencing. The course will rely exclusively on free and user-friendly open-source tools in R/Bioconductor. We hope that this will provide a solid basis for beginners, but will also bring new perspectives to those already familiar with standard data analysis workflows for proteomics and next-generation sequencing applications.

# Target Audience

This course is oriented towards biologists and bioinformaticians with a particular interest in differential analysis for quantitative omics.

# Prerequisites

The prerequisites for the Statistical Genomics analysis course are the successful completion of a basic course of statistics that covers topics on data exploration and descriptive statistics, statistical modeling, and inference: linear models, confidence intervals, t-tests, F-tests, anova, chi-squared test.

The basis concepts may be revisited in the online course at https://gtpb.github.io/PSLS20/ (English) and in https://statomics.github.io/statistiekCursusNotas/ (Dutch).

A primer to R and Data visualization in R can be found at:

 - R Basics: https://dodona.ugent.be/nl/courses/335/
 - R Data Exploration: https://dodona.ugent.be/nl/courses/345/

# Lecturers

 - [Koen Van den Berge](https://koenvandenberge.github.io/)
 - [Lieven Clement](https://statomics.github.io/pages/about.html)
 
# Lecture material

All lecture material will be provided on this website. Please check the 'Lectures' tab at the top of the homepage.
