<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lieven Clement" />


<title>Statistical Methods for Quantitative MS-based Proteomics: Part II. Differential Abundance Analysis</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./docs/intro.pdf">1.1 Position</a>
    </li>
    <li>
      <a href="recapGeneralLinearModel.html">1.2 Recap linear model</a>
    </li>
    <li>
      <a href="./docs/martens_proteomics_bioinformatics.pdf">2.1 PDA - identification</a>
    </li>
    <li>
      <a href="pda_quantification_preprocessing.html">2.2 PDA- Quantification - Preprocessing</a>
    </li>
    <li>
      <a href="pda_robustSummarisation_peptideModels.html">2.2 PDA - Wrap-up Preprocessing - Peptide-Level-Models</a>
    </li>
    <li>
      <a href="pda_quantification_inference.html">2.3 PDA - Quantification - Differential Abundance Analysis</a>
    </li>
    <li>
      <a href="pda_blocking_wrapup.html">2.3 PDA - Wrap-up DA - Blocking</a>
    </li>
    <li>
      <a href="technicalDetailsProteomics.html">2.4. PDA - DA Technical Details</a>
    </li>
    <li>
      <a href="./docs/stagewiseTesting.pdf">2.5 PDA - Stage-wise testing</a>
    </li>
    <li>
      <a href="sequencing_intro.html">3.1 SDA - Intro to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">3.2 SDA - Working with count data</a>
    </li>
    <li>
      <a href="sequencing_rnaseqIntro.html">3.3 SDA - Intro to RNA-seq analysis</a>
    </li>
    <li>
      <a href="sequencing_technicalDE.html">3.4 SDA - Technical topics in DE analysis</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multipleRegression_KPNA2.html">1 Intro KPNA2</a>
    </li>
    <li>
      <a href="https://www.compomics.com/bioinformatics-for-proteomics/">2.1 Identification</a>
    </li>
    <li>
      <a href="pda_tutorialPreprocessing.html">2.2 DA - Preprocessing</a>
    </li>
    <li>
      <a href="pda_tutorialDesign.html">2.3 DA - Design</a>
    </li>
    <li>
      <a href="sequencing_preprocessing.sh">3.1 Preprocessing</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Statistical Methods for Quantitative MS-based Proteomics: Part II. Differential Abundance Analysis</h1>
<h4 class="author">Lieven Clement</h4>
<h4 class="date"><a href="https://statomics.github.io">statOmics</a>, Ghent University</h4>

</div>


<p><a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a></p>
<ul>
<li><a href="https://www.youtube.com/watch?v=1mhg3BCuEm8&amp;list=PLZH1hP8_LbJKqnPSS4hxTkn-tQolSCGjP">Playlist PDA Preprocessing</a></li>
</ul>
<div id="outline" class="section level1 unnumbered">
<h1>Outline</h1>
<ul>
<li>Francisella tularensis Example</li>
<li>Hypothesis testing</li>
<li>Multiple testing</li>
<li>Moderated statistics</li>
<li>Experimental design</li>
</ul>
<p>Note, that the R-code is included for learners who are aiming to develop R/markdown scripts to automate their quantitative proteomics data analyses. According to the target audience of the course we either work with a graphical user interface (GUI) in a R/shiny App msqrob2gui (e.g. Proteomics Bioinformatics course of the EBI and the Proteomics Data Analysis course at the Gulbenkian institute) or with R/markdowns scripts (e.g. Bioinformatics Summer School at UCLouvain or the Statistical Genomics Course at Ghent University).</p>
<hr />
</div>
<div id="francisella-tularensis-experiment" class="section level1">
<h1><span class="header-section-number">1</span> Francisella tularensis experiment</h1>
<p><img src="figures/francisella.jpg" width="50%" /></p>
<p><img src="figures/tularemia_lesion.jpg" /><!-- --></p>
<ul>
<li>Pathogen: causes tularemia</li>
<li>Metabolic adaptation key for intracellular life cycle of pathogenic microorganisms.</li>
<li>Upon entry into host cells quick phasomal escape and active multiplication in cytosolic compartment.</li>
<li>Franciscella is auxotroph for several amino acids, including arginine.</li>
<li>Inactivation of arginine transporter delayed bacterial phagosomal escape and intracellular multiplication.</li>
<li>Experiment to assess difference in proteome using 3 WT vs 3 ArgP KO mutants</li>
</ul>
<div id="import-the-data-in-r" class="section level2">
<h2><span class="header-section-number">1.1</span> Import the data in R</h2>
<details>
<summary> Click to see code </summary>
<p>
<ol style="list-style-type: decimal">
<li>Load libraries</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">library</span>(limma)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">library</span>(QFeatures)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw">library</span>(msqrob2)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="kw">library</span>(plotly)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="kw">library</span>(ggplot2)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>We use a peptides.txt file from MS-data quantified with maxquant that contains MS1 intensities summarized at the peptide level.</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>peptidesFile &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/statOmics/PDA21/data/quantification/francisella/peptides.txt&quot;</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Maxquant stores the intensity data for the different samples in columnns that start with Intensity. We can retreive the column names with the intensity data with the code below:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>ecols &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;Intensity</span><span class="ch">\\</span><span class="st">.&quot;</span>, <span class="kw">names</span>(<span class="kw">read.delim</span>(peptidesFile)))</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Read the data and store it in QFeatures object</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>pe &lt;-<span class="st"> </span><span class="kw">readQFeatures</span>(</span>
<span id="cb4-2"><a href="#cb4-2"></a>  <span class="dt">table =</span> peptidesFile,</span>
<span id="cb4-3"><a href="#cb4-3"></a>  <span class="dt">fnames =</span> <span class="dv">1</span>,</span>
<span id="cb4-4"><a href="#cb4-4"></a>  <span class="dt">ecol =</span> ecols,</span>
<span id="cb4-5"><a href="#cb4-5"></a>  <span class="dt">name =</span> <span class="st">&quot;peptideRaw&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>Update data with information on design</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">colData</span>(pe)<span class="op">$</span>genotype &lt;-<span class="st"> </span>pe[[<span class="dv">1</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="st">  </span>colnames <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="st">  </span><span class="kw">substr</span>(<span class="dv">12</span>,<span class="dv">13</span>) <span class="op">%&gt;%</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="st">  </span>as.factor <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="st">  </span><span class="kw">relevel</span>(<span class="st">&quot;WT&quot;</span>)</span>
<span id="cb5-6"><a href="#cb5-6"></a>pe <span class="op">%&gt;%</span><span class="st"> </span>colData</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 1 column
##                          genotype
##                          &lt;factor&gt;
## Intensity.1WT_20_2h_n3_3       WT
## Intensity.1WT_20_2h_n4_3       WT
## Intensity.1WT_20_2h_n5_3       WT
## Intensity.3D8_20_2h_n3_3       D8
## Intensity.3D8_20_2h_n4_3       D8
## Intensity.3D8_20_2h_n5_3       D8</code></pre>
</p>
</details>
</div>
<div id="preprocessing" class="section level2">
<h2><span class="header-section-number">1.2</span> Preprocessing</h2>
<details>
<summary> Click to see code to log-transfrom the data </summary>
<p>
<ol style="list-style-type: decimal">
<li>Log transform</li>
</ol>
<ul>
<li>Calculate number of non zero intensities for each peptide</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">rowData</span>(pe[[<span class="st">&quot;peptideRaw&quot;</span>]])<span class="op">$</span>nNonZero &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assay</span>(pe[[<span class="st">&quot;peptideRaw&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</span></code></pre></div>
<ul>
<li>Peptides with zero intensities are missing peptides and should be represent with a <code>NA</code> value rather than <code>0</code>.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>pe &lt;-<span class="st"> </span><span class="kw">zeroIsNA</span>(pe, <span class="st">&quot;peptideRaw&quot;</span>) <span class="co"># convert 0 to NA</span></span></code></pre></div>
<ul>
<li>Logtransform data with base 2</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>pe &lt;-<span class="st"> </span><span class="kw">logTransform</span>(pe, <span class="dt">base =</span> <span class="dv">2</span>, <span class="dt">i =</span> <span class="st">&quot;peptideRaw&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;peptideLog&quot;</span>)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Filtering</li>
</ol>
<ul>
<li>Handling overlapping protein groups</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe, <span class="op">~</span><span class="st"> </span>Proteins <span class="op">%in%</span><span class="st"> </span><span class="kw">smallestUniqueGroups</span>(<span class="kw">rowData</span>(pe[[<span class="st">&quot;peptideLog&quot;</span>]])<span class="op">$</span>Proteins))</span></code></pre></div>
<ul>
<li>Remove reverse sequences (decoys) and contaminants. Note that this is indicated by the column names Reverse and depending on the version of maxQuant with Potential.contaminants or Contaminants.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe,<span class="op">~</span>Reverse <span class="op">!=</span><span class="st"> &quot;+&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe,<span class="op">~</span><span class="st"> </span>Contaminant <span class="op">!=</span><span class="st"> &quot;+&quot;</span>)</span></code></pre></div>
<ul>
<li>Drop peptides that were only identified in one sample</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe,<span class="op">~</span><span class="st"> </span>nNonZero <span class="op">&gt;=</span><span class="dv">2</span>)</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="kw">nrow</span>(pe[[<span class="st">&quot;peptideLog&quot;</span>]])</span></code></pre></div>
<pre><code>## [1] 6525</code></pre>
<p>We keep 6525 peptides upon filtering.</p>
<ol start="3" style="list-style-type: decimal">
<li>Normalization by median centering</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>pe &lt;-<span class="st"> </span><span class="kw">normalize</span>(pe, </span>
<span id="cb14-2"><a href="#cb14-2"></a>                <span class="dt">i =</span> <span class="st">&quot;peptideLog&quot;</span>, </span>
<span id="cb14-3"><a href="#cb14-3"></a>                <span class="dt">name =</span> <span class="st">&quot;peptideNorm&quot;</span>, </span>
<span id="cb14-4"><a href="#cb14-4"></a>                <span class="dt">method =</span> <span class="st">&quot;center.median&quot;</span>)</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Summarization. We use the standard sumarisation in aggregateFeatures, which is a robust summarisation method.</li>
</ol>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>pe &lt;-<span class="st"> </span><span class="kw">aggregateFeatures</span>(pe,</span>
<span id="cb15-2"><a href="#cb15-2"></a>    <span class="dt">i =</span> <span class="st">&quot;peptideNorm&quot;</span>, </span>
<span id="cb15-3"><a href="#cb15-3"></a>    <span class="dt">fcol =</span> <span class="st">&quot;Proteins&quot;</span>, </span>
<span id="cb15-4"><a href="#cb15-4"></a>    <span class="dt">na.rm =</span> <span class="ot">TRUE</span>,</span>
<span id="cb15-5"><a href="#cb15-5"></a>    <span class="dt">name =</span> <span class="st">&quot;protein&quot;</span>)</span></code></pre></div>
<pre><code>## Your quantitative and row data contain missing values. Please read the
## relevant section(s) in the aggregateFeatures manual page regarding the
## effects of missing values on data aggregation.</code></pre>
<p>Plot of preprocessed data</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>pe[[<span class="st">&quot;peptideNorm&quot;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">genotype =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;genotype&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> intensity,<span class="dt">group =</span> sample,<span class="dt">color =</span> genotype)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="st">    </span><span class="kw">geom_density</span>() <span class="op">+</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Peptide-level&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 7561 rows containing non-finite values (stat_density).</code></pre>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>pe[[<span class="st">&quot;protein&quot;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">genotype =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;genotype&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> intensity,<span class="dt">group =</span> sample,<span class="dt">color =</span> genotype)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="st">    </span><span class="kw">geom_density</span>() <span class="op">+</span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Protein-level&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 428 rows containing non-finite values (stat_density).</code></pre>
<img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-16-2.png" width="672" />
</p>
</details>
</div>
<div id="summarized-data-structure" class="section level2">
<h2><span class="header-section-number">1.3</span> Summarized data structure</h2>
<div id="design" class="section level3">
<h3><span class="header-section-number">1.3.1</span> Design</h3>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>pe <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="st">  </span>colData <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="st">  </span>knitr<span class="op">::</span><span class="kw">kable</span>()</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">genotype</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Intensity.1WT_20_2h_n3_3</td>
<td align="left">WT</td>
</tr>
<tr class="even">
<td align="left">Intensity.1WT_20_2h_n4_3</td>
<td align="left">WT</td>
</tr>
<tr class="odd">
<td align="left">Intensity.1WT_20_2h_n5_3</td>
<td align="left">WT</td>
</tr>
<tr class="even">
<td align="left">Intensity.3D8_20_2h_n3_3</td>
<td align="left">D8</td>
</tr>
<tr class="odd">
<td align="left">Intensity.3D8_20_2h_n4_3</td>
<td align="left">D8</td>
</tr>
<tr class="even">
<td align="left">Intensity.3D8_20_2h_n5_3</td>
<td align="left">D8</td>
</tr>
</tbody>
</table>
<ul>
<li>WT vs KO</li>
<li>3 vs 3 repeats</li>
</ul>
</div>
<div id="summarized-intensity-matrix" class="section level3">
<h3><span class="header-section-number">1.3.2</span> Summarized intensity matrix</h3>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>pe[[<span class="st">&quot;protein&quot;</span>]] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">assay</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>() <span class="op">%&gt;%</span><span class="st"> </span>knitr<span class="op">::</span><span class="kw">kable</span>()</span></code></pre></div>
<table>
<colgroup>
<col width="7%" />
<col width="15%" />
<col width="15%" />
<col width="15%" />
<col width="15%" />
<col width="15%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Intensity.1WT_20_2h_n3_3</th>
<th align="right">Intensity.1WT_20_2h_n4_3</th>
<th align="right">Intensity.1WT_20_2h_n5_3</th>
<th align="right">Intensity.3D8_20_2h_n3_3</th>
<th align="right">Intensity.3D8_20_2h_n4_3</th>
<th align="right">Intensity.3D8_20_2h_n5_3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">WP_003013731</td>
<td align="right">-0.2748775</td>
<td align="right">-0.0856247</td>
<td align="right">0.1595370</td>
<td align="right">-0.2809009</td>
<td align="right">0.0035526</td>
<td align="right">0.0567110</td>
</tr>
<tr class="even">
<td align="left">WP_003013860</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">-0.2512039</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">-0.4865646</td>
</tr>
<tr class="odd">
<td align="left">WP_003013909</td>
<td align="right">-0.6851118</td>
<td align="right">-0.8161658</td>
<td align="right">-0.7557906</td>
<td align="right">-0.4591476</td>
<td align="right">-0.5449424</td>
<td align="right">-0.4962482</td>
</tr>
<tr class="even">
<td align="left">WP_003014068</td>
<td align="right">0.6495386</td>
<td align="right">0.8522239</td>
<td align="right">1.1344852</td>
<td align="right">0.5459176</td>
<td align="right">0.9187714</td>
<td align="right">0.5974741</td>
</tr>
<tr class="odd">
<td align="left">WP_003014122</td>
<td align="right">-0.7630863</td>
<td align="right">-1.0430741</td>
<td align="right">-0.8091715</td>
<td align="right">-1.1743951</td>
<td align="right">-1.1924725</td>
<td align="right">-1.2565893</td>
</tr>
<tr class="even">
<td align="left">WP_003014123</td>
<td align="right">-0.2051672</td>
<td align="right">-0.3361704</td>
<td align="right">-0.2151930</td>
<td align="right">-0.3855747</td>
<td align="right">-0.2802011</td>
<td align="right">-0.5801771</td>
</tr>
</tbody>
</table>
<ul>
<li>1115 proteins</li>
</ul>
</div>
<div id="hypothesis-testing-a-single-protein" class="section level3">
<h3><span class="header-section-number">1.3.3</span> Hypothesis testing: a single protein</h3>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<div id="t-test" class="section level4">
<h4><span class="header-section-number">1.3.3.1</span> T-test</h4>
<p><span class="math display">\[
 \log_2 \text{FC} = \bar{y}_{p1}-\bar{y}_{p2}
\]</span></p>
<p><span class="math display">\[
T_g=\frac{\log_2 \text{FC}}{\text{se}_{\log_2 \text{FC}}}
\]</span></p>
<p><span class="math display">\[
T_g=\frac{\widehat{\text{signal}}}{\widehat{\text{Noise}}}
\]</span></p>
<p>If we can assume equal variance in both treatment groups:</p>
<p><span class="math display">\[
\text{se}_{\log_2 \text{FC}}=\text{SD}\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}
\]</span></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>WP_<span class="dv">003023392</span> &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb23-2"><a href="#cb23-2"></a>    <span class="dt">intensity =</span> <span class="kw">assay</span>(pe[[<span class="st">&quot;protein&quot;</span>]][<span class="st">&quot;WP_003023392&quot;</span>,]) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">c</span>(), </span>
<span id="cb23-3"><a href="#cb23-3"></a>    <span class="dt">genotype =</span> <span class="kw">colData</span>(pe)[,<span class="dv">1</span>]) </span>
<span id="cb23-4"><a href="#cb23-4"></a></span>
<span id="cb23-5"><a href="#cb23-5"></a>WP_<span class="dv">003023392</span> <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb23-6"><a href="#cb23-6"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>genotype,<span class="dt">y=</span>intensity)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb23-8"><a href="#cb23-8"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Protein WP_003023392&quot;</span>)</span></code></pre></div>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p><span class="math display">\[
t=\frac{\log_2\widehat{\text{FC}}}{\text{se}_{\log_2\widehat{\text{FC}}}}=\frac{-1.43}{0.0577}=-24.7
\]</span></p>
<ul>
<li><p>Is t = -24.7 indicating that there is an effect?</p></li>
<li><p>How likely is it to observe t = -24.7 when there is no effect of the argP KO on the protein expression?</p></li>
</ul>
</div>
<div id="null-hypothesis-h_0-and-alternative-hypothesis-h_1" class="section level4">
<h4><span class="header-section-number">1.3.3.2</span> Null hypothesis (<span class="math inline">\(H_0\)</span>) and alternative hypothesis (<span class="math inline">\(H_1\)</span>)</h4>
<ul>
<li><p>With data we can never prove a hypothesis (falsification principle of Popper)</p></li>
<li><p>With data we can only reject a hypothesis</p></li>
<li><p>In general we start from <em>alternative hypothese</em> <span class="math inline">\(H_1\)</span>: we want to show an effect of the KO on a protein</p></li>
</ul>
<center>
<span class="math inline">\(H_1\)</span>: On average the protein abundance in WT is different from that in KO
</center>
<ul>
<li>But, we will assess this by falsifying the opposite:
<center>
<span class="math inline">\(H_0\)</span>: On average the protein abundance in WT is equal to that in KO&lt;-
</center></li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw">t.test</span>(intensity <span class="op">~</span><span class="st"> </span>genotype, <span class="dt">data =</span> WP_<span class="dv">003023392</span>, <span class="dt">var.equal=</span><span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## 
##  Two Sample t-test
## 
## data:  intensity by genotype
## t = 24.747, df = 4, p-value = 1.582e-05
## alternative hypothesis: true difference in means between group WT and group D8 is not equal to 0
## 95 percent confidence interval:
##  1.267666 1.588058
## sample estimates:
## mean in group WT mean in group D8 
##       -0.1821147       -1.6099769</code></pre>
<ul>
<li><p>How likely is it to observe an equal or more extreme effect than the one observed in the sample when the null hypothesis is true?</p></li>
<li><p>When we make assumptions about the distribution of our test statistic we can quantify this probability: <em>p-value</em>. The p-value will only be calculated correctly if the underlying assumptions hold!</p></li>
<li><p>When we repeat the experiment, the probability to observe a fold change for this gene that is more extreme than a 2.69 fold (<span class="math inline">\(\log_2 FC=-1.43\)</span>) down or up regulation by random change (if <span class="math inline">\(H_0\)</span> is true) is 16 out of 1 000 000.<br />
</p></li>
<li><p>If the p-value is below a significance threshold <span class="math inline">\(\alpha\)</span> we reject the null hypothesis. <em>We control the probability on a false positive result at the <span class="math inline">\(\alpha\)</span>-level (type I error)</em></p></li>
<li><p>Note, that the p-values are uniform under the null hypothesis, i.e. when <span class="math inline">\(H_0\)</span> is true all p-values are equally likely.</p></li>
</ul>
</div>
</div>
</div>
<div id="multiple-hypothesis-testing" class="section level2">
<h2><span class="header-section-number">1.4</span> Multiple hypothesis testing</h2>
<ul>
<li><p>Consider testing DA for all <span class="math inline">\(m=1066\)</span> proteins simultaneously</p></li>
<li><p>What if we assess each individual test at level <span class="math inline">\(\alpha\)</span>? <span class="math inline">\(\rightarrow\)</span> Probability to have a false positive (FP) among all m simultatenous test <span class="math inline">\(&gt;&gt;&gt; \alpha= 0.05\)</span></p></li>
<li><p>Indeed for each non DA protein we have a probability of 5% to return a FP.</p></li>
<li><p>In a typical experiment the majority of the proteins are non DA.</p></li>
<li><p>So an upperbound of the expected FP is <span class="math inline">\(m \times \alpha\)</span> or <span class="math inline">\(1066 \times 0.05=53\)</span>.</p></li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> Hence, we are bound to call many false positive proteins each time we run the experiment.</p>
<div id="multiple-testing" class="section level3">
<h3><span class="header-section-number">1.4.1</span> Multiple testing</h3>
<div id="family-wise-error-rate" class="section level4">
<h4><span class="header-section-number">1.4.1.1</span> Family-wise error rate</h4>
<p>The family-wise error rate (FWER) addresses the multiple testing issue by no longer controlling the individual type I error for each protein, instead it controls:</p>
<p><span class="math display">\[
   \text{FWER} = 
   \text{P}\left[FP \geq 1 \right].
\]</span></p>
<p>The Bonferroni method is widely used to control the type I error:</p>
<ul>
<li>assess each test at <span class="math display">\[\alpha_\text{adj}=\frac{\alpha}{m}\]</span></li>
<li>or use adjusted p-values and compare them to <span class="math inline">\(\alpha\)</span>: <span class="math display">\[p_\text{adj}=\text{min}\left(p \times m,1\right)\]</span></li>
</ul>
<p>Problem, the method is very conservative!</p>
</div>
<div id="false-discovery-rate" class="section level4">
<h4><span class="header-section-number">1.4.1.2</span> False discovery rate</h4>
<ul>
<li>FDR: Expected proportion of false positives on the total number of positives you return.</li>
<li>An FDR of 1% means that on average we expect 1% false positive proteins in the list of proteins that are called significant.</li>
</ul>
<p>The <strong>False Discovery Proportion (FDP)</strong> is the fraction of false positives that are returned, i.e. </p>
<p><span class="math display">\[
FDP = \frac{FP}{R}
\]</span></p>
<ul>
<li>However, this quantity cannot be observed because in practice we only know the number of proteins for which we rejected <span class="math inline">\(H_0\)</span>, <span class="math inline">\(R\)</span>.</li>
<li>But, we do not know the number of false positives, <span class="math inline">\(FP\)</span>.</li>
</ul>
<p>Therefore, Benjamini and Hochberg, 1995, defined The <strong>False Discovery Rate (FDR)</strong> as <span class="math display">\[
   \text{FDR} = \text{E}\left[\frac{FP}{R}\right] =\text{E}\left[\text{FDP}\right]
\]</span> the expected FDP.</p>
<ul>
<li>Controlling the FDR allows for more discoveries (i.e. longer lists with significant results), while the fraction of false discoveries among the significant results in well controlled on average. As a consequence, more of the true positive hypotheses will be detected.</li>
</ul>
</div>
<div id="intuition-of-bh-fdr-procedure" class="section level4">
<h4><span class="header-section-number">1.4.1.3</span> Intuition of BH-FDR procedure</h4>
<p>Consider <span class="math inline">\(m = 1000\)</span> tests</p>
<ul>
<li><p>Suppose that a researcher rejects all null hypotheses for which <span class="math inline">\(p &lt; 0.01\)</span>.</p></li>
<li><p>If we use <span class="math inline">\(p &lt; 0.01\)</span>, we expect <span class="math inline">\(0.01 \times m_0\)</span> tests to return false positives.</p></li>
<li><p>A conservative estimate of the number of false positives that we can expect can be obtained by considering that the null hypotheses are true for all features, <span class="math inline">\(m_0 = m = 1000\)</span>.</p></li>
<li><p>We then would expect <span class="math inline">\(0.01 \times 1000 = 10\)</span> false positives (<span class="math inline">\(FP=10\)</span>).</p></li>
<li><p>Suppose that the researcher found 200 genes with <span class="math inline">\(p&lt;0.01\)</span> (<span class="math inline">\(R=200\)</span>).</p></li>
<li><p>The proportion of false positive results (FDP = false positive proportion) among the list of <span class="math inline">\(R=200\)</span> genes can then be estimated as <span class="math display">\[
 \widehat{\text{FDP}}=\frac{FP}{R}=\frac{10}{200}=\frac{0.01 \times 1000}{200} = 0.05.
 \]</span></p></li>
</ul>
</div>
<div id="benjamini-and-hochberg-1995-procedure-for-controlling-the-fdr-at-alpha" class="section level4">
<h4><span class="header-section-number">1.4.1.4</span> Benjamini and Hochberg (1995) procedure for controlling the FDR at <span class="math inline">\(\alpha\)</span></h4>
<ol style="list-style-type: decimal">
<li><p>Let <span class="math inline">\(p_{(1)}\leq \ldots \leq p_{(m)}\)</span> denote the ordered <span class="math inline">\(p\)</span>-values.</p></li>
<li><p>Find the largest integer <span class="math inline">\(k\)</span> so that <span class="math display">\[
\frac{p_{(k)} \times m}{k} \leq \alpha
\]</span> <span class="math display">\[\text{or}\]</span> <span class="math display">\[
p_{(k)} \leq k \times \alpha/m
\]</span></p></li>
<li><p>If such a <span class="math inline">\(k\)</span> exists, reject the <span class="math inline">\(k\)</span> null hypotheses associated with <span class="math inline">\(p_{(1)}, \ldots, p_{(k)}\)</span>. Otherwise none of the null hypotheses is rejected.</p></li>
</ol>
<p>The adjusted <span class="math inline">\(p\)</span>-value (also known as the <span class="math inline">\(q\)</span>-value in FDR literature): <span class="math display">\[
   q_{(i)}=\tilde{p}_{(i)} = \min\left[\min_{j=i,\ldots, m}\left(m p_{(j)}/j\right), 1 \right].
 \]</span> In the hypothetical example above: <span class="math inline">\(k=200\)</span>, <span class="math inline">\(p_{(k)}=0.01\)</span>, <span class="math inline">\(m=1000\)</span> and <span class="math inline">\(\alpha=0.05\)</span>.</p>
</div>
<div id="francisella-example" class="section level4">
<h4><span class="header-section-number">1.4.1.5</span> Francisella Example</h4>
<details>
<summary> Click to see code </summary>
<p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>ttestMx &lt;-<span class="st"> </span><span class="cf">function</span>(y,group) {</span>
<span id="cb26-2"><a href="#cb26-2"></a>    test &lt;-<span class="st"> </span><span class="kw">try</span>(<span class="kw">t.test</span>(y[group],y[<span class="op">!</span>group],<span class="dt">var.equal=</span><span class="ot">TRUE</span>),<span class="dt">silent=</span><span class="ot">TRUE</span>)</span>
<span id="cb26-3"><a href="#cb26-3"></a>    <span class="cf">if</span>(<span class="kw">is</span>(test,<span class="st">&quot;try-error&quot;</span>)) {</span>
<span id="cb26-4"><a href="#cb26-4"></a>      <span class="kw">return</span>(<span class="kw">c</span>(<span class="dt">log2FC=</span><span class="ot">NA</span>,<span class="dt">se=</span><span class="ot">NA</span>,<span class="dt">tstat=</span><span class="ot">NA</span>,<span class="dt">p=</span><span class="ot">NA</span>))</span>
<span id="cb26-5"><a href="#cb26-5"></a>      } <span class="cf">else</span> {</span>
<span id="cb26-6"><a href="#cb26-6"></a>      <span class="kw">return</span>(<span class="kw">c</span>(<span class="dt">log2FC=</span> (test<span class="op">$</span>estimate<span class="op">%*%</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="op">-</span><span class="dv">1</span>)),<span class="dt">se=</span>test<span class="op">$</span>stderr,<span class="dt">tstat=</span>test<span class="op">$</span>statistic,<span class="dt">pval=</span>test<span class="op">$</span>p.value))</span>
<span id="cb26-7"><a href="#cb26-7"></a>      }</span>
<span id="cb26-8"><a href="#cb26-8"></a> }</span>
<span id="cb26-9"><a href="#cb26-9"></a> </span>
<span id="cb26-10"><a href="#cb26-10"></a> res &lt;-<span class="st"> </span><span class="kw">apply</span>(</span>
<span id="cb26-11"><a href="#cb26-11"></a>    <span class="kw">assay</span>(pe[[<span class="st">&quot;protein&quot;</span>]]), </span>
<span id="cb26-12"><a href="#cb26-12"></a>    <span class="dv">1</span>, </span>
<span id="cb26-13"><a href="#cb26-13"></a>    ttestMx,</span>
<span id="cb26-14"><a href="#cb26-14"></a>    <span class="dt">group =</span> <span class="kw">colData</span>(pe)<span class="op">$</span>genotype<span class="op">==</span><span class="st">&quot;D8&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb26-15"><a href="#cb26-15"></a><span class="st">  </span>t </span>
<span id="cb26-16"><a href="#cb26-16"></a> <span class="kw">colnames</span>(res) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;logFC&quot;</span>,<span class="st">&quot;se&quot;</span>,<span class="st">&quot;tstat&quot;</span>,<span class="st">&quot;pval&quot;</span>)</span>
<span id="cb26-17"><a href="#cb26-17"></a> res &lt;-<span class="st"> </span>res <span class="op">%&gt;%</span><span class="st"> </span>as.data.frame <span class="op">%&gt;%</span><span class="st"> </span>na.exclude <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(pval)</span>
<span id="cb26-18"><a href="#cb26-18"></a> res<span class="op">$</span>adjPval &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(res<span class="op">$</span>pval, <span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb26-19"><a href="#cb26-19"></a> alpha &lt;-<span class="st"> </span><span class="fl">0.05</span></span>
<span id="cb26-20"><a href="#cb26-20"></a>res<span class="op">$</span>adjAlphaForm &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(res),<span class="st">&quot; x &quot;</span>,alpha,<span class="st">&quot;/&quot;</span>,<span class="kw">nrow</span>(res))</span>
<span id="cb26-21"><a href="#cb26-21"></a>res<span class="op">$</span>adjAlpha &lt;-<span class="st"> </span>alpha <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(res))<span class="op">/</span><span class="kw">nrow</span>(res) </span>
<span id="cb26-22"><a href="#cb26-22"></a>res<span class="op">$</span><span class="st">&quot;pval &lt; adjAlpha&quot;</span> &lt;-<span class="st"> </span>res<span class="op">$</span>pval <span class="op">&lt;</span><span class="st"> </span>res<span class="op">$</span>adjAlpha </span>
<span id="cb26-23"><a href="#cb26-23"></a>res<span class="op">$</span><span class="st">&quot;adjPval &lt; alpha&quot;</span> &lt;-<span class="st"> </span>res<span class="op">$</span>adjPval <span class="op">&lt;</span><span class="st"> </span>alpha </span></code></pre></div>
</p>
</details>
<p>FWER: Bonferroni method:<span class="math inline">\(\alpha_\text{adj} = \alpha/m = 0.05 / 1066= 5\times 10^{-5}\)</span></p>
<table>
<colgroup>
<col width="12%" />
<col width="10%" />
<col width="9%" />
<col width="9%" />
<col width="14%" />
<col width="9%" />
<col width="15%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">logFC</th>
<th align="right">pval</th>
<th align="right">adjPval</th>
<th align="left">adjAlphaForm</th>
<th align="right">adjAlpha</th>
<th align="left">pval &lt; adjAlpha</th>
<th align="left">adjPval &lt; alpha</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">WP_003038940</td>
<td align="right">-0.2876290</td>
<td align="right">0.0000146</td>
<td align="right">0.0084347</td>
<td align="left">1 x 0.05/1066</td>
<td align="right">0.0000469</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">WP_003023392</td>
<td align="right">-1.4278622</td>
<td align="right">0.0000158</td>
<td align="right">0.0084347</td>
<td align="left">2 x 0.05/1066</td>
<td align="right">0.0000938</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">WP_003039212</td>
<td align="right">-0.2658247</td>
<td align="right">0.0000820</td>
<td align="right">0.0291520</td>
<td align="left">3 x 0.05/1066</td>
<td align="right">0.0001407</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">WP_003026016</td>
<td align="right">-1.0800305</td>
<td align="right">0.0001395</td>
<td align="right">0.0346124</td>
<td align="left">4 x 0.05/1066</td>
<td align="right">0.0001876</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">WP_003039615</td>
<td align="right">-0.3992190</td>
<td align="right">0.0001623</td>
<td align="right">0.0346124</td>
<td align="left">5 x 0.05/1066</td>
<td align="right">0.0002345</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">WP_011733588</td>
<td align="right">-0.4323262</td>
<td align="right">0.0002291</td>
<td align="right">0.0407034</td>
<td align="left">6 x 0.05/1066</td>
<td align="right">0.0002814</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">WP_003014552</td>
<td align="right">-0.9843865</td>
<td align="right">0.0003224</td>
<td align="right">0.0440266</td>
<td align="left">7 x 0.05/1066</td>
<td align="right">0.0003283</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">WP_003040849</td>
<td align="right">-1.2780743</td>
<td align="right">0.0003304</td>
<td align="right">0.0440266</td>
<td align="left">8 x 0.05/1066</td>
<td align="right">0.0003752</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">WP_003038430</td>
<td align="right">-0.4331987</td>
<td align="right">0.0004505</td>
<td align="right">0.0489078</td>
<td align="left">9 x 0.05/1066</td>
<td align="right">0.0004221</td>
<td align="left">FALSE</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">WP_003033975</td>
<td align="right">-0.2949061</td>
<td align="right">0.0005047</td>
<td align="right">0.0489078</td>
<td align="left">10 x 0.05/1066</td>
<td align="right">0.0004690</td>
<td align="left">FALSE</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">WP_011733645</td>
<td align="right">0.3531405</td>
<td align="right">0.0005171</td>
<td align="right">0.0489078</td>
<td align="left">11 x 0.05/1066</td>
<td align="right">0.0005159</td>
<td align="left">FALSE</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">WP_011733723</td>
<td align="right">-0.3935768</td>
<td align="right">0.0005506</td>
<td align="right">0.0489078</td>
<td align="left">12 x 0.05/1066</td>
<td align="right">0.0005629</td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">WP_003038679</td>
<td align="right">-0.3909725</td>
<td align="right">0.0007083</td>
<td align="right">0.0580821</td>
<td align="left">13 x 0.05/1066</td>
<td align="right">0.0006098</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">WP_003033719</td>
<td align="right">-1.1865453</td>
<td align="right">0.0008426</td>
<td align="right">0.0603810</td>
<td align="left">14 x 0.05/1066</td>
<td align="right">0.0006567</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="right">…</td>
<td align="right">…</td>
<td align="right">…</td>
<td align="left">…</td>
<td align="right">…</td>
<td align="left">…</td>
<td align="left">…</td>
</tr>
<tr class="even">
<td align="left">WP_003040562</td>
<td align="right">0.0039480</td>
<td align="right">0.9976429</td>
<td align="right">0.9985797</td>
<td align="left">1065 x 0.05/1066</td>
<td align="right">0.0499531</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">WP_003041130</td>
<td align="right">0.0002941</td>
<td align="right">0.9992812</td>
<td align="right">0.9992812</td>
<td align="left">1066 x 0.05/1066</td>
<td align="right">0.05</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
</tbody>
</table>
</div>
<div id="results" class="section level4">
<h4><span class="header-section-number">1.4.1.6</span> Results</h4>
<details>
<summary> Click to see code </summary>
<p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>volcanoT &lt;-<span class="st"> </span>res <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> logFC, <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(pval), <span class="dt">color =</span> adjPval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)) <span class="op">+</span></span>
<span id="cb27-3"><a href="#cb27-3"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">cex =</span> <span class="fl">2.5</span>) <span class="op">+</span></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">alpha</span>(<span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="fl">0.5</span>)) <span class="op">+</span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="st">    </span><span class="kw">theme_minimal</span>() </span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>volcanoT</span></code></pre></div>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>
</div>
</div>
</div>
<div id="moderated-statistics" class="section level2">
<h2><span class="header-section-number">1.5</span> Moderated Statistics</h2>
<p>Problems with ordinary t-test</p>
<details>
<summary> Click to see code </summary>
<p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>problemPlots &lt;-<span class="st"> </span><span class="kw">list</span>() </span>
<span id="cb29-2"><a href="#cb29-2"></a>problemPlots[[<span class="dv">1</span>]] &lt;-<span class="st"> </span>res <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb29-3"><a href="#cb29-3"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> logFC, <span class="dt">y =</span> se, <span class="dt">color =</span> adjPval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)) <span class="op">+</span></span>
<span id="cb29-4"><a href="#cb29-4"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">cex =</span> <span class="fl">2.5</span>) <span class="op">+</span></span>
<span id="cb29-5"><a href="#cb29-5"></a><span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">alpha</span>(<span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="fl">0.5</span>)) <span class="op">+</span></span>
<span id="cb29-6"><a href="#cb29-6"></a><span class="st">    </span><span class="kw">theme_minimal</span>() </span>
<span id="cb29-7"><a href="#cb29-7"></a></span>
<span id="cb29-8"><a href="#cb29-8"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)</span>
<span id="cb29-9"><a href="#cb29-9"></a>{</span>
<span id="cb29-10"><a href="#cb29-10"></a>problemPlots[[i]] &lt;-<span class="st"> </span><span class="kw">colData</span>(pe) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb29-11"><a href="#cb29-11"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb29-12"><a href="#cb29-12"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">intensity =</span> pe[[<span class="st">&quot;protein&quot;</span>]][<span class="kw">rownames</span>(res)[i],] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb29-13"><a href="#cb29-13"></a><span class="st">             </span>assay <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb29-14"><a href="#cb29-14"></a><span class="st">             </span>c) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb29-15"><a href="#cb29-15"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>genotype,<span class="dt">y=</span>intensity)) <span class="op">+</span></span>
<span id="cb29-16"><a href="#cb29-16"></a><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb29-17"><a href="#cb29-17"></a><span class="st">    </span><span class="kw">ylim</span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">0</span>) <span class="op">+</span></span>
<span id="cb29-18"><a href="#cb29-18"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="kw">rownames</span>(res)[i])</span>
<span id="cb29-19"><a href="#cb29-19"></a>}</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>problemPlots</span></code></pre></div>
<pre><code>## [[1]]</code></pre>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<pre><code>## 
## [[2]]</code></pre>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-28-2.png" width="672" /></p>
<pre><code>## 
## [[3]]</code></pre>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-28-3.png" width="672" /></p>
<p>A general class of moderated test statistics is given by <span class="math display">\[
   T_g^{mod} = \frac{\bar{Y}_{g1} - \bar{Y}_{g2}}{C \quad \tilde{S}_g} ,
 \]</span> where <span class="math inline">\(\tilde{S}_g\)</span> is a moderated standard deviation estimate.</p>
<ul>
<li><span class="math inline">\(C\)</span> is a constant depending on the design e.g. <span class="math inline">\(\sqrt{1/{n_1}+1/n_2}\)</span> for a t-test and of another form for linear models.</li>
<li><span class="math inline">\(\tilde{S}_g=S_g+S_0\)</span>: add small positive constant to denominator of t-statistic.</li>
<li>This can be adopted in Perseus.</li>
</ul>
<details>
<summary> Click to see code </summary>
<p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>simI&lt;-<span class="kw">sapply</span>(res<span class="op">$</span>se<span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">1</span><span class="op">/</span><span class="dv">3</span><span class="op">+</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>),<span class="cf">function</span>(n,mean,sd) <span class="kw">rnorm</span>(n,mean,sd),<span class="dt">n=</span><span class="dv">6</span>,<span class="dt">mean=</span><span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span>t</span>
<span id="cb34-2"><a href="#cb34-2"></a>resSim &lt;-<span class="st"> </span><span class="kw">apply</span>(</span>
<span id="cb34-3"><a href="#cb34-3"></a>    simI, </span>
<span id="cb34-4"><a href="#cb34-4"></a>    <span class="dv">1</span>, </span>
<span id="cb34-5"><a href="#cb34-5"></a>    ttestMx,</span>
<span id="cb34-6"><a href="#cb34-6"></a>    <span class="dt">group =</span> <span class="kw">colData</span>(pe)<span class="op">$</span>genotype<span class="op">==</span><span class="st">&quot;D8&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb34-7"><a href="#cb34-7"></a><span class="st">  </span>t </span>
<span id="cb34-8"><a href="#cb34-8"></a> <span class="kw">colnames</span>(resSim) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;logFC&quot;</span>,<span class="st">&quot;se&quot;</span>,<span class="st">&quot;tstat&quot;</span>,<span class="st">&quot;pval&quot;</span>)</span>
<span id="cb34-9"><a href="#cb34-9"></a> resSim &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(resSim)</span>
<span id="cb34-10"><a href="#cb34-10"></a> tstatSimPlot &lt;-<span class="st"> </span>resSim <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb34-11"><a href="#cb34-11"></a><span class="st">   </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>tstat)) <span class="op">+</span></span>
<span id="cb34-12"><a href="#cb34-12"></a><span class="st">     </span><span class="kw">geom_histogram</span>(<span class="kw">aes</span>(<span class="dt">y=</span>..density.., <span class="dt">fill=</span>..count..),<span class="dt">bins=</span><span class="dv">30</span>) <span class="op">+</span></span>
<span id="cb34-13"><a href="#cb34-13"></a><span class="st">     </span><span class="kw">stat_function</span>(<span class="dt">fun=</span>dt,</span>
<span id="cb34-14"><a href="#cb34-14"></a>    <span class="dt">color=</span><span class="st">&quot;red&quot;</span>,</span>
<span id="cb34-15"><a href="#cb34-15"></a>    <span class="dt">args=</span><span class="kw">list</span>(<span class="dt">df=</span><span class="dv">4</span>)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb34-16"><a href="#cb34-16"></a><span class="st">   </span><span class="kw">ylim</span>(<span class="dv">0</span>,.<span class="dv">6</span>) <span class="op">+</span></span>
<span id="cb34-17"><a href="#cb34-17"></a><span class="st">   </span><span class="kw">ggtitle</span>(<span class="st">&quot;t-statistic&quot;</span>)</span>
<span id="cb34-18"><a href="#cb34-18"></a></span>
<span id="cb34-19"><a href="#cb34-19"></a> </span>
<span id="cb34-20"><a href="#cb34-20"></a> resSim<span class="op">$</span>C &lt;-<span class="st"> </span><span class="kw">sqrt</span>(<span class="dv">1</span><span class="op">/</span><span class="dv">3</span><span class="op">+</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>) </span>
<span id="cb34-21"><a href="#cb34-21"></a> resSim<span class="op">$</span>sd &lt;-<span class="st"> </span>resSim<span class="op">$</span>se<span class="op">/</span>resSim<span class="op">$</span>C </span>
<span id="cb34-22"><a href="#cb34-22"></a> tstatSimPerseus &lt;-<span class="st"> </span>resSim <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb34-23"><a href="#cb34-23"></a><span class="st">   </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>logFC<span class="op">/</span>((sd<span class="fl">+.1</span>)<span class="op">*</span>C))) <span class="op">+</span></span>
<span id="cb34-24"><a href="#cb34-24"></a><span class="st">     </span><span class="kw">geom_histogram</span>(<span class="kw">aes</span>(<span class="dt">y=</span>..density.., <span class="dt">fill=</span>..count..),<span class="dt">bins=</span><span class="dv">30</span>) <span class="op">+</span></span>
<span id="cb34-25"><a href="#cb34-25"></a><span class="st">     </span><span class="kw">stat_function</span>(<span class="dt">fun=</span>dt,</span>
<span id="cb34-26"><a href="#cb34-26"></a>                   <span class="dt">color=</span><span class="st">&quot;red&quot;</span>,</span>
<span id="cb34-27"><a href="#cb34-27"></a>                  <span class="dt">args=</span><span class="kw">list</span>(<span class="dt">df=</span><span class="dv">4</span>)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb34-28"><a href="#cb34-28"></a><span class="st">     </span><span class="kw">ylim</span>(<span class="dv">0</span>,.<span class="dv">6</span>) <span class="op">+</span></span>
<span id="cb34-29"><a href="#cb34-29"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Persues&quot;</span>)</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>gridExtra<span class="op">::</span><span class="kw">grid.arrange</span>(tstatSimPlot,tstatSimPerseus,<span class="dt">nrow=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-30-1.png" width="672" /></p>
<ul>
<li>The choice of <span class="math inline">\(S_0\)</span> in Perseus is ad hoc and the t-statistic is no-longer t-distributed.</li>
<li>Permutation test, but is difficult for more complex designs.</li>
<li>Allows for Data Dredging because user can choose <span class="math inline">\(S_0\)</span></li>
</ul>
<div id="empirical-bayes" class="section level3">
<h3><span class="header-section-number">1.5.1</span> Empirical Bayes</h3>
<p><img src="figures/limmaShrinkage.png" width="50%" /></p>
<p>Figure courtesy to Rafael Irizarry</p>
<p><span class="math display">\[
   T_g^{mod} = \frac{\bar{Y}_{g1} - \bar{Y}_{g2}}{C \quad \tilde{S}_g} ,
 \]</span></p>
<ul>
<li><strong>empirical Bayes</strong> theory provides formal framework for borrowing strength across proteins,</li>
<li>Implemented in popular bioconductor package <strong>limma</strong> and <strong>msqrob2</strong></li>
</ul>
<p><span class="math display">\[
  \tilde{S}_g=\sqrt{\frac{d_gS_g^2+d_0S_0^2}{d_g+d_0}},
\]</span></p>
<ul>
<li><span class="math inline">\(S_0^2\)</span>: common variance (over all proteins)</li>
<li>Moderated t-statistic is t-distributed with <span class="math inline">\(d_0+d_g\)</span> degrees of freedom.</li>
<li>Note that the degrees of freedom increase by borrowing strength across proteins!
<details>
<summary> Click to see the code </summary>
<p></li>
</ul>
<ol style="list-style-type: decimal">
<li>We model the protein level expression values using the <code>msqrob</code> function. By default <code>msqrob2</code> estimates the model parameters using robust regression.</li>
</ol>
<p>We will model the data with a different group mean for every genotype. The group is incoded in the variable <code>genotype</code> of the colData. We can specify this model by using a formula with the factor <code>genotype</code> as its predictor: <code>formula = ~genotype</code>.</p>
<p>Note, that a formula always starts with a symbol ‘~’.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>pe &lt;-<span class="st"> </span><span class="kw">msqrob</span>(<span class="dt">object =</span> pe, <span class="dt">i =</span> <span class="st">&quot;protein&quot;</span>, <span class="dt">formula =</span> <span class="op">~</span>genotype)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Inference</li>
</ol>
<p>We first explore the design of the model that we specified using the the package <code>ExploreModelMatrix</code></p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="kw">library</span>(ExploreModelMatrix)</span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="kw">VisualizeDesign</span>(<span class="kw">colData</span>(pe),<span class="op">~</span>genotype)<span class="op">$</span>plotlist[[<span class="dv">1</span>]]</span></code></pre></div>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-33-1.png" width="672" /></p>
<p>We have two model parameters, the (Intercept) and genotypeD8. This results in a model with two group means:</p>
<ol style="list-style-type: decimal">
<li>For the wild type (WT) the expected value (mean) of the log2 transformed intensity y for a protein will be modelled using</li>
</ol>
<p><span class="math display">\[\text{E}[Y\vert \text{genotype}=\text{WT}] = \text{(Intercept)}\]</span> 1. For the knockout genotype D8 the expected value (mean) of the log2 transformed intensity y for a protein will be modelled using</p>
<p><span class="math display">\[\text{E}[Y\vert \text{genotype}=\text{D8}] = \text{(Intercept)} + \text{genotypeD8}\]</span></p>
<p>The average log2FC between D8 and WT is thus <span class="math display">\[\log_2\text{FC}_{D8-WT}= \text{E}[Y\vert \text{genotype}=\text{D8}] - \text{E}[Y\vert \text{genotype}=\text{WT}] = \text{genotypeD8}
\]</span></p>
<p>Hence, assessing the null hypothesis that there is no differential abundance between D8 and WT can be reformulated as</p>
<p><span class="math display">\[H_0:  \text{genotypeD8}=0\]</span> We can implement a hypothesis test for each protein in msqrob2 using the code below:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>L &lt;-<span class="st"> </span><span class="kw">makeContrast</span>(<span class="st">&quot;genotypeD8 = 0&quot;</span>, <span class="dt">parameterNames =</span> <span class="kw">c</span>(<span class="st">&quot;genotypeD8&quot;</span>))</span>
<span id="cb38-2"><a href="#cb38-2"></a>pe &lt;-<span class="st"> </span><span class="kw">hypothesisTest</span>(<span class="dt">object =</span> pe, <span class="dt">i =</span> <span class="st">&quot;protein&quot;</span>, <span class="dt">contrast =</span> L)</span></code></pre></div>
<p>We can show the list with all significant DE proteins at the 5% FDR using</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>genotypeD8 <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb39-2"><a href="#cb39-2"></a><span class="st">  </span><span class="kw">arrange</span>(pval) <span class="op">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3"></a><span class="st">  </span><span class="kw">filter</span>(adjPval<span class="op">&lt;</span><span class="fl">0.05</span>)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["logFC"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["se"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["df"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["t"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["pval"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["adjPval"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"-1.4278622","2":"0.09546104","3":"6.353267","4":"-14.957538","5":"3.435824e-06","6":"0.003456439","_rn_":"WP_003023392"},{"1":"-1.2857733","2":"0.12226820","3":"6.236222","4":"-10.516007","5":"3.372825e-05","6":"0.016965307","_rn_":"WP_003040849"},{"1":"-0.9843865","2":"0.10770682","3":"6.353267","4":"-9.139501","5":"6.920701e-05","6":"0.022593404","_rn_":"WP_003014552"},{"1":"-1.1970073","2":"0.13408775","3":"6.220667","4":"-8.927044","5":"8.983461e-05","6":"0.022593404","_rn_":"WP_003033719"},{"1":"-1.3531336","2":"0.16187947","3":"6.353267","4":"-8.358896","5":"1.174333e-04","6":"0.023627585","_rn_":"WP_003040790"},{"1":"-1.0584448","2":"0.11240892","3":"5.442252","4":"-9.416022","5":"1.429981e-04","6":"0.023976009","_rn_":"WP_003026016"},{"1":"-0.8815374","2":"0.12199689","3":"6.353267","4":"-7.225901","5":"2.740575e-04","6":"0.039385984","_rn_":"WP_003033905"},{"1":"-0.7635044","2":"0.10902744","3":"6.353267","4":"-7.002865","5":"3.279672e-04","6":"0.041241874","_rn_":"WP_003039713"},{"1":"-0.9184399","2":"0.13755363","3":"6.353267","4":"-6.676959","5":"4.299645e-04","6":"0.041477282","_rn_":"WP_003039530"},{"1":"-0.9789058","2":"0.12767782","3":"5.353267","4":"-7.667000","5":"4.409838e-04","6":"0.041477282","_rn_":"WP_003014581"},{"1":"-1.3800203","2":"0.20369907","3":"6.080188","4":"-6.774799","5":"4.777868e-04","6":"0.041477282","_rn_":"WP_003038816"},{"1":"-0.7657848","2":"0.11307420","3":"6.033034","4":"-6.772410","5":"4.947588e-04","6":"0.041477282","_rn_":"WP_003033046"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>We can also visualise the results using a volcanoplot</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>volcano &lt;-<span class="st"> </span><span class="kw">ggplot</span>(</span>
<span id="cb40-2"><a href="#cb40-2"></a>    <span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>genotypeD8,</span>
<span id="cb40-3"><a href="#cb40-3"></a>    <span class="kw">aes</span>(<span class="dt">x =</span> logFC, <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(pval), <span class="dt">color =</span> adjPval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb40-4"><a href="#cb40-4"></a>) <span class="op">+</span></span>
<span id="cb40-5"><a href="#cb40-5"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">cex =</span> <span class="fl">2.5</span>) <span class="op">+</span></span>
<span id="cb40-6"><a href="#cb40-6"></a><span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">alpha</span>(<span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="fl">0.5</span>)) <span class="op">+</span></span>
<span id="cb40-7"><a href="#cb40-7"></a><span class="st">    </span><span class="kw">theme_minimal</span>() <span class="op">+</span></span>
<span id="cb40-8"><a href="#cb40-8"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;msqrob2&quot;</span>)</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a>gridExtra<span class="op">::</span><span class="kw">grid.arrange</span>(volcanoT <span class="op">+</span><span class="st">    </span><span class="kw">xlim</span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">3</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;ordinary t-test&quot;</span>)</span>
<span id="cb41-2"><a href="#cb41-2"></a>,volcano <span class="op">+</span><span class="st">     </span><span class="kw">xlim</span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">3</span>)</span>
<span id="cb41-3"><a href="#cb41-3"></a>,<span class="dt">nrow=</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 109 rows containing missing values (geom_point).</code></pre>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-37-1.png" width="672" /></p>
<ul>
<li><p>The volcano plot opens up when using the EB variance estimator</p></li>
<li><p>Borrowing strength to estimate the variance using empirical Bayes solves the issue of returning proteins with a low fold change as significant due to a low variance.</p></li>
</ul>
</div>
<div id="shrinkage-of-the-variance-and-moderated-t-statistics" class="section level3">
<h3><span class="header-section-number">1.5.2</span> Shrinkage of the variance and moderated t-statistics</h3>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="kw">qplot</span>(</span>
<span id="cb43-2"><a href="#cb43-2"></a>  <span class="kw">sapply</span>(<span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>msqrobModels,getSigma),</span>
<span id="cb43-3"><a href="#cb43-3"></a>  <span class="kw">sapply</span>(<span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>msqrobModels,getSigmaPosterior)) <span class="op">+</span></span>
<span id="cb43-4"><a href="#cb43-4"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;SD&quot;</span>) <span class="op">+</span></span>
<span id="cb43-5"><a href="#cb43-5"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;moderated SD&quot;</span>) <span class="op">+</span></span>
<span id="cb43-6"><a href="#cb43-6"></a><span class="st">  </span><span class="kw">geom_abline</span>(<span class="dt">intercept =</span> <span class="dv">0</span>,<span class="dt">slope =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb43-7"><a href="#cb43-7"></a><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> ) </span></code></pre></div>
<pre><code>## Warning: Removed 109 rows containing missing values (geom_point).</code></pre>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-38-1.png" width="672" /></p>
<ul>
<li>Small variances are shrunken towards the common variance resulting in large EB variance estimates</li>
<li>Large variances are shrunken towards the common variance resulting in smaller EB variance estimates</li>
<li>Pooled degrees of freedom of the EB variance estimator are larger because information is borrowed across proteins to estimate the variance</li>
</ul>
</div>
</div>
<div id="plots" class="section level2">
<h2><span class="header-section-number">1.6</span> Plots</h2>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a>sigNames &lt;-<span class="st"> </span><span class="kw">rowData</span>(pe[[<span class="st">&quot;protein&quot;</span>]])<span class="op">$</span>genotypeD8 <span class="op">%&gt;%</span></span>
<span id="cb45-2"><a href="#cb45-2"></a><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;protein&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb45-3"><a href="#cb45-3"></a><span class="st">    </span><span class="kw">filter</span>(adjPval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>) <span class="op">%&gt;%</span></span>
<span id="cb45-4"><a href="#cb45-4"></a><span class="st">    </span><span class="kw">pull</span>(protein)</span>
<span id="cb45-5"><a href="#cb45-5"></a><span class="kw">heatmap</span>(<span class="kw">assay</span>(pe[[<span class="st">&quot;protein&quot;</span>]])[sigNames, ])</span></code></pre></div>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-39-1.png" width="672" /></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a><span class="cf">for</span> (protName <span class="cf">in</span> sigNames)</span>
<span id="cb46-2"><a href="#cb46-2"></a>    {</span>
<span id="cb46-3"><a href="#cb46-3"></a>        pePlot &lt;-<span class="st"> </span>pe[protName, , <span class="kw">c</span>(<span class="st">&quot;peptideNorm&quot;</span>, <span class="st">&quot;protein&quot;</span>)]</span>
<span id="cb46-4"><a href="#cb46-4"></a>        pePlotDf &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">longFormat</span>(pePlot))</span>
<span id="cb46-5"><a href="#cb46-5"></a>        pePlotDf<span class="op">$</span>assay &lt;-<span class="st"> </span><span class="kw">factor</span>(pePlotDf<span class="op">$</span>assay,</span>
<span id="cb46-6"><a href="#cb46-6"></a>            <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;peptideNorm&quot;</span>, <span class="st">&quot;protein&quot;</span>)</span>
<span id="cb46-7"><a href="#cb46-7"></a>        )</span>
<span id="cb46-8"><a href="#cb46-8"></a>        pePlotDf<span class="op">$</span>genotype &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">colData</span>(pePlot)[pePlotDf<span class="op">$</span>colname, <span class="st">&quot;genotype&quot;</span>])</span>
<span id="cb46-9"><a href="#cb46-9"></a></span>
<span id="cb46-10"><a href="#cb46-10"></a>        <span class="co"># plotting</span></span>
<span id="cb46-11"><a href="#cb46-11"></a>        p1 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(</span>
<span id="cb46-12"><a href="#cb46-12"></a>            <span class="dt">data =</span> pePlotDf,</span>
<span id="cb46-13"><a href="#cb46-13"></a>            <span class="kw">aes</span>(<span class="dt">x =</span> colname, <span class="dt">y =</span> value, <span class="dt">group =</span> rowname)</span>
<span id="cb46-14"><a href="#cb46-14"></a>        ) <span class="op">+</span></span>
<span id="cb46-15"><a href="#cb46-15"></a><span class="st">            </span><span class="kw">geom_line</span>() <span class="op">+</span></span>
<span id="cb46-16"><a href="#cb46-16"></a><span class="st">            </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb46-17"><a href="#cb46-17"></a><span class="st">            </span><span class="kw">facet_grid</span>(<span class="op">~</span>assay) <span class="op">+</span></span>
<span id="cb46-18"><a href="#cb46-18"></a><span class="st">            </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">70</span>, <span class="dt">hjust =</span> <span class="dv">1</span>, <span class="dt">vjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span></span>
<span id="cb46-19"><a href="#cb46-19"></a><span class="st">            </span><span class="kw">ggtitle</span>(protName)</span>
<span id="cb46-20"><a href="#cb46-20"></a>        <span class="kw">print</span>(p1)</span>
<span id="cb46-21"><a href="#cb46-21"></a></span>
<span id="cb46-22"><a href="#cb46-22"></a>        <span class="co"># plotting 2</span></span>
<span id="cb46-23"><a href="#cb46-23"></a>        p2 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(pePlotDf, <span class="kw">aes</span>(<span class="dt">x =</span> colname, <span class="dt">y =</span> value, <span class="dt">fill =</span> genotype)) <span class="op">+</span></span>
<span id="cb46-24"><a href="#cb46-24"></a><span class="st">            </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape =</span> <span class="ot">NA</span>) <span class="op">+</span></span>
<span id="cb46-25"><a href="#cb46-25"></a><span class="st">            </span><span class="kw">geom_point</span>(</span>
<span id="cb46-26"><a href="#cb46-26"></a>                <span class="dt">position =</span> <span class="kw">position_jitter</span>(<span class="dt">width =</span> <span class="fl">.1</span>),</span>
<span id="cb46-27"><a href="#cb46-27"></a>                <span class="kw">aes</span>(<span class="dt">shape =</span> rowname)</span>
<span id="cb46-28"><a href="#cb46-28"></a>            ) <span class="op">+</span></span>
<span id="cb46-29"><a href="#cb46-29"></a><span class="st">            </span><span class="kw">scale_shape_manual</span>(<span class="dt">values =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(pePlotDf)) <span class="op">+</span></span>
<span id="cb46-30"><a href="#cb46-30"></a><span class="st">            </span><span class="kw">labs</span>(<span class="dt">title =</span> protName, <span class="dt">x =</span> <span class="st">&quot;sample&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;peptide intensity (log2)&quot;</span>) <span class="op">+</span></span>
<span id="cb46-31"><a href="#cb46-31"></a><span class="st">            </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">70</span>, <span class="dt">hjust =</span> <span class="dv">1</span>, <span class="dt">vjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span></span>
<span id="cb46-32"><a href="#cb46-32"></a><span class="st">            </span><span class="kw">facet_grid</span>(<span class="op">~</span>assay)</span>
<span id="cb46-33"><a href="#cb46-33"></a>        <span class="kw">print</span>(p2)</span>
<span id="cb46-34"><a href="#cb46-34"></a>}</span></code></pre></div>
<p><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-1.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-2.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-3.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-4.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-5.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-6.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-7.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-8.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-9.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-10.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-11.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-12.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-13.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-14.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-15.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-16.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-17.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-18.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-19.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-20.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-21.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-22.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-23.png" width="672" /><img src="pda_quantification_inference_noFrames_files/figure-html/unnamed-chunk-40-24.png" width="672" /></p>
</div>
</div>
<div id="experimental-design" class="section level1">
<h1><span class="header-section-number">2</span> Experimental Design</h1>
<div id="sample-size" class="section level2">
<h2><span class="header-section-number">2.1</span> Sample size</h2>
<p><span class="math display">\[
 \log_2 \text{FC} = \bar{y}_{p1}-\bar{y}_{p2}
\]</span></p>
<p><span class="math display">\[
T_g=\frac{\log_2 \text{FC}}{\text{se}_{\log_2 \text{FC}}}
\]</span></p>
<p><span class="math display">\[
T_g=\frac{\widehat{\text{signal}}}{\widehat{\text{Noise}}}
\]</span></p>
<p>If we can assume equal variance in both treatment groups:</p>
<p><span class="math display">\[
\text{se}_{\log_2 \text{FC}}=\text{SD}\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}
\]</span></p>
<p><span class="math inline">\(\rightarrow\)</span> if number of bio-repeats increases we have a higher power!</p>
<ul>
<li>cfr. Study of tamoxifen treated Estrogen Recepter (ER) positive breast cancer patients</li>
</ul>
</div>
<div id="blocking" class="section level2">
<h2><span class="header-section-number">2.2</span> Blocking</h2>
<p><span class="math display">\[\sigma^2= \sigma^2_{bio}+\sigma^2_\text{lab} +\sigma^2_\text{extraction} + \sigma^2_\text{run} + \ldots\]</span></p>
<ul>
<li>Biological: fluctuations in protein level between mice, fluctations in protein level between cells, …</li>
<li>Technical: cage effect, lab effect, week effect, plasma extraction, MS-run, …</li>
</ul>
</div>
<div id="nature-methods-points-of-significance---blocking" class="section level2">
<h2><span class="header-section-number">2.3</span> Nature methods: Points of significance - Blocking</h2>
<p><a href="https://www.nature.com/articles/nmeth.3005.pdf">https://www.nature.com/articles/nmeth.3005.pdf</a></p>
</div>
<div id="mouse-example" class="section level2">
<h2><span class="header-section-number">2.4</span> Mouse example</h2>
<p><img src="figures/mouseTcell_RCB_design.png" width="50%" /> Duguet et al. (2017) MCP 16(8):1416-1432. doi: 10.1074/mcp.m116.062745</p>
<ul>
<li>All treatments of interest are present within block!</li>
<li>We can estimate the effect of the treatment within block!</li>
<li>We can isolate the between block variability from the analysis using linear model: <span class="math display">\[ 
y \sim \text{type} + \text{mouse}
\]</span></li>
<li>Not possible with Perseus!</li>
</ul>
<div id="assess-the-impact-of-blocking-in-the-tutorial-session" class="section level3">
<h3><span class="header-section-number">2.4.1</span> Assess the impact of blocking in the tutorial session!</h3>
<ul>
<li>Completely randomized design with only one cell type per mouse (Treg and Tconv)</li>
</ul>
<p><span class="math display">\[\updownarrow\]</span></p>
<ul>
<li>Randomized complete block design assessing Treg and Tconv on each mouse</li>
</ul>
</div>
</div>
</div>

<div id="rmd-source-code">---
title: "Statistical Methods for Quantitative MS-based Proteomics: Part II. Differential Abundance Analysis"
author: "Lieven Clement"
date: "[statOmics](https://statomics.github.io), Ghent University"
output:
    html_document:
      code_download: true
      theme: cosmo
      toc: true
      toc_float: true
      highlight: tango
      number_sections: true
bibliography: msqrob2.bib

---

<a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>

- [Playlist PDA Preprocessing](https://www.youtube.com/watch?v=1mhg3BCuEm8&list=PLZH1hP8_LbJKqnPSS4hxTkn-tQolSCGjP)

# Outline {-}

- Francisella tularensis Example
- Hypothesis testing
- Multiple testing
- Moderated statistics
- Experimental design


Note, that the R-code is included for learners who are aiming to develop R/markdown scripts to automate their quantitative proteomics data analyses.
According to the target audience of the course we either work with a graphical user interface (GUI) in a R/shiny App msqrob2gui (e.g. Proteomics Bioinformatics course of the EBI and the Proteomics Data Analysis course at the Gulbenkian institute) or with R/markdowns scripts (e.g. Bioinformatics Summer School at UCLouvain or the Statistical Genomics Course at Ghent University). 

---

# Francisella tularensis experiment

```{r echo=FALSE,out.width="50%"}
knitr::include_graphics("./figures/francisella.jpg")
```

```{r echo=FALSE}
knitr::include_graphics("./figures/tularemia_lesion.jpg")
```

- Pathogen: causes tularemia
- Metabolic adaptation key for intracellular life cycle of pathogenic microorganisms. 
- Upon entry into host cells quick phasomal escape and active multiplication in cytosolic compartment.
- Franciscella is auxotroph for several amino acids, including arginine. 
- Inactivation of arginine transporter delayed bacterial phagosomal escape and intracellular multiplication. 
- Experiment to assess difference in proteome using 3 WT vs 3 ArgP KO mutants


## Import the data in R 


<details><summary> Click to see code </summary><p>
1. Load libraries 

```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(limma)
library(QFeatures)
library(msqrob2)
library(plotly)
library(ggplot2)
```

2. We use a peptides.txt file from MS-data quantified with maxquant that 
contains MS1 intensities summarized at the peptide level. 
```{r}
peptidesFile <- "https://raw.githubusercontent.com/statOmics/PDA21/data/quantification/francisella/peptides.txt"
```

3. Maxquant stores the intensity data for the different samples in columnns that start with Intensity. We can retreive the column names with the intensity data with the code below: 

```{r}
ecols <- grep("Intensity\\.", names(read.delim(peptidesFile)))
```

4. Read the data and store it in  QFeatures object 

```{r}
pe <- readQFeatures(
  table = peptidesFile,
  fnames = 1,
  ecol = ecols,
  name = "peptideRaw", sep="\t")
```

5. Update data with information on design

```{r}
colData(pe)$genotype <- pe[[1]] %>% 
  colnames %>% 
  substr(12,13) %>%
  as.factor %>% 
  relevel("WT")
pe %>% colData
```

</p></details>

## Preprocessing

<details><summary> Click to see code to log-transfrom the data </summary><p>

1. Log transform

  - Calculate number of non zero intensities for each peptide
```{r}
rowData(pe[["peptideRaw"]])$nNonZero <- rowSums(assay(pe[["peptideRaw"]]) > 0)
```

  - Peptides with zero intensities are missing peptides and should be represent
with a `NA` value rather than `0`.

```{r}
pe <- zeroIsNA(pe, "peptideRaw") # convert 0 to NA
```

  - Logtransform data with base 2

```{r}
pe <- logTransform(pe, base = 2, i = "peptideRaw", name = "peptideLog")
```

2. Filtering

  - Handling overlapping protein groups

```{r}
pe <- filterFeatures(pe, ~ Proteins %in% smallestUniqueGroups(rowData(pe[["peptideLog"]])$Proteins))
```

  - Remove reverse sequences (decoys) and contaminants. Note that this is indicated by the column names Reverse and depending on the version of maxQuant with Potential.contaminants or Contaminants.


```{r}
pe <- filterFeatures(pe,~Reverse != "+")
pe <- filterFeatures(pe,~ Contaminant != "+")
```

  - Drop peptides that were only identified in one sample

```{r}
pe <- filterFeatures(pe,~ nNonZero >=2)
nrow(pe[["peptideLog"]])
```

We keep `r nrow(pe[["peptideLog"]])` peptides upon filtering.

3. Normalization by median centering

```{r}
pe <- normalize(pe, 
                i = "peptideLog", 
                name = "peptideNorm", 
                method = "center.median")
```


4. Summarization. We use the standard sumarisation in aggregateFeatures, which is a
robust summarisation method.

```{r,warning=FALSE}
pe <- aggregateFeatures(pe,
    i = "peptideNorm", 
    fcol = "Proteins", 
    na.rm = TRUE,
    name = "protein")
```


Plot of preprocessed data 

```{r}
pe[["peptideNorm"]] %>% 
  assay %>%
  as.data.frame() %>%
  gather(sample, intensity) %>% 
  mutate(genotype = colData(pe)[sample,"genotype"]) %>%
  ggplot(aes(x = intensity,group = sample,color = genotype)) + 
    geom_density() +
    ggtitle("Peptide-level")

pe[["protein"]] %>% 
  assay %>%
  as.data.frame() %>%
  gather(sample, intensity) %>% 
  mutate(genotype = colData(pe)[sample,"genotype"]) %>%
  ggplot(aes(x = intensity,group = sample,color = genotype)) + 
    geom_density() +
    ggtitle("Protein-level")
```
</p></details>

## Summarized data structure

### Design

```{r}
pe %>% 
  colData %>% 
  knitr::kable()
```

- WT vs KO 
- 3 vs 3 repeats 

### Summarized intensity matrix

```{r}
pe[["protein"]] %>% assay() %>% head() %>% knitr::kable()
```

- `r nrow(pe[["protein"]])` proteins 

### Hypothesis testing: a single protein 

```{r echo=FALSE}
if ("pi"%in%ls()) rm("pi")
kopvoeter<-function(x,y,angle=0,l=.2,cex.dot=.5,pch=19,col="black")
{
angle=angle/180*pi
points(x,y,cex=cex.dot,pch=pch,col=col)
lines(c(x,x+l*cos(-pi/2+angle)),c(y,y+l*sin(-pi/2+angle)),col=col)
lines(c(x+l/2*cos(-pi/2+angle),x+l/2*cos(-pi/2+angle)+l/4*cos(angle)),c(y+l/2*sin(-pi/2+angle),y+l/2*sin(-pi/2+angle)+l/4*sin(angle)),col=col)
lines(c(x+l/2*cos(-pi/2+angle),x+l/2*cos(-pi/2+angle)+l/4*cos(pi+angle)),c(y+l/2*sin(-pi/2+angle),y+l/2*sin(-pi/2+angle)+l/4*sin(pi+angle)),col=col)
lines(c(x+l*cos(-pi/2+angle),x+l*cos(-pi/2+angle)+l/2*cos(-pi/2+pi/4+angle)),c(y+l*sin(-pi/2+angle),y+l*sin(-pi/2+angle)+l/2*sin(-pi/2+pi/4+angle)),col=col)
lines(c(x+l*cos(-pi/2+angle),x+l*cos(-pi/2+angle)+l/2*cos(-pi/2-pi/4+angle)),c(y+l*sin(-pi/2+angle),y+l*sin(-pi/2+angle)+l/2*sin(-pi/2-pi/4+angle)),col=col)
}

par(mar=c(0,0,0,0),mai=c(0,0,0,0))
plot(0,0,xlab="",ylab="",xlim=c(0,10),ylim=c(0,10),col=0,xaxt="none",yaxt="none",axes=FALSE)
rect(0,6,10,10,border="red",lwd=2)
text(.5,8,"population",srt=90,col="red",cex=2)
symbols (3, 8, circles=1.2, col="red",add=TRUE,fg="red",inches=FALSE,lwd=2)
set.seed(330)
grid=seq(0,1,.01)

for (i in 1:50)
{
	angle1=runif(n=1,min=0,max=360)
	angle2=runif(n=1,min=0,max=360)
	radius=sample(grid,prob=grid^2*pi/sum(grid^2*pi),size=1)
	kopvoeter(3+radius*cos(angle1/180*pi),8+radius*sin(angle1/180*pi),angle=angle2)
}
text(7.5,8,"Effect of arginine def. in population",col="red",cex=1.2)

rect(0,0,10,4,border="blue",lwd=2)
text(.5,2,"sample",srt=90,col="blue",cex=2)
symbols (3, 2, circles=1.2, col="red",add=TRUE,fg="blue",inches=FALSE,lwd=2)
for (i in 0:1)
	for (j in 0:2)
{

	kopvoeter(2.5+j*(3.9-2.1)/4,1.5+i)
}
text(7.5,2,"Effect of arginine def. in sample",col="blue",cex=1.2)

arrows(3,5.9,3,4.1,col="black",lwd=3)
arrows(7,4.1,7,5.9,col="black",lwd=3)
text(1.5,5,"Exp. Design",col="black",cex=1.2)
text(8.5,5,"Estimation \n Inference ",col="black",cex=1.2)
text(7.5,.5,"Data exploration",col="black",cex=1.2)
```

#### T-test


$$
 \log_2 \text{FC} = \bar{y}_{p1}-\bar{y}_{p2}
$$

$$
T_g=\frac{\log_2 \text{FC}}{\text{se}_{\log_2 \text{FC}}}
$$

$$
T_g=\frac{\widehat{\text{signal}}}{\widehat{\text{Noise}}}
$$

If we can assume equal variance in both treatment groups:

$$
\text{se}_{\log_2 \text{FC}}=\text{SD}\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}
$$

```{r}
WP_003023392 <- data.frame(
    intensity = assay(pe[["protein"]]["WP_003023392",]) %>% c(), 
    genotype = colData(pe)[,1]) 

WP_003023392 %>% 
  ggplot(aes(x=genotype,y=intensity)) + 
  geom_point() +
  ggtitle("Protein WP_003023392")
```

```{r echo=FALSE}
lmHlp <- lm(intensity ~ genotype, data = WP_003023392)
```

$$
t=\frac{\log_2\widehat{\text{FC}}}{\text{se}_{\log_2\widehat{\text{FC}}}}=\frac{`r format(summary(lmHlp)$coef[2,1], digit = 3)`}{`r format(summary(lmHlp)$coef[2,2], digit = 3)`}=`r format(summary(lmHlp)$coef[2,3], digit = 3)`
$$

- Is t = `r format(summary(lmHlp)$coef[2,3], digit = 3)` indicating that
there is an effect?

- How likely is it to observe
t = `r format(summary(lmHlp)$coef[2,3], digit = 3)` when there is no effect of the argP KO on the protein expression?

#### Null hypothesis ($H_0$) and alternative hypothesis ($H_1$)

- With data we can never prove a hypothesis (falsification principle of Popper)
- With data we can only reject a hypothesis 

- In general we start from *alternative hypothese* $H_1$: we want to show an effect of the KO on a protein

<center>
$H_1$: On average the protein abundance in WT is different from that in KO
</center>

- But, we will assess this by falsifying the opposite: 
<center>
$H_0$: On average the protein abundance in WT is equal to that in KO<-
</center>


```{r}
t.test(intensity ~ genotype, data = WP_003023392, var.equal=TRUE)
```

- How likely is it to observe an equal or more extreme effect than the one observed in the sample when the null hypothesis is true?
- When we make assumptions about the distribution of our test statistic we can quantify this probability: *p-value*. 
The p-value will only be calculated correctly if the underlying assumptions hold!
- When we repeat the experiment, the probability to observe a fold change for this gene that is more extreme than a `r format(2^abs(lmHlp$coef[2]),digits=3)` fold ($\log_2 FC=`r format(lmHlp$coef[2],digits=3)`$) down or up regulation by random change (if $H_0$ is true) is `r round(summary(lmHlp)$coef[2,4]*1e6,0)` out of 1 000 000.  
- If the p-value is below a significance threshold $\alpha$ we reject the null hypothesis. *We control the probability on a false positive result at the $\alpha$-level (type I error)*

- Note, that the p-values are uniform under the null hypothesis, i.e. when $H_0$ is true all p-values are equally likely. 

## Multiple hypothesis testing

- Consider testing DA for all $m=1066$ proteins simultaneously
- What if we assess each individual test at level $\alpha$?
$\rightarrow$ Probability to have a false positive (FP) among all m simultatenous
test $>>>  \alpha= 0.05$

- Indeed for each non DA protein we have a probability of 5% to return a FP.
- In a typical experiment the majority of the proteins are non DA. 
- So an upperbound of the expected FP is $m \times \alpha$ or $1066 \times 0.05=`r round(1066*0.05,0)`$. 

$\rightarrow$ Hence, we are bound to call many false positive proteins each time we run the experiment.

### Multiple testing

#### Family-wise error rate

The family-wise error rate (FWER) addresses the multiple testing issue by no longer controlling the individual type I error for each protein, instead it controls:  

\[
   \text{FWER} = 
   \text{P}\left[FP \geq 1 \right].
\]

The Bonferroni method is widely used to control the type I error: 

- assess each test at 
\[\alpha_\text{adj}=\frac{\alpha}{m}\]
- or use adjusted p-values and compare them to $\alpha$: 
\[p_\text{adj}=\text{min}\left(p \times m,1\right)\]

Problem, the method is very conservative! 

#### False discovery rate

- FDR: Expected proportion of false positives on the total number of positives you return.
- An FDR of 1% means that on average we expect 1% false positive proteins in the list of proteins that are called significant.

The **False Discovery Proportion (FDP)** is the fraction of false positives that are returned, i.e. 

\[
FDP = \frac{FP}{R}
\]

- However, this quantity cannot be observed because in practice we only know the number of proteins for which we rejected $H_0$, $R$. 
- But, we do not know the number of false positives, $FP$.

Therefore, Benjamini and Hochberg, 1995, defined The **False Discovery Rate (FDR)** as
\[
   \text{FDR} = \text{E}\left[\frac{FP}{R}\right] =\text{E}\left[\text{FDP}\right]
\]
the expected FDP. 

- Controlling the FDR allows for more discoveries (i.e. longer lists with significant results), while the fraction of false discoveries among the significant results in well controlled on average. As a consequence, more of the true positive hypotheses will be detected.


#### Intuition of BH-FDR procedure

Consider $m = 1000$ tests

- Suppose that a researcher rejects all null hypotheses for which $p < 0.01$. 

- If we use $p < 0.01$, we expect $0.01 \times m_0$ tests to return false positives. 
- A conservative estimate of the number of false positives that we can expect can be obtained by considering that the null hypotheses are true for all features, $m_0 = m =  1000$. 
- We then would expect $0.01 \times 1000 = 10$ false positives ($FP=10$).

- Suppose that the researcher found 200 genes with $p<0.01$ ($R=200$).

- The proportion of false positive results (FDP = false positive proportion) among the list of $R=200$ genes can then be estimated as
 \[
   \widehat{\text{FDP}}=\frac{FP}{R}=\frac{10}{200}=\frac{0.01 \times 1000}{200} = 0.05.
 \]


#### Benjamini and Hochberg (1995) procedure for controlling the FDR at $\alpha$

1. Let $p_{(1)}\leq \ldots \leq p_{(m)}$ denote the ordered $p$-values.

2. Find the largest integer $k$ so that 
$$
\frac{p_{(k)} \times m}{k} \leq \alpha
$$
$$\text{or}$$
$$
p_{(k)} \leq k \times \alpha/m
$$

3. If such a $k$ exists, reject the $k$ null hypotheses associated with $p_{(1)}, \ldots, p_{(k)}$.
Otherwise none of the null hypotheses is rejected.

The adjusted $p$-value (also known as the $q$-value in FDR literature):
 $$
   q_{(i)}=\tilde{p}_{(i)} = \min\left[\min_{j=i,\ldots, m}\left(m p_{(j)}/j\right), 1 \right].
 $$
 In the hypothetical example above: $k=200$, $p_{(k)}=0.01$, $m=1000$ and $\alpha=0.05$.

#### Francisella Example

<details><summary> Click to see code </summary><p>
```{r}
ttestMx <- function(y,group) {
    test <- try(t.test(y[group],y[!group],var.equal=TRUE),silent=TRUE)
    if(is(test,"try-error")) {
      return(c(log2FC=NA,se=NA,tstat=NA,p=NA))
      } else {
      return(c(log2FC= (test$estimate%*%c(1,-1)),se=test$stderr,tstat=test$statistic,pval=test$p.value))
      }
 }
 
 res <- apply(
    assay(pe[["protein"]]), 
    1, 
    ttestMx,
    group = colData(pe)$genotype=="D8") %>% 
  t 
 colnames(res) <- c("logFC","se","tstat","pval")
 res <- res %>% as.data.frame %>% na.exclude %>% arrange(pval)
 res$adjPval <- p.adjust(res$pval, "fdr")
 alpha <- 0.05
res$adjAlphaForm <- paste0(1:nrow(res)," x ",alpha,"/",nrow(res))
res$adjAlpha <- alpha * (1:nrow(res))/nrow(res) 
res$"pval < adjAlpha" <- res$pval < res$adjAlpha 
res$"adjPval < alpha" <- res$adjPval < alpha 
```
</p></details>

FWER: Bonferroni method:$\alpha_\text{adj} = \alpha/m = 0.05 / `r nrow(res)`= `r round(alpha/nrow(res),5)`$

```{r echo=FALSE} 
head(res[,-c(2:3)],sum(res$adjPval < alpha)+2) %>% knitr::kable()
```
| ... | ... | ... | ... | ... | ... | ... | ... |
|WP_003040562 | 0.0039480| 0.9976429| 0.9985797|1065 x 0.05/1066 | 0.0499531|FALSE           |FALSE    
|WP_003041130 | 0.0002941| 0.9992812| 0.9992812|1066 x 0.05/1066 |     0.05|FALSE           |FALSE           |




#### Results
<details><summary> Click to see code </summary><p>
```{r}
volcanoT <- res %>% 
  ggplot(aes(x = logFC, y = -log10(pval), color = adjPval < 0.05)) +
    geom_point(cex = 2.5) +
    scale_color_manual(values = alpha(c("black", "red"), 0.5)) +
    theme_minimal() 
```
</p></details>


```{r}
volcanoT
```


## Moderated Statistics


Problems with ordinary t-test

<details><summary> Click to see code </summary><p>
```{r}
problemPlots <- list() 
problemPlots[[1]] <- res %>% 
  ggplot(aes(x = logFC, y = se, color = adjPval < 0.05)) +
    geom_point(cex = 2.5) +
    scale_color_manual(values = alpha(c("black", "red"), 0.5)) +
    theme_minimal() 

for (i in 2:3)
{
problemPlots[[i]] <- colData(pe) %>% 
    as.data.frame %>% 
    mutate(intensity = pe[["protein"]][rownames(res)[i],] %>% 
             assay %>% 
             c) %>% 
    ggplot(aes(x=genotype,y=intensity)) +
    geom_point() + 
    ylim(-3,0) +
    ggtitle(rownames(res)[i])
}
```
</p></details>

```{r}
problemPlots
```

A general class of moderated test statistics is given by
 \[
   T_g^{mod} = \frac{\bar{Y}_{g1} - \bar{Y}_{g2}}{C \quad \tilde{S}_g} ,
 \]
 where $\tilde{S}_g$ is a moderated standard deviation estimate. 

- $C$ is a constant depending on the design e.g. $\sqrt{1/{n_1}+1/n_2}$ for a t-test and of another form for linear models.
- $\tilde{S}_g=S_g+S_0$: add small positive constant to denominator of t-statistic. 
- This can be adopted in Perseus. 

<details><summary> Click to see code </summary><p>
```{r}
simI<-sapply(res$se/sqrt(1/3+1/3),function(n,mean,sd) rnorm(n,mean,sd),n=6,mean=0) %>% t
resSim <- apply(
    simI, 
    1, 
    ttestMx,
    group = colData(pe)$genotype=="D8") %>% 
  t 
 colnames(resSim) <- c("logFC","se","tstat","pval")
 resSim <- as.data.frame(resSim)
 tstatSimPlot <- resSim %>% 
   ggplot(aes(x=tstat)) +
     geom_histogram(aes(y=..density.., fill=..count..),bins=30) +
     stat_function(fun=dt,
    color="red",
    args=list(df=4)) + 
   ylim(0,.6) +
   ggtitle("t-statistic")

 
 resSim$C <- sqrt(1/3+1/3) 
 resSim$sd <- resSim$se/resSim$C 
 tstatSimPerseus <- resSim %>% 
   ggplot(aes(x=logFC/((sd+.1)*C))) +
     geom_histogram(aes(y=..density.., fill=..count..),bins=30) +
     stat_function(fun=dt,
                   color="red",
                  args=list(df=4)) + 
     ylim(0,.6) +
    ggtitle("Persues")
```

</p></details>

```{r}
gridExtra::grid.arrange(tstatSimPlot,tstatSimPerseus,nrow=1)
```

- The choice of $S_0$ in Perseus is ad hoc and the t-statistic is no-longer t-distributed. 
- Permutation test, but is difficult for more complex designs.
- Allows for Data Dredging because user can choose $S_0$ 


### Empirical Bayes 


```{r echo=FALSE, out.width="50%"}
knitr::include_graphics("./figures/limmaShrinkage.png")
```

Figure courtesy to Rafael Irizarry

$$
   T_g^{mod} = \frac{\bar{Y}_{g1} - \bar{Y}_{g2}}{C \quad \tilde{S}_g} ,
 $$

- **empirical Bayes** theory provides formal framework for borrowing strength across proteins,
- Implemented in popular bioconductor package **limma** and **msqrob2**

$$
  \tilde{S}_g=\sqrt{\frac{d_gS_g^2+d_0S_0^2}{d_g+d_0}},
$$

- $S_0^2$: common variance (over all proteins) 
- Moderated t-statistic is t-distributed with $d_0+d_g$ degrees of freedom. 
- Note that the degrees of freedom increase by borrowing strength across proteins!
<details><summary> Click to see the code </summary><p>   

1. We model the protein level expression values using the `msqrob` function.
By default `msqrob2` estimates the model parameters using robust regression.

We will model the data with a different group mean for every genotype. 
The group is incoded in the variable `genotype` of the colData. 
We can specify this model by using a formula with the factor `genotype` as its predictor: 
`formula = ~genotype`.

Note, that a formula always starts with a symbol '~'.
```{r warning=FALSE}
pe <- msqrob(object = pe, i = "protein", formula = ~genotype)
```

2. Inference 

We first explore the design of the model that we specified using the the package `ExploreModelMatrix` 

```{r}
library(ExploreModelMatrix)
VisualizeDesign(colData(pe),~genotype)$plotlist[[1]]
```

We have two model parameters, the (Intercept) and genotypeD8. 
This results in a model with two group means: 

1. For the wild type (WT) the expected value (mean) of the log2 transformed intensity y for a protein will be modelled using 

$$\text{E}[Y\vert \text{genotype}=\text{WT}] = \text{(Intercept)}$$ 
1. For the knockout genotype D8 the expected value (mean) of the log2 transformed intensity y for a protein will be modelled using 

$$\text{E}[Y\vert \text{genotype}=\text{D8}] = \text{(Intercept)} + \text{genotypeD8}$$ 

The average log2FC between D8 and WT is thus
$$\log_2\text{FC}_{D8-WT}= \text{E}[Y\vert \text{genotype}=\text{D8}] - \text{E}[Y\vert \text{genotype}=\text{WT}] = \text{genotypeD8}
$$

Hence, assessing the null hypothesis that there is no differential abundance between D8 and WT can be reformulated as

$$H_0:  \text{genotypeD8}=0$$
We can implement a hypothesis test for each protein in msqrob2 using the code below: 

```{r}
L <- makeContrast("genotypeD8 = 0", parameterNames = c("genotypeD8"))
pe <- hypothesisTest(object = pe, i = "protein", contrast = L)
```

We can show the list with all significant DE proteins at the 5% FDR using 
```{r}
rowData(pe[["protein"]])$genotypeD8 %>% 
  arrange(pval) %>%
  filter(adjPval<0.05)
```

We can also visualise the results using a volcanoplot

```{r}
volcano <- ggplot(
    rowData(pe[["protein"]])$genotypeD8,
    aes(x = logFC, y = -log10(pval), color = adjPval < 0.05)
) +
    geom_point(cex = 2.5) +
    scale_color_manual(values = alpha(c("black", "red"), 0.5)) +
    theme_minimal() +
    ggtitle("msqrob2")
```
</p></details>


```{r}
gridExtra::grid.arrange(volcanoT +    xlim(-3,3) + ggtitle("ordinary t-test")
,volcano +     xlim(-3,3)
,nrow=2)
```

- The volcano plot opens up when using the EB variance estimator

-  Borrowing strength to estimate the variance using empirical Bayes solves the issue of returning proteins with a low fold change as significant due to a low variance. 


### Shrinkage of the  variance and moderated t-statistics

```{r}
qplot(
  sapply(rowData(pe[["protein"]])$msqrobModels,getSigma),
  sapply(rowData(pe[["protein"]])$msqrobModels,getSigmaPosterior)) +
  xlab("SD") +
  ylab("moderated SD") +
  geom_abline(intercept = 0,slope = 1) +
  geom_hline(yintercept = ) 
```

- Small variances are shrunken towards the common variance resulting in large EB variance estimates
- Large variances are shrunken towards the common variance resulting in smaller EB variance estimates 
- Pooled degrees of freedom of the EB variance estimator are larger because information is borrowed across proteins to estimate the variance

## Plots 

```{r}
sigNames <- rowData(pe[["protein"]])$genotypeD8 %>%
    rownames_to_column("protein") %>%
    filter(adjPval < 0.05) %>%
    pull(protein)
heatmap(assay(pe[["protein"]])[sigNames, ])
```


```{r, warning=FALSE, message=FALSE}
for (protName in sigNames)
    {
        pePlot <- pe[protName, , c("peptideNorm", "protein")]
        pePlotDf <- data.frame(longFormat(pePlot))
        pePlotDf$assay <- factor(pePlotDf$assay,
            levels = c("peptideNorm", "protein")
        )
        pePlotDf$genotype <- as.factor(colData(pePlot)[pePlotDf$colname, "genotype"])

        # plotting
        p1 <- ggplot(
            data = pePlotDf,
            aes(x = colname, y = value, group = rowname)
        ) +
            geom_line() +
            geom_point() +
            facet_grid(~assay) +
            theme(axis.text.x = element_text(angle = 70, hjust = 1, vjust = 0.5)) +
            ggtitle(protName)
        print(p1)

        # plotting 2
        p2 <- ggplot(pePlotDf, aes(x = colname, y = value, fill = genotype)) +
            geom_boxplot(outlier.shape = NA) +
            geom_point(
                position = position_jitter(width = .1),
                aes(shape = rowname)
            ) +
            scale_shape_manual(values = 1:nrow(pePlotDf)) +
            labs(title = protName, x = "sample", y = "peptide intensity (log2)") +
            theme(axis.text.x = element_text(angle = 70, hjust = 1, vjust = 0.5)) +
            facet_grid(~assay)
        print(p2)
}
```


# Experimental Design

## Sample size 

$$
 \log_2 \text{FC} = \bar{y}_{p1}-\bar{y}_{p2}
$$

$$
T_g=\frac{\log_2 \text{FC}}{\text{se}_{\log_2 \text{FC}}}
$$

$$
T_g=\frac{\widehat{\text{signal}}}{\widehat{\text{Noise}}}
$$

If we can assume equal variance in both treatment groups:

$$
\text{se}_{\log_2 \text{FC}}=\text{SD}\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}
$$

$\rightarrow$ if number of bio-repeats increases we have a higher power!

- cfr. Study of tamoxifen treated Estrogen Recepter (ER) positive breast cancer patients

## Blocking 

\[\sigma^2= \sigma^2_{bio}+\sigma^2_\text{lab} +\sigma^2_\text{extraction} + \sigma^2_\text{run} + \ldots\]

- Biological: fluctuations in protein level between mice, fluctations in protein level between cells, ...
- Technical: cage effect, lab effect, week effect, plasma extraction, MS-run, ...

## Nature methods: Points of significance - Blocking 

[https://www.nature.com/articles/nmeth.3005.pdf](https://www.nature.com/articles/nmeth.3005.pdf)


## Mouse example 

```{r echo=FALSE, out.width="50%"}
knitr::include_graphics("./figures/mouseTcell_RCB_design.png")
```
Duguet et al. (2017) MCP 16(8):1416-1432. doi: 10.1074/mcp.m116.062745

- All treatments of interest are present within block!
- We can estimate the effect of the treatment within block!
- We can isolate the between block variability from the analysis using linear model:
$$ 
y \sim \text{type} + \text{mouse}
$$
- Not possible with Perseus!

### Assess the impact of blocking in the tutorial session!

- Completely randomized design with only one cell type per mouse (Treg and Tconv)

$$\updownarrow$$

- Randomized complete block design assessing Treg and Tconv on each mouse


</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("pda_quantification_inference_noFrames.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
