<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lieven Clement" />


<title>Statistical Methods for Quantitative MS-based Proteomics: Part I. Preprocessing</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./docs/intro.pdf">1.1 Position</a>
    </li>
    <li>
      <a href="recapGeneralLinearModel.html">1.2 Recap linear model</a>
    </li>
    <li>
      <a href="./docs/martens_proteomics_bioinformatics.pdf">2.1 PDA - identification</a>
    </li>
    <li>
      <a href="pda_quantification_preprocessing.html">2.2 PDA- Quantification - Preprocessing</a>
    </li>
    <li>
      <a href="pda_robustSummarisation_peptideModels.html">2.2 PDA - Wrap-up Preprocessing - Peptide-Level-Models</a>
    </li>
    <li>
      <a href="pda_quantification_inference.html">2.3 PDA - Quantification - Differential Abundance Analysis</a>
    </li>
    <li>
      <a href="pda_blocking_wrapup.html">2.3 PDA - Wrap-up DA - Blocking</a>
    </li>
    <li>
      <a href="technicalDetailsProteomics.html">2.4. PDA - DA Technical Details</a>
    </li>
    <li>
      <a href="./docs/stagewiseTesting.pdf">2.5 PDA - Stage-wise testing</a>
    </li>
    <li>
      <a href="sequencing_intro.html">3.1 SDA - Intro to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">3.2 SDA - Working with count data</a>
    </li>
    <li>
      <a href="sequencing_rnaseqIntro.html">3.3 SDA - Intro to RNA-seq analysis</a>
    </li>
    <li>
      <a href="sequencing_technicalDE.html">3.4 SDA - Technical topics in DE analysis</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multipleRegression_KPNA2.html">1 Intro KPNA2</a>
    </li>
    <li>
      <a href="https://www.compomics.com/bioinformatics-for-proteomics/">2.1 Identification</a>
    </li>
    <li>
      <a href="pda_tutorialPreprocessing.html">2.2 DA - Preprocessing</a>
    </li>
    <li>
      <a href="pda_tutorialDesign.html">2.3 DA - Design</a>
    </li>
    <li>
      <a href="sequencing_preprocessing.sh">3.1 Preprocessing</a>
    </li>
    <li>
      <a href="sequencing_lab_oneGene.html">3.2 GLM for one gene</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Statistical Methods for Quantitative MS-based Proteomics: Part I. Preprocessing</h1>
<h4 class="author">Lieven Clement</h4>
<h4 class="date"><a href="https://statomics.github.io">statOmics</a>, Ghent University</h4>

</div>


<p><a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a></p>
<ul>
<li><a href="https://www.youtube.com/playlist?list=PLZH1hP8_LbJJXQeQ_KYDNoq-AXyFBG6nX">Playlist PDA Preprocessing</a></li>
</ul>
<div id="outline" class="section level1 unnumbered">
<h1>Outline</h1>
<ol style="list-style-type: decimal">
<li><p>Introduction</p></li>
<li><p>Preprocessing</p>
<ul>
<li>Log-transformation</li>
<li>Filtering</li>
<li>Normalization</li>
<li>Summarization</li>
</ul></li>
</ol>
<p>Note, that the R-code is included for learners who are aiming to develop R/markdown scripts to automate their quantitative proteomics data analyses. According to the target audience of the course we either work with a graphical user interface (GUI) in a R/shiny App msqrob2gui (e.g. Proteomics Bioinformatics course of the EBI and the Proteomics Data Analysis course at the Gulbenkian institute) or with R/markdowns scripts (e.g. Bioinformatics Summer School at UCLouvain or the Statistical Genomics Course at Ghent University).</p>
<hr />
</div>
<div id="intro-challenges-in-label-free-quantitative-proteomics" class="section level1">
<h1><span class="header-section-number">1</span> Intro: Challenges in Label-Free Quantitative Proteomics</h1>
<div id="ms-based-workflow" class="section level2">
<h2><span class="header-section-number">1.1</span> MS-based workflow</h2>
<p><img src="figures/ProteomicsWorkflow.png" width="360" /></p>
<ul>
<li><p>Peptide Characteristics</p>
<ul>
<li>Modifications</li>
<li>Ionisation Efficiency: huge variability</li>
<li>Identification
<ul>
<li>Misidentification <span class="math inline">\(\rightarrow\)</span> outliers</li>
<li>MS<span class="math inline">\(^2\)</span> selection on peptide abundance</li>
<li>Context depending missingness</li>
<li>Non-random missingness</li>
</ul></li>
</ul></li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> Unbalanced pepide identifications across samples and messy data</p>
<hr />
</div>
<div id="level-of-quantification" class="section level2">
<h2><span class="header-section-number">1.2</span> Level of quantification</h2>
<ul>
<li>MS-based proteomics returns peptides: pieces of proteins</li>
</ul>
<p><img src="figures/challenges_peptides.png" width="256" /></p>
<ul>
<li>Quantification commonly required on the protein level</li>
</ul>
<p><img src="figures/challenges_proteins.png" width="258" /></p>
<hr />
</div>
<div id="label-free-quantitative-proteomics-data-analysis-workflows" class="section level2">
<h2><span class="header-section-number">1.3</span> Label-free Quantitative Proteomics Data Analysis Workflows</h2>
<p><img src="figures/proteomicsDataAnalysis.png" width="360" /></p>
<hr />
</div>
<div id="cptac-spike-in-study" class="section level2">
<h2><span class="header-section-number">1.4</span> CPTAC Spike-in Study</h2>
<p><img src="figures/cptacLayoutLudger.png" width="50%" /></p>
<ul>
<li><p>Same trypsin-digested yeast proteome background in each sample</p></li>
<li><p>Trypsin-digested Sigma UPS1 standard: 48 different human proteins spiked in at 5 different concentrations (treatment A-E)</p></li>
<li><p>Samples repeatedly run on different instruments in different labs</p></li>
<li><p>After MaxQuant search with match between runs option</p>
<ul>
<li>41% of all proteins are quantified in all samples</li>
<li>6.6% of all peptides are quantified in all samples</li>
</ul></li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> vast amount of missingness</p>
</div>
<div id="maxquant-output" class="section level2">
<h2><span class="header-section-number">1.5</span> Maxquant output</h2>
<p><img src="figures/maxquantOutputDir.png" width="1204" /></p>
<hr />
</div>
</div>
<div id="import-the-data-in-r" class="section level1">
<h1><span class="header-section-number">2</span> Import the data in R</h1>
<div id="data-infrastructure" class="section level2">
<h2><span class="header-section-number">2.1</span> Data infrastructure</h2>
<details>
<summary> Click to see background on data infrastructure used in R to store proteomics data </summary>
<p>
<ul>
<li>We use the <code>QFeatures</code> package that provides the infrastructure to
<ul>
<li>store,<br />
</li>
<li>process,</li>
<li>manipulate and</li>
<li>analyse quantitative data/features from mass spectrometry experiments.</li>
</ul></li>
<li>It is based on the <code>SummarizedExperiment</code> and <code>MultiAssayExperiment</code> classes.</li>
</ul>
<div class="figure">
<img src="figures/SE.png" alt="Conceptual representation of a `SummarizedExperiment` object.  Assays contain information on the measured omics features (rows) for different samples (columns). The `rowData` contains information on the omics features, the `colData` contains information on the samples, i.e. experimental design etc." width="80%" />
<p class="caption">
Conceptual representation of a <code>SummarizedExperiment</code> object. Assays contain information on the measured omics features (rows) for different samples (columns). The <code>rowData</code> contains information on the omics features, the <code>colData</code> contains information on the samples, i.e. experimental design etc.
</p>
</div>
<ul>
<li><p>Assays in a QFeatures object have a hierarchical relation:</p>
<ul>
<li>proteins are composed of peptides,</li>
<li>themselves produced by spectra</li>
<li>relations between assays are tracked and recorded throughout data processing</li>
</ul></li>
</ul>
<div class="figure">
<img src="pda_quantification_preprocessing_noframes_files/figure-html/featuresplot-1.png" alt="Conceptual representation of a `QFeatures` object and the aggregative relation between different assays." width="672" />
<p class="caption">
Conceptual representation of a <code>QFeatures</code> object and the aggregative relation between different assays.
</p>
</div>
</p>
</details>
</div>
<div id="import-data-in-r" class="section level2">
<h2><span class="header-section-number">2.2</span> Import data in R</h2>
<div id="load-libraries" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Load libraries</h3>
<details>
<summary> Click to see code </summary>
<p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">library</span>(limma)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">library</span>(QFeatures)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw">library</span>(msqrob2)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="kw">library</span>(plotly)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="kw">library</span>(ggplot2)</span></code></pre></div>
</p>
</details>
</div>
<div id="read-data" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Read data</h3>
<details>
<summary> Click to see background and code </summary>
<p>
<ol style="list-style-type: decimal">
<li>We use a peptides.txt file from MS-data quantified with maxquant that contains MS1 intensities summarized at the peptide level.</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>peptidesFile &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/statOmics/PDA21/data/quantification/fullCptacDatasSetNotForTutorial/peptides.txt&quot;</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Maxquant stores the intensity data for the different samples in columnns that start with Intensity. We can retreive the column names with the intensity data with the code below:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>ecols &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;Intensity</span><span class="ch">\\</span><span class="st">.&quot;</span>, <span class="kw">names</span>(<span class="kw">read.delim</span>(peptidesFile)))</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Read the data and store it in QFeatures object</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>pe &lt;-<span class="st"> </span><span class="kw">readQFeatures</span>(</span>
<span id="cb4-2"><a href="#cb4-2"></a>  <span class="dt">table =</span> peptidesFile,</span>
<span id="cb4-3"><a href="#cb4-3"></a>  <span class="dt">fnames =</span> <span class="dv">1</span>,</span>
<span id="cb4-4"><a href="#cb4-4"></a>  <span class="dt">ecol =</span> ecols,</span>
<span id="cb4-5"><a href="#cb4-5"></a>  <span class="dt">name =</span> <span class="st">&quot;peptideRaw&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
</p>
</details>
</div>
<div id="explore-object" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Explore object</h3>
<details>
<summary> Click to see background and code </summary>
<p>
<ul>
<li>The rowData contains information on the features (peptides) in the assay. E.g. Sequence, protein, …</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">rowData</span>(pe[[<span class="st">&quot;peptideRaw&quot;</span>]])</span></code></pre></div>
<pre><code>## DataFrame with 11466 rows and 143 columns
##                          Sequence N.term.cleavage.window C.term.cleavage.window
##                       &lt;character&gt;            &lt;character&gt;            &lt;character&gt;
## AAAAGAGGAGDSGDAVTK  AAAAGAGGAG...          EHQHDEQKAA...          DSGDAVTKIG...
## AAAALAGGK               AAAALAGGK          QQLSKAAKAA...          AAALAGGKKS...
## AAAALAGGKK             AAAALAGGKK          QQLSKAAKAA...          AALAGGKKSK...
## AAADALSDLEIK        AAADALSDLE...          MPKETPSKAA...          ALSDLEIKDS...
## AAADALSDLEIKDSK     AAADALSDLE...          MPKETPSKAA...          DLEIKDSKSN...
## ...                           ...                    ...                    ...
## YYSIYDLGNNAVGLAK    YYSIYDLGNN...          VGDAFLRKYY...          NNAVGLAKAI...
## YYTFNGPNYNENETIR    YYTFNGPNYN...          FKDGSYPKYY...          YNENETIRHI...
## YYTITEVATR             YYTITEVATR          QEWDINERYY...          TITEVATRAK...
## YYTVFDRDNNR         YYTVFDRDNN...          LGDVFIGRYY...          VFDRDNNRVG...
## YYTVFDRDNNRVGFAEAAR YYTVFDRDNN...          LGDVFIGRYY...          VGFAEAARL_...
##                     Amino.acid.before First.amino.acid Second.amino.acid
##                           &lt;character&gt;      &lt;character&gt;       &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                  K                A                 A
## AAAALAGGK                           K                A                 A
## AAAALAGGKK                          K                A                 A
## AAADALSDLEIK                        K                A                 A
## AAADALSDLEIKDSK                     K                A                 A
## ...                               ...              ...               ...
## YYSIYDLGNNAVGLAK                    K                Y                 Y
## YYTFNGPNYNENETIR                    K                Y                 Y
## YYTITEVATR                          R                Y                 Y
## YYTVFDRDNNR                         R                Y                 Y
## YYTVFDRDNNRVGFAEAAR                 R                Y                 Y
##                     Second.last.amino.acid Last.amino.acid Amino.acid.after
##                                &lt;character&gt;     &lt;character&gt;      &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                       T               K                I
## AAAALAGGK                                G               K                K
## AAAALAGGKK                               K               K                S
## AAADALSDLEIK                             I               K                D
## AAADALSDLEIKDSK                          S               K                S
## ...                                    ...             ...              ...
## YYSIYDLGNNAVGLAK                         A               K                A
## YYTFNGPNYNENETIR                         I               R                H
## YYTITEVATR                               T               R                A
## YYTVFDRDNNR                              N               R                V
## YYTVFDRDNNRVGFAEAAR                      A               R                L
##                       A.Count   R.Count   N.Count   D.Count   C.Count   Q.Count
##                     &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK          7         0         0         2         0         0
## AAAALAGGK                   5         0         0         0         0         0
## AAAALAGGKK                  5         0         0         0         0         0
## AAADALSDLEIK                4         0         0         2         0         0
## AAADALSDLEIKDSK             4         0         0         3         0         0
## ...                       ...       ...       ...       ...       ...       ...
## YYSIYDLGNNAVGLAK            2         0         2         1         0         0
## YYTFNGPNYNENETIR            0         1         4         0         0         0
## YYTITEVATR                  1         1         0         0         0         0
## YYTVFDRDNNR                 0         2         2         2         0         0
## YYTVFDRDNNRVGFAEAAR         3         3         2         2         0         0
##                       E.Count   G.Count   H.Count   I.Count   L.Count   K.Count
##                     &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK          0         5         0         0         0         1
## AAAALAGGK                   0         2         0         0         1         1
## AAAALAGGKK                  0         2         0         0         1         2
## AAADALSDLEIK                1         0         0         1         2         1
## AAADALSDLEIKDSK             1         0         0         1         2         2
## ...                       ...       ...       ...       ...       ...       ...
## YYSIYDLGNNAVGLAK            0         2         0         1         2         1
## YYTFNGPNYNENETIR            2         1         0         1         0         0
## YYTITEVATR                  1         0         0         1         0         0
## YYTVFDRDNNR                 0         0         0         0         0         0
## YYTVFDRDNNRVGFAEAAR         1         1         0         0         0         0
##                       M.Count   F.Count   P.Count   S.Count   T.Count   W.Count
##                     &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK          0         0         0         1         1         0
## AAAALAGGK                   0         0         0         0         0         0
## AAAALAGGKK                  0         0         0         0         0         0
## AAADALSDLEIK                0         0         0         1         0         0
## AAADALSDLEIKDSK             0         0         0         2         0         0
## ...                       ...       ...       ...       ...       ...       ...
## YYSIYDLGNNAVGLAK            0         0         0         1         0         0
## YYTFNGPNYNENETIR            0         1         1         0         2         0
## YYTITEVATR                  0         0         0         0         3         0
## YYTVFDRDNNR                 0         1         0         0         1         0
## YYTVFDRDNNRVGFAEAAR         0         2         0         0         1         0
##                       Y.Count   V.Count   U.Count    Length Missed.cleavages
##                     &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;        &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK          0         1         0        18                0
## AAAALAGGK                   0         0         0         9                0
## AAAALAGGKK                  0         0         0        10                1
## AAADALSDLEIK                0         0         0        12                0
## AAADALSDLEIKDSK             0         0         0        15                1
## ...                       ...       ...       ...       ...              ...
## YYSIYDLGNNAVGLAK            3         1         0        16                0
## YYTFNGPNYNENETIR            3         0         0        16                0
## YYTITEVATR                  2         1         0        10                0
## YYTVFDRDNNR                 2         1         0        11                1
## YYTVFDRDNNRVGFAEAAR         2         2         0        19                2
##                          Mass      Proteins Leading.razor.protein
##                     &lt;numeric&gt;   &lt;character&gt;           &lt;character&gt;
## AAAAGAGGAGDSGDAVTK   1445.675 sp|P38915|...         sp|P38915|...
## AAAALAGGK             728.418 sp|Q3E792|...         sp|Q3E792|...
## AAAALAGGKK            856.513 sp|Q3E792|...         sp|Q3E792|...
## AAADALSDLEIK         1215.635 sp|P09938|...         sp|P09938|...
## AAADALSDLEIKDSK      1545.789 sp|P09938|...         sp|P09938|...
## ...                       ...           ...                   ...
## YYSIYDLGNNAVGLAK      1759.88 sp|P07267|...         sp|P07267|...
## YYTFNGPNYNENETIR      1993.88 sp|Q00955|...         sp|Q00955|...
## YYTITEVATR            1215.61 sp|P38891|...         sp|P38891|...
## YYTVFDRDNNR           1461.66 P07339ups|...         P07339ups|...
## YYTVFDRDNNRVGFAEAAR   2263.08 P07339ups|...         P07339ups|...
##                     Start.position End.position Unique..Groups.
##                          &lt;integer&gt;    &lt;integer&gt;     &lt;character&gt;
## AAAAGAGGAGDSGDAVTK              97          114             yes
## AAAALAGGK                       13           21             yes
## AAAALAGGKK                      13           22             yes
## AAADALSDLEIK                     9           20             yes
## AAADALSDLEIKDSK                  9           23             yes
## ...                            ...          ...             ...
## YYSIYDLGNNAVGLAK               388          403             yes
## YYTFNGPNYNENETIR              1275         1290             yes
## YYTITEVATR                     311          320             yes
## YYTVFDRDNNR                    225          235             yes
## YYTVFDRDNNRVGFAEAAR            225          243             yes
##                     Unique..Proteins.     Charges        PEP     Score
##                           &lt;character&gt; &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt;
## AAAAGAGGAGDSGDAVTK                yes           2 1.1843e-05    82.942
## AAAALAGGK                          no           2 7.4562e-06   134.810
## AAAALAGGKK                         no           2 3.3094e-09   143.730
## AAADALSDLEIK                      yes           2 9.1593e-23   182.230
## AAADALSDLEIKDSK                   yes           3 1.5319e-04    73.927
## ...                               ...         ...        ...       ...
## YYSIYDLGNNAVGLAK                  yes           2 7.7415e-37   174.240
## YYTFNGPNYNENETIR                  yes           2 4.2208e-21   147.750
## YYTITEVATR                        yes           2 1.3566e-04   109.160
## YYTVFDRDNNR                       yes           2 6.1425e-04   110.930
## YYTVFDRDNNRVGFAEAAR               yes           3 8.9859e-04    59.728
##                     Identification.type.6A_1 Identification.type.6A_2
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...                 By MS/MS
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                By matchin...            By matchin...
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                          By MS/MS            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6A_3 Identification.type.6A_4
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...                 By MS/MS
## AAAALAGGK                      By matchin...                 By MS/MS
## AAAALAGGKK                     By matchin...                 By MS/MS
## AAADALSDLEIK                   By matchin...                 By MS/MS
## AAADALSDLEIKDSK                By matchin...                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...                 By MS/MS
## YYTFNGPNYNENETIR               By matchin...                 By MS/MS
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6A_5 Identification.type.6A_6
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK                    By MS/MS                 By MS/MS
## YYTFNGPNYNENETIR                    By MS/MS                 By MS/MS
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6A_7 Identification.type.6A_8
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                  By MS/MS                 By MS/MS
## AAAALAGGK                           By MS/MS                 By MS/MS
## AAAALAGGKK                          By MS/MS                 By MS/MS
## AAADALSDLEIK                        By MS/MS            By matchin...
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                          By MS/MS            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6A_9 Identification.type.6B_1
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                  By MS/MS            By matchin...
## AAAALAGGK                           By MS/MS                 By MS/MS
## AAAALAGGKK                          By MS/MS            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS            By matchin...
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                     By matchin...                 By MS/MS
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6B_2 Identification.type.6B_3
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                          By MS/MS                 By MS/MS
## AAADALSDLEIK                        By MS/MS            By matchin...
## AAADALSDLEIKDSK                By matchin...            By matchin...
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6B_4 Identification.type.6B_5
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK                    By MS/MS            By matchin...
## YYTFNGPNYNENETIR                    By MS/MS                 By MS/MS
## YYTITEVATR                          By MS/MS                 By MS/MS
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6B_6 Identification.type.6B_7
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...                 By MS/MS
## AAAALAGGKK                     By matchin...                 By MS/MS
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR                    By MS/MS            By matchin...
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6B_8 Identification.type.6B_9
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                  By MS/MS                 By MS/MS
## AAAALAGGK                           By MS/MS                 By MS/MS
## AAAALAGGKK                          By MS/MS                 By MS/MS
## AAADALSDLEIK                   By matchin...            By matchin...
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                          By MS/MS            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6C_1 Identification.type.6C_2
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...                 By MS/MS
## AAAALAGGKK                     By matchin...                 By MS/MS
## AAADALSDLEIK                        By MS/MS            By matchin...
## AAADALSDLEIKDSK                By matchin...            By matchin...
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6C_3 Identification.type.6C_4
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...                 By MS/MS
## AAAALAGGKK                     By matchin...                 By MS/MS
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                By matchin...                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...                 By MS/MS
## YYTFNGPNYNENETIR               By matchin...                 By MS/MS
## YYTITEVATR                          By MS/MS            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6C_5 Identification.type.6C_6
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                  By MS/MS            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK                    By MS/MS                 By MS/MS
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6C_7 Identification.type.6C_8
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                  By MS/MS            By matchin...
## AAAALAGGK                           By MS/MS                 By MS/MS
## AAAALAGGKK                          By MS/MS                 By MS/MS
## AAADALSDLEIK                   By matchin...                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                     By matchin...                 By MS/MS
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6C_9 Identification.type.6D_1
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                           By MS/MS            By matchin...
## AAAALAGGKK                          By MS/MS            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                          By MS/MS            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6D_2 Identification.type.6D_3
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                   By matchin...            By matchin...
## AAADALSDLEIKDSK                     By MS/MS            By matchin...
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                          By MS/MS                 By MS/MS
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6D_4 Identification.type.6D_5
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                          By MS/MS            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK                    By MS/MS                 By MS/MS
## YYTFNGPNYNENETIR                    By MS/MS                 By MS/MS
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6D_6 Identification.type.6D_7
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK                  By MS/MS            By matchin...
## AAAALAGGK                      By matchin...                 By MS/MS
## AAAALAGGKK                     By matchin...                 By MS/MS
## AAADALSDLEIK                        By MS/MS            By matchin...
## AAADALSDLEIKDSK                By matchin...                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK                    By MS/MS            By matchin...
## YYTFNGPNYNENETIR                    By MS/MS            By matchin...
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...                 By MS/MS
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6D_8 Identification.type.6D_9
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                           By MS/MS                 By MS/MS
## AAAALAGGKK                          By MS/MS                 By MS/MS
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                          By MS/MS            By matchin...
## YYTVFDRDNNR                         By MS/MS            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6E_1 Identification.type.6E_2
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6E_3 Identification.type.6E_4
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...                 By MS/MS
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                   By matchin...                 By MS/MS
## AAADALSDLEIKDSK                     By MS/MS            By matchin...
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...                 By MS/MS
## YYTFNGPNYNENETIR               By matchin...                 By MS/MS
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                    By matchin...                 By MS/MS
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6E_5 Identification.type.6E_6
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                      By matchin...            By matchin...
## AAAALAGGKK                     By matchin...            By matchin...
## AAADALSDLEIK                        By MS/MS            By matchin...
## AAADALSDLEIKDSK                     By MS/MS                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK                    By MS/MS                 By MS/MS
## YYTFNGPNYNENETIR                    By MS/MS                 By MS/MS
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                         By MS/MS            By matchin...
## YYTVFDRDNNRVGFAEAAR            By matchin...                 By MS/MS
##                     Identification.type.6E_7 Identification.type.6E_8
##                                  &lt;character&gt;              &lt;character&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...            By matchin...
## AAAALAGGK                           By MS/MS                 By MS/MS
## AAAALAGGKK                          By MS/MS                 By MS/MS
## AAADALSDLEIK                        By MS/MS                 By MS/MS
## AAADALSDLEIKDSK                By matchin...                 By MS/MS
## ...                                      ...                      ...
## YYSIYDLGNNAVGLAK               By matchin...            By matchin...
## YYTFNGPNYNENETIR               By matchin...            By matchin...
## YYTITEVATR                     By matchin...            By matchin...
## YYTVFDRDNNR                         By MS/MS                 By MS/MS
## YYTVFDRDNNRVGFAEAAR            By matchin...            By matchin...
##                     Identification.type.6E_9 Experiment.6A_1 Experiment.6A_2
##                                  &lt;character&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK             By matchin...              NA               1
## AAAALAGGK                           By MS/MS              NA               1
## AAAALAGGKK                          By MS/MS              NA               1
## AAADALSDLEIK                        By MS/MS               1               1
## AAADALSDLEIKDSK                     By MS/MS               1               1
## ...                                      ...             ...             ...
## YYSIYDLGNNAVGLAK               By matchin...              NA              NA
## YYTFNGPNYNENETIR                    By MS/MS              NA              NA
## YYTITEVATR                     By matchin...               1              NA
## YYTVFDRDNNR                         By MS/MS              NA              NA
## YYTVFDRDNNRVGFAEAAR            By matchin...              NA              NA
##                     Experiment.6A_3 Experiment.6A_4 Experiment.6A_5
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK               NA               1               1
## AAAALAGGK                         2               1               1
## AAAALAGGKK                       NA               1              NA
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                  NA               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA               1               1
## YYTFNGPNYNENETIR                 NA               1               1
## YYTITEVATR                        1              NA              NA
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6A_6 Experiment.6A_7 Experiment.6A_8
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1               1               1
## AAAALAGGK                         1               2               1
## AAAALAGGKK                        1               1               1
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                  1              NA              NA
## YYTFNGPNYNENETIR                  1               1              NA
## YYTITEVATR                        1               1              NA
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6A_9 Experiment.6B_1 Experiment.6B_2
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1              NA              NA
## AAAALAGGK                         1               1               1
## AAAALAGGKK                        1              NA               1
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1              NA               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA              NA              NA
## YYTFNGPNYNENETIR                  1              NA              NA
## YYTITEVATR                       NA               1               1
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6B_3 Experiment.6B_4 Experiment.6B_5
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK               NA              NA               1
## AAAALAGGK                         1               2               1
## AAAALAGGKK                        1               1              NA
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                  NA               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA               1               1
## YYTFNGPNYNENETIR                 NA               1               1
## YYTITEVATR                        1               1               1
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6B_6 Experiment.6B_7 Experiment.6B_8
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1              NA               1
## AAAALAGGK                        NA               2               1
## AAAALAGGKK                       NA               1               1
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                  1              NA              NA
## YYTFNGPNYNENETIR                  1               1              NA
## YYTITEVATR                        1              NA               1
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6B_9 Experiment.6C_1 Experiment.6C_2
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1              NA              NA
## AAAALAGGK                         2              NA               1
## AAAALAGGKK                        1              NA               1
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA              NA              NA
## YYTFNGPNYNENETIR                 NA              NA              NA
## YYTITEVATR                       NA               1               1
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6C_3 Experiment.6C_4 Experiment.6C_5
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK               NA               1               1
## AAAALAGGK                         2               2              NA
## AAAALAGGKK                       NA               1              NA
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA               1               1
## YYTFNGPNYNENETIR                 NA               1               1
## YYTITEVATR                        1               1              NA
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6C_6 Experiment.6C_7 Experiment.6C_8
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1               1               1
## AAAALAGGK                        NA               2               1
## AAAALAGGKK                       NA               1               1
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                  1              NA              NA
## YYTFNGPNYNENETIR                  1               1               1
## YYTITEVATR                        1              NA               1
## YYTVFDRDNNR                       1              NA               1
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6C_9 Experiment.6D_1 Experiment.6D_2
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1              NA              NA
## AAAALAGGK                         1              NA               1
## AAAALAGGKK                        1              NA              NA
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA              NA              NA
## YYTFNGPNYNENETIR                  1              NA              NA
## YYTITEVATR                        1              NA               1
## YYTVFDRDNNR                      NA              NA              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6D_3 Experiment.6D_4 Experiment.6D_5
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK               NA               1               1
## AAAALAGGK                         1               1               1
## AAAALAGGKK                       NA               1              NA
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA               1               1
## YYTFNGPNYNENETIR                 NA               1               1
## YYTITEVATR                        1               1               1
## YYTVFDRDNNR                      NA               1               1
## YYTVFDRDNNRVGFAEAAR              NA               1              NA
##                     Experiment.6D_6 Experiment.6D_7 Experiment.6D_8
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1               1              NA
## AAAALAGGK                        NA               2               1
## AAAALAGGKK                       NA               1               1
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                  1               1              NA
## YYTFNGPNYNENETIR                  1               1               1
## YYTITEVATR                        1              NA               1
## YYTVFDRDNNR                       1               1               1
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6D_9 Experiment.6E_1 Experiment.6E_2
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK               NA              NA               1
## AAAALAGGK                         2              NA               1
## AAAALAGGKK                        1              NA              NA
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                 NA              NA              NA
## YYTFNGPNYNENETIR                  1              NA              NA
## YYTITEVATR                       NA              NA               1
## YYTVFDRDNNR                       1               1              NA
## YYTVFDRDNNRVGFAEAAR              NA              NA              NA
##                     Experiment.6E_3 Experiment.6E_4 Experiment.6E_5
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK               NA              NA               1
## AAAALAGGK                         2               2               1
## AAAALAGGKK                       NA               1              NA
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1               1               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                  1               1               1
## YYTFNGPNYNENETIR                 NA               1               1
## YYTITEVATR                        1               1               1
## YYTVFDRDNNR                       1               1               1
## YYTVFDRDNNRVGFAEAAR              NA               1               1
##                     Experiment.6E_6 Experiment.6E_7 Experiment.6E_8
##                           &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK                1              NA              NA
## AAAALAGGK                        NA               2               2
## AAAALAGGKK                       NA               1               1
## AAADALSDLEIK                      1               1               1
## AAADALSDLEIKDSK                   1              NA               1
## ...                             ...             ...             ...
## YYSIYDLGNNAVGLAK                  1              NA              NA
## YYTFNGPNYNENETIR                  1               1               1
## YYTITEVATR                       NA              NA              NA
## YYTVFDRDNNR                       1               1               1
## YYTVFDRDNNRVGFAEAAR               1               1               1
##                     Experiment.6E_9 Intensity     Reverse Potential.contaminant
##                           &lt;integer&gt; &lt;numeric&gt; &lt;character&gt;           &lt;character&gt;
## AAAAGAGGAGDSGDAVTK               NA   1190800                                  
## AAAALAGGK                         1 280990000                                  
## AAAALAGGKK                        1  33360000                                  
## AAADALSDLEIK                      1  54622000                                  
## AAADALSDLEIKDSK                   1  18910000                                  
## ...                             ...       ...         ...                   ...
## YYSIYDLGNNAVGLAK                 NA   2145900                                  
## YYTFNGPNYNENETIR                  1   5608800                                  
## YYTITEVATR                       NA  13034000                                  
## YYTVFDRDNNR                       1   8702500                                  
## YYTVFDRDNNRVGFAEAAR               1   2391100                                  
##                            id Protein.group.IDs Mod..peptide.IDs  Evidence.IDs
##                     &lt;integer&gt;       &lt;character&gt;      &lt;character&gt;   &lt;character&gt;
## AAAAGAGGAGDSGDAVTK          0               859                0 0;1;2;3;4;...
## AAAALAGGK                   1               230                1 24;25;26;2...
## AAAALAGGKK                  2               230                2 74;75;76;7...
## AAADALSDLEIK                3               229                3 99;100;101...
## AAADALSDLEIKDSK             4               229                4 144;145;14...
## ...                       ...               ...              ...           ...
## YYSIYDLGNNAVGLAK        11461               196            12240 331367;331...
## YYTFNGPNYNENETIR        11462              1254            12241 331384;331...
## YYTITEVATR              11463               854            12242 331411;331...
## YYTVFDRDNNR             11464                34            12243 331439;331...
## YYTVFDRDNNRVGFAEAAR     11465                34            12244 331455;331...
##                         MS.MS.IDs Best.MS.MS Oxidation..M..site.IDs MS.MS.Count
##                       &lt;character&gt;  &lt;integer&gt;            &lt;character&gt;   &lt;integer&gt;
## AAAAGAGGAGDSGDAVTK  0;1;2;3;4;...          0                                 10
## AAAALAGGK           10;11;12;1...         21                                 18
## AAAALAGGKK          30;31;32;3...         31                                 21
## AAADALSDLEIK        51;52;53;5...         72                                 29
## AAADALSDLEIKDSK     85;86;87;8...         94                                 32
## ...                           ...        ...                    ...         ...
## YYSIYDLGNNAVGLAK    169138;169...     169147                                 13
## YYTFNGPNYNENETIR    169151;169...     169159                                 14
## YYTITEVATR          169165;169...     169173                                 12
## YYTVFDRDNNR         169177;169...     169180                                  7
## YYTVFDRDNNRVGFAEAAR        169184     169184                                  1</code></pre>
<ul>
<li>The colData contains information on the samples</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">colData</span>(pe)</span></code></pre></div>
<pre><code>## DataFrame with 45 rows and 0 columns</code></pre>
<ul>
<li>No information is stored yet on the design.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>pe <span class="op">%&gt;%</span><span class="st"> </span>colnames</span></code></pre></div>
<pre><code>## CharacterList of length 1
## [[&quot;peptideRaw&quot;]] Intensity.6A_1 Intensity.6A_2 ... Intensity.6E_9</code></pre>
<ul>
<li><p>Note, that the sample names include the spike-in condition.</p></li>
<li><p>They also end on a number.</p>
<ul>
<li>1-3 is from lab 1,</li>
<li>4-6 from lab 2 and</li>
<li>7-9 from lab 3.</li>
</ul></li>
<li><p>We update the colData with information on the design</p></li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">colData</span>(pe)<span class="op">$</span>lab &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">rep</span>(<span class="kw">paste0</span>(<span class="st">&quot;lab&quot;</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>),<span class="dt">each=</span><span class="dv">3</span>),<span class="dv">5</span>) <span class="op">%&gt;%</span><span class="st"> </span>as.factor</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="kw">colData</span>(pe)<span class="op">$</span>condition &lt;-<span class="st"> </span>pe[[<span class="st">&quot;peptideRaw&quot;</span>]] <span class="op">%&gt;%</span><span class="st"> </span>colnames <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">substr</span>(<span class="dv">12</span>,<span class="dv">12</span>) <span class="op">%&gt;%</span><span class="st"> </span>as.factor</span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="kw">colData</span>(pe)<span class="op">$</span>spikeConcentration &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="dt">A =</span> <span class="fl">0.25</span>, <span class="dt">B =</span> <span class="fl">0.74</span>, <span class="dt">C =</span> <span class="fl">2.22</span>, <span class="dt">D =</span> <span class="fl">6.67</span>, <span class="dt">E =</span> <span class="dv">20</span>),<span class="dt">each =</span> <span class="dv">9</span>)</span></code></pre></div>
<ul>
<li>We explore the colData again</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">colData</span>(pe)</span></code></pre></div>
<pre><code>## DataFrame with 45 rows and 3 columns
##                     lab condition spikeConcentration
##                &lt;factor&gt;  &lt;factor&gt;          &lt;numeric&gt;
## Intensity.6A_1     lab1         A               0.25
## Intensity.6A_2     lab1         A               0.25
## Intensity.6A_3     lab1         A               0.25
## Intensity.6A_4     lab2         A               0.25
## Intensity.6A_5     lab2         A               0.25
## ...                 ...       ...                ...
## Intensity.6E_5     lab2         E                 20
## Intensity.6E_6     lab2         E                 20
## Intensity.6E_7     lab3         E                 20
## Intensity.6E_8     lab3         E                 20
## Intensity.6E_9     lab3         E                 20</code></pre>
</p>
</details>
<hr />
</div>
</div>
</div>
<div id="preprocessing" class="section level1">
<h1><span class="header-section-number">3</span> Preprocessing</h1>
<div id="log-transformation" class="section level2">
<h2><span class="header-section-number">3.1</span> Log-transformation</h2>
<div id="explore-the-data-with-plots" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Explore the data with plots</h3>
<p>Peptide AALEELVK from spiked-in UPS protein P12081. We only show data from lab1.</p>
<details>
<summary> Click to see code to make plot </summary>
<p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>subset &lt;-<span class="st"> </span>pe[<span class="st">&quot;AALEELVK&quot;</span>,<span class="kw">colData</span>(pe)<span class="op">$</span>lab<span class="op">==</span><span class="st">&quot;lab1&quot;</span>]</span>
<span id="cb14-2"><a href="#cb14-2"></a>plotWhyLog &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">concentration =</span> <span class="kw">colData</span>(subset)<span class="op">$</span>spikeConcentration,</span>
<span id="cb14-3"><a href="#cb14-3"></a>           <span class="dt">y =</span> <span class="kw">assay</span>(subset[[<span class="st">&quot;peptideRaw&quot;</span>]]) <span class="op">%&gt;%</span><span class="st"> </span>c</span>
<span id="cb14-4"><a href="#cb14-4"></a>           ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(concentration, y)) <span class="op">+</span></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;concentration (fmol/l)&quot;</span>) <span class="op">+</span></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;peptide AALEELVK in lab1&quot;</span>)</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>plotWhyLog</span></code></pre></div>
<p><img src="pda_quantification_preprocessing_noframes_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<ul>
<li>Variance increases with the mean <span class="math inline">\(\rightarrow\)</span> Multiplicative error structure</li>
</ul>
<details>
<summary> Click to see code to make plot </summary>
<p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>plotLog &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">concentration =</span> <span class="kw">colData</span>(subset)<span class="op">$</span>spikeConcentration,</span>
<span id="cb16-2"><a href="#cb16-2"></a>           <span class="dt">y =</span> <span class="kw">assay</span>(subset[[<span class="st">&quot;peptideRaw&quot;</span>]]) <span class="op">%&gt;%</span><span class="st"> </span>c</span>
<span id="cb16-3"><a href="#cb16-3"></a>           ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(concentration, y)) <span class="op">+</span></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">trans=</span><span class="st">&#39;log2&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb16-7"><a href="#cb16-7"></a><span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="dt">trans=</span><span class="st">&#39;log2&#39;</span>) <span class="op">+</span></span>
<span id="cb16-8"><a href="#cb16-8"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;concentration (fmol/l)&quot;</span>) <span class="op">+</span></span>
<span id="cb16-9"><a href="#cb16-9"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;peptide AALEELVK in lab1 with axes on log scale&quot;</span>)</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>plotLog</span></code></pre></div>
<p><img src="pda_quantification_preprocessing_noframes_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<ul>
<li>Data seems to be homoscedastic on log-scale <span class="math inline">\(\rightarrow\)</span> log transformation of the intensity data</li>
<li>In quantitative proteomics analysis on <span class="math inline">\(\log_2\)</span></li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> Differences on a <span class="math inline">\(\log_2\)</span> scale: <span class="math inline">\(\log_2\)</span> fold changes</p>
<p><span class="math display">\[
\log_2 B - \log_2 A = \log_2 \frac{B}{A} = \log FC_\text{B - A}
\]</span> <span class="math display">\[ 
\begin{array} {l}
log_2 FC = 1 \rightarrow FC = 2^1 =2\\
log_2 FC = 2 \rightarrow FC = 2^2 = 4\\
\end{array}
\]</span></p>
</div>
<div id="log-transformation-of-the-data" class="section level3">
<h3><span class="header-section-number">3.1.2</span> log-transformation of the data</h3>
<details>
<summary> Click to see code to log-transfrom the data </summary>
<p>
<ul>
<li>We calculate how many non zero intensities we have for each peptide and this can be useful for filtering.</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">rowData</span>(pe[[<span class="st">&quot;peptideRaw&quot;</span>]])<span class="op">$</span>nNonZero &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assay</span>(pe[[<span class="st">&quot;peptideRaw&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</span></code></pre></div>
<ul>
<li>Peptides with zero intensities are missing peptides and should be represent with a <code>NA</code> value rather than <code>0</code>.</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>pe &lt;-<span class="st"> </span><span class="kw">zeroIsNA</span>(pe, <span class="st">&quot;peptideRaw&quot;</span>) <span class="co"># convert 0 to NA</span></span></code></pre></div>
<ul>
<li>Logtransform data with base 2</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>pe &lt;-<span class="st"> </span><span class="kw">logTransform</span>(pe, <span class="dt">base =</span> <span class="dv">2</span>, <span class="dt">i =</span> <span class="st">&quot;peptideRaw&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;peptideLog&quot;</span>)</span></code></pre></div>
</p>
</details>
<hr />
</div>
</div>
<div id="filtering" class="section level2">
<h2><span class="header-section-number">3.2</span> Filtering</h2>
<ul>
<li>Reverse sequences</li>
<li>Only identified by modification site (only modified peptides detected)</li>
<li>Razor peptides: non-unique peptides assigned to the protein group with the most other peptides</li>
<li>Contaminants</li>
<li>Peptides few identifications</li>
<li>Proteins that are only identified with one or a few peptides</li>
</ul>
<p>Filtering does not induce bias if the criterion is independent from the downstream data analysis!</p>
<details>
<summary> Click to see code to filter the data </summary>
<p>
<ol style="list-style-type: decimal">
<li>Handling overlapping protein groups</li>
</ol>
<p>In our approach a peptide can map to multiple proteins, as long as there is none of these proteins present in a smaller subgroup.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe, <span class="op">~</span><span class="st"> </span>Proteins <span class="op">%in%</span><span class="st"> </span><span class="kw">smallestUniqueGroups</span>(<span class="kw">rowData</span>(pe[[<span class="st">&quot;peptideLog&quot;</span>]])<span class="op">$</span>Proteins))</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Remove reverse sequences (decoys) and contaminants</li>
</ol>
<p>We now remove the contaminants, peptides that map to decoy sequences, and proteins which were only identified by peptides with modifications.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe,<span class="op">~</span>Reverse <span class="op">!=</span><span class="st"> &quot;+&quot;</span>)</span>
<span id="cb22-2"><a href="#cb22-2"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe,<span class="op">~</span><span class="st"> </span>Potential.contaminant <span class="op">!=</span><span class="st"> &quot;+&quot;</span>)</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Drop peptides that were only identified in one sample</li>
</ol>
<p>We keep peptides that were observed at last twice.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>pe &lt;-<span class="st"> </span><span class="kw">filterFeatures</span>(pe,<span class="op">~</span><span class="st"> </span>nNonZero <span class="op">&gt;=</span><span class="dv">2</span>)</span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="kw">nrow</span>(pe[[<span class="st">&quot;peptideLog&quot;</span>]])</span></code></pre></div>
<pre><code>## [1] 10478</code></pre>
We keep 10478 peptides upon filtering.
</p>
</details>
<hr />
</div>
<div id="normalization" class="section level2">
<h2><span class="header-section-number">3.3</span> Normalization</h2>
<details>
<summary> Click to see code to make plot </summary>
<p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>densityConditionD &lt;-<span class="st"> </span>pe[[<span class="st">&quot;peptideLog&quot;</span>]][,<span class="kw">colData</span>(pe)<span class="op">$</span>condition<span class="op">==</span><span class="st">&quot;D&quot;</span>] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">lab =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;lab&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>intensity,<span class="dt">group=</span>sample,<span class="dt">color=</span>lab)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="st">    </span><span class="kw">geom_density</span>() <span class="op">+</span></span>
<span id="cb25-8"><a href="#cb25-8"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;condition D&quot;</span>)</span>
<span id="cb25-9"><a href="#cb25-9"></a></span>
<span id="cb25-10"><a href="#cb25-10"></a>densityLab2 &lt;-<span class="st"> </span>pe[[<span class="st">&quot;peptideLog&quot;</span>]][,<span class="kw">colData</span>(pe)<span class="op">$</span>lab<span class="op">==</span><span class="st">&quot;lab2&quot;</span>] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb25-11"><a href="#cb25-11"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb25-12"><a href="#cb25-12"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb25-13"><a href="#cb25-13"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb25-14"><a href="#cb25-14"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">condition =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;condition&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb25-15"><a href="#cb25-15"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>intensity,<span class="dt">group=</span>sample,<span class="dt">color=</span>condition)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb25-16"><a href="#cb25-16"></a><span class="st">    </span><span class="kw">geom_density</span>() <span class="op">+</span></span>
<span id="cb25-17"><a href="#cb25-17"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;lab2&quot;</span>)</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>densityConditionD</span></code></pre></div>
<pre><code>## Warning: Removed 39179 rows containing non-finite values (stat_density).</code></pre>
<p><img src="pda_quantification_preprocessing_noframes_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>densityLab2</span></code></pre></div>
<pre><code>## Warning: Removed 44480 rows containing non-finite values (stat_density).</code></pre>
<p><img src="pda_quantification_preprocessing_noframes_files/figure-html/unnamed-chunk-29-1.png" width="672" /> - Even in very clean synthetic dataset (same background, only 48 UPS proteins can be different) the marginal peptide intensity distribution across samples can be quite distinct</p>
<ul>
<li>Considerable effects between and within labs for replicate samples</li>
<li>Considerable effects between samples with different spike-in concentration</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> Normalization is needed</p>
<hr />
<div id="mean-or-median" class="section level3">
<h3><span class="header-section-number">3.3.1</span> Mean or median?</h3>
<ul>
<li>Miller and Fishkin (1997) reported that over a period of 30 years males would like to have on average 64.3 partners and females 2.8.</li>
</ul>
<details>
<summary> </summary>
<p>
<ul>
<li>Miller and Fishkin (1997) reported that the median number of partners someone would like to have over a period of 30 years males is 1 for both males and females.</li>
</ul>
</p>
</details>
<details>
<summary> </summary>
<p>
<p>Mean is very sensitive to outliers!</p>
<img src="figures/partners.png" width="720" />
</p>
</details>
<hr />
</div>
<div id="normalization-of-the-data-by-median-centering" class="section level3">
<h3><span class="header-section-number">3.3.2</span> Normalization of the data by median centering</h3>
<p><span class="math display">\[y_{ip}^\text{norm} = y_{ip} - \hat\mu_i\]</span> with <span class="math inline">\(\hat\mu_i\)</span> the median intensity over all observed peptides in sample <span class="math inline">\(i\)</span>.</p>
<details>
<summary> Click to see R-code to normalize the data </summary>
<p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>pe &lt;-<span class="st"> </span><span class="kw">normalize</span>(pe, </span>
<span id="cb30-2"><a href="#cb30-2"></a>                <span class="dt">i =</span> <span class="st">&quot;peptideLog&quot;</span>, </span>
<span id="cb30-3"><a href="#cb30-3"></a>                <span class="dt">name =</span> <span class="st">&quot;peptideNorm&quot;</span>, </span>
<span id="cb30-4"><a href="#cb30-4"></a>                <span class="dt">method =</span> <span class="st">&quot;center.median&quot;</span>)</span></code></pre></div>
</p>
</details>
</div>
<div id="plots-of-normalized-data" class="section level3">
<h3><span class="header-section-number">3.3.3</span> Plots of normalized data</h3>
<details>
<summary> Click to see code to make plot </summary>
<p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>densityConditionDNorm &lt;-<span class="st"> </span>pe[[<span class="st">&quot;peptideNorm&quot;</span>]][,<span class="kw">colData</span>(pe)<span class="op">$</span>condition<span class="op">==</span><span class="st">&quot;D&quot;</span>] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb31-3"><a href="#cb31-3"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb31-4"><a href="#cb31-4"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb31-5"><a href="#cb31-5"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">lab =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;lab&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb31-6"><a href="#cb31-6"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>intensity,<span class="dt">group=</span>sample,<span class="dt">color=</span>lab)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb31-7"><a href="#cb31-7"></a><span class="st">    </span><span class="kw">geom_density</span>() <span class="op">+</span></span>
<span id="cb31-8"><a href="#cb31-8"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;condition D&quot;</span>)</span>
<span id="cb31-9"><a href="#cb31-9"></a></span>
<span id="cb31-10"><a href="#cb31-10"></a>densityLab2Norm &lt;-<span class="st"> </span>pe[[<span class="st">&quot;peptideNorm&quot;</span>]][,<span class="kw">colData</span>(pe)<span class="op">$</span>lab<span class="op">==</span><span class="st">&quot;lab2&quot;</span>] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb31-11"><a href="#cb31-11"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb31-12"><a href="#cb31-12"></a><span class="st">  </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb31-13"><a href="#cb31-13"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb31-14"><a href="#cb31-14"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">condition =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;condition&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb31-15"><a href="#cb31-15"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>intensity,<span class="dt">group=</span>sample,<span class="dt">color=</span>condition)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb31-16"><a href="#cb31-16"></a><span class="st">    </span><span class="kw">geom_density</span>() <span class="op">+</span></span>
<span id="cb31-17"><a href="#cb31-17"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;lab2&quot;</span>)</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>densityConditionDNorm</span></code></pre></div>
<pre><code>## Warning: Removed 39179 rows containing non-finite values (stat_density).</code></pre>
<p><img src="pda_quantification_preprocessing_noframes_files/figure-html/unnamed-chunk-33-1.png" width="672" /></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>densityLab2Norm</span></code></pre></div>
<pre><code>## Warning: Removed 44480 rows containing non-finite values (stat_density).</code></pre>
<p><img src="pda_quantification_preprocessing_noframes_files/figure-html/unnamed-chunk-34-1.png" width="672" /></p>
<ul>
<li>Upon normalization the marginal distributions of the peptide intensities across samples are much more comparable</li>
<li>We still see deviations</li>
<li>This can be due to technical variability</li>
<li>In micro-array literature, quantile normalisation is used to force the median and all other quantiles to be equal across samples</li>
<li>In proteomics quantile normalisation often introduces artifacts due to a difference in missing peptides across samples</li>
<li>More advanced methods should be developed for normalizing proteomics data</li>
<li>If there are differences in the width of the marginal distributions of the data across samples. They can also be standardized by using a robust estimator for location and scale, i.e.  <span class="math display">\[y_{ip}^\text{norm} = \frac{y_{ip} - \mu_i}{s_i}\]</span></li>
</ul>
<hr />
</div>
</div>
<div id="summarization" class="section level2">
<h2><span class="header-section-number">3.4</span> Summarization</h2>
<ul>
<li>We illustrate summarization issues using a subset of the cptac study (Lab 2, condition A and E) for a spiked protein (UPS P12081).</li>
</ul>
<details>
<summary> Click to see code to make plot </summary>
<p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>summaryPlot &lt;-<span class="st"> </span>pe[[<span class="st">&quot;peptideNorm&quot;</span>]][</span>
<span id="cb36-2"><a href="#cb36-2"></a>    <span class="kw">rowData</span>(pe[[<span class="st">&quot;peptideNorm&quot;</span>]])<span class="op">$</span>Proteins <span class="op">==</span><span class="st"> &quot;P12081ups|SYHC_HUMAN_UPS&quot;</span>,</span>
<span id="cb36-3"><a href="#cb36-3"></a>    <span class="kw">colData</span>(pe)<span class="op">$</span>lab<span class="op">==</span><span class="st">&quot;lab2&quot;</span><span class="op">&amp;</span><span class="kw">colData</span>(pe)<span class="op">$</span>condition <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;E&quot;</span>)] <span class="op">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4"></a><span class="st">  </span>assay <span class="op">%&gt;%</span></span>
<span id="cb36-5"><a href="#cb36-5"></a><span class="st">  </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb36-6"><a href="#cb36-6"></a><span class="st">  </span><span class="kw">rownames_to_column</span>(<span class="dt">var =</span> <span class="st">&quot;peptide&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb36-7"><a href="#cb36-7"></a><span class="st">  </span><span class="kw">gather</span>(sample, intensity, <span class="op">-</span>peptide) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb36-8"><a href="#cb36-8"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">condition =</span> <span class="kw">colData</span>(pe)[sample,<span class="st">&quot;condition&quot;</span>]) <span class="op">%&gt;%</span></span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> peptide, <span class="dt">y =</span> intensity, <span class="dt">color =</span> sample, <span class="dt">group =</span> sample, <span class="dt">label =</span> condition), <span class="dt">show.legend =</span> <span class="ot">FALSE</span>) <span class="op">+</span></span>
<span id="cb36-10"><a href="#cb36-10"></a><span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">show.legend =</span> <span class="ot">FALSE</span>) <span class="op">+</span></span>
<span id="cb36-11"><a href="#cb36-11"></a><span class="st">  </span><span class="kw">geom_text</span>(<span class="dt">show.legend =</span> <span class="ot">FALSE</span>) <span class="op">+</span></span>
<span id="cb36-12"><a href="#cb36-12"></a><span class="st">  </span><span class="kw">theme_minimal</span>() <span class="op">+</span></span>
<span id="cb36-13"><a href="#cb36-13"></a><span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">vjust =</span> <span class="fl">0.5</span>, <span class="dt">hjust =</span> <span class="dv">1</span>)) <span class="op">+</span></span>
<span id="cb36-14"><a href="#cb36-14"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Peptide&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb36-15"><a href="#cb36-15"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Intensity (log2)&quot;</span>)</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>summaryPlot</span></code></pre></div>
<pre><code>## Warning: Removed 10 row(s) containing missing values (geom_path).</code></pre>
<pre><code>## Warning: Removed 90 rows containing missing values (geom_text).</code></pre>
<p><img src="pda_quantification_preprocessing_noframes_files/figure-html/unnamed-chunk-36-1.png" width="672" /></p>
<p>We observe:</p>
<ul>
<li>intensities from multiple peptides for each protein in a sample</li>
<li>Strong peptide effect -Unbalanced peptide identification</li>
<li>Pseudo-replication: peptide intensities from a particular protein in the same sample are correlated, i.e. they more alike than peptide intensities from a particular protein between samples.</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> Summarize all peptide intensities from the same protein in a sample into a single protein expression value</p>
<p>Commonly used methods are</p>
<ul>
<li><p>Mean summarization <span class="math display">\[
y_{ip}=\beta_i^\text{samp} + \epsilon_{ip}
\]</span></p></li>
<li><p>Median summarization</p></li>
<li><p>Maxquant’s maxLFQ summarization (in protein groups file)</p></li>
<li><p>Model based summarization: <span class="math display">\[
y_{ip}=\beta_i^\text{samp} + \beta_p^\text{pep} + \epsilon_{ip}
\]</span></p></li>
</ul>
<details>
<summary> Click to see R-code to normalize the data </summary>
<p>
<p>We use the standard sumarization in aggregateFeatures, which is robust model based summarization.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>pe &lt;-<span class="st"> </span><span class="kw">aggregateFeatures</span>(pe,</span>
<span id="cb40-2"><a href="#cb40-2"></a>    <span class="dt">i =</span> <span class="st">&quot;peptideNorm&quot;</span>, </span>
<span id="cb40-3"><a href="#cb40-3"></a>    <span class="dt">fcol =</span> <span class="st">&quot;Proteins&quot;</span>, </span>
<span id="cb40-4"><a href="#cb40-4"></a>    <span class="dt">na.rm =</span> <span class="ot">TRUE</span>,</span>
<span id="cb40-5"><a href="#cb40-5"></a>    <span class="dt">name =</span> <span class="st">&quot;protein&quot;</span>)</span></code></pre></div>
<pre><code>## Your quantitative and row data contain missing values. Please read the
## relevant section(s) in the aggregateFeatures manual page regarding the
## effects of missing values on data aggregation.</code></pre>
<p>Other summarization methods can be implemented by using the <code>fun</code> argument in the <code>aggregateFeatures</code> function.</p>
<ul>
<li><p><code>fun = MsCoreUtils::medianPolish()</code> to fits an additive model (two way decomposition) using Tukey’s median polish_ procedure using stats::medpolish()</p></li>
<li><p><code>fun = MsCoreUtils::robustSummary()</code> to calculate a robust aggregation using MASS::rlm() (default)</p></li>
<li><p><code>fun = base::colMeans()</code> to use the mean of each column</p></li>
<li><p><code>fun = matrixStats::colMedians()</code> to use the median of each column</p></li>
<li><p><code>fun = base::colSums()</code> to use the sum of each column</p></li>
</ul>
</p>
</details>
<hr />
</div>
</div>
<div id="exercise" class="section level1">
<h1><span class="header-section-number">4</span> Exercise</h1>
<ol style="list-style-type: decimal">
<li>We will evaluate different summarization methods (Maxquant maxLFQ, median and robust model based) in the tutorial session before discussing on their advantages/disadvantages.</li>
<li>Can you anticipate on potential problems related to the summarization?</li>
</ol>
<hr />
</div>
<div id="code" class="section level1">
<h1><span class="header-section-number">5</span> Code</h1>
<ul>
<li><p>Our R/Bioconductor package <code>msqrob2</code> can be used in R markdown scripts or with a GUI/shinyApp in the <code>msqrob2gui</code> package or .</p></li>
<li><p>Users who want to learn how to code and automate proteomics data analyses in reproducible R markdown scripts can get more information on all code used in this script in the video below.</p></li>
</ul>
<ol style="list-style-type: decimal">
<li><p>Data infrastructure</p></li>
<li><p>Import proteomics data</p></li>
<li><p>Preprocessing</p>
<ul>
<li><p>Log-transformation</p></li>
<li><p>Filtering</p></li>
<li><p>Normalisation</p></li>
<li><p>Summarization</p></li>
</ul></li>
</ol>
</div>

<div id="rmd-source-code">---
title: "Statistical Methods for Quantitative MS-based Proteomics: Part I. Preprocessing"
author: "Lieven Clement"
date: "[statOmics](https://statomics.github.io), Ghent University"
output:
    html_document:
      code_download: true
      theme: cosmo
      toc: true
      toc_float: true
      highlight: tango
      number_sections: true
bibliography: msqrob2.bib

---

<a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>

- [Playlist PDA Preprocessing](https://www.youtube.com/playlist?list=PLZH1hP8_LbJJXQeQ_KYDNoq-AXyFBG6nX)

# Outline {-}

1. Introduction

2. Preprocessing

    - Log-transformation
    - Filtering
    - Normalization
    - Summarization

Note, that the R-code is included for learners who are aiming to develop R/markdown scripts to automate their quantitative proteomics data analyses.
According to the target audience of the course we either work with a graphical user interface (GUI) in a R/shiny App msqrob2gui (e.g. Proteomics Bioinformatics course of the EBI and the Proteomics Data Analysis course at the Gulbenkian institute) or with R/markdowns scripts (e.g. Bioinformatics Summer School at UCLouvain or the Statistical Genomics Course at Ghent University). 

---

# Intro: Challenges in Label-Free Quantitative Proteomics



## MS-based workflow

```{r echo=FALSE}
knitr::include_graphics("./figures/ProteomicsWorkflow.png")
```
  
- Peptide Characteristics
  
  - Modifications
  - Ionisation Efficiency: huge variability
  - Identification
    - Misidentification $\rightarrow$ outliers
    - MS$^2$ selection on peptide abundance
    - Context depending missingness
    - Non-random missingness

$\rightarrow$ Unbalanced pepide identifications across samples and messy data

---

## Level of quantification

- MS-based proteomics returns peptides: pieces of proteins

```{r echo=FALSE}
knitr::include_graphics("./figures/challenges_peptides.png")
```

- Quantification commonly required on the protein level

```{r echo=FALSE}
knitr::include_graphics("./figures/challenges_proteins.png")
```

---

## Label-free Quantitative Proteomics Data Analysis Workflows

```{r echo=FALSE}
knitr::include_graphics("./figures/proteomicsDataAnalysis.png")
```

---

## CPTAC Spike-in Study

```{r echo=FALSE, out.width="50%"}
knitr::include_graphics("./figures/cptacLayoutLudger.png")
```

- Same trypsin-digested yeast proteome background in each sample
- Trypsin-digested Sigma UPS1 standard: 48 different human proteins spiked in at 5 different concentrations (treatment A-E) 
- Samples repeatedly run on different instruments in different labs
- After MaxQuant search with match between runs option

  - 41\% of all proteins are quantified in all samples
  - 6.6\% of all peptides are quantified in all samples

$\rightarrow$ vast amount of missingness


## Maxquant output

```{r echo=FALSE}
knitr::include_graphics("./figures/maxquantOutputDir.png")
```

---


# Import the data in R 

## Data infrastructure

<details><summary> Click to see background on data infrastructure used in R to store proteomics data </summary><p>
- We use the `QFeatures` package that provides the infrastructure to
  - store,  
  - process, 
  - manipulate and 
  - analyse quantitative data/features from mass spectrometry
experiments. 

- It is based on the `SummarizedExperiment` and
`MultiAssayExperiment` classes. 



```{r fig.cap = "Conceptual representation of a `SummarizedExperiment` object.  Assays contain information on the measured omics features (rows) for different samples (columns). The `rowData` contains information on the omics features, the `colData` contains information on the samples, i.e. experimental design etc.", echo=FALSE, out.width="80%"}
knitr::include_graphics("./figures/SE.png")
```

- Assays in a QFeatures object have a
hierarchical relation: 
  
  - proteins are composed of peptides, 
  - themselves produced by spectra
  - relations between assays are tracked and recorded throughout data processing

```{r featuresplot, fig.cap = "Conceptual representation of a `QFeatures` object and the aggregative relation between different assays.", echo = FALSE}
par(mar = c(0, 0, 0, 0))
plot(NA, xlim = c(0, 12), ylim = c(0, 20),
     xaxt = "n", yaxt = "n",
     xlab = "", ylab = "", bty = "n")
for (i in 0:7)
    rect(0, i, 3, i+1, col = "lightgrey", border = "white")
for (i in 8:12)
    rect(0, i, 3, i+1, col = "steelblue", border = "white")
for (i in 13:18)
    rect(0, i, 3, i+1, col = "orange", border = "white")
for (i in 19)
    rect(0, i, 3, i+1, col = "darkgrey", border = "white")
for (i in 5:7)
    rect(5, i, 8, i+1, col = "lightgrey", border = "white")
for (i in 8:10)
    rect(5, i, 8, i+1, col = "steelblue", border = "white")
for (i in 11:13)
    rect(5, i, 8, i+1, col = "orange", border = "white")
for (i in 14)
    rect(5, i, 8, i+1, col = "darkgrey", border = "white")
rect(9, 8, 12, 8+1, col = "lightgrey", border = "white")
rect(9, 9, 12, 9+1, col = "steelblue", border = "white")
rect(9, 10, 12, 10+1, col = "orange", border = "white")
rect(9, 11, 12, 11+1, col = "darkgrey", border = "white")
segments(3, 8, 5, 8, lty = "dashed")
segments(3, 6, 5, 7, lty = "dashed")
segments(3, 4, 5, 6, lty = "dashed")
segments(3, 0, 5, 5, lty = "dashed")
segments(3, 10, 5, 9, lty = "dashed")
segments(3, 11, 5, 10, lty = "dashed")
segments(3, 13, 5, 11, lty = "dashed")
segments(3, 14, 5, 12, lty = "dashed")
segments(3, 16, 5, 13, lty = "dashed")
segments(3, 19, 5, 14, lty = "dashed")
segments(3, 20, 5, 15, lty = "dashed")
segments(8, 5, 9, 8, lty = "dashed")
segments(8, 8, 9, 9, lty = "dashed")
segments(8, 11, 9, 10, lty = "dashed")
segments(8, 14, 9, 11, lty = "dashed")
segments(8, 15, 9, 12, lty = "dashed")
```

</p></details>

## Import data in R


### Load libraries 

<details><summary> Click to see code </summary><p>
```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(limma)
library(QFeatures)
library(msqrob2)
library(plotly)
library(ggplot2)
```
</p></details>

### Read data 

<details><summary> Click to see background and code </summary><p>
1. We use a peptides.txt file from MS-data quantified with maxquant that 
contains MS1 intensities summarized at the peptide level. 
```{r}
peptidesFile <- "https://raw.githubusercontent.com/statOmics/PDA21/data/quantification/fullCptacDatasSetNotForTutorial/peptides.txt"
```

2. Maxquant stores the intensity data for the different samples in columnns that start with Intensity. We can retreive the column names with the intensity data with the code below: 

```{r}
ecols <- grep("Intensity\\.", names(read.delim(peptidesFile)))
```

3. Read the data and store it in  QFeatures object 

```{r}
pe <- readQFeatures(
  table = peptidesFile,
  fnames = 1,
  ecol = ecols,
  name = "peptideRaw", sep="\t")
```
</p></details>

### Explore object

<details><summary> Click to see background and code </summary><p>
- The rowData contains information on the features (peptides) in the assay. E.g. Sequence, protein, ...

```{r}
rowData(pe[["peptideRaw"]])
```

- The colData contains information on the samples

```{r} 
colData(pe)
```

- No information is stored yet on the design. 


```{r} 
pe %>% colnames
```

- Note, that the sample names include the spike-in condition. 
- They also end on a number. 
  
  - 1-3 is from lab 1, 
  - 4-6 from lab 2 and 
  - 7-9 from lab 3. 

- We update the colData with information on the design

```{r}
colData(pe)$lab <- rep(rep(paste0("lab",1:3),each=3),5) %>% as.factor
colData(pe)$condition <- pe[["peptideRaw"]] %>% colnames %>% substr(12,12) %>% as.factor
colData(pe)$spikeConcentration <- rep(c(A = 0.25, B = 0.74, C = 2.22, D = 6.67, E = 20),each = 9)
```

- We explore the colData again

```{r}
colData(pe)
```

</p></details>

---

# Preprocessing

## Log-transformation

### Explore the data with plots

Peptide AALEELVK from spiked-in UPS protein P12081. 
We only show data from lab1.

<details><summary> Click to see code to make plot </summary><p>
```{r}
subset <- pe["AALEELVK",colData(pe)$lab=="lab1"]
plotWhyLog <- data.frame(concentration = colData(subset)$spikeConcentration,
           y = assay(subset[["peptideRaw"]]) %>% c
           ) %>% 
  ggplot(aes(concentration, y)) +
  geom_point() +
  xlab("concentration (fmol/l)") +
  ggtitle("peptide AALEELVK in lab1")
```
</p></details>

```{r}
plotWhyLog
```

- Variance increases with the mean
$\rightarrow$ Multiplicative error structure 

<details><summary> Click to see code to make plot </summary><p>
```{r}
plotLog <- data.frame(concentration = colData(subset)$spikeConcentration,
           y = assay(subset[["peptideRaw"]]) %>% c
           ) %>% 
  ggplot(aes(concentration, y)) +
  geom_point() + 
  scale_x_continuous(trans='log2') + 
  scale_y_continuous(trans='log2') +
  xlab("concentration (fmol/l)") +
  ggtitle("peptide AALEELVK in lab1 with axes on log scale")
```
</p></details>

```{r}
plotLog
```

- Data seems to be homoscedastic on log-scale $\rightarrow$ log transformation of the intensity data
- In quantitative proteomics analysis on $\log_2$ 

$\rightarrow$ Differences on a $\log_2$ scale: $\log_2$ fold changes

$$
\log_2 B - \log_2 A = \log_2 \frac{B}{A} = \log FC_\text{B - A}
$$
$$ 
\begin{array} {l}
log_2 FC = 1 \rightarrow FC = 2^1 =2\\
log_2 FC = 2 \rightarrow FC = 2^2 = 4\\
\end{array}
$$


### log-transformation of the data 

<details><summary> Click to see code to log-transfrom the data </summary><p>
- We calculate how many non zero intensities we have for each peptide and this can be useful for filtering.

```{r}
rowData(pe[["peptideRaw"]])$nNonZero <- rowSums(assay(pe[["peptideRaw"]]) > 0)
```


- Peptides with zero intensities are missing peptides and should be represent
with a `NA` value rather than `0`.

```{r}
pe <- zeroIsNA(pe, "peptideRaw") # convert 0 to NA
```

- Logtransform data with base 2

```{r}
pe <- logTransform(pe, base = 2, i = "peptideRaw", name = "peptideLog")
```
</p></details>

---

## Filtering

- Reverse sequences
- Only identified by modification site (only modified peptides detected)
- Razor peptides: non-unique peptides assigned to the protein group with the most other peptides 
- Contaminants
- Peptides few identifications
- Proteins that are only identified with one or a few peptides

Filtering does not induce bias if the criterion is independent from the downstream data analysis!

<details><summary> Click to see code to filter the data </summary><p>

1. Handling overlapping protein groups

In our approach a peptide can map to multiple proteins, as long as there is
none of these proteins present in a smaller subgroup.

```{r}
pe <- filterFeatures(pe, ~ Proteins %in% smallestUniqueGroups(rowData(pe[["peptideLog"]])$Proteins))
```

2. Remove reverse sequences (decoys) and contaminants

We now remove the contaminants, peptides that map to decoy sequences, and proteins
which were only identified by peptides with modifications.

```{r}
pe <- filterFeatures(pe,~Reverse != "+")
pe <- filterFeatures(pe,~ Potential.contaminant != "+")
```

3. Drop peptides that were only identified in one sample

We keep peptides that were observed at last twice.

```{r}
pe <- filterFeatures(pe,~ nNonZero >=2)
nrow(pe[["peptideLog"]])
```

We keep `r nrow(pe[["peptideLog"]])` peptides upon filtering.
</p></details>

---

## Normalization 

<details><summary> Click to see code to make plot </summary><p>

```{r}
densityConditionD <- pe[["peptideLog"]][,colData(pe)$condition=="D"] %>% 
  assay %>%
  as.data.frame() %>%
  gather(sample, intensity) %>% 
  mutate(lab = colData(pe)[sample,"lab"]) %>%
  ggplot(aes(x=intensity,group=sample,color=lab)) + 
    geom_density() +
    ggtitle("condition D")

densityLab2 <- pe[["peptideLog"]][,colData(pe)$lab=="lab2"] %>% 
  assay %>%
  as.data.frame() %>%
  gather(sample, intensity) %>% 
  mutate(condition = colData(pe)[sample,"condition"]) %>%
  ggplot(aes(x=intensity,group=sample,color=condition)) + 
    geom_density() +
    ggtitle("lab2")
```
</p></details>

```{r}
densityConditionD
```

```{r}
densityLab2
```
- Even in very clean synthetic dataset (same background, only 48 UPS
proteins can be different) the marginal peptide intensity distribution
across samples can be quite distinct

- Considerable effects between and within labs for replicate samples
- Considerable effects between samples with different spike-in
concentration

$\rightarrow$ Normalization is needed

---


### Mean or median?

- Miller and Fishkin (1997) reported that over a period of 30 years males would like to have on average 64.3 partners and females 2.8. 


<details><summary> </summary><p>
- Miller and Fishkin (1997) reported that the median number of partners someone would like to have over a period of 30 years males is 1 for both males and females. 
</p></details>

<details><summary> </summary><p>
Mean is very sensitive to outliers!

```{r echo=FALSE}
knitr::include_graphics("./figures/partners.png")
```
</p></details>


---

### Normalization of the data by median centering

$$y_{ip}^\text{norm} = y_{ip} - \hat\mu_i$$ 
with $\hat\mu_i$ the median intensity over all observed peptides in sample $i$.

<details><summary> Click to see R-code to normalize the data </summary><p>
```{r}
pe <- normalize(pe, 
                i = "peptideLog", 
                name = "peptideNorm", 
                method = "center.median")
```
</p></details>

### Plots of normalized data


<details><summary> Click to see code to make plot </summary><p>
```{r}
densityConditionDNorm <- pe[["peptideNorm"]][,colData(pe)$condition=="D"] %>% 
  assay %>%
  as.data.frame() %>%
  gather(sample, intensity) %>% 
  mutate(lab = colData(pe)[sample,"lab"]) %>%
  ggplot(aes(x=intensity,group=sample,color=lab)) + 
    geom_density() +
    ggtitle("condition D")

densityLab2Norm <- pe[["peptideNorm"]][,colData(pe)$lab=="lab2"] %>% 
  assay %>%
  as.data.frame() %>%
  gather(sample, intensity) %>% 
  mutate(condition = colData(pe)[sample,"condition"]) %>%
  ggplot(aes(x=intensity,group=sample,color=condition)) + 
    geom_density() +
    ggtitle("lab2")
```
</p></details>

```{r}
densityConditionDNorm
```

```{r}
densityLab2Norm
```

- Upon normalization the marginal distributions of the peptide intensities across samples are much more comparable
- We still see deviations
- This can be due to technical variability
- In micro-array literature, quantile normalisation is used to force the median and all other quantiles to be equal across samples
- In proteomics quantile normalisation often introduces artifacts due to a difference in missing peptides across samples 
- More advanced methods should be developed for normalizing proteomics data
- If there are differences in the width of the marginal distributions of the data across samples. They can also be standardized by using a robust estimator for location and scale, i.e. 
$$y_{ip}^\text{norm} = \frac{y_{ip} - \mu_i}{s_i}$$ 

---

## Summarization 

- We illustrate summarization issues using a subset of the cptac study (Lab 2, condition A and E) for a spiked protein (UPS P12081).

<details><summary> Click to see code to make plot </summary><p>
```{r plot = FALSE}
summaryPlot <- pe[["peptideNorm"]][
    rowData(pe[["peptideNorm"]])$Proteins == "P12081ups|SYHC_HUMAN_UPS",
    colData(pe)$lab=="lab2"&colData(pe)$condition %in% c("A","E")] %>%
  assay %>%
  as.data.frame %>%
  rownames_to_column(var = "peptide") %>%
  gather(sample, intensity, -peptide) %>% 
  mutate(condition = colData(pe)[sample,"condition"]) %>%
  ggplot(aes(x = peptide, y = intensity, color = sample, group = sample, label = condition), show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  geom_text(show.legend = FALSE) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  xlab("Peptide") + 
  ylab("Intensity (log2)")
```
</p></details>

```{r}
summaryPlot
```

We observe:  

- intensities from multiple peptides for each protein in a sample
- Strong peptide effect 
-Unbalanced peptide identification
- Pseudo-replication: peptide intensities from a particular protein in the same sample are correlated, i.e. they more alike than peptide intensities from a particular protein between samples. 


$\rightarrow$ Summarize all peptide intensities from the same protein in a sample into a single protein expression value

Commonly used methods are 

- Mean summarization
$$
y_{ip}=\beta_i^\text{samp} + \epsilon_{ip}
$$

- Median summarization
- Maxquant's maxLFQ summarization (in protein groups file)
- Model based summarization: 
$$
y_{ip}=\beta_i^\text{samp} + \beta_p^\text{pep} + \epsilon_{ip}
$$


<details><summary> Click to see R-code to normalize the data </summary><p>
We use the standard sumarization in aggregateFeatures, which is
robust model based summarization. 

```{r,warning=FALSE}
pe <- aggregateFeatures(pe,
    i = "peptideNorm", 
    fcol = "Proteins", 
    na.rm = TRUE,
    name = "protein")
```

Other summarization methods can be implemented by using the `fun` argument in the `aggregateFeatures` function. 

- `fun = MsCoreUtils::medianPolish()` to fits an additive model (two way decomposition) using Tukey's median polish_ procedure using stats::medpolish()

- `fun = MsCoreUtils::robustSummary()` to calculate a robust aggregation using MASS::rlm() (default)

- `fun = base::colMeans()` to use the mean of each column

- `fun = matrixStats::colMedians()` to use the median of each column

- `fun = base::colSums()` to use the sum of each column

</p></details>

---

# Exercise

1. We will evaluate different summarization methods (Maxquant maxLFQ, median and robust model based) in the tutorial session before discussing on their advantages/disadvantages.
2. Can you anticipate on potential problems related to the summarization?

---

# Code

- Our R/Bioconductor package `msqrob2` can be used in R markdown scripts or with a GUI/shinyApp in the `msqrob2gui` package or . 

- Users who want to learn how to code and automate proteomics data analyses in reproducible R markdown scripts can get more information on all code used in this script in the video below. 

1. Data infrastructure

2. Import proteomics data

3. Preprocessing

    -   Log-transformation

    - Filtering

    - Normalisation

    - Summarization

</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("pda_quantification_preprocessing_noframes.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
