<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge" />


<title>Sequencing: RNA-seq data intro</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="sequencing_intro.html">Sequencing - Introduction to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">Sequencing - Working with count data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/SGA21">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Sequencing: RNA-seq data intro</h1>
<h4 class="author">Koen Van den Berge</h4>
<h4 class="date">8/6/2021</h4>

</div>


<p>In this lecture we will start working with a real bulk RNA-seq dataset from <a href="https://pubmed.ncbi.nlm.nih.gov/23024189/">Haglund <em>et al.</em> (2012)</a>. After importing the data, we will be working our way through four major challenges which, together, will form a full RNA-seq differential expression (DE) analysis pipeline where the result of our analysis will be a(n ordered) list of genes that we find to be differently expressed between our conditions of interest. The four main challenges we will look into are</p>
<ul>
<li>Choice of modeling assumptions (distribution).</li>
<li>Normalization.</li>
<li>Parameter estimation under a limited information setting.</li>
<li>Statistical inference under high dimensionality (many genes).</li>
</ul>
<div id="data-import" class="section level1">
<h1><span class="header-section-number">1</span> Data import</h1>
<p>We will be importing this dataset using the data package <a href="https://www.bioconductor.org/packages/release/data/experiment/html/parathyroidSE.html">parathyroidSE</a> from <a href="https://bioconductor.org/">Bioconductor</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)){</span>
<span id="cb1-2"><a href="#cb1-2"></a>  <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a>}</span></code></pre></div>
<pre><code>## Installing package into &#39;/Users/runner/work/_temp/Library&#39;
## (as &#39;lib&#39; is unspecified)</code></pre>
<pre><code>## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//RtmpqxXDWW/downloaded_packages</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;SummarizedExperiment&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()){</span>
<span id="cb4-2"><a href="#cb4-2"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SummarizedExperiment&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a>} </span></code></pre></div>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;BiocVersion&#39;, &#39;SummarizedExperiment&#39;</code></pre>
<pre><code>## also installing the dependencies &#39;zlibbioc&#39;, &#39;bitops&#39;, &#39;matrixStats&#39;, &#39;XVector&#39;, &#39;RCurl&#39;, &#39;GenomeInfoDbData&#39;, &#39;MatrixGenerics&#39;, &#39;GenomicRanges&#39;, &#39;Biobase&#39;, &#39;BiocGenerics&#39;, &#39;S4Vectors&#39;, &#39;IRanges&#39;, &#39;GenomeInfoDb&#39;, &#39;DelayedArray&#39;</code></pre>
<pre><code>## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//RtmpqxXDWW/downloaded_packages</code></pre>
<pre><code>## installing the source package &#39;GenomeInfoDbData&#39;</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># install package if not installed.</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;parathyroidSE&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()) BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;parathyroidSE&quot;</span>)</span></code></pre></div>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;parathyroidSE&#39;</code></pre>
<pre><code>## installing the source package &#39;parathyroidSE&#39;</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">library</span>(parathyroidSE)</span></code></pre></div>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: MatrixGenerics</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
##     colWeightedMeans, colWeightedMedians, colWeightedSds,
##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
##     rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## 
## Attaching package: &#39;Biobase&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:MatrixGenerics&#39;:
## 
##     rowMedians</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb42-2"><a href="#cb42-2"></a></span>
<span id="cb42-3"><a href="#cb42-3"></a><span class="co"># import data</span></span>
<span id="cb42-4"><a href="#cb42-4"></a><span class="kw">data</span>(<span class="st">&quot;parathyroidGenesSE&quot;</span>, <span class="dt">package=</span><span class="st">&quot;parathyroidSE&quot;</span>)</span>
<span id="cb42-5"><a href="#cb42-5"></a><span class="co"># rename for convenience</span></span>
<span id="cb42-6"><a href="#cb42-6"></a>se &lt;-<span class="st"> </span>parathyroidGenesSE</span>
<span id="cb42-7"><a href="#cb42-7"></a><span class="kw">rm</span>(parathyroidGenesSE)</span></code></pre></div>
<p><strong>TODO: look into paper for exp design. Possibly take screenshot and try to decipher together with students.</strong></p>
</div>
<div id="choice-of-modeling-assumptions" class="section level1">
<h1><span class="header-section-number">2</span> Choice of modeling assumptions</h1>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a>y &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[<span class="dv">1</span>,]</span>
<span id="cb43-2"><a href="#cb43-2"></a><span class="kw">hist</span>(y, <span class="dt">breaks =</span> <span class="dv">40</span>,</span>
<span id="cb43-3"><a href="#cb43-3"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Gene expression&quot;</span>,</span>
<span id="cb43-4"><a href="#cb43-4"></a>     <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="dt">yaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb43-5"><a href="#cb43-5"></a>     <span class="dt">main =</span> <span class="st">&quot;Data for the first gene&quot;</span>)</span>
<span id="cb43-6"><a href="#cb43-6"></a><span class="kw">axis</span>(<span class="dv">1</span>, <span class="dt">at =</span> <span class="kw">seq</span>(<span class="dv">200</span>, <span class="dv">1200</span>, <span class="dt">by=</span><span class="dv">200</span>))</span>
<span id="cb43-7"><a href="#cb43-7"></a><span class="kw">axis</span>(<span class="dv">2</span>, <span class="dt">at =</span> <span class="dv">0</span><span class="op">:</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
</div>
<div id="junk" class="section level1">
<h1><span class="header-section-number">3</span> junk</h1>
<p>Start with challenges below, and make a section for each. This will simultaneously follow a full RNA-seq analysis pipeline.</p>
<p>Following code should be part of a next file where we start working with RNA-seq data, show the mean-variance trend, discuss normalization, and introduce a DE analysis.</p>
</div>
<div id="properties-of-rna-seq-count-data" class="section level1">
<h1><span class="header-section-number">4</span> Properties of (RNA-seq) count data</h1>
<p>Goals:</p>
<ul>
<li>Work with real RNA-seq count data, gene by sample matrix</li>
<li>Explore the data for a single gene: univariate (large range, zeroes, discrete, skewed) as well as bivariate wrt covariate.</li>
<li></li>
<li>Demonstrate the mean-variance relationship by plotting mean and variance across genes. Show that the Poisson doesn’t hold across biological replicates and use this as introduction for the negative binomial distribution.</li>
</ul>
<p>In this lecture, we will introduce working with count data, using a real bulk RNA-seq dataset</p>
<ul>
<li>Count data are inherently discrete.</li>
</ul>
<div id="mean-variance-relationship" class="section level2">
<h2><span class="header-section-number">4.1</span> Mean-variance relationship</h2>
</div>
<div id="relative-uncertainty-offsets-and-count-scaling" class="section level2">
<h2><span class="header-section-number">4.2</span> Relative uncertainty, offsets and count scaling</h2>
<p>Defer to normalization.</p>
</div>
</div>
<div id="variance-stabilizing-transformations" class="section level1">
<h1><span class="header-section-number">5</span> Variance-stabilizing transformations</h1>
<p>Defer to when talking about dim red.</p>
</div>

<div id="rmd-source-code">LS0tCnRpdGxlOiAnU2VxdWVuY2luZzogUk5BLXNlcSBkYXRhIGludHJvJwphdXRob3I6ICJLb2VuIFZhbiBkZW4gQmVyZ2UiCmRhdGU6ICI4LzYvMjAyMSIKb3V0cHV0OiAKICBwZGZfZG9jdW1lbnQ6CiAgICB0b2M6IHRydWUKICAgIG51bWJlcl9zZWN0aW9uczogdHJ1ZQogICAgbGF0ZXhfZW5naW5lOiB4ZWxhdGV4CiAgaHRtbF9kb2N1bWVudDoKICAgIHRvYzogdHJ1ZQogICAgdG9jX2Zsb2F0OiB0cnVlCi0tLQoKSW4gdGhpcyBsZWN0dXJlIHdlIHdpbGwgc3RhcnQgd29ya2luZyB3aXRoIGEgcmVhbCBidWxrIFJOQS1zZXEgZGF0YXNldCBmcm9tIFtIYWdsdW5kICpldCBhbC4qICgyMDEyKV0oaHR0cHM6Ly9wdWJtZWQubmNiaS5ubG0ubmloLmdvdi8yMzAyNDE4OS8pLiBBZnRlciBpbXBvcnRpbmcgdGhlIGRhdGEsIHdlIHdpbGwgYmUgd29ya2luZyBvdXIgd2F5IHRocm91Z2ggZm91ciBtYWpvciBjaGFsbGVuZ2VzIHdoaWNoLCB0b2dldGhlciwgd2lsbCBmb3JtIGEgZnVsbCBSTkEtc2VxIGRpZmZlcmVudGlhbCBleHByZXNzaW9uIChERSkgYW5hbHlzaXMgcGlwZWxpbmUgd2hlcmUgdGhlIHJlc3VsdCBvZiBvdXIgYW5hbHlzaXMgd2lsbCBiZSBhKG4gb3JkZXJlZCkgbGlzdCBvZiBnZW5lcyB0aGF0IHdlIGZpbmQgdG8gYmUgZGlmZmVyZW50bHkgZXhwcmVzc2VkIGJldHdlZW4gb3VyIGNvbmRpdGlvbnMgb2YgaW50ZXJlc3QuIFRoZSBmb3VyIG1haW4gY2hhbGxlbmdlcyB3ZSB3aWxsIGxvb2sgaW50byBhcmUgCgogLSBDaG9pY2Ugb2YgbW9kZWxpbmcgYXNzdW1wdGlvbnMgKGRpc3RyaWJ1dGlvbikuCiAtIE5vcm1hbGl6YXRpb24uCiAtIFBhcmFtZXRlciBlc3RpbWF0aW9uIHVuZGVyIGEgbGltaXRlZCBpbmZvcm1hdGlvbiBzZXR0aW5nLgogLSBTdGF0aXN0aWNhbCBpbmZlcmVuY2UgdW5kZXIgaGlnaCBkaW1lbnNpb25hbGl0eSAobWFueSBnZW5lcykuCgojIERhdGEgaW1wb3J0CgpXZSB3aWxsIGJlIGltcG9ydGluZyB0aGlzIGRhdGFzZXQgdXNpbmcgdGhlIGRhdGEgcGFja2FnZSBbcGFyYXRoeXJvaWRTRV0oaHR0cHM6Ly93d3cuYmlvY29uZHVjdG9yLm9yZy9wYWNrYWdlcy9yZWxlYXNlL2RhdGEvZXhwZXJpbWVudC9odG1sL3BhcmF0aHlyb2lkU0UuaHRtbCkgZnJvbSBbQmlvY29uZHVjdG9yXShodHRwczovL2Jpb2NvbmR1Y3Rvci5vcmcvKS4KCmBgYHtyLCB3YXJuaW5nPUZBTFNFLCBldmFsPVRSVUV9CmlmICghcmVxdWlyZU5hbWVzcGFjZSgiQmlvY01hbmFnZXIiLCBxdWlldGx5ID0gVFJVRSkpewogIGluc3RhbGwucGFja2FnZXMoIkJpb2NNYW5hZ2VyIikKfQppZighIlN1bW1hcml6ZWRFeHBlcmltZW50IiAlaW4lIGluc3RhbGxlZC5wYWNrYWdlcygpKXsKICBCaW9jTWFuYWdlcjo6aW5zdGFsbCgiU3VtbWFyaXplZEV4cGVyaW1lbnQiKQp9IAojIGluc3RhbGwgcGFja2FnZSBpZiBub3QgaW5zdGFsbGVkLgppZighInBhcmF0aHlyb2lkU0UiICVpbiUgaW5zdGFsbGVkLnBhY2thZ2VzKCkpIEJpb2NNYW5hZ2VyOjppbnN0YWxsKCJwYXJhdGh5cm9pZFNFIikKCmxpYnJhcnkocGFyYXRoeXJvaWRTRSkKbGlicmFyeShTdW1tYXJpemVkRXhwZXJpbWVudCkKCiMgaW1wb3J0IGRhdGEKZGF0YSgicGFyYXRoeXJvaWRHZW5lc1NFIiwgcGFja2FnZT0icGFyYXRoeXJvaWRTRSIpCiMgcmVuYW1lIGZvciBjb252ZW5pZW5jZQpzZSA8LSBwYXJhdGh5cm9pZEdlbmVzU0UKcm0ocGFyYXRoeXJvaWRHZW5lc1NFKQpgYGAKCioqVE9ETzogbG9vayBpbnRvIHBhcGVyIGZvciBleHAgZGVzaWduLiBQb3NzaWJseSB0YWtlIHNjcmVlbnNob3QgYW5kIHRyeSB0byBkZWNpcGhlciB0b2dldGhlciB3aXRoIHN0dWRlbnRzLioqCgoKIyBDaG9pY2Ugb2YgbW9kZWxpbmcgYXNzdW1wdGlvbnMKCmBgYHtyLCBldmFsPVRSVUV9CnkgPC0gYXNzYXlzKHNlKSRjb3VudHNbMSxdCmhpc3QoeSwgYnJlYWtzID0gNDAsCiAgICAgeGxhYiA9ICJHZW5lIGV4cHJlc3Npb24iLAogICAgIHhheHQgPSAibiIsIHlheHQgPSAibiIsCiAgICAgbWFpbiA9ICJEYXRhIGZvciB0aGUgZmlyc3QgZ2VuZSIpCmF4aXMoMSwgYXQgPSBzZXEoMjAwLCAxMjAwLCBieT0yMDApKQpheGlzKDIsIGF0ID0gMDozKQpgYGAKCgoKCgoKCgoKCgoKCgoKCgoKCgojIGp1bmsKU3RhcnQgd2l0aCBjaGFsbGVuZ2VzIGJlbG93LCBhbmQgbWFrZSBhIHNlY3Rpb24gZm9yIGVhY2guClRoaXMgd2lsbCBzaW11bHRhbmVvdXNseSBmb2xsb3cgYSBmdWxsIFJOQS1zZXEgYW5hbHlzaXMgcGlwZWxpbmUuCgoKRm9sbG93aW5nIGNvZGUgc2hvdWxkIGJlIHBhcnQgb2YgYSBuZXh0IGZpbGUgd2hlcmUgd2Ugc3RhcnQgd29ya2luZyB3aXRoIFJOQS1zZXEgZGF0YSwgc2hvdyB0aGUgbWVhbi12YXJpYW5jZSB0cmVuZCwgZGlzY3VzcyBub3JtYWxpemF0aW9uLCBhbmQgaW50cm9kdWNlIGEgREUgYW5hbHlzaXMuCgojIFByb3BlcnRpZXMgb2YgKFJOQS1zZXEpIGNvdW50IGRhdGEKCkdvYWxzOgoKIC0gV29yayB3aXRoIHJlYWwgUk5BLXNlcSBjb3VudCBkYXRhLCBnZW5lIGJ5IHNhbXBsZSBtYXRyaXgKIC0gRXhwbG9yZSB0aGUgZGF0YSBmb3IgYSBzaW5nbGUgZ2VuZTogdW5pdmFyaWF0ZSAobGFyZ2UgcmFuZ2UsIHplcm9lcywgZGlzY3JldGUsIHNrZXdlZCkgYXMgd2VsbCBhcyBiaXZhcmlhdGUgd3J0IGNvdmFyaWF0ZS4KIC0gCiAtIERlbW9uc3RyYXRlIHRoZSBtZWFuLXZhcmlhbmNlIHJlbGF0aW9uc2hpcCBieSBwbG90dGluZyBtZWFuIGFuZCB2YXJpYW5jZSBhY3Jvc3MgZ2VuZXMuIFNob3cgdGhhdCB0aGUgUG9pc3NvbiBkb2Vzbid0IGhvbGQgYWNyb3NzIGJpb2xvZ2ljYWwgcmVwbGljYXRlcyBhbmQgdXNlIHRoaXMgYXMgaW50cm9kdWN0aW9uIGZvciB0aGUgbmVnYXRpdmUgYmlub21pYWwgZGlzdHJpYnV0aW9uLgoKSW4gdGhpcyBsZWN0dXJlLCB3ZSB3aWxsIGludHJvZHVjZSB3b3JraW5nIHdpdGggY291bnQgZGF0YSwgdXNpbmcgYSByZWFsIGJ1bGsgUk5BLXNlcSBkYXRhc2V0IAoKCgogLSBDb3VudCBkYXRhIGFyZSBpbmhlcmVudGx5IGRpc2NyZXRlLgogCiAKCiMjIE1lYW4tdmFyaWFuY2UgcmVsYXRpb25zaGlwCgoKIyMgUmVsYXRpdmUgdW5jZXJ0YWludHksIG9mZnNldHMgYW5kIGNvdW50IHNjYWxpbmcKCkRlZmVyIHRvIG5vcm1hbGl6YXRpb24uCgojIFZhcmlhbmNlLXN0YWJpbGl6aW5nIHRyYW5zZm9ybWF0aW9ucwoKRGVmZXIgdG8gd2hlbiB0YWxraW5nIGFib3V0IGRpbSByZWQuCgoK</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("sequencing_rnaseqIntro.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
