<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge" />


<title>Sequencing: Selected technical topics</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./docs/intro.pdf">1.1 Position</a>
    </li>
    <li>
      <a href="recapGeneralLinearModel.html">1.2 Recap linear model</a>
    </li>
    <li>
      <a href="./docs/martens_proteomics_bioinformatics.pdf">2.1 PDA - identification</a>
    </li>
    <li>
      <a href="pda_quantification_preprocessing.html">2.2 PDA- Quantification - Preprocessing</a>
    </li>
    <li>
      <a href="pda_robustSummarisation_peptideModels.html">2.2 PDA - Wrap-up Preprocessing - Peptide-Level-Models</a>
    </li>
    <li>
      <a href="pda_quantification_inference.html">2.3 PDA - Quantification - Differential Abundance Analysis</a>
    </li>
    <li>
      <a href="pda_blocking_wrapup.html">2.3 PDA - Wrap-up DA - Blocking</a>
    </li>
    <li>
      <a href="technicalDetailsProteomics.html">2.4. PDA - DA Technical Details</a>
    </li>
    <li>
      <a href="./docs/stagewiseTesting.pdf">2.5 PDA - Stage-wise testing</a>
    </li>
    <li>
      <a href="sequencing_intro.html">3.1 SDA - Intro to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">3.2 SDA - Working with count data</a>
    </li>
    <li>
      <a href="sequencing_rnaseqIntro.html">3.3 SDA - Intro to RNA-seq analysis</a>
    </li>
    <li>
      <a href="sequencing_technicalDE.html">3.4 SDA - Technical topics in DE analysis</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multipleRegression_KPNA2.html">1 Intro KPNA2</a>
    </li>
    <li>
      <a href="https://www.compomics.com/bioinformatics-for-proteomics/">2.1 Identification</a>
    </li>
    <li>
      <a href="pda_tutorialPreprocessing.html">2.2 DA - Preprocessing</a>
    </li>
    <li>
      <a href="pda_tutorialDesign.html">2.3 DA - Design</a>
    </li>
    <li>
      <a href="sequencing_preprocessing.sh">3.1 Preprocessing</a>
    </li>
    <li>
      <a href="sequencing_lab_oneGene.html">3.2 GLM for one gene</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Sequencing: Selected technical topics</h1>
<h4 class="author">Koen Van den Berge</h4>
<h4 class="date">Last edited on 12 November, 2021</h4>

</div>


<div id="independent-filtering" class="section level1">
<h1><span class="header-section-number">1</span> Independent filtering</h1>
<p>Independent filtering, where genes/transcripts/proteins are filtered out prior to statistical analysis, is a common practice in ‘omics experiments. Typically, lowly expressed features are filtered out, and one can argue that these features’ expression is too low to be deemed biologically relevant. In addition, low-count features are also associated with a low statistical power for differential expression (remember relative uncertainty of counts, and <code>edgeR</code>’s BCV plot), and will increase the number of tests performed, and therefore lead to a more severe multiple testing correction.</p>
<div class="figure">
<img src="images_sequencing/independentFiltering.png" alt="Figure 1 from Bourgon *et al.* (2010)." width="1440" />
<p class="caption">
Figure 1 from Bourgon <em>et al.</em> (2010).
</p>
</div>
<hr />
<p>Independent filtering has been formalized by <a href="https://www.pnas.org/content/107/21/9546">Bourgon <em>et al.</em> (2010)</a>, and the concept can be summarized as follows.</p>
<ul>
<li>For each feature we calculate two statistics, <span class="math inline">\(S_F\)</span> and <span class="math inline">\(S_T\)</span>, respectively used for two stages: filtering and testing (e.g., differential expression).</li>
<li>In order for a feature to be deemed significant, both of its statistics must be greater than some cut-off.</li>
<li>We want to control the type I error rate of the second stage (testing). But note that <strong>the second stage is conditional on the first stage</strong>, as we only test features passing the filter, and basically ignore the fact that filtering was performed. Indeed, one criticism is that computing and correcting the <span class="math inline">\(p\)</span>-values as if filtering had not been performed may lead to overoptimistic adjusted <span class="math inline">\(p\)</span>-values.</li>
<li><a href="https://www.pnas.org/content/107/21/9546">Bourgon <em>et al.</em> (2010)</a> show that filtering is only appropriate (i.e., does not inflate type I error rate) if the conditional null distribution of test statistics for features passing the filter is the same as the unconditional null distribution. Therefore, <strong>filtering is appropriate if the statistic used for filtering is independent of the statistic used for testing under the null hypothesis</strong>.</li>
</ul>
<div class="figure">
<img src="images_sequencing/independentFiltering2.png" alt="Figure 2 from Bourgon *et al.* (2010)." width="1438" />
<p class="caption">
Figure 2 from Bourgon <em>et al.</em> (2010).
</p>
</div>
<hr />
<p>Let’s try a couple of examples to get some intuition using simulated data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(DESeq2))</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">set.seed</span>(<span class="dv">24</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a>dds &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">makeExampleDESeqDataSet</span>()</span>
<span id="cb1-4"><a href="#cb1-4"></a>simCounts &lt;-<span class="st"> </span><span class="kw">counts</span>(dds)</span>
<span id="cb1-5"><a href="#cb1-5"></a>group &lt;-<span class="st"> </span>dds<span class="op">$</span>condition</span></code></pre></div>
<div id="a-dependent-test-statistic" class="section level2">
<h2><span class="header-section-number">1.1</span> A dependent test statistic</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>filterStatEffectSize &lt;-<span class="st"> </span><span class="kw">abs</span>(<span class="kw">rowMeans</span>(simCounts[,group <span class="op">==</span><span class="st"> &quot;A&quot;</span>]) <span class="op">-</span><span class="st"> </span><span class="kw">rowMeans</span>(simCounts[,group <span class="op">==</span><span class="st"> &quot;B&quot;</span>]))</span>
<span id="cb2-2"><a href="#cb2-2"></a>testStat &lt;-<span class="st"> </span>genefilter<span class="op">::</span><span class="kw">rowttests</span>(simCounts, group)</span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">## unconditional distribution</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="kw">plot</span>(<span class="kw">density</span>(testStat<span class="op">$</span>statistic, <span class="dt">na.rm=</span><span class="ot">TRUE</span>),</span>
<span id="cb2-6"><a href="#cb2-6"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Test statistic&quot;</span>,</span>
<span id="cb2-7"><a href="#cb2-7"></a>     <span class="dt">main =</span> <span class="st">&quot;Unconditional distribution&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_technicalDE_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">## conditional distribution: very different!</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw">mean</span>(filterStatEffectSize <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.792</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">hist</span>(filterStatEffectSize, <span class="dt">breaks=</span><span class="dv">40</span>)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw">abline</span>(<span class="dt">v=</span><span class="dv">1</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_technicalDE_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>keepEffectSize &lt;-<span class="st"> </span>filterStatEffectSize <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">plot</span>(<span class="kw">density</span>(testStat<span class="op">$</span>statistic[keepEffectSize], <span class="dt">na.rm=</span><span class="ot">TRUE</span>),</span>
<span id="cb6-3"><a href="#cb6-3"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Test statistic&quot;</span>,</span>
<span id="cb6-4"><a href="#cb6-4"></a>     <span class="dt">main =</span> <span class="st">&quot;Conditional distribution&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_technicalDE_files/figure-html/unnamed-chunk-5-3.png" width="672" /></p>
</div>
<div id="an-independent-test-statistic" class="section level2">
<h2><span class="header-section-number">1.2</span> An independent test statistic</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>filterStatGlobalMean &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(simCounts)</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw">mean</span>(filterStatGlobalMean <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>) <span class="co"># we remove a similar fraction</span></span></code></pre></div>
<pre><code>## [1] 0.771</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>keepGlobalMean &lt;-<span class="st"> </span>filterStatGlobalMean <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span></span>
<span id="cb9-2"><a href="#cb9-2"></a></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co">## unconditional distribution</span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="kw">plot</span>(<span class="kw">density</span>(testStat<span class="op">$</span>statistic, <span class="dt">na.rm=</span><span class="ot">TRUE</span>),</span>
<span id="cb9-5"><a href="#cb9-5"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Test statistic&quot;</span>,</span>
<span id="cb9-6"><a href="#cb9-6"></a>     <span class="dt">main =</span> <span class="st">&quot;Unconditional distribution&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_technicalDE_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co">## conditional distribution: the same.</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="kw">plot</span>(<span class="kw">density</span>(testStat<span class="op">$</span>statistic[keepGlobalMean], <span class="dt">na.rm=</span><span class="ot">TRUE</span>),</span>
<span id="cb10-3"><a href="#cb10-3"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Test statistic&quot;</span>,</span>
<span id="cb10-4"><a href="#cb10-4"></a>     <span class="dt">main =</span> <span class="st">&quot;Conditional distribution&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_technicalDE_files/figure-html/unnamed-chunk-6-2.png" width="672" /></p>
</div>
</div>
<div id="aliasing" class="section level1">
<h1><span class="header-section-number">2</span> Aliasing</h1>
<p>Suppose we are working with the following experimental design on colon cancer. Studying the effect of a drug on gene expression, researchers gather RNA-seq data from four colon cancer patients and four healthy individuals. For each individual, they obtain RNA-seq data from a blood sample before as well as two weeks after taking a daily dose of the drug. The research question relates to differential expression after vs. before taking the drug, in particular whether this is different for the diseased versus healthy group (i.e., the interaction between time (before/after taking the drug) and disease status (healthy/colon cancer)).</p>
<p>In terms of the model matrix, we could imagine a design such as <code>~ patient + disease*time</code>, where</p>
<ul>
<li><code>disease</code> is a binary indicator referring to colon cancer versus control sample.</li>
<li><code>time</code> defines if the sample is taken before or after taking the drug.</li>
<li><code>patient</code> defines the individual donor the sample comes from.</li>
</ul>
<p>The research question could then amount to testing the <code>disease * time</code> interaction.</p>
<p>Let’s try this, by simulating random data for one gene.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a>patient &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">rep</span>(letters[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>], <span class="dt">each=</span><span class="dv">2</span>)) <span class="co"># 2 samples per patient</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>disease &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;healthy&quot;</span>,<span class="dv">8</span>), <span class="kw">rep</span>(<span class="st">&quot;cancer&quot;</span>,<span class="dv">8</span>)), <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;healthy&quot;</span>, <span class="st">&quot;cancer&quot;</span>)) <span class="co"># first four are healthy, next four are alcoholic</span></span>
<span id="cb11-4"><a href="#cb11-4"></a>time &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;before&quot;</span>, <span class="st">&quot;after&quot;</span>), <span class="dv">8</span>), <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;before&quot;</span>, <span class="st">&quot;after&quot;</span>)) <span class="co"># one before and one after sample for each</span></span>
<span id="cb11-5"><a href="#cb11-5"></a></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="kw">table</span>(patient, disease, time)</span></code></pre></div>
<pre><code>## , , time = before
## 
##        disease
## patient healthy cancer
##       a       1      0
##       b       1      0
##       c       1      0
##       d       1      0
##       e       0      1
##       f       0      1
##       g       0      1
##       h       0      1
## 
## , , time = after
## 
##        disease
## patient healthy cancer
##       a       1      0
##       b       1      0
##       c       1      0
##       d       1      0
##       e       0      1
##       f       0      1
##       g       0      1
##       h       0      1</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co">## simulate data for one gene</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>n &lt;-<span class="st"> </span><span class="dv">16</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>y &lt;-<span class="st"> </span><span class="kw">rpois</span>(<span class="dt">n =</span> n, <span class="dt">lambda =</span> <span class="dv">50</span>)</span>
<span id="cb13-4"><a href="#cb13-4"></a></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="co">## fit a Poisson model</span></span>
<span id="cb13-6"><a href="#cb13-6"></a>m &lt;-<span class="st"> </span><span class="kw">glm</span>(y <span class="op">~</span><span class="st"> </span>patient <span class="op">+</span><span class="st"> </span>disease<span class="op">*</span>time,</span>
<span id="cb13-7"><a href="#cb13-7"></a>         <span class="dt">family =</span> <span class="st">&quot;poisson&quot;</span>)</span>
<span id="cb13-8"><a href="#cb13-8"></a><span class="kw">summary</span>(m)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = y ~ patient + disease * time, family = &quot;poisson&quot;)
## 
## Deviance Residuals: 
##      Min        1Q    Median        3Q       Max  
## -1.52772  -0.43544   0.00013   0.44162   1.34650  
## 
## Coefficients: (1 not defined because of singularities)
##                         Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept)              3.76900    0.11916  31.631   &lt;2e-16 ***
## patientb                 0.06744    0.14999   0.450   0.6530    
## patientc                 0.06744    0.14999   0.450   0.6530    
## patientd                 0.27304    0.14310   1.908   0.0564 .  
## patiente                 0.16449    0.16224   1.014   0.3107    
## patientf                 0.02565    0.16644   0.154   0.8775    
## patientg                -0.01784    0.16785  -0.106   0.9154    
## patienth                 0.05706    0.16544   0.345   0.7302    
## diseasecancer                 NA         NA      NA       NA    
## timeafter               -0.01567    0.10220  -0.153   0.8782    
## diseasecancer:timeafter  0.12374    0.14407   0.859   0.3904    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for poisson family taken to be 1)
## 
##     Null deviance: 16.1200  on 15  degrees of freedom
## Residual deviance:  8.8417  on  6  degrees of freedom
## AIC: 120.16
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<hr />
<p>We find that one of the coefficients is <code>NA</code>! This is obviously not because we’re dealing with <code>NA</code> values in the data as we’ve just simulated the response variable. What’s going on?</p>
<p>One of the parameters, in this case the parameter distinguishing cancer from healthy patients <strong>cannot be estimated as it is a linear combination of other parameters</strong>. In our case, estimating the diseased effect would use information that is already used to estimate the patient-level intercepts. In other words, <strong>once you know the patient, you immediately also know the disease status</strong>, so estimating the diseased vs healthy effect on top of the patient effect provides no additional information if we have already estimated the patient-level effects. This concept is called aliasing, and is a common technical issue in ’omics experiments with complex experimental designs.</p>
<hr />
<p>While to understand the origin of the aliasing it is crucial to understand the relationship between the variables in the experimental design, we can also investigate it in detail using the <code>alias</code> function, to give us an idea.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">alias</span>(m)</span></code></pre></div>
<pre><code>## Model :
## y ~ patient + disease * time
## 
## Complete :
##               (Intercept) patientb patientc patientd patiente patientf patientg
## diseasecancer 0           0        0        0        1        1        1       
##               patienth timeafter diseasecancer:timeafter
## diseasecancer 1        0         0</code></pre>
<p>We see that the effect <code>diseasecancer</code> is a linear combination of the patient-specific effects of the cancer patients. This makes sense!</p>
<hr />
<p>For clarity, let’s reproduce this using our design matrix.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>X &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>patient <span class="op">+</span><span class="st"> </span>disease<span class="op">*</span>time) <span class="co"># this is the design used in glm()</span></span>
<span id="cb17-2"><a href="#cb17-2"></a></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="co">## these are indeed identical.</span></span>
<span id="cb17-4"><a href="#cb17-4"></a>X[,<span class="st">&quot;diseasecancer&quot;</span>]</span></code></pre></div>
<pre><code>##  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 
##  0  0  0  0  0  0  0  0  1  1  1  1  1  1  1  1</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>X[,<span class="st">&quot;patiente&quot;</span>] <span class="op">+</span><span class="st"> </span>X[,<span class="st">&quot;patientf&quot;</span>] <span class="op">+</span><span class="st"> </span>X[,<span class="st">&quot;patientg&quot;</span>] <span class="op">+</span><span class="st"> </span>X[,<span class="st">&quot;patienth&quot;</span>]</span></code></pre></div>
<pre><code>##  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 
##  0  0  0  0  0  0  0  0  1  1  1  1  1  1  1  1</code></pre>
<p>Since one of our parameters is a linear combination of other parameters, it cannot be estimated simultaneously with the other parameters. In this case, we can actually drop the <code>disease</code> main effect from the model, since we know that it is already included in the <code>patient</code> effect.</p>
<hr />
<p>We will have to carefully construct our design matrix in order to account for all important sources of variation while still allowing us to answer the research question of interest. The aliasing exploration above has made it clear we may drop the <code>disease</code> main effect, so let’s start by constructing this design matrix.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>X &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>patient <span class="op">+</span><span class="st"> </span>time <span class="op">+</span><span class="st"> </span>disease<span class="op">:</span>time)</span>
<span id="cb21-2"><a href="#cb21-2"></a></span>
<span id="cb21-3"><a href="#cb21-3"></a>m2 &lt;-<span class="st"> </span><span class="kw">glm</span>(y <span class="op">~</span><span class="st"> </span><span class="dv">-1</span> <span class="op">+</span><span class="st"> </span>X,</span>
<span id="cb21-4"><a href="#cb21-4"></a>         <span class="dt">family =</span> <span class="st">&quot;poisson&quot;</span>)</span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="kw">summary</span>(m2)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = y ~ -1 + X, family = &quot;poisson&quot;)
## 
## Deviance Residuals: 
##      Min        1Q    Median        3Q       Max  
## -1.52772  -0.43544   0.00013   0.44162   1.34650  
## 
## Coefficients: (1 not defined because of singularities)
##                           Estimate Std. Error z value Pr(&gt;|z|)    
## X(Intercept)               3.76900    0.11916  31.631   &lt;2e-16 ***
## Xpatientb                  0.06744    0.14999   0.450   0.6530    
## Xpatientc                  0.06744    0.14999   0.450   0.6530    
## Xpatientd                  0.27304    0.14310   1.908   0.0564 .  
## Xpatiente                  0.28823    0.16077   1.793   0.0730 .  
## Xpatientf                  0.14939    0.16500   0.905   0.3653    
## Xpatientg                  0.10590    0.16643   0.636   0.5246    
## Xpatienth                  0.18081    0.16400   1.102   0.2703    
## Xtimeafter                -0.01567    0.10220  -0.153   0.8782    
## Xtimebefore:diseasecancer -0.12374    0.14407  -0.859   0.3904    
## Xtimeafter:diseasecancer        NA         NA      NA       NA    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for poisson family taken to be 1)
## 
##     Null deviance: 4489.2752  on 16  degrees of freedom
## Residual deviance:    8.8417  on  6  degrees of freedom
## AIC: 120.16
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="kw">alias</span>(m2)</span></code></pre></div>
<pre><code>## Model :
## y ~ -1 + X
## 
## Complete :
##                          X(Intercept) Xpatientb Xpatientc Xpatientd Xpatiente
## Xtimeafter:diseasecancer  0            0         0         0         1       
##                          Xpatientf Xpatientg Xpatienth Xtimeafter
## Xtimeafter:diseasecancer  1         1         1         0        
##                          Xtimebefore:diseasecancer
## Xtimeafter:diseasecancer -1</code></pre>
<p>We are still confronted with aliasing as the model matrix contains an interaction effect <code>timebefore:diseasecancer</code> as well as <code>timeafter:diseasecancer</code>, while only the latter is relevant. Indeed, we know that we can derive the <code>timebefore:diseasecancer</code> effect by averaging the patient effects of the cancer patients.</p>
<hr />
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>X &lt;-<span class="st"> </span>X[,<span class="op">!</span><span class="kw">colnames</span>(X) <span class="op">%in%</span><span class="st"> &quot;timebefore:diseasecancer&quot;</span>]</span>
<span id="cb25-2"><a href="#cb25-2"></a></span>
<span id="cb25-3"><a href="#cb25-3"></a></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="co">## fit a Poisson model</span></span>
<span id="cb25-5"><a href="#cb25-5"></a>m2 &lt;-<span class="st"> </span><span class="kw">glm</span>(y <span class="op">~</span><span class="st"> </span><span class="dv">-1</span> <span class="op">+</span><span class="st"> </span>X,</span>
<span id="cb25-6"><a href="#cb25-6"></a>         <span class="dt">family =</span> <span class="st">&quot;poisson&quot;</span>)</span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="kw">summary</span>(m2)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = y ~ -1 + X, family = &quot;poisson&quot;)
## 
## Deviance Residuals: 
##      Min        1Q    Median        3Q       Max  
## -1.52772  -0.43544   0.00013   0.44162   1.34650  
## 
## Coefficients:
##                          Estimate Std. Error z value Pr(&gt;|z|)    
## X(Intercept)              3.76900    0.11916  31.631   &lt;2e-16 ***
## Xpatientb                 0.06744    0.14999   0.450   0.6530    
## Xpatientc                 0.06744    0.14999   0.450   0.6530    
## Xpatientd                 0.27304    0.14310   1.908   0.0564 .  
## Xpatiente                 0.16449    0.16224   1.014   0.3107    
## Xpatientf                 0.02565    0.16644   0.154   0.8775    
## Xpatientg                -0.01784    0.16785  -0.106   0.9154    
## Xpatienth                 0.05706    0.16544   0.345   0.7302    
## Xtimeafter               -0.01567    0.10220  -0.153   0.8782    
## Xtimeafter:diseasecancer  0.12374    0.14407   0.859   0.3904    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for poisson family taken to be 1)
## 
##     Null deviance: 4489.2752  on 16  degrees of freedom
## Residual deviance:    8.8417  on  6  degrees of freedom
## AIC: 120.16
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<p>We see that all coefficients can now be estimated. The <code>timeafter</code> effect may be interpreted as the time effect for healthy patients, while the <code>timeafter:diseasecancer</code> effect may be interpreted as the difference in the time effect for cancer patients, with the time effect for healthy patients, i.e., it is the relevant interaction effect we are interested in.</p>
<p><strong>Question</strong>. Taking this further, suppose that we can safely assume that there is no interaction effect between disease status and time. How would you now test for differential expression between healthy and cancer patients at the first timepoint? Specify the experimental design and contrast used.</p>
<details>
<summary> Answer. </summary>
<p>
<p>Assuming no interaction, we can specify the design as follows:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>XMain &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>patient <span class="op">+</span><span class="st"> </span>time)</span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="kw">head</span>(XMain)</span></code></pre></div>
<pre><code>##   (Intercept) patientb patientc patientd patiente patientf patientg patienth
## 1           1        0        0        0        0        0        0        0
## 2           1        0        0        0        0        0        0        0
## 3           1        1        0        0        0        0        0        0
## 4           1        1        0        0        0        0        0        0
## 5           1        0        1        0        0        0        0        0
## 6           1        0        1        0        0        0        0        0
##   timeafter
## 1         0
## 2         1
## 3         0
## 4         1
## 5         0
## 6         1</code></pre>
In order to set up the contrast testing for healthy versus diseased patients at the first timepoint, we need to take the average of the appropriate patient-level intercepts. The average expression for healthy patients is <span class="math display">\[ \log \mu_{healthy} = \frac{1}{4} \left\{ \beta_0 + (\beta_0 + \beta_1) + (\beta_0 + \beta_2) + (\beta_0 + \beta_3) \right\}. \]</span> Similar, for the diseased patients it equals <span class="math display">\[ \log \mu_{diseased} = \frac{1}{4} \left\{ (\beta_0 + \beta_4) + (\beta_0 + \beta_5) + (\beta_0 + \beta_6) + (\beta_0 + \beta_7) \right\}. \]</span> And thus the relevant contrast <span class="math display">\[ \log \frac{\mu_{diseased}}{\mu_{healthy}} = \frac{1}{4} (\beta_4 + \beta_5 + \beta_6 + \beta_7 ) - \frac{1}{4} (\beta_1 + \beta_2 + \beta_3).\]</span>
</p>
</details>
</div>
<div id="limma-voom-as-an-alternative-approach-to-modeling-counts" class="section level1">
<h1><span class="header-section-number">3</span> limma-voom as an alternative approach to modeling counts</h1>
<p><code>limma</code> is a powerful linear model based framework for modeling microarray gene expression data and inferring differential expression results, and has been introduced in the proteomics module of this course. With the inception of RNA-seq, the <code>limma</code> developers got creative and extended their framework to also model count data, hence creating limma-voom.</p>
<div id="the-limma-framework-for-the-analysis-of-microarrays" class="section level2">
<h2><span class="header-section-number">3.1</span> The limma framework for the analysis of microarrays</h2>
<p>In the proteomics module, we have previously introduced the powerful linear model based framework <code>limma</code>, and how it uses an empirical Bayes strategy to borrow information across proteins to derive a posterior variance estimate. In its default implementation, <code>limma</code> cannot be used to model count data, as it can not account for their mean-variance relationship. The developers, however, came up with a creative approach to use the <code>limma</code> framework to model coutn data.</p>
</div>
<div id="limma-voom-extending-limma-for-rna-seq-data" class="section level2">
<h2><span class="header-section-number">3.2</span> limma-voom: extending limma for RNA-seq data</h2>
<ul>
<li>Count models such as <code>edgeR</code> and <code>DESeq2</code> automatically account for the mean-variance relationship of the data, given that the observed mean-variance relationship is close to the one assumed by the distribution. However, they are also more complex, both computationally as well as statistically and conceptually.</li>
<li>limma-voom (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">Law <em>et al.</em> (2014)</a>) is a method that unlocks Gaussian linear models to analyze count data in the context of RNA-seq, by first estimating the mean-variance relationship of the dataset at hand, and subsequently incorporating it in the analysis through observation-level weights in a linear regression model.</li>
</ul>
<div class="figure">
<img src="images_sequencing/limmaVoomMeanVariance.png" alt="Figure 1 from Law *et al.* (2014)." width="1319" />
<p class="caption">
Figure 1 from Law <em>et al.</em> (2014).
</p>
</div>
<hr />
<ul>
<li>The mean-variance relationship is dataset-specific and needs to be estimated separately for each dataset.</li>
<li>The mean-variance trend is estimated nonparametrically across all genes, using a global mean and variance for each gene. Using this trend, observation-level variances are estimated for each individual observation.</li>
<li>These observation-level variances are then used as inverse weights in the linear modeling framework, to account for heteroscedasticity.</li>
</ul>
<div class="figure">
<img src="images_sequencing/limmaVoomWeights.png" alt="Figure 1 from Law *et al.* (2014)." width="1329" />
<p class="caption">
Figure 1 from Law <em>et al.</em> (2014).
</p>
</div>
<p><strong>Add limma-voom analysis of RNA-seq data</strong></p>
</div>
</div>

<div id="rmd-source-code">---
title: 'Sequencing: Selected technical topics'
author: "Koen Van den Berge"
date: "Last edited on `r format(Sys.time(), '%d %B, %Y')`"
output: 
  pdf_document:
    toc: true
    number_sections: true
    latex_engine: xelatex
  html_document:
    toc: true
    toc_float: true
---

```{r functions, include=FALSE}
# A function for captioning and referencing images
fig <- local({
    i <- 0
    ref <- list()
    list(
        cap=function(refName, text) {
            i <<- i + 1
            ref[[refName]] <<- i
            paste("Figure ", i, ": ", text, sep="")
        },
        ref=function(refName) {
            ref[[refName]]
        })
})
``` 

```{r, echo=FALSE, message=FALSE, eval=TRUE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(ggplot2)
})
```

# Independent filtering

Independent filtering, where genes/transcripts/proteins are filtered out prior to statistical analysis, is a common practice in 'omics experiments. Typically, lowly expressed features are filtered out, and one can argue that these features' expression is too low to be deemed biologically relevant. In addition, low-count features are also associated with a low statistical power for differential expression (remember relative uncertainty of counts, and `edgeR`'s BCV plot), and will increase the number of tests performed, and therefore lead to a more severe multiple testing correction.

```{r, echo=FALSE, fig.cap=paste("Figure 1 from Bourgon *et al.* (2010).")}
# All defaults
include_graphics("./images_sequencing/independentFiltering.png")
```

 ---

Independent filtering has been formalized by [Bourgon *et al.* (2010)](https://www.pnas.org/content/107/21/9546), and the concept can be summarized as follows.

 - For each feature we calculate two statistics, $S_F$ and $S_T$, respectively used for two stages: filtering and testing (e.g., differential expression).
 - In order for a feature to be deemed significant, both of its statistics must be greater than some cut-off.
 - We want to control the type I error rate of the second stage (testing). But note that **the second stage is conditional on the first stage**, as we only test features passing the filter, and basically ignore the fact that filtering was performed. Indeed, one criticism is that computing and correcting the $p$-values as if filtering had not been performed may lead to overoptimistic adjusted $p$-values.
 - [Bourgon *et al.* (2010)](https://www.pnas.org/content/107/21/9546) show that filtering is only appropriate (i.e., does not inflate type I error rate) if the conditional null distribution of test statistics for features passing the filter is the same as the unconditional null distribution. Therefore, **filtering is appropriate if the statistic used for filtering is independent of the statistic used for testing under the null hypothesis**.
 
```{r, echo=FALSE, fig.cap=paste("Figure 2 from Bourgon *et al.* (2010).")}
# All defaults
include_graphics("./images_sequencing/independentFiltering2.png")
```
  ---
  
Let's try a couple of examples to get some intuition using simulated data.

```{r}
suppressPackageStartupMessages(library(DESeq2))
set.seed(24)
dds <- DESeq2::makeExampleDESeqDataSet()
simCounts <- counts(dds)
group <- dds$condition
```

## A dependent test statistic

```{r}
filterStatEffectSize <- abs(rowMeans(simCounts[,group == "A"]) - rowMeans(simCounts[,group == "B"]))
testStat <- genefilter::rowttests(simCounts, group)

## unconditional distribution
plot(density(testStat$statistic, na.rm=TRUE),
     xlab = "Test statistic",
     main = "Unconditional distribution")

## conditional distribution: very different!
mean(filterStatEffectSize > 1)
hist(filterStatEffectSize, breaks=40)
abline(v=1, col="red")
keepEffectSize <- filterStatEffectSize > 1
plot(density(testStat$statistic[keepEffectSize], na.rm=TRUE),
     xlab = "Test statistic",
     main = "Conditional distribution")
```


## An independent test statistic

```{r}
filterStatGlobalMean <- rowMeans(simCounts)
mean(filterStatGlobalMean > 5) # we remove a similar fraction
keepGlobalMean <- filterStatGlobalMean > 5

## unconditional distribution
plot(density(testStat$statistic, na.rm=TRUE),
     xlab = "Test statistic",
     main = "Unconditional distribution")

## conditional distribution: the same.
plot(density(testStat$statistic[keepGlobalMean], na.rm=TRUE),
     xlab = "Test statistic",
     main = "Conditional distribution")
```


# Aliasing

Suppose we are working with the following experimental design on colon cancer. Studying the effect of a drug on gene expression, researchers gather RNA-seq data from four colon cancer patients and four healthy individuals. For each individual, they obtain RNA-seq data from a blood sample before as well as two weeks after taking a daily dose of the drug. The research question relates to differential expression after vs. before taking the drug, in particular whether this is different for the diseased versus healthy group (i.e., the interaction between time (before/after taking the drug) and disease status (healthy/colon cancer)).

In terms of the model matrix, we could imagine a design such as ` ~ patient + disease*time`, where 

 - `disease` is a binary indicator referring to colon cancer versus control sample.
 - `time` defines if the sample is taken before or after taking the drug.
 - `patient` defines the individual donor the sample comes from.

The research question could then amount to testing the `disease * time` interaction.

Let's try this, by simulating random data for one gene.

```{r}
set.seed(2)
patient <- factor(rep(letters[1:8], each=2)) # 2 samples per patient
disease <- factor(c(rep("healthy",8), rep("cancer",8)), levels=c("healthy", "cancer")) # first four are healthy, next four are alcoholic
time <- factor(rep(c("before", "after"), 8), levels=c("before", "after")) # one before and one after sample for each

table(patient, disease, time)

## simulate data for one gene
n <- 16
y <- rpois(n = n, lambda = 50)

## fit a Poisson model
m <- glm(y ~ patient + disease*time,
         family = "poisson")
summary(m)
```

 ---
 
We find that one of the coefficients is `NA`! This is obviously not because we're dealing with `NA` values in the data as we've just simulated the response variable. What's going on?

One of the parameters, in this case the parameter distinguishing cancer from healthy patients **cannot be estimated as it is a linear combination of other parameters**. In our case, estimating the diseased effect would use information that is already used to estimate the patient-level intercepts. In other words, **once you know the patient, you immediately also know the disease status**, so estimating the diseased vs healthy effect on top of the patient effect provides no additional information if we have already estimated the patient-level effects. This concept is called aliasing, and is a common technical issue in 'omics experiments with complex experimental designs. 

 ---

While to understand the origin of the aliasing it is crucial to understand the relationship between the variables in the experimental design, we can also investigate it in detail using the `alias` function, to give us an idea.

```{r}
alias(m)
```

We see that the effect `diseasecancer` is a linear combination of the patient-specific effects of the cancer patients. This makes sense!

 --- 

For clarity, let's reproduce this using our design matrix.

```{r}
X <- model.matrix(~ patient + disease*time) # this is the design used in glm()

## these are indeed identical.
X[,"diseasecancer"]
X[,"patiente"] + X[,"patientf"] + X[,"patientg"] + X[,"patienth"]
```

Since one of our parameters is a linear combination of other parameters, it cannot be estimated simultaneously with the other parameters. In this case, we can actually drop the `disease` main effect from the model, since we know that it is already included in the `patient` effect.

 ---

We will have to carefully construct our design matrix in order to account for all important sources of variation while still allowing us to answer the research question of interest. The aliasing exploration above has made it clear we may drop the `disease` main effect, so let's start by constructing this design matrix.


```{r}
X <- model.matrix(~ patient + time + disease:time)

m2 <- glm(y ~ -1 + X,
         family = "poisson")
summary(m2)
alias(m2)
```

We are still confronted with aliasing as the model matrix contains an interaction effect `timebefore:diseasecancer` as well as `timeafter:diseasecancer`, while only the latter is relevant. Indeed, we know that we can derive the `timebefore:diseasecancer` effect by averaging the patient effects of the cancer patients.

 ---

```{r}
X <- X[,!colnames(X) %in% "timebefore:diseasecancer"]


## fit a Poisson model
m2 <- glm(y ~ -1 + X,
         family = "poisson")
summary(m2)
```

We see that all coefficients can now be estimated. The `timeafter` effect may be interpreted as the time effect for healthy patients, while the `timeafter:diseasecancer` effect may be interpreted as the difference in the time effect for cancer patients, with the time effect for healthy patients, i.e., it is the relevant interaction effect we are interested in. 

**Question**. Taking this further, suppose that we can safely assume that there is no interaction effect between disease status and time. How would you now test for differential expression between healthy and cancer patients at the first timepoint? Specify the experimental design and contrast used.

<details><summary> Answer. </summary><p>
Assuming no interaction, we can specify the design as follows:

```{r}
XMain <- model.matrix(~ patient + time)
head(XMain)
```

In order to set up the contrast testing for healthy versus diseased patients at the first timepoint, we need to take the average of the appropriate patient-level intercepts.
The average expression for healthy patients is
$$ \log \mu_{healthy} = \frac{1}{4} \left\{ \beta_0 + (\beta_0 + \beta_1) + (\beta_0 + \beta_2) + (\beta_0 + \beta_3) \right\}. $$
Similar, for the diseased patients it equals
$$ \log \mu_{diseased} = \frac{1}{4} \left\{ (\beta_0 + \beta_4) + (\beta_0 + \beta_5) + (\beta_0 + \beta_6) + (\beta_0 + \beta_7) \right\}. $$
And thus the relevant contrast
$$ \log \frac{\mu_{diseased}}{\mu_{healthy}} = \frac{1}{4} (\beta_4 + \beta_5 + \beta_6 + \beta_7 ) - \frac{1}{4} (\beta_1 + \beta_2 + \beta_3).$$
</p></details>

# limma-voom as an alternative approach to modeling counts

`limma` is a powerful linear model based framework for modeling microarray gene expression data and inferring differential expression results, and has been introduced in the proteomics module of this course. With the inception of RNA-seq, the `limma` developers got creative and extended their framework to also model count data, hence creating limma-voom.

## The limma framework for the analysis of microarrays

In the proteomics module, we have previously introduced the powerful linear model based framework `limma`, and how it uses an empirical Bayes strategy to borrow information across proteins to derive a posterior variance estimate. In its default implementation, `limma` cannot be used to model count data, as it can not account for their mean-variance relationship. The developers, however, came up with a creative approach to use the `limma` framework to model coutn data.

## limma-voom: extending limma for RNA-seq data

 - Count models such as `edgeR` and `DESeq2` automatically account for the mean-variance relationship of the data, given that the observed mean-variance relationship is close to the one assumed by the distribution. However, they are also more complex, both computationally as well as statistically and conceptually.
 - limma-voom ([Law *et al.* (2014)](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)) is a method that unlocks Gaussian linear models to analyze count data in the context of RNA-seq, by first estimating the mean-variance relationship of the dataset at hand, and subsequently incorporating it in the analysis through observation-level weights in a linear regression model.
 
```{r, echo=FALSE, fig.cap=paste("Figure 1 from Law *et al.* (2014).")}
# All defaults
include_graphics("./images_sequencing/limmaVoomMeanVariance.png")
```
 
 ---
 
 - The mean-variance relationship is dataset-specific and needs to be estimated separately for each dataset.
 - The mean-variance trend is estimated nonparametrically across all genes, using a global mean and variance for each gene. Using this trend, observation-level variances are estimated for each individual observation.
 - These observation-level variances are then used as inverse weights in the linear modeling framework, to account for heteroscedasticity.

```{r, echo=FALSE, fig.cap=paste("Figure 1 from Law *et al.* (2014).")}
# All defaults
include_graphics("./images_sequencing/limmaVoomWeights.png")
```


**Add limma-voom analysis of RNA-seq data**
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("sequencing_technicalDE.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
