<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge" />


<title>Sequencing: RNA-seq data intro</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="sequencing_intro.html">Sequencing - Introduction to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">Sequencing - Working with count data</a>
    </li>
    <li>
      <a href="sequencing_rnaseqIntro.html">Sequencing - RNA-seq analysis challenges</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/SGA21">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Sequencing: RNA-seq data intro</h1>
<h4 class="author">Koen Van den Berge</h4>
<h4 class="date">8/6/2021</h4>

</div>


<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;DESeq2&#39;</code></pre>
<pre><code>## also installing the dependencies &#39;fastmap&#39;, &#39;sys&#39;, &#39;bit&#39;, &#39;cachem&#39;, &#39;askpass&#39;, &#39;bitops&#39;, &#39;formatR&#39;, &#39;bit64&#39;, &#39;blob&#39;, &#39;memoise&#39;, &#39;plogr&#39;, &#39;curl&#39;, &#39;openssl&#39;, &#39;RCurl&#39;, &#39;GenomeInfoDbData&#39;, &#39;zlibbioc&#39;, &#39;matrixStats&#39;, &#39;lambda.r&#39;, &#39;futile.options&#39;, &#39;DBI&#39;, &#39;RSQLite&#39;, &#39;XML&#39;, &#39;xtable&#39;, &#39;httr&#39;, &#39;GenomeInfoDb&#39;, &#39;XVector&#39;, &#39;MatrixGenerics&#39;, &#39;DelayedArray&#39;, &#39;futile.logger&#39;, &#39;snow&#39;, &#39;BH&#39;, &#39;AnnotationDbi&#39;, &#39;annotate&#39;, &#39;S4Vectors&#39;, &#39;IRanges&#39;, &#39;GenomicRanges&#39;, &#39;SummarizedExperiment&#39;, &#39;BiocGenerics&#39;, &#39;Biobase&#39;, &#39;BiocParallel&#39;, &#39;genefilter&#39;, &#39;geneplotter&#39;, &#39;RcppArmadillo&#39;</code></pre>
<pre><code>## 
##   There are binary versions available but the source versions are later:
##         binary   source needs_compilation
## RCurl 1.98-1.3 1.98-1.4              TRUE
## XML   3.99-0.6 3.99-0.7              TRUE
## 
## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmp46qEwa/downloaded_packages</code></pre>
<pre><code>## installing the source packages &#39;RCurl&#39;, &#39;GenomeInfoDbData&#39;, &#39;XML&#39;</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<p>In this lecture we will start working with a real bulk RNA-seq dataset from <a href="https://academic.oup.com/jcem/article/97/12/4631/2536573">Haglund <em>et al.</em> (2012)</a>. After importing the data, we will be working our way through four major challenges which, together, will form a full RNA-seq differential expression (DE) analysis pipeline where the result of our analysis will be a(n ordered) list of genes that we find to be differently expressed between our conditions of interest. The four main challenges we will look into are</p>
<ul>
<li>Choice of modeling assumptions (distribution).</li>
<li>Normalization.</li>
<li>Parameter estimation under a limited information setting.</li>
<li>Statistical inference under high dimensionality (many genes).</li>
</ul>
<div id="experimental-design-data-import-and-data-exploration" class="section level1">
<h1><span class="header-section-number">1</span> Experimental design, data import and data exploration</h1>
<div id="experimental-design" class="section level2">
<h2><span class="header-section-number">1.1</span> Experimental design</h2>
<p>Let’s try to work out the experimental design using the following paragraph from the Methods section of the paper.</p>
<div class="figure">
<img src="images_sequencing/expDesign_para.png" alt="" />
<p class="caption">Figure: A paragraph from the Methods section.</p>
</div>
</div>
<div id="data-import-and-exploration" class="section level2">
<h2><span class="header-section-number">1.2</span> Data import and exploration</h2>
<p>We will be importing the dataset using the <a href="https://www.bioconductor.org/packages/release/data/experiment/html/parathyroidSE.html">parathyroidSE</a> data package from <a href="https://bioconductor.org/">Bioconductor</a>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)){</span>
<span id="cb8-2"><a href="#cb8-2"></a>  <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3"></a>}</span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;SummarizedExperiment&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]){</span>
<span id="cb8-5"><a href="#cb8-5"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SummarizedExperiment&quot;</span>)</span>
<span id="cb8-6"><a href="#cb8-6"></a>} </span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co"># install package if not installed.</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;parathyroidSE&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]) BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;parathyroidSE&quot;</span>)</span></code></pre></div>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;parathyroidSE&#39;</code></pre>
<pre><code>## installing the source package &#39;parathyroidSE&#39;</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">suppressPackageStartupMessages</span>({</span>
<span id="cb14-2"><a href="#cb14-2"></a>  <span class="kw">library</span>(parathyroidSE)</span>
<span id="cb14-3"><a href="#cb14-3"></a>  <span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb14-4"><a href="#cb14-4"></a>})</span>
<span id="cb14-5"><a href="#cb14-5"></a></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="co"># import data</span></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="kw">data</span>(<span class="st">&quot;parathyroidGenesSE&quot;</span>, <span class="dt">package=</span><span class="st">&quot;parathyroidSE&quot;</span>)</span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co"># rename for convenience</span></span>
<span id="cb14-9"><a href="#cb14-9"></a>se1 &lt;-<span class="st"> </span>parathyroidGenesSE</span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="kw">rm</span>(parathyroidGenesSE)</span>
<span id="cb14-11"><a href="#cb14-11"></a></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="co"># three treatments</span></span>
<span id="cb14-13"><a href="#cb14-13"></a>treatment1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>treatment</span>
<span id="cb14-14"><a href="#cb14-14"></a><span class="kw">table</span>(treatment1)</span></code></pre></div>
<pre><code>## treatment1
## Control     DPN     OHT 
##       7      10      10</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># two timepoints</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>time1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>time</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="kw">table</span>(time1)</span></code></pre></div>
<pre><code>## time1
## 24h 48h 
##  13  14</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># four donor patients</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>patient1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>patient</span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="kw">table</span>(patient1)</span></code></pre></div>
<pre><code>## patient1
## 1 2 3 4 
## 6 8 6 7</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">table</span>(patient1, treatment1, time1)</span></code></pre></div>
<pre><code>## , , time1 = 24h
## 
##         treatment1
## patient1 Control DPN OHT
##        1       1   1   1
##        2       1   2   2
##        3       1   1   1
##        4       0   1   1
## 
## , , time1 = 48h
## 
##         treatment1
## patient1 Control DPN OHT
##        1       1   1   1
##        2       1   1   1
##        3       1   1   1
##        4       1   2   2</code></pre>
<ul>
<li>We observe that the number of samples that we are observing here is larger than what is described in the paper. As also described in the <a href="https://www.bioconductor.org/packages/release/data/experiment/vignettes/parathyroidSE/inst/doc/parathyroidSE.pdf">parathyroidSE vignette</a>, some samples were spread over multiple sequencing runs (i.e., the same sample being sequenced repeatedly) and therefore constitute <strong>technical replication</strong>, rather than biological replication.</li>
<li>We have previously seen that technical replicates can be considered to be distributed according to a Poisson distribution. One important property of Poisson random variables is that a sum of Poisson random variables still follow a Poisson distribution. Indeed, if <span class="math inline">\(X \sim Poi(\mu_X)\)</span> and <span class="math inline">\(Y \sim Poi(\mu_Y)\)</span>, then $ X + Y = Z Poi(_X + _Y)$.</li>
<li>For this reason, it is often suggested to sum technical replicates rather than, for example, averaging, which does not retain the Poisson property (try for yourself!). We’ll therefore first sum the technical replicates.</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>dupExps &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment[<span class="kw">duplicated</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment)])</span>
<span id="cb22-2"><a href="#cb22-2"></a>dupExps</span></code></pre></div>
<pre><code>## [1] &quot;SRX140511&quot; &quot;SRX140513&quot; &quot;SRX140523&quot; &quot;SRX140525&quot;</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>counts &lt;-<span class="st"> </span><span class="kw">assays</span>(se1)<span class="op">$</span>counts</span>
<span id="cb24-2"><a href="#cb24-2"></a>newCounts &lt;-<span class="st"> </span>counts</span>
<span id="cb24-3"><a href="#cb24-3"></a>cd &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)</span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="cf">for</span>(ss <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dupExps)){</span>
<span id="cb24-5"><a href="#cb24-5"></a>  <span class="co"># check which samples are duplicates</span></span>
<span id="cb24-6"><a href="#cb24-6"></a>  relevantId &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment <span class="op">==</span><span class="st"> </span>dupExps[ss])</span>
<span id="cb24-7"><a href="#cb24-7"></a>  <span class="co"># sum counts</span></span>
<span id="cb24-8"><a href="#cb24-8"></a>  newCounts[,relevantId[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">rowSums</span>(counts[,relevantId])</span>
<span id="cb24-9"><a href="#cb24-9"></a>  <span class="co"># keep which columns / rows to remove.</span></span>
<span id="cb24-10"><a href="#cb24-10"></a>  <span class="cf">if</span>(ss <span class="op">==</span><span class="st"> </span><span class="dv">1</span>){</span>
<span id="cb24-11"><a href="#cb24-11"></a>    toRemove &lt;-<span class="st"> </span>relevantId[<span class="dv">2</span>]</span>
<span id="cb24-12"><a href="#cb24-12"></a>  } <span class="cf">else</span> {</span>
<span id="cb24-13"><a href="#cb24-13"></a>    toRemove &lt;-<span class="st"> </span><span class="kw">c</span>(toRemove, relevantId[<span class="dv">2</span>])</span>
<span id="cb24-14"><a href="#cb24-14"></a>  }</span>
<span id="cb24-15"><a href="#cb24-15"></a>}</span>
<span id="cb24-16"><a href="#cb24-16"></a></span>
<span id="cb24-17"><a href="#cb24-17"></a><span class="co"># remove after summing counts (otherwise IDs get mixed up)</span></span>
<span id="cb24-18"><a href="#cb24-18"></a>newCounts &lt;-<span class="st"> </span>newCounts[,<span class="op">-</span>toRemove]</span>
<span id="cb24-19"><a href="#cb24-19"></a>newCD &lt;-<span class="st"> </span>cd[<span class="op">-</span>toRemove,]</span>
<span id="cb24-20"><a href="#cb24-20"></a></span>
<span id="cb24-21"><a href="#cb24-21"></a><span class="co"># Create new SummarizedExperiment</span></span>
<span id="cb24-22"><a href="#cb24-22"></a>se &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="st">&quot;counts&quot;</span> =<span class="st"> </span>newCounts),</span>
<span id="cb24-23"><a href="#cb24-23"></a>                            <span class="dt">colData =</span> newCD,</span>
<span id="cb24-24"><a href="#cb24-24"></a>                            <span class="dt">metadata =</span> <span class="kw">metadata</span>(se1))</span>
<span id="cb24-25"><a href="#cb24-25"></a></span>
<span id="cb24-26"><a href="#cb24-26"></a>treatment &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>treatment</span>
<span id="cb24-27"><a href="#cb24-27"></a><span class="kw">table</span>(treatment)</span></code></pre></div>
<pre><code>## treatment
## Control     DPN     OHT 
##       7       8       8</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>time &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>time</span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="kw">table</span>(time)</span></code></pre></div>
<pre><code>## time
## 24h 48h 
##  11  12</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>patient &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>patient</span>
<span id="cb28-2"><a href="#cb28-2"></a><span class="kw">table</span>(patient)</span></code></pre></div>
<pre><code>## patient
## 1 2 3 4 
## 6 6 6 5</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="kw">table</span>(patient, treatment, time) <span class="co"># agrees with paper.</span></span></code></pre></div>
<pre><code>## , , time = 24h
## 
##        treatment
## patient Control DPN OHT
##       1       1   1   1
##       2       1   1   1
##       3       1   1   1
##       4       0   1   1
## 
## , , time = 48h
## 
##        treatment
## patient Control DPN OHT
##       1       1   1   1
##       2       1   1   1
##       3       1   1   1
##       4       1   1   1</code></pre>
<ul>
<li>After summing the technical replicates and appropriately updating the sample information, we again create a <code>SummarizedExperiment</code> object, which is essentially a data container that contains all relevant information about your experiment. Please see the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">vignette</a> for more information on how to use this class.</li>
<li>By directly matching columns (samples) and rows (genes) to their relevant metadata, the <code>SummarizedExperiment</code> class avoids mistakes by mis-matching columns and rows with each other (provided you haven’t mismatched them when you creat the object).</li>
<li>The <code>SummarizedExperiment</code> class is modular and extendable, and extensions exist for example for the analysis of single-cell RNA-seq data.</li>
<li>Due to their convenient organization and widely supported usage within Bioconductor, we will typically work with such containers in the analysis of RNA-seq data.</li>
</ul>
</div>
<div id="data-exploration" class="section level2">
<h2><span class="header-section-number">1.3</span> Data exploration</h2>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="kw">suppressPackageStartupMessages</span>({</span>
<span id="cb32-2"><a href="#cb32-2"></a>  <span class="kw">library</span>(limma)</span>
<span id="cb32-3"><a href="#cb32-3"></a>  <span class="kw">library</span>(edgeR)</span>
<span id="cb32-4"><a href="#cb32-4"></a>})</span>
<span id="cb32-5"><a href="#cb32-5"></a></span>
<span id="cb32-6"><a href="#cb32-6"></a><span class="co"># library size distribution</span></span>
<span id="cb32-7"><a href="#cb32-7"></a><span class="kw">hist</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span>, <span class="dt">breaks=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>treatment)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>time)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-3.png" width="672" /></p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>patient)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-4.png" width="672" /></p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span><span class="kw">interaction</span>(treatment, time))</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-5.png" width="672" /></p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="co"># MDS plot</span></span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="kw">plotMDS</span>(se, </span>
<span id="cb37-3"><a href="#cb37-3"></a>        <span class="dt">labels =</span> treatment, </span>
<span id="cb37-4"><a href="#cb37-4"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(patient))</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-6.png" width="672" /></p>
</div>
</div>
<div id="challenge-i-choice-of-modeling-assumptions" class="section level1">
<h1><span class="header-section-number">2</span> Challenge I: Choice of modeling assumptions</h1>
<p>When working with a GLM, as part of the choices of modeling assumptions, we need to pick an appropriate distribution for the expression counts. Below we perform some exploratory analyses to investigate.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>y &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[<span class="dv">1</span>,]</span>
<span id="cb38-2"><a href="#cb38-2"></a><span class="kw">hist</span>(y, <span class="dt">breaks =</span> <span class="dv">40</span>,</span>
<span id="cb38-3"><a href="#cb38-3"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Gene expression&quot;</span>,</span>
<span id="cb38-4"><a href="#cb38-4"></a>     <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="dt">yaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb38-5"><a href="#cb38-5"></a>     <span class="dt">main =</span> <span class="st">&quot;Data for the first gene&quot;</span>)</span>
<span id="cb38-6"><a href="#cb38-6"></a><span class="kw">axis</span>(<span class="dv">1</span>, <span class="dt">at =</span> <span class="kw">seq</span>(<span class="dv">200</span>, <span class="dv">1200</span>, <span class="dt">by=</span><span class="dv">200</span>))</span>
<span id="cb38-7"><a href="#cb38-7"></a><span class="kw">axis</span>(<span class="dv">2</span>, <span class="dt">at =</span> <span class="dv">0</span><span class="op">:</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="co"># Mean-variance trend within each experimental condition</span></span>
<span id="cb39-2"><a href="#cb39-2"></a>cont24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;Control&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb39-3"><a href="#cb39-3"></a>DPN24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;DPN&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb39-4"><a href="#cb39-4"></a>OHT24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;OHT&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb39-5"><a href="#cb39-5"></a>cont48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;Control&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb39-6"><a href="#cb39-6"></a>DPN48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;DPN&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb39-7"><a href="#cb39-7"></a>OHT48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;OHT&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb39-8"><a href="#cb39-8"></a>idList &lt;-<span class="st"> </span><span class="kw">list</span>(cont24ID, DPN24ID, OHT24ID, </span>
<span id="cb39-9"><a href="#cb39-9"></a>               cont48ID, DPN48ID, OHT48ID)</span>
<span id="cb39-10"><a href="#cb39-10"></a><span class="kw">names</span>(idList) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="kw">rep</span>(<span class="kw">levels</span>(treatment),<span class="dv">2</span>), <span class="kw">rep</span>(<span class="kw">levels</span>(time), <span class="dt">each=</span><span class="dv">3</span>))</span>
<span id="cb39-11"><a href="#cb39-11"></a></span>
<span id="cb39-12"><a href="#cb39-12"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb39-13"><a href="#cb39-13"></a><span class="cf">for</span>(kk <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(idList)){</span>
<span id="cb39-14"><a href="#cb39-14"></a>  <span class="co"># extract counts for each condition</span></span>
<span id="cb39-15"><a href="#cb39-15"></a>  curCounts &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[,idList[[kk]]]</span>
<span id="cb39-16"><a href="#cb39-16"></a>  <span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">rowMeans</span>(curCounts)<span class="op">+</span><span class="dv">1</span>,</span>
<span id="cb39-17"><a href="#cb39-17"></a>       <span class="dt">y =</span> <span class="kw">rowVars</span>(curCounts)<span class="op">+</span><span class="dv">1</span>,</span>
<span id="cb39-18"><a href="#cb39-18"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>,</span>
<span id="cb39-19"><a href="#cb39-19"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Mean&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance&quot;</span>,</span>
<span id="cb39-20"><a href="#cb39-20"></a>       <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>)</span>
<span id="cb39-21"><a href="#cb39-21"></a>  <span class="kw">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb39-22"><a href="#cb39-22"></a>  </span>
<span id="cb39-23"><a href="#cb39-23"></a>  <span class="kw">smoothScatter</span>(<span class="dt">x =</span> <span class="kw">log1p</span>(<span class="kw">rowMeans</span>(curCounts)),</span>
<span id="cb39-24"><a href="#cb39-24"></a>       <span class="dt">y =</span> <span class="kw">log1p</span>(<span class="kw">rowVars</span>(curCounts)),</span>
<span id="cb39-25"><a href="#cb39-25"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>,</span>
<span id="cb39-26"><a href="#cb39-26"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Mean&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance&quot;</span>)</span>
<span id="cb39-27"><a href="#cb39-27"></a>  <span class="kw">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb39-28"><a href="#cb39-28"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-7-2.png" width="672" /><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-7-3.png" width="672" /></p>
<ul>
<li>Having data on thousands of genes provides the opportunity to empirically assess the mean-variance relationship.</li>
<li>It is clear that the data is overdispersed with respect to the Poisson distribution. There also seems to be a quadratic trend of the variance as a function of the mean. This has motivated the <strong>negative binomial distribution as the most popular choice to model (bulk) RNA-seq gene expression data</strong>.</li>
</ul>
<hr />
<ul>
<li>The negative binomial distribution is also referred to as the Gamma-Poisson distribution as it can be formulated as such. Indeed, if <span class="math display">\[
 \lambda \sim \Gamma(\alpha, \beta) \\
 Y | \lambda \sim Poi(\lambda),
 \]</span> then this is equivalent to <span class="math display">\[ Y \sim NB(\mu = \alpha / \beta, \phi = 1/\alpha). \]</span></li>
<li>This can be shown analytically, but is considered out of the scope of this course. Below, we show it empirically using simulation.</li>
<li>This theoretical result has got some practical consequences. The Gamma-Poisson formulation makes it clear why we can sum technical replicates as the sum of Poisson random variables is again a Poisson random variable.</li>
<li>The Poisson statement can thus be considered as capturing technical variation, while the Gamma statement can be considered to capture biological variation, i.e., variation in the expression mean across biological replicates.</li>
</ul>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>alpha &lt;-<span class="st"> </span><span class="dv">20</span></span>
<span id="cb40-2"><a href="#cb40-2"></a>beta &lt;-<span class="st"> </span><span class="dv">10</span></span>
<span id="cb40-3"><a href="#cb40-3"></a>lambda &lt;-<span class="st"> </span><span class="kw">rgamma</span>(<span class="dt">n =</span> <span class="fl">1e5</span>, <span class="dt">shape =</span> alpha, <span class="dt">rate =</span> beta)</span>
<span id="cb40-4"><a href="#cb40-4"></a>y1 &lt;-<span class="st"> </span><span class="kw">rpois</span>(<span class="dt">n =</span> <span class="fl">1e5</span>, <span class="dt">lambda =</span> lambda)</span>
<span id="cb40-5"><a href="#cb40-5"></a></span>
<span id="cb40-6"><a href="#cb40-6"></a><span class="co"># note phi = 1 / size</span></span>
<span id="cb40-7"><a href="#cb40-7"></a>y2 &lt;-<span class="st"> </span><span class="kw">rnbinom</span>(<span class="dt">n=</span><span class="fl">1e5</span>, <span class="dt">mu=</span>alpha <span class="op">/</span><span class="st"> </span>(beta), <span class="dt">size=</span>alpha)</span>
<span id="cb40-8"><a href="#cb40-8"></a></span>
<span id="cb40-9"><a href="#cb40-9"></a></span>
<span id="cb40-10"><a href="#cb40-10"></a><span class="kw">plot</span>(<span class="kw">density</span>(y1))</span>
<span id="cb40-11"><a href="#cb40-11"></a><span class="kw">lines</span>(<span class="kw">density</span>(y2), <span class="dt">col=</span><span class="st">&quot;steelblue&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="challenge-ii-normalization" class="section level1">
<h1><span class="header-section-number">3</span> Challenge II: Normalization</h1>
<p>Normalization is necessary to correct for several sources of technical variation:</p>
<ul>
<li><strong>Differences in sequencing depth</strong> between samples. Some samples get sequenced deeper in the sense that they consist of more (mapped) reads and therefore can be considered to contain a higher amount of information, which we should be taking into account. In addition, if a sample is sequenced deeper, it is natural that the counts for each gene will be higher, jeopardizing a direct comparison of the expression counts.</li>
<li><strong>Differences in RNA population composition</strong> between samples. Suppose that two samples have been sequenced to the exact same depth. One sample is contaminated and has a very high concentration of the contaminant cDNA being sequenced, but otherwise the two samples are identical. Since the contaminant will be taking up a significant proportion of the reads being sequenced, the counts will not be directly comparable between the samples. Hence, we may also want to correct for differences in the composition of the RNA population of the samples.</li>
<li><strong>Other technical variation</strong> such as sample-specific GC-content or transcript length effects may also be accounted for.</li>
</ul>
<hr />
<p>Let’s take a look at how comparable different replicates are in the Control condition at 48h in our dataset. We will investigate this using MD-plots (mean-difference plots as introduced by <a href="https://www.jstor.org/stable/24307038">Dudoit <em>et al.</em> (2002)</a>), also sometimes referred to as MA-plots.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a>cont48ID <span class="co"># relevant samples</span></span></code></pre></div>
<pre><code>## [1]  2  8 14 19</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts[,cont48ID]) <span class="op">/</span><span class="st"> </span><span class="fl">1e6</span></span></code></pre></div>
<pre><code>## [1] 10.827109  6.844144  8.064268  7.701432</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a>combs &lt;-<span class="st"> </span><span class="kw">combn</span>(cont48ID, <span class="dt">m=</span><span class="dv">2</span>) <span class="co">#pairwise combinations between samples</span></span>
<span id="cb45-2"><a href="#cb45-2"></a></span>
<span id="cb45-3"><a href="#cb45-3"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb45-4"><a href="#cb45-4"></a><span class="cf">for</span>(cc <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(combs)){</span>
<span id="cb45-5"><a href="#cb45-5"></a>  curSamples &lt;-<span class="st"> </span>combs[,cc]</span>
<span id="cb45-6"><a href="#cb45-6"></a>  M &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts[,curSamples])</span>
<span id="cb45-7"><a href="#cb45-7"></a>  D &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[,curSamples[<span class="dv">2</span>]] <span class="op">/</span><span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[,curSamples[<span class="dv">1</span>]]</span>
<span id="cb45-8"><a href="#cb45-8"></a>  <span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">log</span>(M), <span class="dt">y =</span> <span class="kw">log2</span>(D),</span>
<span id="cb45-9"><a href="#cb45-9"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb45-10"><a href="#cb45-10"></a>       <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Sample &quot;</span>, curSamples[<span class="dv">2</span>], <span class="st">&quot; vs sample &quot;</span>, curSamples[<span class="dv">1</span>]),</span>
<span id="cb45-11"><a href="#cb45-11"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Log mean&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Log2 fold-change&quot;</span>,</span>
<span id="cb45-12"><a href="#cb45-12"></a>       <span class="dt">bty =</span> <span class="st">&#39;l&#39;</span>)</span>
<span id="cb45-13"><a href="#cb45-13"></a>  <span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span>
<span id="cb45-14"><a href="#cb45-14"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<ul>
<li>We see clear bias for some pairwise comparisons. For example, in the first plot comparing sample 8 versus sample 2, the log fold-changes are biased downwards. This means that, on average, a gene is lower expressed in sample 8 versus sample 2. Looking at the library sizes, we can indeed see that the library size for sample 2 is about <span class="math inline">\(11 \times 10^6\)</span> while it is only about <span class="math inline">\(7 \times 10^6\)</span> for sample 8! This is clear library size effect that we should take into account.</li>
</ul>
<div id="count-scaling-versus-glm-offsets" class="section level2">
<h2><span class="header-section-number">3.1</span> Count scaling versus GLM offsets</h2>
<ul>
<li>We have previously discussed count scaling transformations such as CPM and TPM as normalization methods that directly scale counts. However, these transformations ignore inherent information such as the sequencing depth. A measured gene expression of zero in a sample where only 1000 reads are sequenced is very different from a sample where 10 million reads are sequenced.</li>
<li>A more appropriate and natural way when working with GLMs is through the use of <strong>offsets</strong>. The general use of an offset is to account for the ‘effort’ performed in order to gather that observation of the response variable. Two examples:
<ol style="list-style-type: decimal">
<li>A biologist studying whale migration has one fixed spot where, in the migration season, she counts migrating whales day after day, over several years. For each day she records the number of spotted whales. Of course, the time spent whale-watching may differ from day to day and it is natural that you are more likely to spot more whales if you spend more time looking for them. The time spent spotting whales can then be used as an offset.</li>
<li>In our case, a sample being sequenced deeper contains more information, i.e., more ‘effort’ has been performed, as compared to a sample being sequenced relatively shallow. We have more confidence of a count from a deeply sequenced sample than from a shallowly sequenced sample. We can therefore use the sequencing depth <span class="math inline">\(N_i = \sum_g Y_{gi}\)</span> as off set in the model.</li>
</ol></li>
<li>Adding an offset to the model is different from adding a new variable to the model. For each new variable we add, we will estimate its average effect <span class="math inline">\(\beta\)</span> on the response variable. When adding an offset, however, we are implicitly assuming that <span class="math inline">\(\beta=1\)</span>.</li>
<li>Offsets are typically added on the scale of the linear predictor. Suppose we have a gene <span class="math inline">\(g\)</span> and sample <span class="math inline">\(i\)</span> specific offset <span class="math inline">\(O_{ij}\)</span>, then we can define a negative binomial GLM including the offset as <span class="math display">\[
  \left\{
  \begin{array}{ccc}
  Y_{gi} &amp; \sim &amp; NB(\mu_{gi}, \phi_g) \\
  \log \mu_{gi} &amp; = &amp; \eta_{gi} \\
  \eta_{gi} &amp; = &amp; \mathbf{X}^T_i \beta_g + \log(O_{gi}). \\
  \end{array}
  \right.
  \]</span></li>
</ul>
</div>
<div id="how-to-normalize" class="section level2">
<h2><span class="header-section-number">3.2</span> How to normalize?</h2>
<p>Many approaches are available for normalizing RNA-seq data. Most methods basically calculate an offset that is added to the GLM used to model gene expression. One notable method, full-quantile normalization, does not calculate an offset.</p>
<div id="tmm-method-default-of-edger" class="section level3">
<h3><span class="header-section-number">3.2.1</span> TMM method (default of <code>edgeR</code>)</h3>
<p>The trimmed mean of M-values (TMM) method introduced by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">Robinson &amp; Oshlack (2010)</a> is a global-scaling normalization procedure. As the name suggests, it is based on a trimmed mean of fold-changes (<span class="math inline">\(M\)</span>-values) as the scaling factor. A trimmed mean is an average after removing a set of ``extreme’’ values. Specifically, TMM calculates a normalization factor <span class="math inline">\(F_i^{(r)}\)</span> across genes <span class="math inline">\(g\)</span> for each sample <span class="math inline">\(i\)</span> as compared to a reference sample <span class="math inline">\(r\)</span>, <span class="math display">\[
\log_2(F_i^{(r)}) = \frac{\sum_{g \in {\cal G}^*} w_{gi}^r M_{gi}^r}{\sum_{g \in {\cal G}^*} w_{gi}^r},
\]</span> where <span class="math inline">\(M_{gi}^r\)</span> represents the <span class="math inline">\(\log_2\)</span>-fold-change of the gene expression fraction as compared to a reference sample <span class="math inline">\(r\)</span>, i.e., <span class="math display">\[ M_{gi}^r = \log_2\left( \frac{Y_{gi} / N_i}{ Y_{gr} / N_r} \right), \]</span> <span class="math inline">\(w_{gi}^r\)</span> represents a weight calculated as <span class="math display">\[
 w_{gi}^r = \frac{N_i - Y_{gi}}{N_i Y_{gi}} + \frac{N_r - Y_{gr}}{N_r Y_{gr}},
\]</span> and <span class="math inline">\({\cal G}^*\)</span> represents the set of peaks after trimming those with the most extreme values.</p>
<p>The procedure only takes peaks into account where both <span class="math inline">\(Y_{gi}&gt;0\)</span> and <span class="math inline">\(Y_{gr}&gt;0\)</span>. By default, TMM trims peaks with the <span class="math inline">\(30\%\)</span> most extreme <span class="math inline">\(M\)</span>-values and <span class="math inline">\(5\%\)</span> most extreme average gene expression, and chooses as reference <span class="math inline">\(r\)</span> the sample whose upper-quartile is closest to the across-sample average upper-quartile. The normalized counts are then given by <span class="math inline">\(\tilde{Y}_{gi} = Y_{gi} / N_i^s\)</span>, where <span class="math display">\[N_i^s = \frac{N_i F_i^{(r)}}{\sum_i N_i F_i^{(r)}/n}.\]</span></p>
<p>TMM normalization may be performed from the <code>calcNormFactors</code> function implemented in <code>edgeR</code>:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a>dge &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">calcNormFactors</span>(se)</span>
<span id="cb46-2"><a href="#cb46-2"></a>dge<span class="op">$</span>samples <span class="co">#normalization factors added to colData</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["group"],"name":[1],"type":["fct"],"align":["left"]},{"label":["lib.size"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["norm.factors"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["run"],"name":[4],"type":["chr"],"align":["left"]},{"label":["experiment"],"name":[5],"type":["fct"],"align":["left"]},{"label":["patient"],"name":[6],"type":["fct"],"align":["left"]},{"label":["treatment"],"name":[7],"type":["fct"],"align":["left"]},{"label":["time"],"name":[8],"type":["fct"],"align":["left"]},{"label":["submission"],"name":[9],"type":["fct"],"align":["left"]},{"label":["study"],"name":[10],"type":["fct"],"align":["left"]},{"label":["sample"],"name":[11],"type":["fct"],"align":["left"]}],"data":[{"1":"1","2":"9102683","3":"0.9782830","4":"SRR479052","5":"SRX140503","6":"1","7":"Control","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308865","_rn_":"Sample1"},{"1":"1","2":"10827109","3":"0.9728700","4":"SRR479053","5":"SRX140504","6":"1","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308866","_rn_":"Sample2"},{"1":"1","2":"5217761","3":"0.9898593","4":"SRR479054","5":"SRX140505","6":"1","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308867","_rn_":"Sample3"},{"1":"1","2":"9706035","3":"0.9930169","4":"SRR479055","5":"SRX140506","6":"1","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308868","_rn_":"Sample4"},{"1":"1","2":"5700022","3":"0.9850867","4":"SRR479056","5":"SRX140507","6":"1","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308869","_rn_":"Sample5"},{"1":"1","2":"7854568","3":"0.9897270","4":"SRR479057","5":"SRX140508","6":"1","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308870","_rn_":"Sample6"},{"1":"1","2":"8610014","3":"0.9266581","4":"SRR479058","5":"SRX140509","6":"2","7":"Control","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308871","_rn_":"Sample7"},{"1":"1","2":"6844144","3":"0.9544240","4":"SRR479059","5":"SRX140510","6":"2","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308872","_rn_":"Sample8"},{"1":"1","2":"24584280","3":"0.9188545","4":"SRR479060","5":"SRX140511","6":"2","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308873","_rn_":"Sample9"},{"1":"1","2":"8267977","3":"0.9398000","4":"SRR479062","5":"SRX140512","6":"2","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308874","_rn_":"Sample10"},{"1":"1","2":"23590411","3":"0.9096695","4":"SRR479063","5":"SRX140513","6":"2","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308875","_rn_":"Sample11"},{"1":"1","2":"8247122","3":"0.9369050","4":"SRR479065","5":"SRX140514","6":"2","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308876","_rn_":"Sample12"},{"1":"1","2":"7341000","3":"1.0668032","4":"SRR479066","5":"SRX140515","6":"3","7":"Control","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308877","_rn_":"Sample13"},{"1":"1","2":"8064268","3":"1.0552688","4":"SRR479067","5":"SRX140516","6":"3","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308878","_rn_":"Sample14"},{"1":"1","2":"12481958","3":"1.0461698","4":"SRR479068","5":"SRX140517","6":"3","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308879","_rn_":"Sample15"},{"1":"1","2":"16310090","3":"1.0260056","4":"SRR479069","5":"SRX140518","6":"3","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308880","_rn_":"Sample16"},{"1":"1","2":"23697329","3":"1.0268459","4":"SRR479070","5":"SRX140519","6":"3","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308881","_rn_":"Sample17"},{"1":"1","2":"7642648","3":"1.0409451","4":"SRR479071","5":"SRX140520","6":"3","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308882","_rn_":"Sample18"},{"1":"1","2":"7701432","3":"1.0559132","4":"SRR479072","5":"SRX140521","6":"4","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308883","_rn_":"Sample19"},{"1":"1","2":"7135899","3":"1.0675040","4":"SRR479073","5":"SRX140522","6":"4","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308884","_rn_":"Sample20"},{"1":"1","2":"13818393","3":"1.0327004","4":"SRR479074","5":"SRX140523","6":"4","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308885","_rn_":"Sample21"},{"1":"1","2":"6099942","3":"1.0890994","4":"SRR479076","5":"SRX140524","6":"4","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308886","_rn_":"Sample22"},{"1":"1","2":"15825211","3":"1.0286470","4":"SRR479077","5":"SRX140525","6":"4","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308887","_rn_":"Sample23"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Let’s check how our MD-plots look like after normalization. Note that, we can rewrite the GLM as <span class="math display">\[ \log\left( \frac{\mu_{gi}}{O_{gi}} \right) = \mathbf{X}_i^T \beta_g \]</span> and so <span class="math inline">\(\frac{\mu_{gi}}{O_{gi}}\)</span> can be considered as an ‘offset-corrected count’.</p>
<p>We see that all MD-plots are now nicely centered around a log-fold-change of zero!</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="co">## normalize</span></span>
<span id="cb47-2"><a href="#cb47-2"></a>effLibSize &lt;-<span class="st"> </span>dge<span class="op">$</span>samples<span class="op">$</span>lib.size <span class="op">*</span><span class="st"> </span>dge<span class="op">$</span>samples<span class="op">$</span>norm.factors</span>
<span id="cb47-3"><a href="#cb47-3"></a>normCountTMM &lt;-<span class="st"> </span><span class="kw">sweep</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts, <span class="dv">2</span>, <span class="dt">FUN=</span><span class="st">&quot;/&quot;</span>, effLibSize)</span>
<span id="cb47-4"><a href="#cb47-4"></a></span>
<span id="cb47-5"><a href="#cb47-5"></a></span>
<span id="cb47-6"><a href="#cb47-6"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb47-7"><a href="#cb47-7"></a><span class="cf">for</span>(cc <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(combs)){</span>
<span id="cb47-8"><a href="#cb47-8"></a>  curSamples &lt;-<span class="st"> </span>combs[,cc]</span>
<span id="cb47-9"><a href="#cb47-9"></a>  M &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(normCountTMM[,curSamples])</span>
<span id="cb47-10"><a href="#cb47-10"></a>  D &lt;-<span class="st"> </span>normCountTMM[,curSamples[<span class="dv">2</span>]] <span class="op">/</span><span class="st"> </span>normCountTMM[,curSamples[<span class="dv">1</span>]]</span>
<span id="cb47-11"><a href="#cb47-11"></a>  <span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">log</span>(M), <span class="dt">y =</span> <span class="kw">log2</span>(D),</span>
<span id="cb47-12"><a href="#cb47-12"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb47-13"><a href="#cb47-13"></a>       <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Sample &quot;</span>, curSamples[<span class="dv">2</span>], <span class="st">&quot; vs sample &quot;</span>, curSamples[<span class="dv">1</span>]),</span>
<span id="cb47-14"><a href="#cb47-14"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Log mean&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Log2 fold-change&quot;</span>,</span>
<span id="cb47-15"><a href="#cb47-15"></a>       <span class="dt">bty =</span> <span class="st">&#39;l&#39;</span>)</span>
<span id="cb47-16"><a href="#cb47-16"></a>  <span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span>
<span id="cb47-17"><a href="#cb47-17"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
</div>
</div>
<div id="median-of-ratios-method-default-of-deseq2" class="section level2">
<h2><span class="header-section-number">3.3</span> Median-of-ratios method (default of <code>DESeq2</code>)</h2>
<p>The median-of-ratios method is used in <code>DESeq2</code> as described in <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">Love <em>et al.</em> (2014)</a>. It assumes that the expected value <span class="math inline">\(\mu_{gi} = E(Y_{gi})\)</span> is proportional to the true expression of the gene, <span class="math inline">\(q_{gi}\)</span>, scaled by a normalization factor <span class="math inline">\(s_{i}\)</span> for each sample, <span class="math display">\[ \mu_{gi} = s_{i}q_{gi}. \]</span></p>
<p>The normalization factor <span class="math inline">\(s_{i}\)</span> is then estimated using the median-of-ratios method compared to a synthetic reference sample <span class="math inline">\(r\)</span> defined based on geometric means of counts across samples <span class="math display">\[
s_i = \text{median}_{\{{g:Y^{*}_{gr} \ne 0}\}} \frac{Y_{gi}}{Y^{*}_{gr}},
\]</span> with <span class="math display">\[ Y^{*}_{gr} = \left( \prod_{i=1}^n Y_{gi} \right)^{1/n}. \]</span> From this, we calculate the normalized count as <span class="math inline">\(\tilde{Y}_{gi} = Y_{gi} / s_{i}\)</span>.</p>
<p>Median-of-ratios normalization is implemented in the <code>DESeq2</code> package:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a>dds &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">assays</span>(se)<span class="op">$</span>counts,</span>
<span id="cb48-2"><a href="#cb48-2"></a>                                      <span class="dt">colData =</span> <span class="kw">colData</span>(se),</span>
<span id="cb48-3"><a href="#cb48-3"></a>                                      <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span><span class="dv">1</span>) <span class="co">#just add intercept to showcase normalization</span></span></code></pre></div>
<pre><code>## converting counts to integer mode</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a>dds &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">estimateSizeFactors</span>(dds)</span>
<span id="cb50-2"><a href="#cb50-2"></a><span class="kw">sizeFactors</span>(dds)</span></code></pre></div>
<pre><code>##  [1] 0.9187624 1.0644717 0.5232069 0.9826698 0.5676212 0.7807662 0.8284089
##  [8] 0.6600848 2.3955069 0.7906917 2.2784591 0.7860386 0.7807842 0.8445577
## [15] 1.3345483 1.7011973 2.5001652 0.7889061 0.8138720 0.7632306 1.4518320
## [22] 0.6553677 1.6696064</code></pre>
</div>
<div id="conditional-quantile-normalization-cqn" class="section level2">
<h2><span class="header-section-number">3.4</span> Conditional quantile normalization (<code>cqn</code>)</h2>
<p>The <code>cqn</code> method, introduced by <a href="https://academic.oup.com/biostatistics/article/13/2/204/1746212">Hansen <em>et al.</em> (2012)</a> and implemented in the <code>cqn</code> <code>R</code> package, starts by assuming a Poisson model for the gene expression counts <span class="math inline">\(Y_{gi}\)</span>. Median regression is used to model, for each sample, the log-transformed accessibility count as a smooth function of GC-content as well as gene length, focusing on genes with high average count (above <span class="math inline">\(50\)</span> by default). Next, subset quantile normalization (<a href="https://www.liebertpub.com/doi/10.1089/cmb.2010.0049">Wu <em>et al.</em> (2010)</a>) is performed on the residuals of that model (i.e., on the counts adjusted for GC-content) for between-sample normalization. The method could intuitively be thought of as full-quantile normalization after removing a smoothed sample-specific GC-content effect. Normalized counts are calculated as recommended in the <code>cqn</code> vignette, i.e., <span class="math display">\[
\tilde{Y}_{gi} = \left(\frac{(Y_{gi}+1) 10^6}{\sum_g Y_{gi}}\right)  2^{O_{gi}},
\]</span> with <span class="math inline">\(O_{gi}\)</span> the normalization offset estimated by <code>cqn</code>, which is on the <span class="math inline">\(\log_2\)</span> scale. Note that <code>cqn</code> normalization works with a gene- and sample-specific offset, unlike TMM and median-of-ratios normalization which only work with a sample-specific offset.</p>
</div>
<div id="full-quantile-fq-normalization" class="section level2">
<h2><span class="header-section-number">3.5</span> Full quantile (FQ) normalization</h2>
<p>In full-quantile normalization, originally introduced in the context of microarrays by <a href="https://academic.oup.com/bioinformatics/article/19/2/185/372664">Bolstad <em>et al.</em> (2003)</a>, the samples are forced to each have a distribution identical to the distribution of the median/average of the quantiles across samples. In practice, we implement full-quantile normalization using the following procedure</p>
<ol style="list-style-type: decimal">
<li>Given a data matrix <span class="math inline">\(\mathbf{Y}_{G \times n}\)</span> for <span class="math inline">\(G\)</span> genes (rows) and <span class="math inline">\(n\)</span> samples (columns),</li>
<li>sort each column to get <span class="math inline">\(\mathbf{Y}^S\)</span>,</li>
<li>replace all elements of each row by the median (or average) for that row,</li>
<li>obtain the normalized counts <span class="math inline">\(\tilde{\mathbf{Y}}\)</span> by re-arranging (i.e., unsorting) each column.</li>
</ol>
</div>
</div>
<div id="challenge-iii-parameter-estimation-under-limited-information-setting" class="section level1">
<h1><span class="header-section-number">4</span> Challenge III: Parameter estimation (under limited information setting)</h1>
<p>Empirical Bayes, but also the factors we should incorporate into the mean model: show screenshot of paper that doesn’t include interaction effect. Discuss with students.</p>
</div>
<div id="challenge-iv-statistical-inference-across-many-genes" class="section level1">
<h1><span class="header-section-number">5</span> Challenge IV: Statistical inference across many genes</h1>
</div>
<div id="the-importance-of-reproducible-analysis" class="section level1">
<h1><span class="header-section-number">6</span> The importance of reproducible analysis</h1>
<p><a href="https://www.youtube.com/watch?v=8QJfNS7XXwA" class="uri">https://www.youtube.com/watch?v=8QJfNS7XXwA</a></p>
</div>

<div id="rmd-source-code">---
title: 'Sequencing: RNA-seq data intro'
author: "Koen Van den Berge"
date: "8/6/2021"
output: 
  pdf_document:
    toc: true
    number_sections: true
    latex_engine: xelatex
  html_document:
    toc: true
    toc_float: true
---

```{r functions, include=FALSE}
# A function for captioning and referencing images
fig <- local({
    i <- 0
    ref <- list()
    list(
        cap=function(refName, text) {
            i <<- i + 1
            ref[[refName]] <<- i
            paste("Figure ", i, ": ", text, sep="")
        },
        ref=function(refName) {
            ref[[refName]]
        })
})
``` 

```{r, echo=FALSE, eval=TRUE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(ggplot2)
})
if(!"BiocManager" %in% installed.packages()[,1]){
  install.packages("BiocManager")
}
if(!"limma" %in% installed.packages()[,1]){
  BiocManager::install("limma")
}
if(!"edgeR" %in% installed.packages()[,1]){
  BiocManager::install("edgeR")
}
if(!"DESeq2" %in% installed.packages()[,1]){
  BiocManager::install("DESeq2")
}
```

In this lecture we will start working with a real bulk RNA-seq dataset from [Haglund *et al.* (2012)](https://academic.oup.com/jcem/article/97/12/4631/2536573). After importing the data, we will be working our way through four major challenges which, together, will form a full RNA-seq differential expression (DE) analysis pipeline where the result of our analysis will be a(n ordered) list of genes that we find to be differently expressed between our conditions of interest. The four main challenges we will look into are 

 - Choice of modeling assumptions (distribution).
 - Normalization.
 - Parameter estimation under a limited information setting.
 - Statistical inference under high dimensionality (many genes).

# Experimental design, data import and data exploration

## Experimental design

Let's try to work out the experimental design using the following paragraph from the Methods section of the paper.

```{r, echo=FALSE, fig.cap=paste("Figure: A paragraph from the Methods section.")}
# All defaults
include_graphics("./images_sequencing/expDesign_para.png")
```

## Data import and exploration

We will be importing the dataset using the [parathyroidSE](https://www.bioconductor.org/packages/release/data/experiment/html/parathyroidSE.html) data package from [Bioconductor](https://bioconductor.org/).

```{r, warning=FALSE, eval=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}
if(!"SummarizedExperiment" %in% installed.packages()[,1]){
  BiocManager::install("SummarizedExperiment")
} 
# install package if not installed.
if(!"parathyroidSE" %in% installed.packages()[,1]) BiocManager::install("parathyroidSE")

suppressPackageStartupMessages({
  library(parathyroidSE)
  library(SummarizedExperiment)
})

# import data
data("parathyroidGenesSE", package="parathyroidSE")
# rename for convenience
se1 <- parathyroidGenesSE
rm(parathyroidGenesSE)

# three treatments
treatment1 <- colData(se1)$treatment
table(treatment1)
# two timepoints
time1 <- colData(se1)$time
table(time1)
# four donor patients
patient1 <- colData(se1)$patient
table(patient1)

table(patient1, treatment1, time1)
```

- We observe that the number of samples that we are observing here is larger than what is described in the paper. As also described in the [parathyroidSE vignette](https://www.bioconductor.org/packages/release/data/experiment/vignettes/parathyroidSE/inst/doc/parathyroidSE.pdf), some samples were spread over multiple sequencing runs (i.e., the same sample being sequenced repeatedly) and therefore constitute **technical replication**, rather than biological replication.
- We have previously seen that technical replicates can be considered to be distributed according to a Poisson distribution. One important property of Poisson random variables is that a sum of Poisson random variables still follow a Poisson distribution. Indeed, if $X \sim Poi(\mu_X)$ and $Y \sim Poi(\mu_Y)$, then $ X + Y = Z \sim Poi(\mu_X + \mu_Y)$.
- For this reason, it is often suggested to sum technical replicates rather than, for example, averaging, which does not retain the Poisson property (try for yourself!). We'll therefore first sum the technical replicates.

```{r}
dupExps <- as.character(colData(se1)$experiment[duplicated(colData(se1)$experiment)])
dupExps

counts <- assays(se1)$counts
newCounts <- counts
cd <- colData(se1)
for(ss in 1:length(dupExps)){
  # check which samples are duplicates
  relevantId <- which(colData(se1)$experiment == dupExps[ss])
  # sum counts
  newCounts[,relevantId[1]] <- rowSums(counts[,relevantId])
  # keep which columns / rows to remove.
  if(ss == 1){
    toRemove <- relevantId[2]
  } else {
    toRemove <- c(toRemove, relevantId[2])
  }
}

# remove after summing counts (otherwise IDs get mixed up)
newCounts <- newCounts[,-toRemove]
newCD <- cd[-toRemove,]

# Create new SummarizedExperiment
se <- SummarizedExperiment(assays = list("counts" = newCounts),
                            colData = newCD,
                            metadata = metadata(se1))

treatment <- colData(se)$treatment
table(treatment)
time <- colData(se)$time
table(time)
patient <- colData(se)$patient
table(patient)

table(patient, treatment, time) # agrees with paper.
```

 - After summing the technical replicates and appropriately updating the sample information, we again create a `SummarizedExperiment` object, which is essentially a data container that contains all relevant information about your experiment. Please see the [vignette](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) for more information on how to use this class.
 - By directly matching columns (samples) and rows (genes) to their relevant metadata, the `SummarizedExperiment` class avoids mistakes by mis-matching columns and rows with each other (provided you haven't mismatched them when you creat the object).
 - The `SummarizedExperiment` class is modular and extendable, and extensions exist for example for the analysis of single-cell RNA-seq data.
 - Due to their convenient organization and widely supported usage within Bioconductor, we will typically work with such containers in the analysis of RNA-seq data.

## Data exploration

```{r}

suppressPackageStartupMessages({
  library(limma)
  library(edgeR)
})

# library size distribution
hist(colSums(assays(se)$counts)/1e6, breaks=10)
boxplot(colSums(assays(se)$counts)/1e6 ~ treatment)
boxplot(colSums(assays(se)$counts)/1e6 ~ time)
boxplot(colSums(assays(se)$counts)/1e6 ~ patient)
boxplot(colSums(assays(se)$counts)/1e6 ~ interaction(treatment, time))

# MDS plot
plotMDS(se, 
        labels = treatment, 
        col=as.numeric(patient))
```

```{r, eval=FALSE, echo=FALSE}
# Explain concept of MDS: preserve Euclidean distance from high to low dim.
```

# Challenge I: Choice of modeling assumptions

When working with a GLM, as part of the choices of modeling assumptions, we need to pick an appropriate distribution for the expression counts. Below we perform some exploratory analyses to investigate.

```{r, eval=TRUE}
y <- assays(se)$counts[1,]
hist(y, breaks = 40,
     xlab = "Gene expression",
     xaxt = "n", yaxt = "n",
     main = "Data for the first gene")
axis(1, at = seq(200, 1200, by=200))
axis(2, at = 0:3)

# Mean-variance trend within each experimental condition
cont24ID <- which(treatment == "Control" & time == "24h")
DPN24ID <- which(treatment == "DPN" & time == "24h")
OHT24ID <- which(treatment == "OHT" & time == "24h")
cont48ID <- which(treatment == "Control" & time == "48h")
DPN48ID <- which(treatment == "DPN" & time == "48h")
OHT48ID <- which(treatment == "OHT" & time == "48h")
idList <- list(cont24ID, DPN24ID, OHT24ID, 
               cont48ID, DPN48ID, OHT48ID)
names(idList) <- paste0(rep(levels(treatment),2), rep(levels(time), each=3))

par(mfrow=c(3,2), mar=c(2,2,2,1))
for(kk in 1:length(idList)){
  # extract counts for each condition
  curCounts <- assays(se)$counts[,idList[[kk]]]
  plot(x = rowMeans(curCounts)+1,
       y = rowVars(curCounts)+1,
       pch = 16, cex=1/2,
       xlab = "Mean", ylab="Variance",
       log="xy")
  abline(0,1, col="red")
  
  smoothScatter(x = log1p(rowMeans(curCounts)),
       y = log1p(rowVars(curCounts)),
       pch = 16, cex=1/2,
       xlab = "Mean", ylab="Variance")
  abline(0,1, col="red")
}
```


```{r, eval=FALSE, echo=FALSE}
# Ask students why I separate according to time and treatment, and not patient?
# Relate to model definition of a GLM. (Note caveat of library size N_i)
```

 - Having data on thousands of genes provides the opportunity to empirically assess the mean-variance relationship.
 - It is clear that the data is overdispersed with respect to the Poisson distribution. There also seems to be a quadratic trend of the variance as a function of the mean. This has motivated the **negative binomial distribution as the most popular choice to model (bulk) RNA-seq gene expression data**.
 
 ---
 
 - The negative binomial distribution is also referred to as the Gamma-Poisson distribution as it can be formulated as such. Indeed, if 
 $$
 \lambda \sim \Gamma(\alpha, \beta) \\
 Y | \lambda \sim Poi(\lambda),
 $$
 then this is equivalent to
 $$ Y \sim NB(\mu = \alpha / \beta, \phi = 1/\alpha). $$
 - This can be shown analytically, but is considered out of the scope of this course. Below, we show it empirically using simulation.
 - This theoretical result has got some practical consequences. The Gamma-Poisson formulation makes it clear why we can sum technical replicates as the sum of Poisson random variables is again a Poisson random variable. 
 - The Poisson statement can thus be considered as capturing technical variation, while the Gamma statement can be considered to capture biological variation, i.e., variation in the expression mean across biological replicates.
 

```{r}
alpha <- 20
beta <- 10
lambda <- rgamma(n = 1e5, shape = alpha, rate = beta)
y1 <- rpois(n = 1e5, lambda = lambda)

# note phi = 1 / size
y2 <- rnbinom(n=1e5, mu=alpha / (beta), size=alpha)


plot(density(y1))
lines(density(y2), col="steelblue")
```

# Challenge II: Normalization

Normalization is necessary to correct for several sources of technical variation:

 - **Differences in sequencing depth** between samples. Some samples get sequenced deeper in the sense that they consist of more (mapped) reads and therefore can be considered to contain a higher amount of information, which we should be taking into account. In addition, if a sample is sequenced deeper, it is natural that the counts for each gene will be higher, jeopardizing a direct comparison of the expression counts.
 - **Differences in RNA population composition** between samples. Suppose that two samples have been sequenced to the exact same depth. One sample is contaminated and has a very high concentration of the contaminant cDNA being sequenced, but otherwise the two samples are identical. Since the contaminant will be taking up a significant proportion of the reads being sequenced, the counts will not be directly comparable between the samples. Hence, we may also want to correct for differences in the composition of the RNA population of the samples.
 - **Other technical variation** such as sample-specific GC-content or transcript length effects may also be accounted for.
 
 ---
 
Let's take a look at how comparable different replicates are in the Control condition at 48h in our dataset. We will investigate this using MD-plots (mean-difference plots as introduced by [Dudoit *et al.* (2002)](https://www.jstor.org/stable/24307038)), also sometimes referred to as MA-plots.

```{r}
cont48ID # relevant samples
colSums(assays(se)$counts[,cont48ID]) / 1e6
combs <- combn(cont48ID, m=2) #pairwise combinations between samples

par(mfrow=c(3,2), mar=c(4,4,2,1))
for(cc in 1:ncol(combs)){
  curSamples <- combs[,cc]
  M <- rowMeans(assays(se)$counts[,curSamples])
  D <- assays(se)$counts[,curSamples[2]] / assays(se)$counts[,curSamples[1]]
  plot(x = log(M), y = log2(D),
       pch = 16, cex=1/3,
       main = paste0("Sample ", curSamples[2], " vs sample ", curSamples[1]),
       xlab = "Log mean", ylab = "Log2 fold-change",
       bty = 'l')
  abline(h = 0, col="orange", lwd=2)
}
```
 
 - We see clear bias for some pairwise comparisons. For example, in the first plot comparing sample 8 versus sample 2, the log fold-changes are biased downwards. This means that, on average, a gene is lower expressed in sample 8 versus sample 2. Looking at the library sizes, we can indeed see that the library size for sample 2 is about $11 \times 10^6$ while it is only about $7 \times 10^6$ for sample 8! This is clear library size effect that we should take into account.
 
## Count scaling versus GLM offsets

 - We have previously discussed count scaling transformations such as CPM and TPM as normalization methods that directly scale counts. However, these transformations ignore inherent information such as the sequencing depth. A measured gene expression of zero in a sample where only 1000 reads are sequenced is very different from a sample where 10 million reads are sequenced.
 - A more appropriate and natural way when working with GLMs is through the use of **offsets**. The general use of an offset is to account for the 'effort' performed in order to gather that observation of the response variable. Two examples:
   1. A biologist studying whale migration has one fixed spot where, in the migration season, she counts migrating whales day after day, over several years. For each day she records the number of spotted whales. Of course, the time spent whale-watching may differ from day to day and it is natural that you are more likely to spot more whales if you spend more time looking for them. The time spent spotting whales can then be used as an offset.
   2. In our case, a sample being sequenced deeper contains more information, i.e., more 'effort' has been performed, as compared to a sample being sequenced relatively shallow. We have more confidence of a count from a deeply sequenced sample than from a shallowly sequenced sample. We can therefore use the sequencing depth $N_i = \sum_g Y_{gi}$ as off set in the model.
  - Adding an offset to the model is different from adding a new variable to the model. For each new variable we add, we will estimate its average effect $\beta$ on the response variable. When adding an offset, however, we are implicitly assuming that $\beta=1$.
  - Offsets are typically added on the scale of the linear predictor. Suppose we have a gene $g$ and sample $i$ specific offset $O_{ij}$, then we can define a negative binomial GLM including the offset as
  $$
  \left\{
  \begin{array}{ccc}
  Y_{gi} & \sim & NB(\mu_{gi}, \phi_g) \\
  \log \mu_{gi} & = & \eta_{gi} \\
  \eta_{gi} & = & \mathbf{X}^T_i \beta_g + \log(O_{gi}). \\
  \end{array}
  \right.
  $$
  
## How to normalize?

Many approaches are available for normalizing RNA-seq data. Most methods basically calculate an offset that is added to the GLM used to model gene expression. One notable method, full-quantile normalization, does not calculate an offset.
 
### TMM method (default of `edgeR`)

The trimmed mean of M-values (TMM) method introduced by [Robinson & Oshlack (2010)](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25) is a global-scaling normalization procedure. As the name suggests, it is based on a trimmed mean of fold-changes ($M$-values) as the scaling factor. A trimmed mean is an average after removing a set of ``extreme'' values. 
Specifically, TMM calculates a normalization factor $F_i^{(r)}$ across genes $g$ for each sample $i$ as compared to a reference sample $r$,
$$
\log_2(F_i^{(r)}) = \frac{\sum_{g \in {\cal G}^*} w_{gi}^r M_{gi}^r}{\sum_{g \in {\cal G}^*} w_{gi}^r},
$$
where $M_{gi}^r$ represents the $\log_2$-fold-change of the gene expression fraction as compared to a reference sample $r$, i.e.,
\[ M_{gi}^r = \log_2\left( \frac{Y_{gi} / N_i}{ Y_{gr} / N_r} \right), \]
$w_{gi}^r$ represents a weight calculated as
$$
 w_{gi}^r = \frac{N_i - Y_{gi}}{N_i Y_{gi}} + \frac{N_r - Y_{gr}}{N_r Y_{gr}},
$$
and ${\cal G}^*$ represents the set of peaks after trimming those with the most extreme values.

The procedure only takes peaks into account where both $Y_{gi}>0$ and $Y_{gr}>0$. By default, TMM trims peaks with the $30\%$ most extreme $M$-values and $5\%$ most extreme average gene expression, and chooses as reference $r$ the sample whose upper-quartile is closest to the across-sample average upper-quartile. The normalized counts are then given by $\tilde{Y}_{gi} = Y_{gi} / N_i^s$, where 
$$N_i^s = \frac{N_i F_i^{(r)}}{\sum_i N_i F_i^{(r)}/n}.$$

TMM normalization may be performed from the `calcNormFactors` function implemented in `edgeR`:

```{r}
dge <- edgeR::calcNormFactors(se)
dge$samples #normalization factors added to colData
```

Let's check how our MD-plots look like after normalization. Note that, we can rewrite the GLM as
$$ \log\left( \frac{\mu_{gi}}{O_{gi}} \right) = \mathbf{X}_i^T \beta_g $$
and so $\frac{\mu_{gi}}{O_{gi}}$ can be considered as an 'offset-corrected count'.

We see that all MD-plots are now nicely centered around a log-fold-change of zero!

```{r}
## normalize
effLibSize <- dge$samples$lib.size * dge$samples$norm.factors
normCountTMM <- sweep(assays(se)$counts, 2, FUN="/", effLibSize)


par(mfrow=c(3,2), mar=c(4,4,2,1))
for(cc in 1:ncol(combs)){
  curSamples <- combs[,cc]
  M <- rowMeans(normCountTMM[,curSamples])
  D <- normCountTMM[,curSamples[2]] / normCountTMM[,curSamples[1]]
  plot(x = log(M), y = log2(D),
       pch = 16, cex=1/3,
       main = paste0("Sample ", curSamples[2], " vs sample ", curSamples[1]),
       xlab = "Log mean", ylab = "Log2 fold-change",
       bty = 'l')
  abline(h = 0, col="orange", lwd=2)
}
```


## Median-of-ratios method (default of `DESeq2`)

The median-of-ratios method is used in `DESeq2` as described in [Love *et al.* (2014)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8).
It assumes that the expected value $\mu_{gi} = E(Y_{gi})$ is proportional to the true expression of the gene, $q_{gi}$, scaled by a normalization factor $s_{i}$ for each sample,
\[ \mu_{gi} = s_{i}q_{gi}. \]

The normalization factor $s_{i}$ is then estimated using the median-of-ratios method compared to a synthetic reference sample $r$ defined based on geometric means of counts across samples
$$
s_i = \text{median}_{\{{g:Y^{*}_{gr} \ne 0}\}} \frac{Y_{gi}}{Y^{*}_{gr}},
$$
with 
\[ Y^{*}_{gr} = \left( \prod_{i=1}^n Y_{gi} \right)^{1/n}. \]
From this, we calculate the normalized count as $\tilde{Y}_{gi} = Y_{gi} / s_{i}$.

Median-of-ratios normalization is implemented in the `DESeq2` package:

```{r}
dds <- DESeq2::DESeqDataSetFromMatrix(countData = assays(se)$counts,
                                      colData = colData(se),
                                      design = ~ 1) #just add intercept to showcase normalization
dds <- DESeq2::estimateSizeFactors(dds)
sizeFactors(dds)
```

## Conditional quantile normalization (`cqn`)

The `cqn` method, introduced by [Hansen *et al.* (2012)](https://academic.oup.com/biostatistics/article/13/2/204/1746212) and implemented in the `cqn` `R` package, starts by assuming a Poisson model for the gene expression counts $Y_{gi}$. Median regression is used to model, for each sample, the log-transformed accessibility count as a smooth function of GC-content as well as gene length, focusing on genes with high average count (above $50$ by default). Next, subset quantile normalization ([Wu *et al.* (2010)](https://www.liebertpub.com/doi/10.1089/cmb.2010.0049)) is performed on the residuals of that model (i.e., on the counts adjusted for GC-content) for between-sample normalization. The method could intuitively be thought of as full-quantile normalization after removing a smoothed sample-specific GC-content effect. Normalized counts are calculated as recommended in the `cqn` vignette, i.e.,
$$
\tilde{Y}_{gi} = \left(\frac{(Y_{gi}+1) 10^6}{\sum_g Y_{gi}}\right)  2^{O_{gi}},
$$
with $O_{gi}$ the normalization offset estimated by `cqn`, which is on the $\log_2$ scale. Note that `cqn` normalization works with a gene- and sample-specific offset, unlike TMM and median-of-ratios normalization which only work with a sample-specific offset.

## Full quantile (FQ) normalization

In full-quantile normalization, originally introduced in the context of microarrays by [Bolstad *et al.* (2003)](https://academic.oup.com/bioinformatics/article/19/2/185/372664), the samples are forced to each have a distribution identical to the distribution of the median/average of the quantiles across samples.
In practice, we implement full-quantile normalization using the following procedure

 1. Given a data matrix $\mathbf{Y}_{G \times n}$ for $G$ genes (rows) and $n$ samples (columns),
 2. sort each column to get $\mathbf{Y}^S$,
 3. replace all elements of each row by the median (or average) for that row,
 4. obtain the normalized counts $\tilde{\mathbf{Y}}$ by re-arranging (i.e., unsorting) each column. 

# Challenge III: Parameter estimation (under limited information setting)

Empirical Bayes, but also the factors we should incorporate into the mean model: show screenshot of paper that doesn't include interaction effect. Discuss with students.

# Challenge IV: Statistical inference across many genes










# The importance of reproducible analysis

https://www.youtube.com/watch?v=8QJfNS7XXwA



</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("sequencing_rnaseqIntro.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
