---
title: 'Sequencing: Bulk RNA-seq homework'
author: "Koen Van den Berge"
date: "8/6/2021"
output: 
  pdf_document:
    toc: true
    number_sections: true
    latex_engine: xelatex
  html_document:
    toc: true
    toc_float: true
---

```{r functions, include=FALSE}
# A function for captioning and referencing images
fig <- local({
    i <- 0
    ref <- list()
    list(
        cap=function(refName, text) {
            i <<- i + 1
            ref[[refName]] <<- i
            paste("Figure ", i, ": ", text, sep="")
        },
        ref=function(refName) {
            ref[[refName]]
        })
})
``` 

```{r, echo=FALSE, eval=TRUE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(ggplot2)
})
if(!"BiocManager" %in% installed.packages()[,1]){
  install.packages("BiocManager")
}
if(!"limma" %in% installed.packages()[,1]){
  BiocManager::install("limma")
}
if(!"edgeR" %in% installed.packages()[,1]){
  BiocManager::install("edgeR")
}
```


# Default `edgeR` analysis

Analyze the data using `edgeR`, by using the code in the RNA-seq analysis intro lecture.
Focus on the contrast comparing DPN to control at 48h.

# Impact of blocking

Assess the difference in number of DE genes when not blocking on patient, i.e., removing the `patient` effect of the model.
Compare the p-value distributions between these two models.


# Analyze using full-quantile normalization

## Implement and apply full-quantile normalization

```{r}
### implement FQ normalization
FQnorm <- function(counts){
  ...
}

### normalize the data using FQ

```


## `edgeR` analysis using full-quantile normalized data

Don't forgot to remove the `calcNormFactors` step as data have already been normalized!

```{r}
### use FQ-normalized data as input to the edgeR analysis
```


## Compare DE genes at 5% FDR

