<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge" />


<title>Sequencing: RNA-seq data intro</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="sequencing_intro.html">Sequencing - Introduction to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">Sequencing - Working with count data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/SGA21">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Sequencing: RNA-seq data intro</h1>
<h4 class="author">Koen Van den Berge</h4>
<h4 class="date">8/6/2021</h4>

</div>


<pre><code>## Installing package into &#39;/Users/runner/work/_temp/Library&#39;
## (as &#39;lib&#39; is unspecified)</code></pre>
<pre><code>## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmp6Wjp8k/downloaded_packages</code></pre>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;BiocVersion&#39;, &#39;limma&#39;</code></pre>
<pre><code>## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmp6Wjp8k/downloaded_packages</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;edgeR&#39;</code></pre>
<pre><code>## also installing the dependencies &#39;locfit&#39;, &#39;Rcpp&#39;</code></pre>
<pre><code>## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmp6Wjp8k/downloaded_packages</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<p>In this lecture we will start working with a real bulk RNA-seq dataset from <a href="https://academic.oup.com/jcem/article/97/12/4631/2536573">Haglund <em>et al.</em> (2012)</a>. After importing the data, we will be working our way through four major challenges which, together, will form a full RNA-seq differential expression (DE) analysis pipeline where the result of our analysis will be a(n ordered) list of genes that we find to be differently expressed between our conditions of interest. The four main challenges we will look into are</p>
<ul>
<li>Choice of modeling assumptions (distribution).</li>
<li>Normalization.</li>
<li>Parameter estimation under a limited information setting.</li>
<li>Statistical inference under high dimensionality (many genes).</li>
</ul>
<div id="experimental-design-data-import-and-data-exploration" class="section level1">
<h1><span class="header-section-number">1</span> Experimental design, data import and data exploration</h1>
<div id="experimental-design" class="section level2">
<h2><span class="header-section-number">1.1</span> Experimental design</h2>
<p>Let’s try to work out the experimental design using the following paragraph from the Methods section of the paper.</p>
<div class="figure">
<img src="images_sequencing/expDesign_para.png" alt="" />
<p class="caption">Figure: The data output revolution of sequencing machines. Image from Illumina documentation.</p>
</div>
</div>
<div id="data-import-and-exploration" class="section level2">
<h2><span class="header-section-number">1.2</span> Data import and exploration</h2>
<p>We will be importing the dataset using the <a href="https://www.bioconductor.org/packages/release/data/experiment/html/parathyroidSE.html">parathyroidSE</a> data package from <a href="https://bioconductor.org/">Bioconductor</a>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)){</span>
<span id="cb14-2"><a href="#cb14-2"></a>  <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3"></a>}</span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;SummarizedExperiment&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]){</span>
<span id="cb14-5"><a href="#cb14-5"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SummarizedExperiment&quot;</span>)</span>
<span id="cb14-6"><a href="#cb14-6"></a>} </span></code></pre></div>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;SummarizedExperiment&#39;</code></pre>
<pre><code>## also installing the dependencies &#39;zlibbioc&#39;, &#39;bitops&#39;, &#39;matrixStats&#39;, &#39;XVector&#39;, &#39;RCurl&#39;, &#39;GenomeInfoDbData&#39;, &#39;MatrixGenerics&#39;, &#39;GenomicRanges&#39;, &#39;Biobase&#39;, &#39;BiocGenerics&#39;, &#39;S4Vectors&#39;, &#39;IRanges&#39;, &#39;GenomeInfoDb&#39;, &#39;DelayedArray&#39;</code></pre>
<pre><code>## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmp6Wjp8k/downloaded_packages</code></pre>
<pre><code>## installing the source package &#39;GenomeInfoDbData&#39;</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co"># install package if not installed.</span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;parathyroidSE&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]) BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;parathyroidSE&quot;</span>)</span></code></pre></div>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)</code></pre>
<pre><code>## Installing package(s) &#39;parathyroidSE&#39;</code></pre>
<pre><code>## installing the source package &#39;parathyroidSE&#39;</code></pre>
<pre><code>## Old packages: &#39;boot&#39;, &#39;class&#39;, &#39;cluster&#39;, &#39;KernSmooth&#39;, &#39;lattice&#39;, &#39;MASS&#39;,
##   &#39;Matrix&#39;, &#39;mgcv&#39;, &#39;nnet&#39;, &#39;spatial&#39;, &#39;survival&#39;</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw">suppressPackageStartupMessages</span>({</span>
<span id="cb28-2"><a href="#cb28-2"></a>  <span class="kw">library</span>(parathyroidSE)</span>
<span id="cb28-3"><a href="#cb28-3"></a>  <span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb28-4"><a href="#cb28-4"></a>})</span>
<span id="cb28-5"><a href="#cb28-5"></a></span>
<span id="cb28-6"><a href="#cb28-6"></a><span class="co"># import data</span></span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="kw">data</span>(<span class="st">&quot;parathyroidGenesSE&quot;</span>, <span class="dt">package=</span><span class="st">&quot;parathyroidSE&quot;</span>)</span>
<span id="cb28-8"><a href="#cb28-8"></a><span class="co"># rename for convenience</span></span>
<span id="cb28-9"><a href="#cb28-9"></a>se1 &lt;-<span class="st"> </span>parathyroidGenesSE</span>
<span id="cb28-10"><a href="#cb28-10"></a><span class="kw">rm</span>(parathyroidGenesSE)</span>
<span id="cb28-11"><a href="#cb28-11"></a></span>
<span id="cb28-12"><a href="#cb28-12"></a><span class="co"># three treatments</span></span>
<span id="cb28-13"><a href="#cb28-13"></a>treatment1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>treatment</span>
<span id="cb28-14"><a href="#cb28-14"></a><span class="kw">table</span>(treatment1)</span></code></pre></div>
<pre><code>## treatment1
## Control     DPN     OHT 
##       7      10      10</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co"># two timepoints</span></span>
<span id="cb30-2"><a href="#cb30-2"></a>time1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>time</span>
<span id="cb30-3"><a href="#cb30-3"></a><span class="kw">table</span>(time1)</span></code></pre></div>
<pre><code>## time1
## 24h 48h 
##  13  14</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="co"># four donor patients</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>patient1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>patient</span>
<span id="cb32-3"><a href="#cb32-3"></a><span class="kw">table</span>(patient1)</span></code></pre></div>
<pre><code>## patient1
## 1 2 3 4 
## 6 8 6 7</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="kw">table</span>(patient1, treatment1, time1)</span></code></pre></div>
<pre><code>## , , time1 = 24h
## 
##         treatment1
## patient1 Control DPN OHT
##        1       1   1   1
##        2       1   2   2
##        3       1   1   1
##        4       0   1   1
## 
## , , time1 = 48h
## 
##         treatment1
## patient1 Control DPN OHT
##        1       1   1   1
##        2       1   1   1
##        3       1   1   1
##        4       1   2   2</code></pre>
<ul>
<li>We observe that the number of samples that we are observing here is larger than what is described in the paper. As also described in the <a href="https://www.bioconductor.org/packages/release/data/experiment/vignettes/parathyroidSE/inst/doc/parathyroidSE.pdf">parathyroidSE vignette</a>, some samples were spread over multiple sequencing runs (i.e., the same sample being sequenced repeatedly) and therefore constitute <strong>technical replication</strong>, rather than biological replication.</li>
<li>We have previously seen that technical replicates can be considered to be distributed according to a Poisson distribution. One important property of Poisson random variables is that a sum of Poisson random variables still follow a Poisson distribution. Indeed, if <span class="math inline">\(X \sim Poi(\mu_X)\)</span> and <span class="math inline">\(Y \sim Poi(\mu_Y)\)</span>, then $ X + Y = Z Poi(_X + _Y)$.</li>
<li>For this reason, it is often suggested to sum technical replicates rather than, for example, averaging, which does not retain the Poisson property (try for yourself!). We’ll therefore first sum the technical replicates.</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>dupExps &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment[<span class="kw">duplicated</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment)])</span>
<span id="cb36-2"><a href="#cb36-2"></a>dupExps</span></code></pre></div>
<pre><code>## [1] &quot;SRX140511&quot; &quot;SRX140513&quot; &quot;SRX140523&quot; &quot;SRX140525&quot;</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>counts &lt;-<span class="st"> </span><span class="kw">assays</span>(se1)<span class="op">$</span>counts</span>
<span id="cb38-2"><a href="#cb38-2"></a>newCounts &lt;-<span class="st"> </span>counts</span>
<span id="cb38-3"><a href="#cb38-3"></a>cd &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)</span>
<span id="cb38-4"><a href="#cb38-4"></a><span class="cf">for</span>(ss <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dupExps)){</span>
<span id="cb38-5"><a href="#cb38-5"></a>  <span class="co"># check which samples are duplicates</span></span>
<span id="cb38-6"><a href="#cb38-6"></a>  relevantId &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment <span class="op">==</span><span class="st"> </span>dupExps[ss])</span>
<span id="cb38-7"><a href="#cb38-7"></a>  <span class="co"># sum counts</span></span>
<span id="cb38-8"><a href="#cb38-8"></a>  newCounts[,relevantId[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">rowSums</span>(counts[,relevantId])</span>
<span id="cb38-9"><a href="#cb38-9"></a>  <span class="co"># keep which columns / rows to remove.</span></span>
<span id="cb38-10"><a href="#cb38-10"></a>  <span class="cf">if</span>(ss <span class="op">==</span><span class="st"> </span><span class="dv">1</span>){</span>
<span id="cb38-11"><a href="#cb38-11"></a>    toRemove &lt;-<span class="st"> </span>relevantId[<span class="dv">2</span>]</span>
<span id="cb38-12"><a href="#cb38-12"></a>  } <span class="cf">else</span> {</span>
<span id="cb38-13"><a href="#cb38-13"></a>    toRemove &lt;-<span class="st"> </span><span class="kw">c</span>(toRemove, relevantId[<span class="dv">2</span>])</span>
<span id="cb38-14"><a href="#cb38-14"></a>  }</span>
<span id="cb38-15"><a href="#cb38-15"></a>}</span>
<span id="cb38-16"><a href="#cb38-16"></a></span>
<span id="cb38-17"><a href="#cb38-17"></a><span class="co"># remove after summing counts (otherwise IDs get mixed up)</span></span>
<span id="cb38-18"><a href="#cb38-18"></a>newCounts &lt;-<span class="st"> </span>newCounts[,<span class="op">-</span>toRemove]</span>
<span id="cb38-19"><a href="#cb38-19"></a>newCD &lt;-<span class="st"> </span>cd[<span class="op">-</span>toRemove,]</span>
<span id="cb38-20"><a href="#cb38-20"></a></span>
<span id="cb38-21"><a href="#cb38-21"></a><span class="co"># Create new SummarizedExperiment</span></span>
<span id="cb38-22"><a href="#cb38-22"></a>se &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="st">&quot;counts&quot;</span> =<span class="st"> </span>newCounts),</span>
<span id="cb38-23"><a href="#cb38-23"></a>                            <span class="dt">colData =</span> newCD,</span>
<span id="cb38-24"><a href="#cb38-24"></a>                            <span class="dt">metadata =</span> <span class="kw">metadata</span>(se1))</span>
<span id="cb38-25"><a href="#cb38-25"></a></span>
<span id="cb38-26"><a href="#cb38-26"></a>treatment &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>treatment</span>
<span id="cb38-27"><a href="#cb38-27"></a><span class="kw">table</span>(treatment)</span></code></pre></div>
<pre><code>## treatment
## Control     DPN     OHT 
##       7       8       8</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>time &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>time</span>
<span id="cb40-2"><a href="#cb40-2"></a><span class="kw">table</span>(time)</span></code></pre></div>
<pre><code>## time
## 24h 48h 
##  11  12</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>patient &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>patient</span>
<span id="cb42-2"><a href="#cb42-2"></a><span class="kw">table</span>(patient)</span></code></pre></div>
<pre><code>## patient
## 1 2 3 4 
## 6 6 6 5</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="kw">table</span>(patient, treatment, time) <span class="co"># agrees with paper.</span></span></code></pre></div>
<pre><code>## , , time = 24h
## 
##        treatment
## patient Control DPN OHT
##       1       1   1   1
##       2       1   1   1
##       3       1   1   1
##       4       0   1   1
## 
## , , time = 48h
## 
##        treatment
## patient Control DPN OHT
##       1       1   1   1
##       2       1   1   1
##       3       1   1   1
##       4       1   1   1</code></pre>
<ul>
<li>After summing the technical replicates and appropriately updating the sample information, we again create a <code>SummarizedExperiment</code> object, which is essentially a data container that contains all relevant information about your experiment. Please see the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">vignette</a> for more information on how to use this class.</li>
<li>By directly matching columns (samples) and rows (genes) to their relevant metadata, the <code>SummarizedExperiment</code> class avoids mistakes by mis-matching columns and rows with each other (provided you haven’t mismatched them when you creat the object).</li>
<li>The <code>SummarizedExperiment</code> class is modular and extendable, and extensions exist for example for the analysis of single-cell RNA-seq data.</li>
<li>Due to their convenient organization and widely supported usage within Bioconductor, we will typically work with such containers in the analysis of RNA-seq data.</li>
</ul>
</div>
<div id="data-exploration" class="section level2">
<h2><span class="header-section-number">1.3</span> Data exploration</h2>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a><span class="kw">suppressPackageStartupMessages</span>({</span>
<span id="cb46-2"><a href="#cb46-2"></a>  <span class="kw">library</span>(limma)</span>
<span id="cb46-3"><a href="#cb46-3"></a>  <span class="kw">library</span>(edgeR)</span>
<span id="cb46-4"><a href="#cb46-4"></a>})</span>
<span id="cb46-5"><a href="#cb46-5"></a></span>
<span id="cb46-6"><a href="#cb46-6"></a><span class="co"># library size distribution</span></span>
<span id="cb46-7"><a href="#cb46-7"></a><span class="kw">hist</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span>, <span class="dt">breaks=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>treatment)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>time)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-3.png" width="672" /></p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>patient)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-4.png" width="672" /></p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span><span class="kw">interaction</span>(treatment, time))</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-5.png" width="672" /></p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="co"># MDS plot</span></span>
<span id="cb51-2"><a href="#cb51-2"></a><span class="kw">plotMDS</span>(se, </span>
<span id="cb51-3"><a href="#cb51-3"></a>        <span class="dt">labels =</span> treatment, </span>
<span id="cb51-4"><a href="#cb51-4"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(patient))</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-5-6.png" width="672" /></p>
</div>
</div>
<div id="challenge-i-choice-of-modeling-assumptions" class="section level1">
<h1><span class="header-section-number">2</span> Challenge I: Choice of modeling assumptions</h1>
<p>When working with a GLM, as part of the choices of modeling assumptions, we need to pick an appropriate distribution for the expression counts. Below we perform some exploratory analyses to investigate.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1"></a>y &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[<span class="dv">1</span>,]</span>
<span id="cb52-2"><a href="#cb52-2"></a><span class="kw">hist</span>(y, <span class="dt">breaks =</span> <span class="dv">40</span>,</span>
<span id="cb52-3"><a href="#cb52-3"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Gene expression&quot;</span>,</span>
<span id="cb52-4"><a href="#cb52-4"></a>     <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="dt">yaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb52-5"><a href="#cb52-5"></a>     <span class="dt">main =</span> <span class="st">&quot;Data for the first gene&quot;</span>)</span>
<span id="cb52-6"><a href="#cb52-6"></a><span class="kw">axis</span>(<span class="dv">1</span>, <span class="dt">at =</span> <span class="kw">seq</span>(<span class="dv">200</span>, <span class="dv">1200</span>, <span class="dt">by=</span><span class="dv">200</span>))</span>
<span id="cb52-7"><a href="#cb52-7"></a><span class="kw">axis</span>(<span class="dv">2</span>, <span class="dt">at =</span> <span class="dv">0</span><span class="op">:</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="co"># Mean-variance trend within each experimental condition</span></span>
<span id="cb53-2"><a href="#cb53-2"></a>cont24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;Control&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb53-3"><a href="#cb53-3"></a>DPN24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;DPN&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb53-4"><a href="#cb53-4"></a>OHT24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;OHT&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb53-5"><a href="#cb53-5"></a>cont48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;Control&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb53-6"><a href="#cb53-6"></a>DPN48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;DPN&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb53-7"><a href="#cb53-7"></a>OHT48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;OHT&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb53-8"><a href="#cb53-8"></a>idList &lt;-<span class="st"> </span><span class="kw">list</span>(cont24ID, DPN24ID, OHT24ID, </span>
<span id="cb53-9"><a href="#cb53-9"></a>               cont48ID, DPN48ID, OHT48ID)</span>
<span id="cb53-10"><a href="#cb53-10"></a><span class="kw">names</span>(idList) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="kw">rep</span>(<span class="kw">levels</span>(treatment),<span class="dv">2</span>), <span class="kw">rep</span>(<span class="kw">levels</span>(time), <span class="dt">each=</span><span class="dv">3</span>))</span>
<span id="cb53-11"><a href="#cb53-11"></a></span>
<span id="cb53-12"><a href="#cb53-12"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb53-13"><a href="#cb53-13"></a><span class="cf">for</span>(kk <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(idList)){</span>
<span id="cb53-14"><a href="#cb53-14"></a>  <span class="co"># extract counts for each condition</span></span>
<span id="cb53-15"><a href="#cb53-15"></a>  curCounts &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[,idList[[kk]]]</span>
<span id="cb53-16"><a href="#cb53-16"></a>  <span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">rowMeans</span>(curCounts)<span class="op">+</span><span class="dv">1</span>,</span>
<span id="cb53-17"><a href="#cb53-17"></a>       <span class="dt">y =</span> <span class="kw">rowVars</span>(curCounts)<span class="op">+</span><span class="dv">1</span>,</span>
<span id="cb53-18"><a href="#cb53-18"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>,</span>
<span id="cb53-19"><a href="#cb53-19"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Mean&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance&quot;</span>,</span>
<span id="cb53-20"><a href="#cb53-20"></a>       <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>)</span>
<span id="cb53-21"><a href="#cb53-21"></a>  <span class="kw">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb53-22"><a href="#cb53-22"></a>  </span>
<span id="cb53-23"><a href="#cb53-23"></a>  <span class="kw">smoothScatter</span>(<span class="dt">x =</span> <span class="kw">log1p</span>(<span class="kw">rowMeans</span>(curCounts)),</span>
<span id="cb53-24"><a href="#cb53-24"></a>       <span class="dt">y =</span> <span class="kw">log1p</span>(<span class="kw">rowVars</span>(curCounts)),</span>
<span id="cb53-25"><a href="#cb53-25"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>,</span>
<span id="cb53-26"><a href="#cb53-26"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Mean&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance&quot;</span>)</span>
<span id="cb53-27"><a href="#cb53-27"></a>  <span class="kw">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb53-28"><a href="#cb53-28"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-7-2.png" width="672" /><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-7-3.png" width="672" /></p>
<ul>
<li>Having data on thousands of genes provides the opportunity to empirically assess the mean-variance relationship.</li>
<li>It is clear that the data is overdispersed with respect to the Poisson distribution. There also seems to be a quadratic trend of the variance as a function of the mean. This has motivated the <strong>negative binomial distribution as the most popular choice to model (bulk) RNA-seq gene expression data</strong>.</li>
</ul>
<hr />
<ul>
<li>The negative binomial distribution is also referred to as the Gamma-Poisson distribution as it can be formulated as such. Indeed, if <span class="math display">\[
 \lambda \sim \Gamma(\alpha, \beta) \\
 Y | \lambda \sim Poi(\lambda),
 \]</span> then this is equivalent to <span class="math display">\[ Y \sim NB(\mu = \alpha / \beta, \phi = 1/\alpha). \]</span></li>
<li>This can be shown analytically, but is considered out of the scope of this course. Below, we show it empirically using simulation.</li>
<li>This theoretical result has got some practical consequences. The Gamma-Poisson formulation makes it clear why we can sum technical replicates as the sum of Poisson random variables is again a Poisson random variable.</li>
<li>The Poisson statement can thus be considered as capturing technical variation, while the Gamma statement can be considered to capture biological variation, i.e., variation in the expression mean across biological replicates.</li>
</ul>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1"></a>alpha &lt;-<span class="st"> </span><span class="dv">20</span></span>
<span id="cb54-2"><a href="#cb54-2"></a>beta &lt;-<span class="st"> </span><span class="dv">10</span></span>
<span id="cb54-3"><a href="#cb54-3"></a>lambda &lt;-<span class="st"> </span><span class="kw">rgamma</span>(<span class="dt">n =</span> <span class="fl">1e5</span>, <span class="dt">shape =</span> alpha, <span class="dt">rate =</span> beta)</span>
<span id="cb54-4"><a href="#cb54-4"></a>y1 &lt;-<span class="st"> </span><span class="kw">rpois</span>(<span class="dt">n =</span> <span class="fl">1e5</span>, <span class="dt">lambda =</span> lambda)</span>
<span id="cb54-5"><a href="#cb54-5"></a></span>
<span id="cb54-6"><a href="#cb54-6"></a><span class="co"># note phi = 1 / size</span></span>
<span id="cb54-7"><a href="#cb54-7"></a>y2 &lt;-<span class="st"> </span><span class="kw">rnbinom</span>(<span class="dt">n=</span><span class="fl">1e5</span>, <span class="dt">mu=</span>alpha <span class="op">/</span><span class="st"> </span>(beta), <span class="dt">size=</span>alpha)</span>
<span id="cb54-8"><a href="#cb54-8"></a></span>
<span id="cb54-9"><a href="#cb54-9"></a></span>
<span id="cb54-10"><a href="#cb54-10"></a><span class="kw">plot</span>(<span class="kw">density</span>(y1))</span>
<span id="cb54-11"><a href="#cb54-11"></a><span class="kw">lines</span>(<span class="kw">density</span>(y2), <span class="dt">col=</span><span class="st">&quot;steelblue&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="challenge-ii-normalization" class="section level1">
<h1><span class="header-section-number">3</span> Challenge II: Normalization</h1>
<p>Normalization is necessary to correct for two source of technical variation</p>
<ul>
<li><strong>Differences in sequencing depth</strong> between samples.</li>
<li><strong>Differences in RNA population composition</strong> between samples.</li>
</ul>
</div>
<div id="challenge-iii-parameter-estimation-under-limited-information-setting" class="section level1">
<h1><span class="header-section-number">4</span> Challenge III: Parameter estimation (under limited information setting)</h1>
<p>Empirical Bayes, but also the factors we should incorporate into the mean model: show screenshot of paper that doesn’t include interaction effect. Discuss with students.</p>
</div>
<div id="challenge-iv-statistical-inference-across-many-genes" class="section level1">
<h1><span class="header-section-number">5</span> Challenge IV: Statistical inference across many genes</h1>
</div>
<div id="the-importance-of-reproducible-analysis" class="section level1">
<h1><span class="header-section-number">6</span> The importance of reproducible analysis</h1>
<p><a href="https://www.youtube.com/watch?v=8QJfNS7XXwA" class="uri">https://www.youtube.com/watch?v=8QJfNS7XXwA</a></p>
</div>

<div id="rmd-source-code">---
title: 'Sequencing: RNA-seq data intro'
author: "Koen Van den Berge"
date: "8/6/2021"
output: 
  pdf_document:
    toc: true
    number_sections: true
    latex_engine: xelatex
  html_document:
    toc: true
    toc_float: true
---

```{r functions, include=FALSE}
# A function for captioning and referencing images
fig <- local({
    i <- 0
    ref <- list()
    list(
        cap=function(refName, text) {
            i <<- i + 1
            ref[[refName]] <<- i
            paste("Figure ", i, ": ", text, sep="")
        },
        ref=function(refName) {
            ref[[refName]]
        })
})
``` 

```{r, echo=FALSE, eval=TRUE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(ggplot2)
})
if(!"BiocManager" %in% installed.packages()[,1]){
  install.packages("BiocManager")
}
if(!"limma" %in% installed.packages()[,1]){
  BiocManager::install("limma")
}
if(!"edgeR" %in% installed.packages()[,1]){
  BiocManager::install("edgeR")
}
```

In this lecture we will start working with a real bulk RNA-seq dataset from [Haglund *et al.* (2012)](https://academic.oup.com/jcem/article/97/12/4631/2536573). After importing the data, we will be working our way through four major challenges which, together, will form a full RNA-seq differential expression (DE) analysis pipeline where the result of our analysis will be a(n ordered) list of genes that we find to be differently expressed between our conditions of interest. The four main challenges we will look into are 

 - Choice of modeling assumptions (distribution).
 - Normalization.
 - Parameter estimation under a limited information setting.
 - Statistical inference under high dimensionality (many genes).

# Experimental design, data import and data exploration

## Experimental design

Let's try to work out the experimental design using the following paragraph from the Methods section of the paper.

```{r, echo=FALSE, fig.cap=paste("Figure: The data output revolution of sequencing machines. Image from Illumina documentation.")}
# All defaults
include_graphics("./images_sequencing/expDesign_para.png")
```

## Data import and exploration

We will be importing the dataset using the [parathyroidSE](https://www.bioconductor.org/packages/release/data/experiment/html/parathyroidSE.html) data package from [Bioconductor](https://bioconductor.org/).

```{r, warning=FALSE, eval=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}
if(!"SummarizedExperiment" %in% installed.packages()[,1]){
  BiocManager::install("SummarizedExperiment")
} 
# install package if not installed.
if(!"parathyroidSE" %in% installed.packages()[,1]) BiocManager::install("parathyroidSE")

suppressPackageStartupMessages({
  library(parathyroidSE)
  library(SummarizedExperiment)
})

# import data
data("parathyroidGenesSE", package="parathyroidSE")
# rename for convenience
se1 <- parathyroidGenesSE
rm(parathyroidGenesSE)

# three treatments
treatment1 <- colData(se1)$treatment
table(treatment1)
# two timepoints
time1 <- colData(se1)$time
table(time1)
# four donor patients
patient1 <- colData(se1)$patient
table(patient1)

table(patient1, treatment1, time1)
```

- We observe that the number of samples that we are observing here is larger than what is described in the paper. As also described in the [parathyroidSE vignette](https://www.bioconductor.org/packages/release/data/experiment/vignettes/parathyroidSE/inst/doc/parathyroidSE.pdf), some samples were spread over multiple sequencing runs (i.e., the same sample being sequenced repeatedly) and therefore constitute **technical replication**, rather than biological replication.
- We have previously seen that technical replicates can be considered to be distributed according to a Poisson distribution. One important property of Poisson random variables is that a sum of Poisson random variables still follow a Poisson distribution. Indeed, if $X \sim Poi(\mu_X)$ and $Y \sim Poi(\mu_Y)$, then $ X + Y = Z \sim Poi(\mu_X + \mu_Y)$.
- For this reason, it is often suggested to sum technical replicates rather than, for example, averaging, which does not retain the Poisson property (try for yourself!). We'll therefore first sum the technical replicates.

```{r}
dupExps <- as.character(colData(se1)$experiment[duplicated(colData(se1)$experiment)])
dupExps

counts <- assays(se1)$counts
newCounts <- counts
cd <- colData(se1)
for(ss in 1:length(dupExps)){
  # check which samples are duplicates
  relevantId <- which(colData(se1)$experiment == dupExps[ss])
  # sum counts
  newCounts[,relevantId[1]] <- rowSums(counts[,relevantId])
  # keep which columns / rows to remove.
  if(ss == 1){
    toRemove <- relevantId[2]
  } else {
    toRemove <- c(toRemove, relevantId[2])
  }
}

# remove after summing counts (otherwise IDs get mixed up)
newCounts <- newCounts[,-toRemove]
newCD <- cd[-toRemove,]

# Create new SummarizedExperiment
se <- SummarizedExperiment(assays = list("counts" = newCounts),
                            colData = newCD,
                            metadata = metadata(se1))

treatment <- colData(se)$treatment
table(treatment)
time <- colData(se)$time
table(time)
patient <- colData(se)$patient
table(patient)

table(patient, treatment, time) # agrees with paper.
```

 - After summing the technical replicates and appropriately updating the sample information, we again create a `SummarizedExperiment` object, which is essentially a data container that contains all relevant information about your experiment. Please see the [vignette](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) for more information on how to use this class.
 - By directly matching columns (samples) and rows (genes) to their relevant metadata, the `SummarizedExperiment` class avoids mistakes by mis-matching columns and rows with each other (provided you haven't mismatched them when you creat the object).
 - The `SummarizedExperiment` class is modular and extendable, and extensions exist for example for the analysis of single-cell RNA-seq data.
 - Due to their convenient organization and widely supported usage within Bioconductor, we will typically work with such containers in the analysis of RNA-seq data.

## Data exploration

```{r}

suppressPackageStartupMessages({
  library(limma)
  library(edgeR)
})

# library size distribution
hist(colSums(assays(se)$counts)/1e6, breaks=10)
boxplot(colSums(assays(se)$counts)/1e6 ~ treatment)
boxplot(colSums(assays(se)$counts)/1e6 ~ time)
boxplot(colSums(assays(se)$counts)/1e6 ~ patient)
boxplot(colSums(assays(se)$counts)/1e6 ~ interaction(treatment, time))

# MDS plot
plotMDS(se, 
        labels = treatment, 
        col=as.numeric(patient))
```

```{r, eval=FALSE, echo=FALSE}
# Explain concept of MDS: preserve Euclidean distance from high to low dim.
```

# Challenge I: Choice of modeling assumptions

When working with a GLM, as part of the choices of modeling assumptions, we need to pick an appropriate distribution for the expression counts. Below we perform some exploratory analyses to investigate.

```{r, eval=TRUE}
y <- assays(se)$counts[1,]
hist(y, breaks = 40,
     xlab = "Gene expression",
     xaxt = "n", yaxt = "n",
     main = "Data for the first gene")
axis(1, at = seq(200, 1200, by=200))
axis(2, at = 0:3)

# Mean-variance trend within each experimental condition
cont24ID <- which(treatment == "Control" & time == "24h")
DPN24ID <- which(treatment == "DPN" & time == "24h")
OHT24ID <- which(treatment == "OHT" & time == "24h")
cont48ID <- which(treatment == "Control" & time == "48h")
DPN48ID <- which(treatment == "DPN" & time == "48h")
OHT48ID <- which(treatment == "OHT" & time == "48h")
idList <- list(cont24ID, DPN24ID, OHT24ID, 
               cont48ID, DPN48ID, OHT48ID)
names(idList) <- paste0(rep(levels(treatment),2), rep(levels(time), each=3))

par(mfrow=c(3,2), mar=c(2,2,2,1))
for(kk in 1:length(idList)){
  # extract counts for each condition
  curCounts <- assays(se)$counts[,idList[[kk]]]
  plot(x = rowMeans(curCounts)+1,
       y = rowVars(curCounts)+1,
       pch = 16, cex=1/2,
       xlab = "Mean", ylab="Variance",
       log="xy")
  abline(0,1, col="red")
  
  smoothScatter(x = log1p(rowMeans(curCounts)),
       y = log1p(rowVars(curCounts)),
       pch = 16, cex=1/2,
       xlab = "Mean", ylab="Variance")
  abline(0,1, col="red")
}
```


```{r, eval=FALSE, echo=FALSE}
# Ask students why I separate according to time and treatment, and not patient?
# Relate to model definition of a GLM. (Note caveat of library size N_i)
```

 - Having data on thousands of genes provides the opportunity to empirically assess the mean-variance relationship.
 - It is clear that the data is overdispersed with respect to the Poisson distribution. There also seems to be a quadratic trend of the variance as a function of the mean. This has motivated the **negative binomial distribution as the most popular choice to model (bulk) RNA-seq gene expression data**.
 
 ---
 
 - The negative binomial distribution is also referred to as the Gamma-Poisson distribution as it can be formulated as such. Indeed, if 
 $$
 \lambda \sim \Gamma(\alpha, \beta) \\
 Y | \lambda \sim Poi(\lambda),
 $$
 then this is equivalent to
 $$ Y \sim NB(\mu = \alpha / \beta, \phi = 1/\alpha). $$
 - This can be shown analytically, but is considered out of the scope of this course. Below, we show it empirically using simulation.
 - This theoretical result has got some practical consequences. The Gamma-Poisson formulation makes it clear why we can sum technical replicates as the sum of Poisson random variables is again a Poisson random variable. 
 - The Poisson statement can thus be considered as capturing technical variation, while the Gamma statement can be considered to capture biological variation, i.e., variation in the expression mean across biological replicates.
 

```{r}
alpha <- 20
beta <- 10
lambda <- rgamma(n = 1e5, shape = alpha, rate = beta)
y1 <- rpois(n = 1e5, lambda = lambda)

# note phi = 1 / size
y2 <- rnbinom(n=1e5, mu=alpha / (beta), size=alpha)


plot(density(y1))
lines(density(y2), col="steelblue")
```

# Challenge II: Normalization

Normalization is necessary to correct for two source of technical variation

 - **Differences in sequencing depth** between samples.
 - **Differences in RNA population composition** between samples.



# Challenge III: Parameter estimation (under limited information setting)

Empirical Bayes, but also the factors we should incorporate into the mean model: show screenshot of paper that doesn't include interaction effect. Discuss with students.

# Challenge IV: Statistical inference across many genes










# The importance of reproducible analysis

https://www.youtube.com/watch?v=8QJfNS7XXwA



</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("sequencing_rnaseqIntro.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
