---
title: "Introduction to sequencing"
author: "Koen Van den Berge"
date: "6/25/2021"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r functions, include=FALSE}
# A function for captioning and referencing images
fig <- local({
    i <- 0
    ref <- list()
    list(
        cap=function(refName, text) {
            i <<- i + 1
            ref[[refName]] <<- i
            paste("Figure ", i, ": ", text, sep="")
        },
        ref=function(refName) {
            ref[[refName]]
        })
})
``` 

```{r, echo=FALSE, eval=TRUE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(ggplot2)
})
```

# The study of gene expression

The first part of this course has focussed on proteomics, studying the concentration of proteins in biological samples. We have seen that identification of proteins and measuring their respective concentrations are extremely challenging, leading to many technological and statistical challenges in order to interpret these data.
In the second part of the course, we will now focus on measuring gene expression, i.e., measuring the concentration of mRNA molecules, that may eventually be translated into proteins, but may also have functions on their own.

# Sequencing technology

 - Measuring mRNA molecules typically happens through sequencing.
 - The technology continues to evolve at an incredible speed. The data output of so-called `next generation sequencing' machines has more than doubled each year! Simultaneously, the cost of sequencing (in terms of $ per Gigabase) is dropping. Each year, we're able to sequence more for less money.
 - This tremendous technological revolution has revolutionized biology, and genomic sequencing is now a core component of the modern-day biologist's toolkit.
 - The large majority of sequencing data is generated using sequencing-by-synthesis using machines produced by the company Illumina. While new players such as Pacific Biosciences and Oxford Nanopore have entered the scene, these are typically most useful for DNA sequencing rather than gene expression studies, owing to their capability of sequencing long molecules.

```{r, echo=FALSE, fig.cap=paste("Figure: The data output revolution of sequencing machines. Image from Illumina documentation.")}
# All defaults
include_graphics("./images_sequencing/seqTechnology_throughput.png")
```

## The sequencing workflow

Library preparation steps:

1. First, the biological samples of interest are collected. Owing to the maturity of different protocols for sequencing, several types of biological input samples are amenable to sequencing, such as frozen tissues or FFPE-preserved samples.
2. The mRNA molecules from our sample are captured. If the tissue sample used as input is still intact, this also involves cell lysis in order to release the mRNA molecules from within the cells. The mRNA molecules are mosten often captured using (i) polyA-capture to select for polyadenylated RNA, or (ii) ribosomal deplation, where ribosomal and transfer RNAs are depleted. In the case of `targeted sequencing', where relevant molecules are of main interest (e.g., a gene panel), these targets can be specifically targeted in this step.
3. Fragmentation of captured molecules. The captured molecules are fragmented, either chemically or mechanically. The appropriate size of fragments depends on the sequencing machines (often in the range of 300 - 500bp).
4. Reverse transcription. Current dominant sequencing machines only sequence double-stranded DNA molecules. Therefore, in order to measure single-stranded mRNA, we must first isolate the mRNA molecules from the biological sample(s) of interest, and reverse transcribe these molecules to a double-stranded complementary (cDNA) molecule.
5. Addition of adapters. Adapters are oligonucleotides that are platform-specific sequences for fragment recognition by the sequencing machines. These are added either to the 3' or 5' end of the cDNA molecules or used as primers in the reverse transcription reaction. The final cDNA library consists of cDNA inserts flanked by an adapter sequence on each end. 
6. PCR amplification. To increase concentration, several PCR reactions are performed.
7. Loading the amplified cDNA library on the sequencing machine. Find out how sequencing-by-synthesis works through [this video](https://www.youtube.com/watch?v=fCd6B5HRaZ8). Note that the video shows paired-end sequencing, where a number of basepairs are sequenced at each end of the fragment. All previous steps together are described as `sample prep' in that video.

```{r, echo=FALSE, fig.cap=paste("Figure: The sequencing workflow. Image adapted from Van den Berge et al. (2019).")}
include_graphics("./images_sequencing/seqWorkflow1_clean.png")
```

Note that several variants of library preparation protocols are available. The most important ones are:
 - Single-end vs paired-end sequecing: In single-end sequencing, a single end (3' or 5') of the cDNA fragment is sequenced. In paired-end sequencing, both ends are sequenced.
 - Strand specific protocols: Some library preparation protocols allow measuring strand specificity, where the strand information of each read can be preserved.

## The sequencing output files

- The typical output of a sequencing machine we will be working with are FASTA or FASTQ files for each sample. Each of these files are several gigbases large and contain millions of reads. For paired-end sequencing, there are two files for each sample, one for each end of the sequenced fragments.
 - The difference between a FASTA file and a FASTQ file, is that while FASTA files only store the results of base calls (sequences), FASTQ files also store the quality score of each base call (i.e., each called nucleotide), which can be useful in downstream analyses such as mapping or variant calling.
 - A fASTQ file contains four lines for each sequenced read:
    1. Sequence identifier line, starting with @.
    2. The sequence.
    3. Another sequence identifier line, now starting with +.
    4. Quality scores.

```{r, echo=FALSE, fig.cap=paste("Figure: One read in a FASTQ file. Slide courtesy by Charlotte Soneson.")}
include_graphics("./images_sequencing/fastqLine.png")
```

As you'll have noticed, the base call quality scores are encoded as ASCII characters for efficient storage. These ASCII characters can be convered into integers called Phred scores, which are logarithmically related to the probability of an erroneous base call.

# Preprocessing of sequencing data

After sequencing, we typically do a quality control (QC) check to verify the quality of the samples. During QC check, aberrant samples due to e.g. degraded mRNA can be detected.

The sequencing reads on their own contain a lot of information, but are most useful if we would be able to assign sequencing reads to genes, i.e., for each sequencing read we will try to derive the (set of) gene(s) that could have plausibly produced the fragment through the process of gene expression. This process is called 'mapping'. 

```{r, echo=FALSE, fig.cap=paste("Figure: An updated sequencing workflow. Image adapted from Van den Berge et al. (2019).")}
include_graphics("./images_sequencing/seqWorkflow2.png")
```

## Quality control

During quality control, diagnostic plots are created for each sample in order to determine its quality. The most popular QC tool for bulk RNA-seq data is [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). If many samples are sequenced, then [MultiQC](https://multiqc.info/) can be used to aggregate the QC checks across samples in a conveniently organized overview.

The FastQC website provides interesting example reports for us to look at and compare against. Here are example reports of [high-quality Illumina data](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html) and [low-quality Illumina](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html) data.

## Mapping

Note that we are typically not able to assign each individual read uniquely to one specific gene; some reads cannot be unambiguously mapped and are compatible with multiple genes. These reads are said to be `multi-mapping'.

# Terminology

- An oligo is short for oligonucleotide, which is a short sequence of nucleotides.
