<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge" />


<title>Sequencing: RNA-seq data intro</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./docs/intro.pdf">1.1 Position</a>
    </li>
    <li>
      <a href="recapGeneralLinearModel.html">1.2 Recap linear model</a>
    </li>
    <li>
      <a href="./docs/martens_proteomics_bioinformatics.pdf">2.1 PDA - identification</a>
    </li>
    <li>
      <a href="pda_quantification_preprocessing.html">2.2 PDA- Quantification - Preprocessing</a>
    </li>
    <li>
      <a href="pda_robustSummarisation_peptideModels.html">2.2 PDA - Wrap-up Preprocessing - Peptide-Level-Models</a>
    </li>
    <li>
      <a href="pda_quantification_inference.html">2.3 PDA - Quantification - Differential Abundance Analysis</a>
    </li>
    <li>
      <a href="pda_blocking_wrapup.html">2.3 PDA - Wrap-up DA - Blocking</a>
    </li>
    <li>
      <a href="technicalDetailsProteomics.html">2.4. PDA - DA Technical Details</a>
    </li>
    <li>
      <a href="./docs/stagewiseTesting.pdf">2.5 PDA - Stage-wise testing</a>
    </li>
    <li>
      <a href="sequencing_intro.html">3.1 SDA - Intro to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">3.2 SDA - Working with count data</a>
    </li>
    <li>
      <a href="sequencing_rnaseqIntro.html">3.3 SDA - Intro to RNA-seq analysis</a>
    </li>
    <li>
      <a href="sequencing_technicalDE.html">3.4 SDA - Technical topics in DE analysis</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multipleRegression_KPNA2.html">1 Intro KPNA2</a>
    </li>
    <li>
      <a href="https://www.compomics.com/bioinformatics-for-proteomics/">2.1 Identification</a>
    </li>
    <li>
      <a href="pda_tutorialPreprocessing.html">2.2 DA - Preprocessing</a>
    </li>
    <li>
      <a href="pda_tutorialDesign.html">2.3 DA - Design</a>
    </li>
    <li>
      <a href="sequencing_preprocessing.sh">3.1 Preprocessing</a>
    </li>
    <li>
      <a href="sequencing_lab_oneGene.html">3.2 GLM for one gene</a>
    </li>
    <li>
      <a href="sequencing_hw.html">3.3 Bulk RNA-seq homework</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Sequencing: RNA-seq data intro</h1>
<h4 class="author">Koen Van den Berge</h4>
<h4 class="date">Last compiled on 16 November, 2021</h4>

</div>


<p>In this lecture we will start working with a real bulk RNA-seq dataset from <a href="https://academic.oup.com/jcem/article/97/12/4631/2536573">Haglund <em>et al.</em> (2012)</a>. After importing the data, we will be working our way through four major challenges which, together, will form a full RNA-seq differential expression (DE) analysis pipeline where the result of our analysis will be a(n ordered) list of genes that we find to be differently expressed between our conditions of interest. The four main challenges we will look into are</p>
<ul>
<li>Choice of modeling assumptions (distribution).</li>
<li>Normalization.</li>
<li>Parameter estimation under a limited information setting.</li>
<li>Statistical inference under high dimensionality (many genes).</li>
</ul>
<div id="experimental-design-data-import-and-data-exploration" class="section level1">
<h1><span class="header-section-number">1</span> Experimental design, data import and data exploration</h1>
<div id="experimental-design" class="section level2">
<h2><span class="header-section-number">1.1</span> Experimental design</h2>
<p>Let’s try to work out the experimental design using the following paragraph from the Methods section of the paper.</p>
<div class="figure">
<img src="images_sequencing/expDesign_para.png" alt="Figure: A paragraph from the Methods section." width="689" />
<p class="caption">
Figure: A paragraph from the Methods section.
</p>
</div>
</div>
<div id="data-import-and-exploration" class="section level2">
<h2><span class="header-section-number">1.2</span> Data import and exploration</h2>
<p>We will be importing the dataset using the <a href="https://www.bioconductor.org/packages/release/data/experiment/html/parathyroidSE.html">parathyroidSE</a> data package from <a href="https://bioconductor.org/">Bioconductor</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)){</span>
<span id="cb1-2"><a href="#cb1-2"></a>  <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a>}</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;SummarizedExperiment&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]){</span>
<span id="cb1-5"><a href="#cb1-5"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SummarizedExperiment&quot;</span>)</span>
<span id="cb1-6"><a href="#cb1-6"></a>} </span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co"># install package if not installed.</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="cf">if</span>(<span class="op">!</span><span class="st">&quot;parathyroidSE&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]) BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;parathyroidSE&quot;</span>)</span>
<span id="cb1-9"><a href="#cb1-9"></a></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="kw">suppressPackageStartupMessages</span>({</span>
<span id="cb1-11"><a href="#cb1-11"></a>  <span class="kw">library</span>(parathyroidSE)</span>
<span id="cb1-12"><a href="#cb1-12"></a>  <span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb1-13"><a href="#cb1-13"></a>})</span>
<span id="cb1-14"><a href="#cb1-14"></a></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="co"># import data</span></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="kw">data</span>(<span class="st">&quot;parathyroidGenesSE&quot;</span>, <span class="dt">package=</span><span class="st">&quot;parathyroidSE&quot;</span>)</span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="co"># rename for convenience</span></span>
<span id="cb1-18"><a href="#cb1-18"></a>se1 &lt;-<span class="st"> </span>parathyroidGenesSE</span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="kw">rm</span>(parathyroidGenesSE)</span>
<span id="cb1-20"><a href="#cb1-20"></a></span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="co"># three treatments</span></span>
<span id="cb1-22"><a href="#cb1-22"></a>treatment1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>treatment</span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="kw">table</span>(treatment1)</span></code></pre></div>
<pre><code>## treatment1
## Control     DPN     OHT 
##       7      10      10</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># two timepoints</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>time1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>time</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="kw">table</span>(time1)</span></code></pre></div>
<pre><code>## time1
## 24h 48h 
##  13  14</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># four donor patients</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>patient1 &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)<span class="op">$</span>patient</span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="kw">table</span>(patient1)</span></code></pre></div>
<pre><code>## patient1
## 1 2 3 4 
## 6 8 6 7</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">table</span>(patient1, treatment1, time1)</span></code></pre></div>
<pre><code>## , , time1 = 24h
## 
##         treatment1
## patient1 Control DPN OHT
##        1       1   1   1
##        2       1   2   2
##        3       1   1   1
##        4       0   1   1
## 
## , , time1 = 48h
## 
##         treatment1
## patient1 Control DPN OHT
##        1       1   1   1
##        2       1   1   1
##        3       1   1   1
##        4       1   2   2</code></pre>
<ul>
<li>We observe that the number of samples that we are observing here is larger than what is described in the paper. As also described in the <a href="https://www.bioconductor.org/packages/release/data/experiment/vignettes/parathyroidSE/inst/doc/parathyroidSE.pdf">parathyroidSE vignette</a>, some samples were spread over multiple sequencing runs (i.e., the same sample being sequenced repeatedly) and therefore constitute <strong>technical replication</strong>, rather than biological replication.</li>
<li>We have previously seen that technical replicates can be considered to be distributed according to a Poisson distribution. One important property of Poisson random variables is that a sum of Poisson random variables still follow a Poisson distribution. Indeed, if <span class="math inline">\(X \sim Poi(\mu_X)\)</span> and <span class="math inline">\(Y \sim Poi(\mu_Y)\)</span>, then <span class="math inline">\(X + Y = Z \sim Poi(\mu_X + \mu_Y)\)</span>.</li>
<li>For this reason, it is often suggested to sum technical replicates rather than, for example, averaging, which does not retain the Poisson property (try for yourself!). We’ll therefore first sum the technical replicates.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>dupExps &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment[<span class="kw">duplicated</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment)])</span>
<span id="cb9-2"><a href="#cb9-2"></a>dupExps</span></code></pre></div>
<pre><code>## [1] &quot;SRX140511&quot; &quot;SRX140513&quot; &quot;SRX140523&quot; &quot;SRX140525&quot;</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>counts &lt;-<span class="st"> </span><span class="kw">assays</span>(se1)<span class="op">$</span>counts</span>
<span id="cb11-2"><a href="#cb11-2"></a>newCounts &lt;-<span class="st"> </span>counts</span>
<span id="cb11-3"><a href="#cb11-3"></a>cd &lt;-<span class="st"> </span><span class="kw">colData</span>(se1)</span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="cf">for</span>(ss <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dupExps)){</span>
<span id="cb11-5"><a href="#cb11-5"></a>  <span class="co"># check which samples are duplicates</span></span>
<span id="cb11-6"><a href="#cb11-6"></a>  relevantId &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colData</span>(se1)<span class="op">$</span>experiment <span class="op">==</span><span class="st"> </span>dupExps[ss])</span>
<span id="cb11-7"><a href="#cb11-7"></a>  <span class="co"># sum counts</span></span>
<span id="cb11-8"><a href="#cb11-8"></a>  newCounts[,relevantId[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">rowSums</span>(counts[,relevantId])</span>
<span id="cb11-9"><a href="#cb11-9"></a>  <span class="co"># keep which columns / rows to remove.</span></span>
<span id="cb11-10"><a href="#cb11-10"></a>  <span class="cf">if</span>(ss <span class="op">==</span><span class="st"> </span><span class="dv">1</span>){</span>
<span id="cb11-11"><a href="#cb11-11"></a>    toRemove &lt;-<span class="st"> </span>relevantId[<span class="dv">2</span>]</span>
<span id="cb11-12"><a href="#cb11-12"></a>  } <span class="cf">else</span> {</span>
<span id="cb11-13"><a href="#cb11-13"></a>    toRemove &lt;-<span class="st"> </span><span class="kw">c</span>(toRemove, relevantId[<span class="dv">2</span>])</span>
<span id="cb11-14"><a href="#cb11-14"></a>  }</span>
<span id="cb11-15"><a href="#cb11-15"></a>}</span>
<span id="cb11-16"><a href="#cb11-16"></a></span>
<span id="cb11-17"><a href="#cb11-17"></a><span class="co"># remove after summing counts (otherwise IDs get mixed up)</span></span>
<span id="cb11-18"><a href="#cb11-18"></a>newCounts &lt;-<span class="st"> </span>newCounts[,<span class="op">-</span>toRemove]</span>
<span id="cb11-19"><a href="#cb11-19"></a>newCD &lt;-<span class="st"> </span>cd[<span class="op">-</span>toRemove,]</span>
<span id="cb11-20"><a href="#cb11-20"></a></span>
<span id="cb11-21"><a href="#cb11-21"></a><span class="co"># Create new SummarizedExperiment</span></span>
<span id="cb11-22"><a href="#cb11-22"></a>se &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="st">&quot;counts&quot;</span> =<span class="st"> </span>newCounts),</span>
<span id="cb11-23"><a href="#cb11-23"></a>                            <span class="dt">colData =</span> newCD,</span>
<span id="cb11-24"><a href="#cb11-24"></a>                            <span class="dt">metadata =</span> <span class="kw">metadata</span>(se1))</span>
<span id="cb11-25"><a href="#cb11-25"></a></span>
<span id="cb11-26"><a href="#cb11-26"></a>treatment &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>treatment</span>
<span id="cb11-27"><a href="#cb11-27"></a><span class="kw">table</span>(treatment)</span></code></pre></div>
<pre><code>## treatment
## Control     DPN     OHT 
##       7       8       8</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>time &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>time</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="kw">table</span>(time)</span></code></pre></div>
<pre><code>## time
## 24h 48h 
##  11  12</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>patient &lt;-<span class="st"> </span><span class="kw">colData</span>(se)<span class="op">$</span>patient</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="kw">table</span>(patient)</span></code></pre></div>
<pre><code>## patient
## 1 2 3 4 
## 6 6 6 5</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="kw">table</span>(patient, treatment, time) <span class="co"># agrees with paper.</span></span></code></pre></div>
<pre><code>## , , time = 24h
## 
##        treatment
## patient Control DPN OHT
##       1       1   1   1
##       2       1   1   1
##       3       1   1   1
##       4       0   1   1
## 
## , , time = 48h
## 
##        treatment
## patient Control DPN OHT
##       1       1   1   1
##       2       1   1   1
##       3       1   1   1
##       4       1   1   1</code></pre>
<ul>
<li>After summing the technical replicates and appropriately updating the sample information, we again create a <code>SummarizedExperiment</code> object, which is essentially a data container that contains all relevant information about your experiment. Please see the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">vignette</a> for more information on how to use this class.</li>
<li>By directly matching columns (samples) and rows (genes) to their relevant metadata, the <code>SummarizedExperiment</code> class avoids mistakes by mis-matching columns and rows with each other (provided you haven’t mismatched them when you create the object).</li>
<li>The <code>SummarizedExperiment</code> class is modular and extendable, and extensions exist for example for the analysis of single-cell RNA-seq data, i.e., the <code>SingleCellExperiment</code> class.</li>
<li>Due to their convenient organization and widely supported usage within Bioconductor, we will typically work with such classes in the analysis of RNA-seq data.</li>
</ul>
</div>
<div id="independent-filtering" class="section level2">
<h2><span class="header-section-number">1.3</span> Independent filtering</h2>
<p>Independent filtering is a strategy to remove features (in this case, genes) prior to the analysis. Removal of these features may lower the multiple testing correction for other genes that pass the filter. We try to remove genes that have a low power to be found statistically significant, and/or that are biologically less relevant. A common filtering strategy is to remove genes with a generally low expression, as low counts have lower relative uncertainty (hence lower statistical power), and may be considered biologically less relevant.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="kw">suppressPackageStartupMessages</span>({</span>
<span id="cb19-2"><a href="#cb19-2"></a>  <span class="kw">library</span>(limma)</span>
<span id="cb19-3"><a href="#cb19-3"></a>  <span class="kw">library</span>(edgeR)</span>
<span id="cb19-4"><a href="#cb19-4"></a>})</span>
<span id="cb19-5"><a href="#cb19-5"></a></span>
<span id="cb19-6"><a href="#cb19-6"></a>keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">cpm</span>(se) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>) <span class="op">&gt;=</span><span class="st"> </span><span class="dv">3</span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="kw">table</span>(keep)</span></code></pre></div>
<pre><code>## keep
## FALSE  TRUE 
## 47837 15356</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>se &lt;-<span class="st"> </span>se[keep,]</span></code></pre></div>
</div>
<div id="data-exploration" class="section level2">
<h2><span class="header-section-number">1.4</span> Data exploration</h2>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co"># library size distribution</span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="kw">hist</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span>, <span class="dt">breaks=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>treatment)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-2.png" width="672" /></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>time)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-3.png" width="672" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span>patient)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-4.png" width="672" /></p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="kw">boxplot</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts)<span class="op">/</span><span class="fl">1e6</span> <span class="op">~</span><span class="st"> </span><span class="kw">interaction</span>(treatment, time))</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-5.png" width="672" /></p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># MDS plot</span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="kw">plotMDS</span>(se, </span>
<span id="cb27-3"><a href="#cb27-3"></a>        <span class="dt">labels =</span> treatment, </span>
<span id="cb27-4"><a href="#cb27-4"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(patient))</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-6.png" width="672" /></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="co">## hard to see influence of experimental factors due to large between-patient variation</span></span>
<span id="cb28-2"><a href="#cb28-2"></a><span class="co">## we could also make an MDS plot per patient to take a look.</span></span>
<span id="cb28-3"><a href="#cb28-3"></a><span class="cf">for</span>(kk <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>){</span>
<span id="cb28-4"><a href="#cb28-4"></a>  id &lt;-<span class="st"> </span><span class="kw">which</span>(patient <span class="op">==</span><span class="st"> </span>kk)</span>
<span id="cb28-5"><a href="#cb28-5"></a>  <span class="kw">plotMDS</span>(se[,id], </span>
<span id="cb28-6"><a href="#cb28-6"></a>        <span class="dt">labels =</span> <span class="kw">paste0</span>(treatment[id],<span class="st">&quot;_&quot;</span>,time[id]), </span>
<span id="cb28-7"><a href="#cb28-7"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(time[id]))</span>
<span id="cb28-8"><a href="#cb28-8"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-7.png" width="672" /><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-8.png" width="672" /><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-9.png" width="672" /><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-6-10.png" width="672" /></p>
<p>Observations based on MDS plot:</p>
<ul>
<li>There is a very large between-patient variability, which is the major source of variation in this dataset. The samples from each patient cluster together tightly.</li>
<li>Within patient, time consistently explains more variation than the treatments.</li>
<li>Relative to patients and time, the treatment seems to have a fairly small effect.</li>
</ul>
</div>
</div>
<div id="challenge-i-choice-of-modeling-assumptions" class="section level1">
<h1><span class="header-section-number">2</span> Challenge I: Choice of modeling assumptions</h1>
<p>When working with a GLM, as part of the choices of modeling assumptions, we need to pick an appropriate distribution for the expression counts. Below we perform some exploratory analyses to investigate.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>y &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[<span class="dv">1</span>,]</span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="kw">hist</span>(y, <span class="dt">breaks =</span> <span class="dv">40</span>,</span>
<span id="cb29-3"><a href="#cb29-3"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Gene expression&quot;</span>,</span>
<span id="cb29-4"><a href="#cb29-4"></a>     <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="dt">yaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb29-5"><a href="#cb29-5"></a>     <span class="dt">main =</span> <span class="st">&quot;Data for the first gene&quot;</span>)</span>
<span id="cb29-6"><a href="#cb29-6"></a><span class="kw">axis</span>(<span class="dv">1</span>, <span class="dt">at =</span> <span class="kw">seq</span>(<span class="dv">200</span>, <span class="dv">1200</span>, <span class="dt">by=</span><span class="dv">200</span>))</span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="kw">axis</span>(<span class="dv">2</span>, <span class="dt">at =</span> <span class="dv">0</span><span class="op">:</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co"># Mean-variance trend within each experimental condition</span></span>
<span id="cb30-2"><a href="#cb30-2"></a>cont24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;Control&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb30-3"><a href="#cb30-3"></a>DPN24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;DPN&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb30-4"><a href="#cb30-4"></a>OHT24ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;OHT&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;24h&quot;</span>)</span>
<span id="cb30-5"><a href="#cb30-5"></a>cont48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;Control&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb30-6"><a href="#cb30-6"></a>DPN48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;DPN&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb30-7"><a href="#cb30-7"></a>OHT48ID &lt;-<span class="st"> </span><span class="kw">which</span>(treatment <span class="op">==</span><span class="st"> &quot;OHT&quot;</span> <span class="op">&amp;</span><span class="st"> </span>time <span class="op">==</span><span class="st"> &quot;48h&quot;</span>)</span>
<span id="cb30-8"><a href="#cb30-8"></a>idList &lt;-<span class="st"> </span><span class="kw">list</span>(cont24ID, DPN24ID, OHT24ID, </span>
<span id="cb30-9"><a href="#cb30-9"></a>               cont48ID, DPN48ID, OHT48ID)</span>
<span id="cb30-10"><a href="#cb30-10"></a><span class="kw">names</span>(idList) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="kw">rep</span>(<span class="kw">levels</span>(treatment),<span class="dv">2</span>), <span class="kw">rep</span>(<span class="kw">levels</span>(time), <span class="dt">each=</span><span class="dv">3</span>))</span>
<span id="cb30-11"><a href="#cb30-11"></a></span>
<span id="cb30-12"><a href="#cb30-12"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb30-13"><a href="#cb30-13"></a><span class="cf">for</span>(kk <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(idList)){</span>
<span id="cb30-14"><a href="#cb30-14"></a>  <span class="co"># extract counts for each condition</span></span>
<span id="cb30-15"><a href="#cb30-15"></a>  curCounts &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[,idList[[kk]]]</span>
<span id="cb30-16"><a href="#cb30-16"></a>  <span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">rowMeans</span>(curCounts)<span class="op">+</span><span class="dv">1</span>,</span>
<span id="cb30-17"><a href="#cb30-17"></a>       <span class="dt">y =</span> <span class="kw">rowVars</span>(curCounts)<span class="op">+</span><span class="dv">1</span>,</span>
<span id="cb30-18"><a href="#cb30-18"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>,</span>
<span id="cb30-19"><a href="#cb30-19"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Mean&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance&quot;</span>,</span>
<span id="cb30-20"><a href="#cb30-20"></a>       <span class="dt">main =</span> <span class="kw">names</span>(idList)[kk],</span>
<span id="cb30-21"><a href="#cb30-21"></a>       <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>)</span>
<span id="cb30-22"><a href="#cb30-22"></a>  <span class="kw">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb30-23"><a href="#cb30-23"></a>  lw1 &lt;-<span class="st"> </span><span class="kw">loess</span>((<span class="kw">rowVars</span>(curCounts)<span class="op">+</span><span class="dv">1</span>) <span class="op">~</span><span class="st"> </span>(<span class="kw">rowMeans</span>(curCounts)<span class="op">+</span><span class="dv">1</span>), <span class="dt">span=</span><span class="dv">1</span><span class="op">/</span><span class="dv">4</span>, <span class="dt">lwd=</span><span class="dv">3</span>)</span>
<span id="cb30-24"><a href="#cb30-24"></a>  oo &lt;-<span class="st"> </span><span class="kw">order</span>(<span class="kw">rowMeans</span>(curCounts)<span class="op">+</span><span class="dv">1</span>)</span>
<span id="cb30-25"><a href="#cb30-25"></a>  <span class="kw">lines</span>(<span class="kw">rowMeans</span>(curCounts)[oo]<span class="op">+</span><span class="dv">1</span>, lw1<span class="op">$</span>fitted[oo], <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>)</span>
<span id="cb30-26"><a href="#cb30-26"></a></span>
<span id="cb30-27"><a href="#cb30-27"></a>  <span class="kw">smoothScatter</span>(<span class="dt">x =</span> <span class="kw">log1p</span>(<span class="kw">rowMeans</span>(curCounts)),</span>
<span id="cb30-28"><a href="#cb30-28"></a>       <span class="dt">y =</span> <span class="kw">log1p</span>(<span class="kw">rowVars</span>(curCounts)),</span>
<span id="cb30-29"><a href="#cb30-29"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>,</span>
<span id="cb30-30"><a href="#cb30-30"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Mean&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Variance&quot;</span>)</span>
<span id="cb30-31"><a href="#cb30-31"></a>  <span class="kw">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb30-32"><a href="#cb30-32"></a>  <span class="kw">lines</span>(<span class="kw">log</span>(<span class="kw">rowMeans</span>(curCounts)[oo]<span class="op">+</span><span class="dv">1</span>), <span class="kw">log</span>(lw1<span class="op">$</span>fitted[oo]<span class="op">+</span><span class="dv">1</span>), <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>)</span>
<span id="cb30-33"><a href="#cb30-33"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-8-2.png" width="672" /><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-8-3.png" width="672" /></p>
<ul>
<li>Having data on thousands of genes provides the opportunity to empirically assess the mean-variance relationship.</li>
<li>It is clear that the data is overdispersed with respect to the Poisson distribution (red <span class="math inline">\(y=x\)</span> line). There also seems to be a quadratic trend of the variance as a function of the mean. This has motivated the <strong>negative binomial distribution as the most popular choice to model (bulk) RNA-seq gene expression data</strong>.</li>
</ul>
<hr />
<ul>
<li>The negative binomial distribution is also referred to as the Gamma-Poisson distribution as it can be formulated as such. Indeed, if <span class="math display">\[
 \lambda \sim \Gamma(\alpha, \beta) \\
 Y | \lambda \sim Poi(\lambda),
 \]</span> then this is equivalent to <span class="math display">\[ Y \sim NB(\mu = \alpha / \beta, \phi = 1/\alpha). \]</span></li>
<li>This can be shown analytically, but is considered out of the scope of this course. Below, we show it empirically using simulation.</li>
<li>This theoretical result has got some practical consequences. The Gamma-Poisson formulation makes it clear why we can sum technical replicates as the sum of Poisson random variables is again a Poisson random variable.</li>
<li>The Poisson statement can thus be considered as capturing technical variation, while the Gamma statement can be considered to capture biological variation, i.e., variation in the mean expression across biological replicates.</li>
</ul>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>alpha &lt;-<span class="st"> </span><span class="dv">20</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>beta &lt;-<span class="st"> </span><span class="dv">10</span></span>
<span id="cb31-3"><a href="#cb31-3"></a>lambda &lt;-<span class="st"> </span><span class="kw">rgamma</span>(<span class="dt">n =</span> <span class="fl">1e5</span>, <span class="dt">shape =</span> alpha, <span class="dt">rate =</span> beta)</span>
<span id="cb31-4"><a href="#cb31-4"></a>y1 &lt;-<span class="st"> </span><span class="kw">rpois</span>(<span class="dt">n =</span> <span class="fl">1e5</span>, <span class="dt">lambda =</span> lambda)</span>
<span id="cb31-5"><a href="#cb31-5"></a></span>
<span id="cb31-6"><a href="#cb31-6"></a><span class="co"># note phi = 1 / size</span></span>
<span id="cb31-7"><a href="#cb31-7"></a>y2 &lt;-<span class="st"> </span><span class="kw">rnbinom</span>(<span class="dt">n=</span><span class="fl">1e5</span>, <span class="dt">mu=</span>alpha <span class="op">/</span><span class="st"> </span>(beta), <span class="dt">size=</span>alpha)</span>
<span id="cb31-8"><a href="#cb31-8"></a></span>
<span id="cb31-9"><a href="#cb31-9"></a></span>
<span id="cb31-10"><a href="#cb31-10"></a><span class="kw">plot</span>(<span class="kw">density</span>(y1))</span>
<span id="cb31-11"><a href="#cb31-11"></a><span class="kw">lines</span>(<span class="kw">density</span>(y2), <span class="dt">col=</span><span class="st">&quot;steelblue&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="challenge-ii-normalization" class="section level1">
<h1><span class="header-section-number">3</span> Challenge II: Normalization</h1>
<p>Normalization is necessary to correct for several sources of technical variation:</p>
<ul>
<li><strong>Differences in sequencing depth</strong> between samples. Some samples get sequenced deeper in the sense that they consist of more (mapped) reads and therefore can be considered to contain a higher amount of information, which we should be taking into account. In addition, if a sample is sequenced deeper, it is natural that the counts for each gene will be higher, jeopardizing a direct comparison of the expression counts.</li>
<li><strong>Differences in RNA population composition</strong> between samples. As an extreme example, suppose that two samples have been sequenced to the exact same depth. One sample is contaminated and has a very high concentration of the contaminant cDNA being sequenced, but otherwise the two samples are identical. Since the contaminant will be taking up a significant proportion of the reads being sequenced, the counts will not be directly comparable between the samples. Hence, we may also want to correct for differences in the composition of the RNA population of the samples.</li>
<li><strong>Other technical variation</strong> such as sample-specific GC-content or transcript length effects may also be accounted for.</li>
</ul>
<hr />
<p>Let’s take a look at how comparable different replicates are in the Control condition at 48h in our dataset. We will investigate this using MD-plots (mean-difference plots as introduced by <a href="https://www.jstor.org/stable/24307038">Dudoit <em>et al.</em> (2002)</a>), also sometimes referred to as MA-plots.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>cont48ID <span class="co"># relevant samples</span></span></code></pre></div>
<pre><code>## [1]  2  8 14 19</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="kw">colSums</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts[,cont48ID]) <span class="op">/</span><span class="st"> </span><span class="fl">1e6</span></span></code></pre></div>
<pre><code>## [1] 10.801193  6.828259  8.038079  7.678421</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>combs &lt;-<span class="st"> </span><span class="kw">combn</span>(cont48ID, <span class="dt">m=</span><span class="dv">2</span>) <span class="co">#pairwise combinations between samples</span></span>
<span id="cb36-2"><a href="#cb36-2"></a></span>
<span id="cb36-3"><a href="#cb36-3"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb36-4"><a href="#cb36-4"></a><span class="cf">for</span>(cc <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(combs)){</span>
<span id="cb36-5"><a href="#cb36-5"></a>  curSamples &lt;-<span class="st"> </span>combs[,cc]</span>
<span id="cb36-6"><a href="#cb36-6"></a>  M &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts[,curSamples])</span>
<span id="cb36-7"><a href="#cb36-7"></a>  D &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[,curSamples[<span class="dv">2</span>]] <span class="op">/</span><span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts[,curSamples[<span class="dv">1</span>]]</span>
<span id="cb36-8"><a href="#cb36-8"></a>  <span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">log</span>(M), <span class="dt">y =</span> <span class="kw">log2</span>(D),</span>
<span id="cb36-9"><a href="#cb36-9"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb36-10"><a href="#cb36-10"></a>       <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Sample &quot;</span>, curSamples[<span class="dv">2</span>], <span class="st">&quot; vs sample &quot;</span>, curSamples[<span class="dv">1</span>]),</span>
<span id="cb36-11"><a href="#cb36-11"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Log mean&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Log2 fold-change&quot;</span>,</span>
<span id="cb36-12"><a href="#cb36-12"></a>       <span class="dt">bty =</span> <span class="st">&#39;l&#39;</span>)</span>
<span id="cb36-13"><a href="#cb36-13"></a>  <span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span>
<span id="cb36-14"><a href="#cb36-14"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<ul>
<li>We see clear bias for some pairwise comparisons. For example, in the first plot comparing sample 8 versus sample 2, the log fold-changes are biased downwards. This means that, on average, a gene is lower expressed in sample 8 versus sample 2. Looking at the library sizes, we can indeed see that the library size for sample 2 is about <span class="math inline">\(11 \times 10^6\)</span> while it is only about <span class="math inline">\(7 \times 10^6\)</span> for sample 8! This is a clear library size effect that we should take into account.</li>
</ul>
<div id="count-scaling-versus-glm-offsets" class="section level2">
<h2><span class="header-section-number">3.1</span> Count scaling versus GLM offsets</h2>
<ul>
<li>We have previously discussed count scaling transformations such as CPM and TPM.</li>
<li>A more appropriate and natural way when working with GLMs is through the use of <strong>offsets</strong>. The general use of an offset is to account for the ‘effort’ performed in order to gather that observation of the response variable. Two examples:
<ol style="list-style-type: decimal">
<li>A biologist studying whale migration has one fixed spot where, in the migration season, she counts migrating whales day after day, over several years. For each day she records the number of spotted whales. Of course, the time spent whale-watching may differ from day to day and it is natural that you are more likely to spot more whales if you spend more time looking for them. The time spent spotting whales can then be used as an offset.</li>
<li>In our case, a sample being sequenced deeper contains more information, i.e., more ‘effort’ has been performed, as compared to a sample being sequenced relatively shallow. We have more confidence of a count from a deeply sequenced sample than from a shallowly sequenced sample. We can therefore use the sequencing depth <span class="math inline">\(N_i = \sum_g Y_{gi}\)</span> as offset in the model.</li>
</ol></li>
<li>Adding an offset to the model is different from adding a new variable to the model. For each new variable we add, we will estimate its average effect <span class="math inline">\(\beta\)</span> on the response variable. When adding an offset, however, we are implicitly assuming that <span class="math inline">\(\beta=1\)</span>.</li>
<li>Offsets are typically added on the scale of the linear predictor. Suppose we have a gene <span class="math inline">\(g\)</span> and sample <span class="math inline">\(i\)</span> specific offset <span class="math inline">\(O_{gi}\)</span>, then we can define a negative binomial GLM including the offset as <span class="math display">\[
  \left\{
  \begin{array}{ccc}
  Y_{gi} &amp; \sim &amp; NB(\mu_{gi}, \phi_g) \\
  \log \mu_{gi} &amp; = &amp; \eta_{gi} \\
  \eta_{gi} &amp; = &amp; \mathbf{X}^T_i \beta_g + \log(O_{gi}). \\
  \end{array}
  \right.
  \]</span></li>
<li>Please <a href="https://statomics.github.io/SGA21/sequencing_scalingNormalization.html">read this page</a> for an intuitive reasoning as to why offsets are preferred over count scaling.</li>
</ul>
</div>
<div id="how-to-normalize" class="section level2">
<h2><span class="header-section-number">3.2</span> How to normalize?</h2>
<p>Many approaches are available for normalizing RNA-seq data, and we will review a couple of these. Most methods calculate an offset that is used in the GLM used to model gene expression. One notable method, full-quantile normalization, does not calculate an offset, and rather normalizes counts directly, immediately enforcing the same sequencing depth for all samples.</p>
<div id="tmm-method-default-of-edger" class="section level3">
<h3><span class="header-section-number">3.2.1</span> TMM method (default of <code>edgeR</code>)</h3>
<p>The trimmed mean of M-values (TMM) method introduced by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">Robinson &amp; Oshlack (2010)</a> is a normalization procedure that calculates a single normalization factor for each sample. As the name suggests, it is based on a trimmed mean of fold-changes (<span class="math inline">\(M\)</span>-values) as the scaling factor. A trimmed mean is an average after removing a set of ‘extreme’ values. Specifically, TMM calculates a normalization factor <span class="math inline">\(F_i^{(r)}\)</span> across genes <span class="math inline">\(g\)</span> for each sample <span class="math inline">\(i\)</span> as compared to a reference sample <span class="math inline">\(r\)</span>, <span class="math display">\[
\log_2(F_i^{(r)}) = \frac{\sum_{g \in {\cal G}^*} w_{gi}^r M_{gi}^r}{\sum_{g \in {\cal G}^*} w_{gi}^r},
\]</span> where <span class="math inline">\(M_{gi}^r\)</span> represents the <span class="math inline">\(\log_2\)</span>-fold-change of the gene expression fraction as compared to a reference sample <span class="math inline">\(r\)</span>, i.e., <span class="math display">\[ M_{gi}^r = \log_2\left( \frac{Y_{gi} / N_i}{ Y_{gr} / N_r} \right), \]</span> and <span class="math inline">\(w_{gi}^r\)</span> represents a precision weight calculated as <span class="math display">\[
 w_{gi}^r = \frac{N_i - Y_{gi}}{N_i Y_{gi}} + \frac{N_r - Y_{gr}}{N_r Y_{gr}},
\]</span> and <span class="math inline">\({\cal G}^*\)</span> represents the set of genes after trimming those with the most extreme average expression and fold-change. The weights serve to account for the fact that fold-changes for genes with lower read counts are more variable.</p>
<p>The procedure only takes genes into account where both <span class="math inline">\(Y_{gi}&gt;0\)</span> and <span class="math inline">\(Y_{gr}&gt;0\)</span>. By default, TMM trims genes with the <span class="math inline">\(30\%\)</span> most extreme <span class="math inline">\(M\)</span>-values and <span class="math inline">\(5\%\)</span> most extreme average gene expression, and chooses as reference <span class="math inline">\(r\)</span> the sample whose upper-quartile is closest to the across-sample average upper-quartile. The normalization factor is then used in conjunction with the library size to calculate an <strong>effective library size</strong> <span class="math display">\[
N_i^{eff} = N_i F_i^{(r)}
\]</span> which is used as offset in the GLM. The normalized counts may be given by <span class="math inline">\(\tilde{Y}_{gi} = Y_{gi} / N_i^s\)</span>, with size factor <span class="math display">\[N_i^s = \frac{N_i F_i^{(r)}}{\frac{1}{n}\sum_{i=1}^n N_i F_i^{(r)}}.\]</span></p>
<p>TMM normalization may be performed from the <code>calcNormFactors</code> function implemented in <code>edgeR</code>:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>dge &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">calcNormFactors</span>(se)</span>
<span id="cb37-2"><a href="#cb37-2"></a>dge<span class="op">$</span>samples <span class="co">#normalization factors added to colData</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["group"],"name":[1],"type":["fct"],"align":["left"]},{"label":["lib.size"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["norm.factors"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["run"],"name":[4],"type":["chr"],"align":["left"]},{"label":["experiment"],"name":[5],"type":["fct"],"align":["left"]},{"label":["patient"],"name":[6],"type":["fct"],"align":["left"]},{"label":["treatment"],"name":[7],"type":["fct"],"align":["left"]},{"label":["time"],"name":[8],"type":["fct"],"align":["left"]},{"label":["submission"],"name":[9],"type":["fct"],"align":["left"]},{"label":["study"],"name":[10],"type":["fct"],"align":["left"]},{"label":["sample"],"name":[11],"type":["fct"],"align":["left"]}],"data":[{"1":"1","2":"9079958","3":"0.9824006","4":"SRR479052","5":"SRX140503","6":"1","7":"Control","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308865","_rn_":"Sample1"},{"1":"1","2":"10801193","3":"0.9730905","4":"SRR479053","5":"SRX140504","6":"1","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308866","_rn_":"Sample2"},{"1":"1","2":"5205034","3":"0.9771680","4":"SRR479054","5":"SRX140505","6":"1","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308867","_rn_":"Sample3"},{"1":"1","2":"9681399","3":"0.9935090","4":"SRR479055","5":"SRX140506","6":"1","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308868","_rn_":"Sample4"},{"1":"1","2":"5685671","3":"0.9730827","4":"SRR479056","5":"SRX140507","6":"1","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308869","_rn_":"Sample5"},{"1":"1","2":"7835627","3":"0.9837862","4":"SRR479057","5":"SRX140508","6":"1","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308870","_rn_":"Sample6"},{"1":"1","2":"8590615","3":"0.9350218","4":"SRR479058","5":"SRX140509","6":"2","7":"Control","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308871","_rn_":"Sample7"},{"1":"1","2":"6828259","3":"0.9436147","4":"SRR479059","5":"SRX140510","6":"2","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308872","_rn_":"Sample8"},{"1":"1","2":"24525588","3":"0.9445726","4":"SRR479060","5":"SRX140511","6":"2","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308873","_rn_":"Sample9"},{"1":"1","2":"8248985","3":"0.9320020","4":"SRR479062","5":"SRX140512","6":"2","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308874","_rn_":"Sample10"},{"1":"1","2":"23535205","3":"0.9384823","4":"SRR479063","5":"SRX140513","6":"2","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308875","_rn_":"Sample11"},{"1":"1","2":"8228473","3":"0.9311185","4":"SRR479065","5":"SRX140514","6":"2","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308876","_rn_":"Sample12"},{"1":"1","2":"7317690","3":"1.0567909","4":"SRR479066","5":"SRX140515","6":"3","7":"Control","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308877","_rn_":"Sample13"},{"1":"1","2":"8038079","3":"1.0431594","4":"SRR479067","5":"SRX140516","6":"3","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308878","_rn_":"Sample14"},{"1":"1","2":"12443101","3":"1.0607802","4":"SRR479068","5":"SRX140517","6":"3","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308879","_rn_":"Sample15"},{"1":"1","2":"16260235","3":"1.0407639","4":"SRR479069","5":"SRX140518","6":"3","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308880","_rn_":"Sample16"},{"1":"1","2":"23624835","3":"1.0437936","4":"SRR479070","5":"SRX140519","6":"3","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308881","_rn_":"Sample17"},{"1":"1","2":"7619275","3":"1.0354159","4":"SRR479071","5":"SRX140520","6":"3","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308882","_rn_":"Sample18"},{"1":"1","2":"7678421","3":"1.0440457","4":"SRR479072","5":"SRX140521","6":"4","7":"Control","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308883","_rn_":"Sample19"},{"1":"1","2":"7114567","3":"1.0541970","4":"SRR479073","5":"SRX140522","6":"4","7":"DPN","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308884","_rn_":"Sample20"},{"1":"1","2":"13777391","3":"1.0365701","4":"SRR479074","5":"SRX140523","6":"4","7":"DPN","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308885","_rn_":"Sample21"},{"1":"1","2":"6081314","3":"1.0611051","4":"SRR479076","5":"SRX140524","6":"4","7":"OHT","8":"24h","9":"SRA051611","10":"SRP012167","11":"SRS308886","_rn_":"Sample22"},{"1":"1","2":"15778115","3":"1.0414101","4":"SRR479077","5":"SRX140525","6":"4","7":"OHT","8":"48h","9":"SRA051611","10":"SRP012167","11":"SRS308887","_rn_":"Sample23"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Let’s check how our MD-plots look like after normalization. Note that, we can rewrite the GLM as <span class="math display">\[ \log\left( \frac{\mu_{gi}}{N_i^s} \right) = \mathbf{X}_i^T \beta_g \]</span> and so <span class="math inline">\(\frac{\mu_{gi}}{N_i^s}\)</span> can be considered as an ‘offset-corrected average count’.</p>
<p>We see that all MD-plots are now nicely centered around a log-fold-change of zero!</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="co">## normalize</span></span>
<span id="cb38-2"><a href="#cb38-2"></a>effLibSize &lt;-<span class="st"> </span>dge<span class="op">$</span>samples<span class="op">$</span>lib.size <span class="op">*</span><span class="st"> </span>dge<span class="op">$</span>samples<span class="op">$</span>norm.factors</span>
<span id="cb38-3"><a href="#cb38-3"></a>normCountTMM &lt;-<span class="st"> </span><span class="kw">sweep</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts, <span class="dv">2</span>, <span class="dt">FUN=</span><span class="st">&quot;/&quot;</span>, effLibSize)</span>
<span id="cb38-4"><a href="#cb38-4"></a></span>
<span id="cb38-5"><a href="#cb38-5"></a></span>
<span id="cb38-6"><a href="#cb38-6"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb38-7"><a href="#cb38-7"></a><span class="cf">for</span>(cc <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(combs)){</span>
<span id="cb38-8"><a href="#cb38-8"></a>  curSamples &lt;-<span class="st"> </span>combs[,cc]</span>
<span id="cb38-9"><a href="#cb38-9"></a>  M &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(normCountTMM[,curSamples])</span>
<span id="cb38-10"><a href="#cb38-10"></a>  D &lt;-<span class="st"> </span>normCountTMM[,curSamples[<span class="dv">2</span>]] <span class="op">/</span><span class="st"> </span>normCountTMM[,curSamples[<span class="dv">1</span>]]</span>
<span id="cb38-11"><a href="#cb38-11"></a>  <span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">log</span>(M), <span class="dt">y =</span> <span class="kw">log2</span>(D),</span>
<span id="cb38-12"><a href="#cb38-12"></a>       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex=</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb38-13"><a href="#cb38-13"></a>       <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Sample &quot;</span>, curSamples[<span class="dv">2</span>], <span class="st">&quot; vs sample &quot;</span>, curSamples[<span class="dv">1</span>]),</span>
<span id="cb38-14"><a href="#cb38-14"></a>       <span class="dt">xlab =</span> <span class="st">&quot;Log mean&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Log2 fold-change&quot;</span>,</span>
<span id="cb38-15"><a href="#cb38-15"></a>       <span class="dt">bty =</span> <span class="st">&#39;l&#39;</span>)</span>
<span id="cb38-16"><a href="#cb38-16"></a>  <span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span>
<span id="cb38-17"><a href="#cb38-17"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
</div>
</div>
<div id="median-of-ratios-method-default-of-deseq2" class="section level2">
<h2><span class="header-section-number">3.3</span> Median-of-ratios method (default of <code>DESeq2</code>)</h2>
<p>The median-of-ratios method is used in <code>DESeq2</code> as described in <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">Love <em>et al.</em> (2014)</a>. It assumes that the expected value <span class="math inline">\(\mu_{gi} = E(Y_{gi})\)</span> is proportional to the true expression of the gene, <span class="math inline">\(q_{gi}\)</span>, scaled by a size factor <span class="math inline">\(s_{i}\)</span> for each sample, <span class="math display">\[ \mu_{gi} = s_{i}q_{gi}. \]</span></p>
<p>The size factor <span class="math inline">\(s_{i}\)</span> is then estimated using the median-of-ratios method compared to a synthetic reference sample <span class="math inline">\(r\)</span> defined based on geometric means of counts across samples <span class="math display">\[
s_i = \text{median}_{\{{g:Y^{*}_{gr} \ne 0}\}} \frac{Y_{gi}}{Y^{*}_{gr}},
\]</span> with <span class="math display">\[ Y^{*}_{gr} = \left( \prod_{i=1}^n Y_{gi} \right)^{1/n}. \]</span></p>
<p>We can then use the size factors <span class="math inline">\(s_i\)</span> as offsets to the GLM.</p>
<p><strong>Question</strong>. Do you see any issues with the procedure described as such?</p>
<details>
<summary> Answer. </summary>
<p>
The procedure relies on genes being expressed in all samples. As sample size increases, the number of genes with this property steadily decreases. This will become crucial in single-cell RNA-seq data analysis.
</p>
</details>
<p>Median-of-ratios normalization is implemented in the <code>DESeq2</code> package:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>dds &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">assays</span>(se)<span class="op">$</span>counts,</span>
<span id="cb39-2"><a href="#cb39-2"></a>                                      <span class="dt">colData =</span> <span class="kw">colData</span>(se),</span>
<span id="cb39-3"><a href="#cb39-3"></a>                                      <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span><span class="dv">1</span>) <span class="co">#just add intercept to showcase normalization</span></span></code></pre></div>
<pre><code>## converting counts to integer mode</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a>dds &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">estimateSizeFactors</span>(dds)</span>
<span id="cb41-2"><a href="#cb41-2"></a><span class="kw">sizeFactors</span>(dds)</span></code></pre></div>
<pre><code>##  [1] 0.9166999 1.0640199 0.5224518 0.9806075 0.5664996 0.7802413 0.8282554
##  [8] 0.6592136 2.3949029 0.7898898 2.2781598 0.7858861 0.7775688 0.8399826
## [15] 1.3311498 1.6977165 2.4949218 0.7875376 0.8106158 0.7607974 1.4476678
## [22] 0.6529423 1.6653456</code></pre>
<p>You may also want to check out the <a href="https://www.youtube.com/watch?v=UFB993xufUU">StatQuest video on DESeq2 normalization</a>.</p>
<div id="comparing-tmm-with-deseq2-normalization" class="section level3">
<h3><span class="header-section-number">3.3.1</span> Comparing TMM with DESeq2 normalization</h3>
<p>We can compare the size factors for both normalizations to verify if they agree on the normalization factors. Note we need to scale the effective library sizes from <code>edgeR</code> to enforce a similar scale as the size factors from <code>DESeq2</code>. While below we are using an arithmetic mean, a geometric mean may be used as well, which will be more robust to outlying effective library sizes.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="kw">plot</span>(effLibSize <span class="op">/</span><span class="st"> </span><span class="kw">mean</span>(effLibSize), <span class="kw">sizeFactors</span>(dds),</span>
<span id="cb43-2"><a href="#cb43-2"></a>     <span class="dt">xlab =</span> <span class="st">&quot;edgeR size factor&quot;</span>,</span>
<span id="cb43-3"><a href="#cb43-3"></a>     <span class="dt">ylab =</span> <span class="st">&quot;DESeq2 size factor&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
</div>
</div>
<div id="full-quantile-fq-normalization" class="section level2">
<h2><span class="header-section-number">3.4</span> Full-quantile (FQ) normalization</h2>
<p>In full-quantile normalization, originally introduced in the context of microarrays by <a href="https://academic.oup.com/bioinformatics/article/19/2/185/372664">Bolstad <em>et al.</em> (2003)</a>, the samples are forced to each have a distribution identical to the distribution of the median/average of the quantiles across samples. In practice, we can implement full-quantile normalization using the following procedure</p>
<ol style="list-style-type: decimal">
<li>Given a data matrix <span class="math inline">\(\mathbf{Y}_{G \times n}\)</span> for <span class="math inline">\(G\)</span> genes (rows) and <span class="math inline">\(n\)</span> samples (columns),</li>
<li>sort each column to get <span class="math inline">\(\mathbf{Y}^S\)</span>,</li>
<li>replace all elements of each row by the median (or average) for that row,</li>
<li>obtain the normalized counts <span class="math inline">\(\tilde{\mathbf{Y}}\)</span> by re-arranging (i.e., unsorting) each column.</li>
</ol>
</div>
<div id="uncertainty-in-normalization" class="section level2">
<h2><span class="header-section-number">3.5</span> Uncertainty in normalization</h2>
<ul>
<li>The offsets that are being cacluated for normalization purposes are estimates.</li>
<li>The downstream analyses incorporates these estimates as if they are known, i.e., conditions on the estimates.</li>
<li>The analysis therefore <strong>ignores the uncertainty</strong> in their estimation.</li>
</ul>
<div class="figure">
<img src="images_sequencing/normalizationUncertainty.png" alt="Figure: Box 1 from Vallejos et al. (2017)." width="1414" />
<p class="caption">
Figure: Box 1 from Vallejos et al. (2017).
</p>
</div>
</div>
</div>
<div id="challenge-iii-parameter-estimation-under-limited-information-setting" class="section level1">
<h1><span class="header-section-number">4</span> Challenge III: Parameter estimation (under limited information setting)</h1>
<p>There are two challenges to be overcome here:</p>
<ul>
<li>First, we need to get the structure of our mean model right, this is, which covariates to include, and how to include them, such that we are capable of capturing important sources of variation in our experiment, in order to derive a correct interpretation of the data in terms of our research question.</li>
<li>Second, we need to be able to estimate the parameters of our model in an efficient way, while having limited information (i.e., we are often confronted with only a small number of replicates).</li>
</ul>
<div id="defining-the-model-for-the-mean" class="section level2">
<h2><span class="header-section-number">4.1</span> Defining the model for the mean</h2>
<p>Let’s first check how the authors of the original study parameterized the mean model.</p>
<div class="figure">
<img src="images_sequencing/deAnalysis_para.png" alt="Figure: Another paragraph from the Methods section." width="683" />
<p class="caption">
Figure: Another paragraph from the Methods section.
</p>
</div>
<p>The authors write that they are “employing treatment type, time point and sample ID as factors in the model”. Concerning experimental variables, this suggests they have added covariates defining the treatment and time point for each sample. The sample ID in the text refers to the original tissue sample and therefore corresponds to the donor patient. While there is also a variable called <code>sample</code> in the <code>colData</code>, this is not what the authors refer to. Note the ambiguity here and since the authors didn’t share their code, this is hard to check! But, more on reproducibility later…</p>
<p><strong>Question</strong>. What are the authors assuming when using this structure for the mean model? Do you think that there are extensions or simplifications of the mean model that would be relevant?</p>
<details>
<summary> Answer. </summary>
<p>
The authors are acknowledging the relatedness of samples derived from the same donor patient by adding it as a fixed effect to the model, which is great. This blocking strategy has been extensively discussed in the proteomics part of this course. However, by only adding a main effect for treatment and time, they are assuming that the effect of time is identical for all treatments, i.e., the average gene expression in-/decrease at 48h versus 24h is identical for the DPN, OHT or the control samples, which seems like a quite stringent assumption. We can make the model more flexible by allowing for a <code>treatment * time</code> interaction.
</p>
</details>
</div>
<div id="parameter-estimation-and-empirical-bayes" class="section level2">
<h2><span class="header-section-number">4.2</span> Parameter estimation and empirical Bayes</h2>
<ul>
<li><p>Even in limited sample size settings, the parameters <span class="math inline">\(\beta\)</span> of the mean model may be estimated reasonably efficiently, and we have previously discussed the IRLS algorithm to do so.</p></li>
<li><p>However, estimating parameters for the variance (this is, the dispersion parameter <span class="math inline">\(\phi\)</span> from the negative binomial or the variance parameter <span class="math inline">\(\sigma\)</span> from the Gaussian) is typically quite a bit harder.</p></li>
<li><p>In genomics, we often take advantage of the parallel structure of the thousands of regression models (one for each gene) to <strong>borrow information across genes</strong> in a procedure called <strong>empirical Bayes</strong>, as also seen in the proteomics part of this course.</p></li>
<li><p>In the Bayesian setting, we use not only the data, but also a prior distribution, to derive our parameter estimates. In a traditional Bayesian analysis, one assumes an <em>a priori</em> known prior distribution, which reflects our prior belief into all possible values of the parameter. This prior distribution is completely independent of the observed data and the idea is that one specifies the prior distribution before observing the data. One can then use the data and prior distribution to derive a <strong>posterior distribution</strong> for the parameter(s) <span class="math inline">\(\theta\)</span> of interest through Bayes rule <span class="math display">\[
 p(\theta | \mathbf{Y}) = \frac{p(\mathbf{Y} | \theta) p(\theta)}{\int_{\theta \in \Theta} p(\mathbf{Y} | \theta) p(\theta) d\theta}.
 \]</span> Here, the posterior distribution <span class="math inline">\(p(\theta | \mathbf{Y})\)</span> is calculated using the data likelihood <span class="math inline">\(p(\mathbf{Y} | \theta)\)</span>, prior distribution <span class="math inline">\(p(\theta)\)</span> and the ‘marginal likelihood’ <span class="math inline">\(\int_{\theta \in \Theta} p(\mathbf{Y} | \theta) p(\theta) d\theta\)</span>, where <span class="math inline">\(\Theta\)</span> denotes the parameter space of <span class="math inline">\(\theta\)</span>. We can see that the posterior probability for a specific value of the parameter <span class="math inline">\(\theta\)</span> will be high if both the data likelihood as well as prior probability are high.</p></li>
<li><p>In empirical Bayes, we basically take a semi-Bayesian approach to parameter estimation. Indeed, we do not assume a known prior distribution, but estimate it empirically using the data. This empirically estimated prior <span class="math inline">\(\hat{p}(\theta)\)</span> is then used to calculate the posterior distribution.</p></li>
<li><p>While in some settings one can easily calculate the posterior distribution, sometimes it can be a hard problem. In such cases, it may be useful to restrict ourselves to calculating the <strong>maximum a posteriori (MAP)</strong> estimate, which corresponds to the mode of the posterior distribution. This can be considered to be analogous to point estimation in the frequentist setting.</p></li>
<li><p>In our setting, we use genes with a similar average expression to moderate the dispersion estimate for a particular gene. The basic assumption for this to make sense is that genes with similar means might have similar dispersion parameters (or variances), owing to the mean-variance trend.</p></li>
</ul>
<hr />
<ul>
<li>Once initial estimates for the gene-wise dispersions have been derived (<span class="math inline">\(\hat{\Phi}_g^{ML}\)</span> in the figure), we use a parametric model to estimate its distribution (typically as a function of the mean) across genes. This distribution is the <strong>prior distribution</strong>.</li>
<li>Then, each initial estimate is shrunken towards that empirically estimated prior distribution. The amount of shrinkage being performed is data-driven, and depends on the data, taking into account the precision of our initial estimate (i.e., shape of the likelihood) and the variability of the prior distribution.</li>
<li>These strategies result in impressive performance gains in terms of differential expression analysis and are implemented in all popular differential expression analysis software packages (though in slightly differing ways) like <code>limma</code>, <code>edgeR</code> and <code>DESeq2</code>.</li>
</ul>
<div class="figure">
<img src="images_sequencing/empiricalBayes.png" alt="Figure 8 from Van den Berge *et al.* (2019)." width="1040" />
<p class="caption">
Figure 8 from Van den Berge <em>et al.</em> (2019).
</p>
</div>
<p>The blog post on <a href="http://varianceexplained.org/r/empirical_bayes_baseball/">understanding empirical Bayes estimation using baseball statistics</a> is a great primer for further reading, as well as the <a href="https://drob.gumroad.com/l/empirical-bayes">accompanying book</a> by David Robinson.</p>
</div>
<div id="in-practice" class="section level2">
<h2><span class="header-section-number">4.3</span> In practice</h2>
<p>Let’s fit the model using <code>edgeR</code>.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a>design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>treatment<span class="op">*</span>time <span class="op">+</span><span class="st"> </span>patient, <span class="dt">data=</span><span class="kw">colData</span>(se))</span>
<span id="cb44-2"><a href="#cb44-2"></a></span>
<span id="cb44-3"><a href="#cb44-3"></a></span>
<span id="cb44-4"><a href="#cb44-4"></a>dge &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(se)</span>
<span id="cb44-5"><a href="#cb44-5"></a>dge &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(dge, design) <span class="co"># estimate dispersion estimates</span></span>
<span id="cb44-6"><a href="#cb44-6"></a><span class="kw">plotBCV</span>(dge)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a>fit &lt;-<span class="st"> </span><span class="kw">glmFit</span>(dge, design)</span>
<span id="cb45-2"><a href="#cb45-2"></a><span class="kw">head</span>(fit<span class="op">$</span>coefficients)</span></code></pre></div>
<pre><code>##                 (Intercept) treatmentDPN treatmentOHT     time48h   patient2
## ENSG00000000003   -9.332660   0.11215980   0.08617711  0.13277338 -0.5185631
## ENSG00000000419  -10.374585  -0.05233370  -0.03544105 -0.09301681  0.1376257
## ENSG00000000457  -10.935441  -0.12815863  -0.13380545 -0.07347351  0.4574027
## ENSG00000000460  -10.098383   0.08875322   0.13754925 -0.89357390  0.5407554
## ENSG00000000971  -13.689209   0.29526136   0.26736723  0.49224506 -0.1468142
## ENSG00000001036   -8.280283   0.02211192  -0.02416563 -0.08638397 -0.4676508
##                    patient3   patient4 treatmentDPN:time48h
## ENSG00000000003 -0.82994270 -0.6245673          -0.12278804
## ENSG00000000419  0.10515134  0.1023421           0.04810092
## ENSG00000000457 -0.05731193  0.2089687           0.06405330
## ENSG00000000460 -0.33459640 -0.1321134           0.15904868
## ENSG00000000971  0.93857349 -0.2676066          -0.60487781
## ENSG00000001036 -1.52141542 -1.0046446           0.04238597
##                 treatmentOHT:time48h
## ENSG00000000003          -0.11955914
## ENSG00000000419           0.11561301
## ENSG00000000457           0.13101511
## ENSG00000000460           0.18156925
## ENSG00000000971          -0.71520710
## ENSG00000001036           0.08331487</code></pre>
</div>
</div>
<div id="challenge-iv-statistical-inference-across-many-genes" class="section level1">
<h1><span class="header-section-number">5</span> Challenge IV: Statistical inference across many genes</h1>
<div id="contrasts-on-the-treatment-effects" class="section level2">
<h2><span class="header-section-number">5.1</span> Contrasts on the treatment effects</h2>
<p>We will first derive all contrasts that are also investigated in the original manuscript, using our extended model where we are allowing for an interaction effect between treatment and time.</p>
<p>The mean model is <span class="math display">\[ \log(\mu_{gi}) = \beta_{g0} + \beta_{g1} x_{DPN} + \beta_{g2} x_{OHT} + \beta_{g3} x_{48h} + \beta_{g4} x_{pat2} + \beta_{g5} x_{pat3} + \beta_{g6} x_{pat4} + \\ \beta_{g7} x_{DPN:48h} + \beta_{g8} x_{OHT:48h}. \]</span></p>
<p>The intercept corresponds to the log average gene expression in the control group at 24h for patient 1.</p>
<p><strong>DPN 24h vs control 24h.</strong> The respective means are <span class="math display">\[\log \mu_{g,DPN,24h} = \beta_{g0} + \beta_{g1},\]</span> <span class="math display">\[\log \mu_{g,con,24h} = \beta_{g0}.\]</span> And their difference is <span class="math display">\[ \delta_g = \beta_{g1}. \]</span></p>
<p><strong>DPN 48h vs control 48h.</strong> The respective means are <span class="math display">\[\log \mu_{g,DPN,48h} = \beta_{g0} + \beta_{g1} + \beta_{g3} + \beta_{g7},\]</span> <span class="math display">\[\log \mu_{g,con,48h} = \beta_{g0} + \beta_{g3}.\]</span> And their difference is <span class="math display">\[ \delta_g = \beta_{g1} + \beta_{g7}. \]</span></p>
<p><strong>OHT 24h vs control 24h.</strong> The respective means are <span class="math display">\[\log \mu_{g,OHT,24h} = \beta_{g0} + \beta_{g2} ,\]</span> <span class="math display">\[\log \mu_{g,con,24h} = \beta_{g0}.\]</span> And their difference is <span class="math display">\[ \delta_g = \beta_{g2}. \]</span></p>
<p><strong>OHT 48h vs control 48h.</strong> The respective means are <span class="math display">\[\log \mu_{g,OHT,48h} = \beta_{g0} + \beta_{g2} + \beta_{g3} + \beta_{g8},\]</span> <span class="math display">\[\log \mu_{g,con,48h} = \beta_{g0} + \beta_{g3}.\]</span> And their difference is <span class="math display">\[ \delta_g = \beta_{g2} + \beta_{g8}. \]</span></p>
<p>However, we can also assess the interaction effects: is the time effect different between DPN and OHT treatment versus the control? And how about the DPN vs OHT treatments?</p>
<p><strong>DPN vs control interaction.</strong> The time effect for each condition is <span class="math display">\[ \delta_{DPN} = \log \mu_{g,DPN,48h} - \log \mu_{g,DPN,24h} = \beta_{g3} + \beta_{g7},\]</span> <span class="math display">\[ \delta_{con} = \log \mu_{g,con,48h} - \log \mu_{g,con,24h} = \beta_{g3}. \]</span> So the interaction effect is <span class="math display">\[\delta_{DPN-con} = \beta_{g7}.\]</span></p>
<p><strong>OHT vs control interaction.</strong> The time effect for each condition is <span class="math display">\[ \delta_{OHT} = \log \mu_{g,OHT,48h} - \log \mu_{g,OHT,24h} = \beta_{g3} + \beta_{g8},\]</span> <span class="math display">\[ \delta_{con} = \log \mu_{g,con,48h} - \log \mu_{g,con,24h} = \beta_{g3}. \]</span> So the interaction effect is <span class="math display">\[\delta_{DPN-con} = \beta_{g8}.\]</span></p>
<p><strong>OHT vs DPN interaction.</strong> The time effect for each condition is <span class="math display">\[ \delta_{OHT} = \log \mu_{g,OHT,48h} - \log \mu_{g,OHT,24h} = \beta_{g3} + \beta_{g8},\]</span> <span class="math display">\[ \delta_{DPN} = \log \mu_{g,DPN,48h} - \log \mu_{g,DPN,24h} = \beta_{g3} + \beta_{g7},\]</span> So the interaction effect is <span class="math display">\[\delta_{OHT-DPN} = \beta_{g8} - \beta_{g7}.\]</span></p>
<p>Let’s implement all of these in a contrast matrix.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a>L &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> <span class="kw">ncol</span>(fit<span class="op">$</span>coefficients), <span class="dt">ncol =</span> <span class="dv">7</span>)</span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="kw">rownames</span>(L) &lt;-<span class="st"> </span><span class="kw">colnames</span>(fit<span class="op">$</span>coefficients)</span>
<span id="cb47-3"><a href="#cb47-3"></a><span class="kw">colnames</span>(L) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;DPNvsCON24&quot;</span>, <span class="st">&quot;DPNvsCON48&quot;</span>,</span>
<span id="cb47-4"><a href="#cb47-4"></a>                 <span class="st">&quot;OHTvsCON24&quot;</span>, <span class="st">&quot;OHTvsCON48&quot;</span>,</span>
<span id="cb47-5"><a href="#cb47-5"></a>                 <span class="st">&quot;DPNvsCONInt&quot;</span>, <span class="st">&quot;OHTvsCONInt&quot;</span>,</span>
<span id="cb47-6"><a href="#cb47-6"></a>                 <span class="st">&quot;OHTvsDPNInt&quot;</span>)</span>
<span id="cb47-7"><a href="#cb47-7"></a><span class="co"># DPN vs control at 24h</span></span>
<span id="cb47-8"><a href="#cb47-8"></a>L[<span class="dv">2</span>,<span class="st">&quot;DPNvsCON24&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb47-9"><a href="#cb47-9"></a><span class="co"># DPN vs control at 48h</span></span>
<span id="cb47-10"><a href="#cb47-10"></a>L[<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">8</span>),<span class="st">&quot;DPNvsCON48&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb47-11"><a href="#cb47-11"></a><span class="co"># OHT vs control at 24h</span></span>
<span id="cb47-12"><a href="#cb47-12"></a>L[<span class="dv">3</span>,<span class="st">&quot;OHTvsCON24&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb47-13"><a href="#cb47-13"></a><span class="co"># OHT vs control at 48h</span></span>
<span id="cb47-14"><a href="#cb47-14"></a>L[<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">9</span>),<span class="st">&quot;OHTvsCON48&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb47-15"><a href="#cb47-15"></a><span class="co"># DPN control interaction</span></span>
<span id="cb47-16"><a href="#cb47-16"></a>L[<span class="dv">8</span>,<span class="st">&quot;DPNvsCONInt&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb47-17"><a href="#cb47-17"></a><span class="co"># OHT control interaction</span></span>
<span id="cb47-18"><a href="#cb47-18"></a>L[<span class="dv">9</span>,<span class="st">&quot;OHTvsCONInt&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb47-19"><a href="#cb47-19"></a><span class="co"># OHT DPN interaction</span></span>
<span id="cb47-20"><a href="#cb47-20"></a>L[<span class="kw">c</span>(<span class="dv">9</span>,<span class="dv">8</span>),<span class="st">&quot;OHTvsDPNInt&quot;</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">-1</span>)</span>
<span id="cb47-21"><a href="#cb47-21"></a></span>
<span id="cb47-22"><a href="#cb47-22"></a>L</span></code></pre></div>
<pre><code>##                      DPNvsCON24 DPNvsCON48 OHTvsCON24 OHTvsCON48 DPNvsCONInt
## (Intercept)                   0          0          0          0           0
## treatmentDPN                  1          1          0          0           0
## treatmentOHT                  0          0          1          1           0
## time48h                       0          0          0          0           0
## patient2                      0          0          0          0           0
## patient3                      0          0          0          0           0
## patient4                      0          0          0          0           0
## treatmentDPN:time48h          0          1          0          0           1
## treatmentOHT:time48h          0          0          0          1           0
##                      OHTvsCONInt OHTvsDPNInt
## (Intercept)                    0           0
## treatmentDPN                   0           0
## treatmentOHT                   0           0
## time48h                        0           0
## patient2                       0           0
## patient3                       0           0
## patient4                       0           0
## treatmentDPN:time48h           0          -1
## treatmentOHT:time48h           1           1</code></pre>
<p>And, finally, we can assess each hypothesis using the <code>glmLRT</code> function implemented in <code>edgeR</code>. We can assess each hypothesis separately by looping over the contrasts.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a>lrtList &lt;-<span class="st"> </span><span class="kw">list</span>() <span class="co">#list of results</span></span>
<span id="cb49-2"><a href="#cb49-2"></a><span class="cf">for</span>(cc <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(L)) lrtList[[cc]] &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit, <span class="dt">contrast =</span> L[,cc])</span>
<span id="cb49-3"><a href="#cb49-3"></a></span>
<span id="cb49-4"><a href="#cb49-4"></a><span class="co"># p-value histograms</span></span>
<span id="cb49-5"><a href="#cb49-5"></a>pvalList &lt;-<span class="st"> </span><span class="kw">lapply</span>(lrtList, <span class="cf">function</span>(x) x<span class="op">$</span>table<span class="op">$</span>PValue)</span>
<span id="cb49-6"><a href="#cb49-6"></a>pvalMat &lt;-<span class="st"> </span><span class="kw">do.call</span>(cbind,  pvalList)</span>
<span id="cb49-7"><a href="#cb49-7"></a><span class="kw">colnames</span>(pvalMat) &lt;-<span class="st"> </span><span class="kw">colnames</span>(L)</span>
<span id="cb49-8"><a href="#cb49-8"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>))</span>
<span id="cb49-9"><a href="#cb49-9"></a><span class="kw">sapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(pvalMat), <span class="cf">function</span>(ii) <span class="kw">hist</span>(pvalMat[,ii], </span>
<span id="cb49-10"><a href="#cb49-10"></a>                                          <span class="dt">main =</span> <span class="kw">colnames</span>(pvalMat)[ii],</span>
<span id="cb49-11"><a href="#cb49-11"></a>                                          <span class="dt">xlab =</span> <span class="st">&quot;p-value&quot;</span>))</span></code></pre></div>
<pre><code>##          [,1]            [,2]            [,3]            [,4]           
## breaks   numeric,21      numeric,21      numeric,21      numeric,21     
## counts   integer,20      integer,20      integer,20      integer,20     
## density  numeric,20      numeric,20      numeric,20      numeric,20     
## mids     numeric,20      numeric,20      numeric,20      numeric,20     
## xname    &quot;pvalMat[, ii]&quot; &quot;pvalMat[, ii]&quot; &quot;pvalMat[, ii]&quot; &quot;pvalMat[, ii]&quot;
## equidist TRUE            TRUE            TRUE            TRUE           
##          [,5]            [,6]            [,7]           
## breaks   numeric,21      numeric,21      numeric,21     
## counts   integer,20      integer,20      integer,20     
## density  numeric,20      numeric,20      numeric,20     
## mids     numeric,20      numeric,20      numeric,20     
## xname    &quot;pvalMat[, ii]&quot; &quot;pvalMat[, ii]&quot; &quot;pvalMat[, ii]&quot;
## equidist TRUE            TRUE            TRUE</code></pre>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<div id="multiple-testing" class="section level3">
<h3><span class="header-section-number">5.1.1</span> Multiple testing</h3>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="co"># number of DE genes</span></span>
<span id="cb51-2"><a href="#cb51-2"></a>padjMat &lt;-<span class="st"> </span><span class="kw">apply</span>(pvalMat, <span class="dv">2</span>, p.adjust, <span class="dt">method=</span><span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb51-3"><a href="#cb51-3"></a><span class="kw">colSums</span>(padjMat <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span> )</span></code></pre></div>
<pre><code>##  DPNvsCON24  DPNvsCON48  OHTvsCON24  OHTvsCON48 DPNvsCONInt OHTvsCONInt 
##           2          62           0          22           0           0 
## OHTvsDPNInt 
##           0</code></pre>
<p>We are finding low numbers of DE genes between treatments at a 5% FDR level. This was already reflected in the the MDS plots.</p>
</div>
<div id="visualization" class="section level3">
<h3><span class="header-section-number">5.1.2</span> Visualization</h3>
<p>Let’s visualize some results for the DPN vs control at 48h contrast.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="kw">library</span>(scales) <span class="co"># for scales::alpha()</span></span>
<span id="cb53-2"><a href="#cb53-2"></a>deGenes &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table<span class="op">$</span>PValue, <span class="st">&quot;fdr&quot;</span>) <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span></span>
<span id="cb53-3"><a href="#cb53-3"></a></span>
<span id="cb53-4"><a href="#cb53-4"></a><span class="co">## volcano plot</span></span>
<span id="cb53-5"><a href="#cb53-5"></a><span class="kw">plot</span>(<span class="dt">x =</span> lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table<span class="op">$</span>logFC,</span>
<span id="cb53-6"><a href="#cb53-6"></a>     <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table<span class="op">$</span>PValue),</span>
<span id="cb53-7"><a href="#cb53-7"></a>     <span class="dt">xlab =</span> <span class="st">&quot;log Fold-change&quot;</span>,</span>
<span id="cb53-8"><a href="#cb53-8"></a>     <span class="dt">ylab =</span> <span class="st">&quot;-log10 P-value&quot;</span>,</span>
<span id="cb53-9"><a href="#cb53-9"></a>     <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">col =</span> <span class="kw">alpha</span>(deGenes<span class="op">+</span><span class="dv">1</span>, <span class="fl">.4</span>), </span>
<span id="cb53-10"><a href="#cb53-10"></a>     <span class="dt">cex=</span><span class="dv">2</span><span class="op">/</span><span class="dv">3</span>, <span class="dt">bty=</span><span class="st">&#39;l&#39;</span>)</span>
<span id="cb53-11"><a href="#cb53-11"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="kw">c</span>(<span class="st">&quot;DE&quot;</span>, <span class="st">&quot;not DE&quot;</span>),</span>
<span id="cb53-12"><a href="#cb53-12"></a>       <span class="dt">col =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">1</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1"></a><span class="co">## MD-plot</span></span>
<span id="cb54-2"><a href="#cb54-2"></a><span class="kw">plot</span>(<span class="dt">x =</span> lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table<span class="op">$</span>logCPM,</span>
<span id="cb54-3"><a href="#cb54-3"></a>     <span class="dt">y =</span> lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table<span class="op">$</span>logFC,</span>
<span id="cb54-4"><a href="#cb54-4"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Average log CPM&quot;</span>,</span>
<span id="cb54-5"><a href="#cb54-5"></a>     <span class="dt">ylab =</span> <span class="st">&quot;Log fold-change&quot;</span>,</span>
<span id="cb54-6"><a href="#cb54-6"></a>     <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">col =</span> <span class="kw">alpha</span>(deGenes<span class="op">+</span><span class="dv">1</span>, <span class="fl">.4</span>), </span>
<span id="cb54-7"><a href="#cb54-7"></a>     <span class="dt">cex=</span><span class="dv">2</span><span class="op">/</span><span class="dv">3</span>, <span class="dt">bty=</span><span class="st">&#39;l&#39;</span>)</span>
<span id="cb54-8"><a href="#cb54-8"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="kw">c</span>(<span class="st">&quot;DE&quot;</span>, <span class="st">&quot;not DE&quot;</span>),</span>
<span id="cb54-9"><a href="#cb54-9"></a>       <span class="dt">col =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">1</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb54-10"><a href="#cb54-10"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">lty=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-22-2.png" width="672" /></p>
<hr />
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a><span class="co"># extract all DE genes</span></span>
<span id="cb55-2"><a href="#cb55-2"></a>deGenes &lt;-<span class="st"> </span><span class="kw">rownames</span>(lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table)[<span class="kw">p.adjust</span>(lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table<span class="op">$</span>PValue, <span class="st">&quot;fdr&quot;</span>) <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>]</span>
<span id="cb55-3"><a href="#cb55-3"></a>deGenes</span></code></pre></div>
<pre><code>##  [1] &quot;ENSG00000035928&quot; &quot;ENSG00000044574&quot; &quot;ENSG00000070882&quot; &quot;ENSG00000075790&quot;
##  [5] &quot;ENSG00000091137&quot; &quot;ENSG00000092621&quot; &quot;ENSG00000099864&quot; &quot;ENSG00000099875&quot;
##  [9] &quot;ENSG00000100219&quot; &quot;ENSG00000100867&quot; &quot;ENSG00000101255&quot; &quot;ENSG00000101974&quot;
## [13] &quot;ENSG00000102547&quot; &quot;ENSG00000103064&quot; &quot;ENSG00000103257&quot; &quot;ENSG00000103449&quot;
## [17] &quot;ENSG00000107796&quot; &quot;ENSG00000111181&quot; &quot;ENSG00000111790&quot; &quot;ENSG00000119242&quot;
## [21] &quot;ENSG00000120129&quot; &quot;ENSG00000120217&quot; &quot;ENSG00000133935&quot; &quot;ENSG00000135473&quot;
## [25] &quot;ENSG00000135541&quot; &quot;ENSG00000138685&quot; &quot;ENSG00000143127&quot; &quot;ENSG00000145050&quot;
## [29] &quot;ENSG00000145244&quot; &quot;ENSG00000149428&quot; &quot;ENSG00000149485&quot; &quot;ENSG00000152952&quot;
## [33] &quot;ENSG00000155111&quot; &quot;ENSG00000155330&quot; &quot;ENSG00000155660&quot; &quot;ENSG00000164597&quot;
## [37] &quot;ENSG00000166813&quot; &quot;ENSG00000167608&quot; &quot;ENSG00000167703&quot; &quot;ENSG00000168014&quot;
## [41] &quot;ENSG00000169174&quot; &quot;ENSG00000169239&quot; &quot;ENSG00000169762&quot; &quot;ENSG00000170122&quot;
## [45] &quot;ENSG00000170837&quot; &quot;ENSG00000171798&quot; &quot;ENSG00000173210&quot; &quot;ENSG00000175198&quot;
## [49] &quot;ENSG00000181790&quot; &quot;ENSG00000182704&quot; &quot;ENSG00000182836&quot; &quot;ENSG00000183401&quot;
## [53] &quot;ENSG00000185818&quot; &quot;ENSG00000187908&quot; &quot;ENSG00000188783&quot; &quot;ENSG00000197976&quot;
## [57] &quot;ENSG00000219481&quot; &quot;ENSG00000226887&quot; &quot;ENSG00000233705&quot; &quot;ENSG00000234431&quot;
## [61] &quot;ENSG00000236044&quot; &quot;ENSG00000243927&quot;</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="co"># order according to absolute fold-change</span></span>
<span id="cb57-2"><a href="#cb57-2"></a>orderedDEGenes &lt;-<span class="st"> </span>deGenes[<span class="kw">order</span>(<span class="kw">abs</span>(lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table[deGenes, <span class="st">&quot;logFC&quot;</span>]), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)]</span>
<span id="cb57-3"><a href="#cb57-3"></a></span>
<span id="cb57-4"><a href="#cb57-4"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>))</span>
<span id="cb57-5"><a href="#cb57-5"></a><span class="cf">for</span>(kk <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">9</span>){</span>
<span id="cb57-6"><a href="#cb57-6"></a>  <span class="kw">boxplot</span>(<span class="kw">log1p</span>(<span class="kw">assays</span>(se)<span class="op">$</span>counts[orderedDEGenes[kk],]) <span class="op">~</span><span class="st"> </span><span class="kw">interaction</span>(treatment, time))</span>
<span id="cb57-7"><a href="#cb57-7"></a>  <span class="kw">boxplot</span>(fit<span class="op">$</span>fitted.values[orderedDEGenes[kk],] <span class="op">~</span><span class="st"> </span><span class="kw">interaction</span>(treatment, time))</span>
<span id="cb57-8"><a href="#cb57-8"></a>}</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-23-1.png" width="672" /><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-23-2.png" width="672" /></p>
</div>
</div>
<div id="contrast-on-the-time-effect" class="section level2">
<h2><span class="header-section-number">5.2</span> Contrast on the time effect</h2>
<p>Based on the MDS plot, we can expect comparatively more DE genes for the time effect. For didactic purposes, here we assess an <strong>average time effect across the three treatments</strong>. The analysis shows how flexible one can be when using contrasts.</p>
<p>Mean of time 24h: <span class="math display">\[\log \mu_{g,24h} = \frac{1}{3} \left\{ \underbrace{(\beta_{g0})}_\text{Control, 24h} + \underbrace{(\beta_{g0} + \beta_{g1})}_\text{DPN, 24h} + \underbrace{(\beta_{g0} + \beta_{g2})}_\text{OHT, 24h} \right\}\]</span></p>
<p>Mean of time 48h: <span class="math display">\[\log \mu_{g,48h} = \frac{1}{3} \left\{ \underbrace{(\beta_{g0} + \beta_{g3})}_\text{Control, 48h} + \underbrace{(\beta_{g0} + \beta_{g1} + \beta_{g3} + \beta_{g7})}_\text{DPN, 48h} + \underbrace{(\beta_{g0} + \beta_{g2}+ \beta_{g3} + \beta_{g8})}_\text{OHT, 48h} \right\}\]</span></p>
<p>Difference: <span class="math display">\[ \log \left( \frac{\mu_{g,48h}}{\mu_{g,24h}} \right) = \beta_{g3} + \frac{1}{3}(\beta_{g7} + \beta_{g8}) \]</span></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1"></a>Ltime &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> <span class="kw">ncol</span>(fit<span class="op">$</span>coefficients), <span class="dt">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb58-2"><a href="#cb58-2"></a><span class="kw">rownames</span>(Ltime) &lt;-<span class="st"> </span><span class="kw">colnames</span>(fit<span class="op">$</span>coefficients)</span>
<span id="cb58-3"><a href="#cb58-3"></a>Ltime[<span class="kw">c</span>(<span class="st">&quot;time48h&quot;</span>, <span class="st">&quot;treatmentDPN:time48h&quot;</span>, <span class="st">&quot;treatmentOHT:time48h&quot;</span>),<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span><span class="op">/</span><span class="dv">3</span>, <span class="dv">1</span><span class="op">/</span><span class="dv">3</span>)</span>
<span id="cb58-4"><a href="#cb58-4"></a></span>
<span id="cb58-5"><a href="#cb58-5"></a>lrtTime &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit, <span class="dt">contrast=</span>Ltime)</span>
<span id="cb58-6"><a href="#cb58-6"></a><span class="kw">hist</span>(lrtTime<span class="op">$</span>table<span class="op">$</span>PValue) <span class="co"># very different p-value distribution</span></span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1"></a><span class="kw">sum</span>(<span class="kw">p.adjust</span>(lrtTime<span class="op">$</span>table<span class="op">$</span>PValue, <span class="st">&quot;fdr&quot;</span>) <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>) <span class="co"># many DE genes</span></span></code></pre></div>
<pre><code>## [1] 5938</code></pre>
</div>
</div>
<div id="alternative-parameterizations" class="section level1">
<h1><span class="header-section-number">6</span> Alternative parameterizations</h1>
<p>While our design matrix here was parameterized as <code>~ treatment*time + patient</code> alternative, equivalent parameterizations are also possible. Below, we demonstrate another parameterization that could work, too, and can be more intuitive. In this parameterization, we estimate a mean for each experimental condition, without an intercept, which can be convenient to think about how to set up contrasts.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1"></a>treatTime &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">paste0</span>(treatment, time))</span>
<span id="cb61-2"><a href="#cb61-2"></a><span class="kw">table</span>(treatTime)</span></code></pre></div>
<pre><code>## treatTime
## Control24h Control48h     DPN24h     DPN48h     OHT24h     OHT48h 
##          3          4          4          4          4          4</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1"></a>design2 &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="dv">0</span> <span class="op">+</span><span class="st"> </span>treatTime <span class="op">+</span><span class="st"> </span>patient)</span>
<span id="cb63-2"><a href="#cb63-2"></a></span>
<span id="cb63-3"><a href="#cb63-3"></a>dge2 &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(se)</span>
<span id="cb63-4"><a href="#cb63-4"></a>dge2 &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(dge2, design2)</span>
<span id="cb63-5"><a href="#cb63-5"></a><span class="kw">plotBCV</span>(dge2)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1"></a>fit2 &lt;-<span class="st"> </span><span class="kw">glmFit</span>(dge2, design2)</span>
<span id="cb64-2"><a href="#cb64-2"></a><span class="kw">head</span>(fit2<span class="op">$</span>coefficients)</span></code></pre></div>
<pre><code>##                 treatTimeControl24h treatTimeControl48h treatTimeDPN24h
## ENSG00000000003           -9.332660           -9.199887       -9.220500
## ENSG00000000419          -10.374585          -10.467601      -10.426918
## ENSG00000000457          -10.935441          -11.008915      -11.063600
## ENSG00000000460          -10.098383          -10.991957      -10.009630
## ENSG00000000971          -13.689209          -13.196964      -13.393947
## ENSG00000001036           -8.280283           -8.366667       -8.258171
##                 treatTimeDPN48h treatTimeOHT24h treatTimeOHT48h   patient2
## ENSG00000000003       -9.210515       -9.246483       -9.233269 -0.5185631
## ENSG00000000419      -10.471834      -10.410026      -10.387429  0.1376257
## ENSG00000000457      -11.073020      -11.069247      -11.011705  0.4574027
## ENSG00000000460      -10.744155       -9.960834      -10.672839  0.5407554
## ENSG00000000971      -13.506580      -13.421841      -13.644803 -0.1468142
## ENSG00000001036       -8.302169       -8.304449       -8.307518 -0.4676508
##                    patient3   patient4
## ENSG00000000003 -0.82994270 -0.6245673
## ENSG00000000419  0.10515134  0.1023421
## ENSG00000000457 -0.05731193  0.2089687
## ENSG00000000460 -0.33459640 -0.1321134
## ENSG00000000971  0.93857349 -0.2676066
## ENSG00000001036 -1.52141542 -1.0046446</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1"></a><span class="co">## for example: the estimate for the DPN24h vs control 24h is still the same,</span></span>
<span id="cb66-2"><a href="#cb66-2"></a><span class="co">## but requires a different combination of parameters</span></span>
<span id="cb66-3"><a href="#cb66-3"></a><span class="kw">plot</span>(fit<span class="op">$</span>coefficients[,<span class="st">&quot;treatmentDPN&quot;</span>], </span>
<span id="cb66-4"><a href="#cb66-4"></a>     fit2<span class="op">$</span>coefficients[,<span class="st">&quot;treatTimeDPN24h&quot;</span>] <span class="op">-</span><span class="st"> </span>fit2<span class="op">$</span>coefficients[,<span class="st">&quot;treatTimeControl24h&quot;</span>],</span>
<span id="cb66-5"><a href="#cb66-5"></a>     <span class="dt">xlab=</span><span class="st">&quot;Intercept model estimate&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;No intercept model estimate&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-25-2.png" width="672" /></p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1"></a><span class="co"># Let&#39;s implement the DPNvsCON48 contrast</span></span>
<span id="cb67-2"><a href="#cb67-2"></a>L2 &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> <span class="kw">ncol</span>(fit2<span class="op">$</span>coefficients), <span class="dt">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb67-3"><a href="#cb67-3"></a><span class="kw">rownames</span>(L2) &lt;-<span class="st"> </span><span class="kw">colnames</span>(fit2<span class="op">$</span>coefficients)</span>
<span id="cb67-4"><a href="#cb67-4"></a>L2[<span class="kw">c</span>(<span class="st">&quot;treatTimeDPN48h&quot;</span>, <span class="st">&quot;treatTimeControl48h&quot;</span>),<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">-1</span>)</span>
<span id="cb67-5"><a href="#cb67-5"></a>lrt2 &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit2, <span class="dt">contrast=</span>L2[,<span class="dv">1</span>])</span>
<span id="cb67-6"><a href="#cb67-6"></a><span class="kw">hist</span>(lrt2<span class="op">$</span>table<span class="op">$</span>PValue)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-25-3.png" width="672" /></p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a><span class="kw">plot</span>(<span class="dt">x=</span>lrt2<span class="op">$</span>table<span class="op">$</span>PValue, <span class="dt">y=</span>lrtList[[<span class="dv">2</span>]]<span class="op">$</span>table<span class="op">$</span>PValue,</span>
<span id="cb68-2"><a href="#cb68-2"></a>     <span class="dt">xlab=</span><span class="st">&quot;No intercept model p-value&quot;</span>,</span>
<span id="cb68-3"><a href="#cb68-3"></a>     <span class="dt">ylab=</span><span class="st">&quot;Intercept model p-value&quot;</span>)</span></code></pre></div>
<p><img src="sequencing_rnaseqIntro_files/figure-html/unnamed-chunk-25-4.png" width="672" /></p>
</div>
<div id="additional-challenge-opportunity-the-importance-of-reproducible-analysis" class="section level1">
<h1><span class="header-section-number">7</span> Additional Challenge (Opportunity?): The importance of reproducible analysis</h1>
<p>Finally, it is crucial to make your analysis reproducible using tools such as RMarkdown and GitHub. Please sit back and watch this <a href="https://www.youtube.com/watch?v=8QJfNS7XXwA">amazing lecture from Professor Keith Baggerly on "The Importance of Reproducible Research in High-Throughput Biology</a>.</p>
</div>

<div id="rmd-source-code">---
title: 'Sequencing: RNA-seq data intro'
author: "Koen Van den Berge"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output: 
  pdf_document:
    toc: true
    number_sections: true
    latex_engine: xelatex
  html_document:
    toc: true
    toc_float: true
---

```{r functions, include=FALSE}
# A function for captioning and referencing images
fig <- local({
    i <- 0
    ref <- list()
    list(
        cap=function(refName, text) {
            i <<- i + 1
            ref[[refName]] <<- i
            paste("Figure ", i, ": ", text, sep="")
        },
        ref=function(refName) {
            ref[[refName]]
        })
})
``` 

```{r, echo=FALSE, message=FALSE, eval=TRUE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(ggplot2)
})
if(!"BiocManager" %in% installed.packages()[,1]){
  install.packages("BiocManager")
}
if(!"limma" %in% installed.packages()[,1]){
  BiocManager::install("limma")
}
if(!"edgeR" %in% installed.packages()[,1]){
  BiocManager::install("edgeR")
}
if(!"DESeq2" %in% installed.packages()[,1]){
  BiocManager::install("DESeq2")
}
if(!"scales" %in% installed.packages()[,1]){
  install.packages("scales")
}
```

In this lecture we will start working with a real bulk RNA-seq dataset from [Haglund *et al.* (2012)](https://academic.oup.com/jcem/article/97/12/4631/2536573). After importing the data, we will be working our way through four major challenges which, together, will form a full RNA-seq differential expression (DE) analysis pipeline where the result of our analysis will be a(n ordered) list of genes that we find to be differently expressed between our conditions of interest. The four main challenges we will look into are 

 - Choice of modeling assumptions (distribution).
 - Normalization.
 - Parameter estimation under a limited information setting.
 - Statistical inference under high dimensionality (many genes).

# Experimental design, data import and data exploration

## Experimental design

Let's try to work out the experimental design using the following paragraph from the Methods section of the paper.

```{r, echo=FALSE, fig.cap=paste("Figure: A paragraph from the Methods section.")}
# All defaults
include_graphics("./images_sequencing/expDesign_para.png")
```

## Data import and exploration

We will be importing the dataset using the [parathyroidSE](https://www.bioconductor.org/packages/release/data/experiment/html/parathyroidSE.html) data package from [Bioconductor](https://bioconductor.org/).

```{r, warning=FALSE, eval=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}
if(!"SummarizedExperiment" %in% installed.packages()[,1]){
  BiocManager::install("SummarizedExperiment")
} 
# install package if not installed.
if(!"parathyroidSE" %in% installed.packages()[,1]) BiocManager::install("parathyroidSE")

suppressPackageStartupMessages({
  library(parathyroidSE)
  library(SummarizedExperiment)
})

# import data
data("parathyroidGenesSE", package="parathyroidSE")
# rename for convenience
se1 <- parathyroidGenesSE
rm(parathyroidGenesSE)

# three treatments
treatment1 <- colData(se1)$treatment
table(treatment1)
# two timepoints
time1 <- colData(se1)$time
table(time1)
# four donor patients
patient1 <- colData(se1)$patient
table(patient1)

table(patient1, treatment1, time1)
```

- We observe that the number of samples that we are observing here is larger than what is described in the paper. As also described in the [parathyroidSE vignette](https://www.bioconductor.org/packages/release/data/experiment/vignettes/parathyroidSE/inst/doc/parathyroidSE.pdf), some samples were spread over multiple sequencing runs (i.e., the same sample being sequenced repeatedly) and therefore constitute **technical replication**, rather than biological replication.
- We have previously seen that technical replicates can be considered to be distributed according to a Poisson distribution. One important property of Poisson random variables is that a sum of Poisson random variables still follow a Poisson distribution. Indeed, if $X \sim Poi(\mu_X)$ and $Y \sim Poi(\mu_Y)$, then $X + Y = Z \sim Poi(\mu_X + \mu_Y)$.
- For this reason, it is often suggested to sum technical replicates rather than, for example, averaging, which does not retain the Poisson property (try for yourself!). We'll therefore first sum the technical replicates.

```{r}
dupExps <- as.character(colData(se1)$experiment[duplicated(colData(se1)$experiment)])
dupExps

counts <- assays(se1)$counts
newCounts <- counts
cd <- colData(se1)
for(ss in 1:length(dupExps)){
  # check which samples are duplicates
  relevantId <- which(colData(se1)$experiment == dupExps[ss])
  # sum counts
  newCounts[,relevantId[1]] <- rowSums(counts[,relevantId])
  # keep which columns / rows to remove.
  if(ss == 1){
    toRemove <- relevantId[2]
  } else {
    toRemove <- c(toRemove, relevantId[2])
  }
}

# remove after summing counts (otherwise IDs get mixed up)
newCounts <- newCounts[,-toRemove]
newCD <- cd[-toRemove,]

# Create new SummarizedExperiment
se <- SummarizedExperiment(assays = list("counts" = newCounts),
                            colData = newCD,
                            metadata = metadata(se1))

treatment <- colData(se)$treatment
table(treatment)
time <- colData(se)$time
table(time)
patient <- colData(se)$patient
table(patient)

table(patient, treatment, time) # agrees with paper.

```

 - After summing the technical replicates and appropriately updating the sample information, we again create a `SummarizedExperiment` object, which is essentially a data container that contains all relevant information about your experiment. Please see the [vignette](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) for more information on how to use this class.
 - By directly matching columns (samples) and rows (genes) to their relevant metadata, the `SummarizedExperiment` class avoids mistakes by mis-matching columns and rows with each other (provided you haven't mismatched them when you create the object).
 - The `SummarizedExperiment` class is modular and extendable, and extensions exist for example for the analysis of single-cell RNA-seq data, i.e., the `SingleCellExperiment` class.
 - Due to their convenient organization and widely supported usage within Bioconductor, we will typically work with such classes in the analysis of RNA-seq data.
 
## Independent filtering

Independent filtering is a strategy to remove features (in this case, genes) prior to the analysis. Removal of these features may lower the multiple testing correction for other genes that pass the filter. We try to remove genes that have a low power to be found statistically significant, and/or that are biologically less relevant. 
A common filtering strategy is to remove genes with a generally low expression, as low counts have lower relative uncertainty (hence lower statistical power), and may be considered biologically less relevant.

```{r}
suppressPackageStartupMessages({
  library(limma)
  library(edgeR)
})

keep <- rowSums(cpm(se) > 2) >= 3
table(keep)
se <- se[keep,]
```

## Data exploration

```{r}
# library size distribution
hist(colSums(assays(se)$counts)/1e6, breaks=10)
boxplot(colSums(assays(se)$counts)/1e6 ~ treatment)
boxplot(colSums(assays(se)$counts)/1e6 ~ time)
boxplot(colSums(assays(se)$counts)/1e6 ~ patient)
boxplot(colSums(assays(se)$counts)/1e6 ~ interaction(treatment, time))

# MDS plot
plotMDS(se, 
        labels = treatment, 
        col=as.numeric(patient))

## hard to see influence of experimental factors due to large between-patient variation
## we could also make an MDS plot per patient to take a look.
for(kk in 1:4){
  id <- which(patient == kk)
  plotMDS(se[,id], 
        labels = paste0(treatment[id],"_",time[id]), 
        col=as.numeric(time[id]))
}
```

```{r, eval=FALSE, echo=FALSE}
# Explain concept of MDS: preserve Euclidean distance from high to low dim.
```

Observations based on MDS plot:

 - There is a very large between-patient variability, which is the major source of variation in this dataset. The samples from each patient cluster together tightly.
 - Within patient, time consistently explains more variation than the treatments.
 - Relative to patients and time, the treatment seems to have a fairly small effect.

# Challenge I: Choice of modeling assumptions

When working with a GLM, as part of the choices of modeling assumptions, we need to pick an appropriate distribution for the expression counts. Below we perform some exploratory analyses to investigate.

```{r, eval=TRUE}
y <- assays(se)$counts[1,]
hist(y, breaks = 40,
     xlab = "Gene expression",
     xaxt = "n", yaxt = "n",
     main = "Data for the first gene")
axis(1, at = seq(200, 1200, by=200))
axis(2, at = 0:3)

# Mean-variance trend within each experimental condition
cont24ID <- which(treatment == "Control" & time == "24h")
DPN24ID <- which(treatment == "DPN" & time == "24h")
OHT24ID <- which(treatment == "OHT" & time == "24h")
cont48ID <- which(treatment == "Control" & time == "48h")
DPN48ID <- which(treatment == "DPN" & time == "48h")
OHT48ID <- which(treatment == "OHT" & time == "48h")
idList <- list(cont24ID, DPN24ID, OHT24ID, 
               cont48ID, DPN48ID, OHT48ID)
names(idList) <- paste0(rep(levels(treatment),2), rep(levels(time), each=3))

par(mfrow=c(3,2), mar=c(2,2,2,1))
for(kk in 1:length(idList)){
  # extract counts for each condition
  curCounts <- assays(se)$counts[,idList[[kk]]]
  plot(x = rowMeans(curCounts)+1,
       y = rowVars(curCounts)+1,
       pch = 16, cex=1/2,
       xlab = "Mean", ylab="Variance",
       main = names(idList)[kk],
       log="xy")
  abline(0,1, col="red")
  lw1 <- loess((rowVars(curCounts)+1) ~ (rowMeans(curCounts)+1), span=1/4, lwd=3)
  oo <- order(rowMeans(curCounts)+1)
  lines(rowMeans(curCounts)[oo]+1, lw1$fitted[oo], col="orange")

  smoothScatter(x = log1p(rowMeans(curCounts)),
       y = log1p(rowVars(curCounts)),
       pch = 16, cex=1/2,
       xlab = "Mean", ylab="Variance")
  abline(0,1, col="red")
  lines(log(rowMeans(curCounts)[oo]+1), log(lw1$fitted[oo]+1), col="orange")
}
```

 - Having data on thousands of genes provides the opportunity to empirically assess the mean-variance relationship.
 - It is clear that the data is overdispersed with respect to the Poisson distribution (red $y=x$ line). There also seems to be a quadratic trend of the variance as a function of the mean. This has motivated the **negative binomial distribution as the most popular choice to model (bulk) RNA-seq gene expression data**.
 
 ---
 
 - The negative binomial distribution is also referred to as the Gamma-Poisson distribution as it can be formulated as such. Indeed, if 
 $$
 \lambda \sim \Gamma(\alpha, \beta) \\
 Y | \lambda \sim Poi(\lambda),
 $$
 then this is equivalent to
 $$ Y \sim NB(\mu = \alpha / \beta, \phi = 1/\alpha). $$
 - This can be shown analytically, but is considered out of the scope of this course. Below, we show it empirically using simulation.
 - This theoretical result has got some practical consequences. The Gamma-Poisson formulation makes it clear why we can sum technical replicates as the sum of Poisson random variables is again a Poisson random variable. 
 - The Poisson statement can thus be considered as capturing technical variation, while the Gamma statement can be considered to capture biological variation, i.e., variation in the mean expression across biological replicates.

```{r}
alpha <- 20
beta <- 10
lambda <- rgamma(n = 1e5, shape = alpha, rate = beta)
y1 <- rpois(n = 1e5, lambda = lambda)

# note phi = 1 / size
y2 <- rnbinom(n=1e5, mu=alpha / (beta), size=alpha)


plot(density(y1))
lines(density(y2), col="steelblue")
```


# Challenge II: Normalization

Normalization is necessary to correct for several sources of technical variation:

 - **Differences in sequencing depth** between samples. Some samples get sequenced deeper in the sense that they consist of more (mapped) reads and therefore can be considered to contain a higher amount of information, which we should be taking into account. In addition, if a sample is sequenced deeper, it is natural that the counts for each gene will be higher, jeopardizing a direct comparison of the expression counts.
 - **Differences in RNA population composition** between samples. As an extreme example, suppose that two samples have been sequenced to the exact same depth. One sample is contaminated and has a very high concentration of the contaminant cDNA being sequenced, but otherwise the two samples are identical. Since the contaminant will be taking up a significant proportion of the reads being sequenced, the counts will not be directly comparable between the samples. Hence, we may also want to correct for differences in the composition of the RNA population of the samples.
 - **Other technical variation** such as sample-specific GC-content or transcript length effects may also be accounted for.
 
 ---
 
Let's take a look at how comparable different replicates are in the Control condition at 48h in our dataset. We will investigate this using MD-plots (mean-difference plots as introduced by [Dudoit *et al.* (2002)](https://www.jstor.org/stable/24307038)), also sometimes referred to as MA-plots.

```{r}
cont48ID # relevant samples
colSums(assays(se)$counts[,cont48ID]) / 1e6
combs <- combn(cont48ID, m=2) #pairwise combinations between samples

par(mfrow=c(3,2), mar=c(4,4,2,1))
for(cc in 1:ncol(combs)){
  curSamples <- combs[,cc]
  M <- rowMeans(assays(se)$counts[,curSamples])
  D <- assays(se)$counts[,curSamples[2]] / assays(se)$counts[,curSamples[1]]
  plot(x = log(M), y = log2(D),
       pch = 16, cex=1/3,
       main = paste0("Sample ", curSamples[2], " vs sample ", curSamples[1]),
       xlab = "Log mean", ylab = "Log2 fold-change",
       bty = 'l')
  abline(h = 0, col="orange", lwd=2)
}
```
 
 - We see clear bias for some pairwise comparisons. For example, in the first plot comparing sample 8 versus sample 2, the log fold-changes are biased downwards. This means that, on average, a gene is lower expressed in sample 8 versus sample 2. Looking at the library sizes, we can indeed see that the library size for sample 2 is about $11 \times 10^6$ while it is only about $7 \times 10^6$ for sample 8! This is a clear library size effect that we should take into account.
 
## Count scaling versus GLM offsets

 - We have previously discussed count scaling transformations such as CPM and TPM.
 - A more appropriate and natural way when working with GLMs is through the use of **offsets**. The general use of an offset is to account for the 'effort' performed in order to gather that observation of the response variable. Two examples:
   1. A biologist studying whale migration has one fixed spot where, in the migration season, she counts migrating whales day after day, over several years. For each day she records the number of spotted whales. Of course, the time spent whale-watching may differ from day to day and it is natural that you are more likely to spot more whales if you spend more time looking for them. The time spent spotting whales can then be used as an offset.
   2. In our case, a sample being sequenced deeper contains more information, i.e., more 'effort' has been performed, as compared to a sample being sequenced relatively shallow. We have more confidence of a count from a deeply sequenced sample than from a shallowly sequenced sample. We can therefore use the sequencing depth $N_i = \sum_g Y_{gi}$ as offset in the model.
  - Adding an offset to the model is different from adding a new variable to the model. For each new variable we add, we will estimate its average effect $\beta$ on the response variable. When adding an offset, however, we are implicitly assuming that $\beta=1$.
  - Offsets are typically added on the scale of the linear predictor. Suppose we have a gene $g$ and sample $i$ specific offset $O_{gi}$, then we can define a negative binomial GLM including the offset as
  $$
  \left\{
  \begin{array}{ccc}
  Y_{gi} & \sim & NB(\mu_{gi}, \phi_g) \\
  \log \mu_{gi} & = & \eta_{gi} \\
  \eta_{gi} & = & \mathbf{X}^T_i \beta_g + \log(O_{gi}). \\
  \end{array}
  \right.
  $$
 - Please [read this page](https://statomics.github.io/SGA21/sequencing_scalingNormalization.html) for an intuitive reasoning as to why offsets are preferred over count scaling.
  
## How to normalize?

Many approaches are available for normalizing RNA-seq data, and we will review a couple of these. Most methods calculate an offset that is used in the GLM used to model gene expression. One notable method, full-quantile normalization, does not calculate an offset, and rather normalizes counts directly, immediately enforcing the same sequencing depth for all samples.
 
### TMM method (default of `edgeR`)

The trimmed mean of M-values (TMM) method introduced by [Robinson & Oshlack (2010)](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25) is a normalization procedure that calculates a single normalization factor for each sample. As the name suggests, it is based on a trimmed mean of fold-changes ($M$-values) as the scaling factor. A trimmed mean is an average after removing a set of 'extreme' values. 
Specifically, TMM calculates a normalization factor $F_i^{(r)}$ across genes $g$ for each sample $i$ as compared to a reference sample $r$,
$$
\log_2(F_i^{(r)}) = \frac{\sum_{g \in {\cal G}^*} w_{gi}^r M_{gi}^r}{\sum_{g \in {\cal G}^*} w_{gi}^r},
$$
where $M_{gi}^r$ represents the $\log_2$-fold-change of the gene expression fraction as compared to a reference sample $r$, i.e.,
\[ M_{gi}^r = \log_2\left( \frac{Y_{gi} / N_i}{ Y_{gr} / N_r} \right), \]
and $w_{gi}^r$ represents a precision weight calculated as
$$
 w_{gi}^r = \frac{N_i - Y_{gi}}{N_i Y_{gi}} + \frac{N_r - Y_{gr}}{N_r Y_{gr}},
$$
and ${\cal G}^*$ represents the set of genes after trimming those with the most extreme average expression and fold-change. The weights serve to account for the fact that fold-changes for genes with lower read counts are more variable.

The procedure only takes genes into account where both $Y_{gi}>0$ and $Y_{gr}>0$. By default, TMM trims genes with the $30\%$ most extreme $M$-values and $5\%$ most extreme average gene expression, and chooses as reference $r$ the sample whose upper-quartile is closest to the across-sample average upper-quartile. The normalization factor is then used in conjunction with the library size to calculate an **effective library size**
$$
N_i^{eff} = N_i F_i^{(r)}
$$
which is used as offset in the GLM.
The normalized counts may be given by $\tilde{Y}_{gi} = Y_{gi} / N_i^s$, with size factor 
$$N_i^s = \frac{N_i F_i^{(r)}}{\frac{1}{n}\sum_{i=1}^n N_i F_i^{(r)}}.$$

TMM normalization may be performed from the `calcNormFactors` function implemented in `edgeR`:

```{r}
dge <- edgeR::calcNormFactors(se)
dge$samples #normalization factors added to colData
```

Let's check how our MD-plots look like after normalization. Note that, we can rewrite the GLM as
$$ \log\left( \frac{\mu_{gi}}{N_i^s} \right) = \mathbf{X}_i^T \beta_g $$
and so $\frac{\mu_{gi}}{N_i^s}$ can be considered as an 'offset-corrected average count'.

We see that all MD-plots are now nicely centered around a log-fold-change of zero!

```{r}
## normalize
effLibSize <- dge$samples$lib.size * dge$samples$norm.factors
normCountTMM <- sweep(assays(se)$counts, 2, FUN="/", effLibSize)


par(mfrow=c(3,2), mar=c(4,4,2,1))
for(cc in 1:ncol(combs)){
  curSamples <- combs[,cc]
  M <- rowMeans(normCountTMM[,curSamples])
  D <- normCountTMM[,curSamples[2]] / normCountTMM[,curSamples[1]]
  plot(x = log(M), y = log2(D),
       pch = 16, cex=1/3,
       main = paste0("Sample ", curSamples[2], " vs sample ", curSamples[1]),
       xlab = "Log mean", ylab = "Log2 fold-change",
       bty = 'l')
  abline(h = 0, col="orange", lwd=2)
}
```


## Median-of-ratios method (default of `DESeq2`)

The median-of-ratios method is used in `DESeq2` as described in [Love *et al.* (2014)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8).
It assumes that the expected value $\mu_{gi} = E(Y_{gi})$ is proportional to the true expression of the gene, $q_{gi}$, scaled by a size factor $s_{i}$ for each sample,
\[ \mu_{gi} = s_{i}q_{gi}. \]

The size factor $s_{i}$ is then estimated using the median-of-ratios method compared to a synthetic reference sample $r$ defined based on geometric means of counts across samples
$$
s_i = \text{median}_{\{{g:Y^{*}_{gr} \ne 0}\}} \frac{Y_{gi}}{Y^{*}_{gr}},
$$
with 
\[ Y^{*}_{gr} = \left( \prod_{i=1}^n Y_{gi} \right)^{1/n}. \]

We can then use the size factors $s_i$ as offsets to the GLM.

**Question**. Do you see any issues with the procedure described as such?

<details><summary> Answer. </summary><p>
The procedure relies on genes being expressed in all samples.
As sample size increases, the number of genes with this property steadily decreases.
This will become crucial in single-cell RNA-seq data analysis.
</p></details>

Median-of-ratios normalization is implemented in the `DESeq2` package:

```{r}
dds <- DESeq2::DESeqDataSetFromMatrix(countData = assays(se)$counts,
                                      colData = colData(se),
                                      design = ~ 1) #just add intercept to showcase normalization
dds <- DESeq2::estimateSizeFactors(dds)
sizeFactors(dds)
```

You may also want to check out the [StatQuest video on DESeq2 normalization](https://www.youtube.com/watch?v=UFB993xufUU).

### Comparing TMM with DESeq2 normalization

We can compare the size factors for both normalizations to verify if they agree on the normalization factors. Note we need to scale the effective library sizes from `edgeR` to enforce a similar scale as the size factors from `DESeq2`. While below we are using an arithmetic mean, a geometric mean may be used as well, which will be more robust to outlying effective library sizes.

```{r}
plot(effLibSize / mean(effLibSize), sizeFactors(dds),
     xlab = "edgeR size factor",
     ylab = "DESeq2 size factor")
```


## Full-quantile (FQ) normalization

In full-quantile normalization, originally introduced in the context of microarrays by [Bolstad *et al.* (2003)](https://academic.oup.com/bioinformatics/article/19/2/185/372664), the samples are forced to each have a distribution identical to the distribution of the median/average of the quantiles across samples.
In practice, we can implement full-quantile normalization using the following procedure

 1. Given a data matrix $\mathbf{Y}_{G \times n}$ for $G$ genes (rows) and $n$ samples (columns),
 2. sort each column to get $\mathbf{Y}^S$,
 3. replace all elements of each row by the median (or average) for that row,
 4. obtain the normalized counts $\tilde{\mathbf{Y}}$ by re-arranging (i.e., unsorting) each column.

## Uncertainty in normalization

 - The offsets that are being cacluated for normalization purposes are estimates.
 - The downstream analyses incorporates these estimates as if they are known, i.e., conditions on the estimates.
 - The analysis therefore **ignores the uncertainty** in their estimation.
 
```{r, echo=FALSE, fig.cap=paste("Figure: Box 1 from Vallejos et al. (2017).")}
# All defaults
include_graphics("./images_sequencing/normalizationUncertainty.png")
```


# Challenge III: Parameter estimation (under limited information setting)

There are two challenges to be overcome here:

 - First, we need to get the structure of our mean model right, this is, which covariates to include, and how to include them, such that we are capable of capturing important sources of variation in our experiment, in order to derive a correct interpretation of the data in terms of our research question.
 - Second, we need to be able to estimate the parameters of our model in an efficient way, while having limited information (i.e., we are often confronted with only a small number of replicates).

## Defining the model for the mean

Let's first check how the authors of the original study parameterized the mean model.

```{r, echo=FALSE, fig.cap=paste("Figure: Another paragraph from the Methods section.")}
# All defaults
include_graphics("./images_sequencing/deAnalysis_para.png")
```

The authors write that they are "employing treatment type, time point and sample ID as factors in the model". Concerning experimental variables, this suggests they have added covariates defining the treatment and time point for each sample. The sample ID in the text refers to the original tissue sample and therefore corresponds to the donor patient. While there is also a variable called `sample` in the `colData`, this is not what the authors refer to. Note the ambiguity here and since the authors didn't share their code, this is hard to check! But, more on reproducibility later...

**Question**. What are the authors assuming when using this structure for the mean model? Do you think that there are extensions or simplifications of the mean model that would be relevant?

<details><summary> Answer. </summary><p>
The authors are acknowledging the relatedness of samples derived from the same donor patient by adding it as a fixed effect to the model, which is great. This blocking strategy has been extensively discussed in the proteomics part of this course.
However, by only adding a main effect for treatment and time, they are assuming that the effect of time is identical for all treatments, i.e., the average gene expression in-/decrease at 48h versus 24h is identical for the DPN, OHT or the control samples, which seems like a quite stringent assumption.
We can make the model more flexible by allowing for a `treatment * time` interaction.
</p></details>

## Parameter estimation and empirical Bayes

 - Even in limited sample size settings, the parameters $\beta$ of the mean model may be estimated reasonably efficiently, and we have previously discussed the IRLS algorithm to do so. 
 - However, estimating parameters for the variance (this is, the dispersion parameter $\phi$ from the negative binomial or the variance parameter $\sigma$ from the Gaussian) is typically quite a bit harder.
 - In genomics, we often take advantage of the parallel structure of the thousands of regression models (one for each gene) to **borrow information across genes** in a procedure called **empirical Bayes**, as also seen in the proteomics part of this course. 
 - In the Bayesian setting, we use not only the data, but also a prior distribution, to derive our parameter estimates. In a traditional Bayesian analysis, one assumes an *a priori* known prior distribution, which reflects our prior belief into all possible values of the parameter. This prior distribution is completely independent of the observed data and the idea is that one specifies the prior distribution before observing the data. One can then use the data and prior distribution to derive a **posterior distribution** for the parameter(s) $\theta$ of interest through Bayes rule
 $$
 p(\theta | \mathbf{Y}) = \frac{p(\mathbf{Y} | \theta) p(\theta)}{\int_{\theta \in \Theta} p(\mathbf{Y} | \theta) p(\theta) d\theta}.
 $$
    Here, the posterior distribution $p(\theta | \mathbf{Y})$ is calculated using the data likelihood $p(\mathbf{Y} | \theta)$, prior distribution $p(\theta)$ and the 'marginal likelihood' $\int_{\theta \in \Theta} p(\mathbf{Y} | \theta) p(\theta) d\theta$, where $\Theta$ denotes the parameter space of $\theta$. We can see that the posterior probability for a specific value of the parameter $\theta$ will be high if both the data likelihood as well as prior probability are high.
    
 - In empirical Bayes, we basically take a semi-Bayesian approach to parameter estimation. Indeed, we do not assume a known prior distribution, but estimate it empirically using the data. This empirically estimated prior $\hat{p}(\theta)$ is then used to calculate the posterior distribution.
 - While in some settings one can easily calculate the posterior distribution, sometimes it can be a hard problem. In such cases, it may be useful to restrict ourselves to calculating the **maximum a posteriori (MAP)** estimate, which corresponds to the mode of the posterior distribution. This can be considered to be analogous to point estimation in the frequentist setting.
 - In our setting, we use genes with a similar average expression to moderate the dispersion estimate for a particular gene. The basic assumption for this to make sense is that genes with similar means might have similar dispersion parameters (or variances), owing to the mean-variance trend. 

 ---
 
 - Once initial estimates for the gene-wise dispersions have been derived ($\hat{\Phi}_g^{ML}$ in the figure), we use a parametric model to estimate its distribution (typically as a function of the mean) across genes. This distribution is the **prior distribution**.
 - Then, each initial estimate is shrunken towards that empirically estimated prior distribution. The amount of shrinkage being performed is data-driven, and depends on the data, taking into account the precision of our initial estimate (i.e., shape of the likelihood) and the variability of the prior distribution.
 - These strategies result in impressive performance gains in terms of differential expression analysis and are implemented in all popular differential expression analysis software packages (though in slightly differing ways) like `limma`, `edgeR` and `DESeq2`.

```{r, echo=FALSE, fig.cap=paste("Figure 8 from Van den Berge *et al.* (2019).")}
# All defaults
include_graphics("./images_sequencing/empiricalBayes.png")
```

The blog post on [understanding empirical Bayes estimation using baseball statistics](http://varianceexplained.org/r/empirical_bayes_baseball/) is a great primer for further reading, as well as the [accompanying book](https://drob.gumroad.com/l/empirical-bayes) by David Robinson.

## In practice

Let's fit the model using `edgeR`.

```{r}
design <- model.matrix(~ treatment*time + patient, data=colData(se))


dge <- calcNormFactors(se)
dge <- estimateDisp(dge, design) # estimate dispersion estimates
plotBCV(dge)
fit <- glmFit(dge, design)
head(fit$coefficients)


```

# Challenge IV: Statistical inference across many genes

## Contrasts on the treatment effects

We will first derive all contrasts that are also investigated in the original manuscript, using our extended model where we are allowing for an interaction effect between treatment and time.

The mean model is
$$ \log(\mu_{gi}) = \beta_{g0} + \beta_{g1} x_{DPN} + \beta_{g2} x_{OHT} + \beta_{g3} x_{48h} + \beta_{g4} x_{pat2} + \beta_{g5} x_{pat3} + \beta_{g6} x_{pat4} + \\ \beta_{g7} x_{DPN:48h} + \beta_{g8} x_{OHT:48h}. $$

The intercept corresponds to the log average gene expression in the control group at 24h for patient 1.

**DPN 24h vs control 24h.**
The respective means are
$$\log \mu_{g,DPN,24h} = \beta_{g0} + \beta_{g1},$$
$$\log \mu_{g,con,24h} = \beta_{g0}.$$
And their difference is
$$ \delta_g = \beta_{g1}. $$

**DPN 48h vs control 48h.**
The respective means are
$$\log \mu_{g,DPN,48h} = \beta_{g0} + \beta_{g1} + \beta_{g3} + \beta_{g7},$$
$$\log \mu_{g,con,48h} = \beta_{g0} + \beta_{g3}.$$
And their difference is
$$ \delta_g = \beta_{g1} + \beta_{g7}. $$

**OHT 24h vs control 24h.**
The respective means are
$$\log \mu_{g,OHT,24h} = \beta_{g0} + \beta_{g2} ,$$
$$\log \mu_{g,con,24h} = \beta_{g0}.$$
And their difference is
$$ \delta_g = \beta_{g2}. $$

**OHT 48h vs control 48h.**
The respective means are
$$\log \mu_{g,OHT,48h} = \beta_{g0} + \beta_{g2} + \beta_{g3} + \beta_{g8},$$
$$\log \mu_{g,con,48h} = \beta_{g0} + \beta_{g3}.$$
And their difference is
$$ \delta_g = \beta_{g2} + \beta_{g8}. $$

However, we can also assess the interaction effects: is the time effect different between DPN and OHT treatment versus the control? And how about the DPN vs OHT treatments?

**DPN vs control interaction.**
The time effect for each condition is
$$ \delta_{DPN} = \log \mu_{g,DPN,48h} - \log \mu_{g,DPN,24h} = \beta_{g3} + \beta_{g7},$$
$$ \delta_{con} = \log \mu_{g,con,48h} - \log \mu_{g,con,24h} = \beta_{g3}. $$
So the interaction effect is
$$\delta_{DPN-con} = \beta_{g7}.$$

**OHT vs control interaction.**
The time effect for each condition is
$$ \delta_{OHT} = \log \mu_{g,OHT,48h} - \log \mu_{g,OHT,24h} = \beta_{g3} + \beta_{g8},$$
$$ \delta_{con} = \log \mu_{g,con,48h} - \log \mu_{g,con,24h} = \beta_{g3}. $$
So the interaction effect is
$$\delta_{DPN-con} = \beta_{g8}.$$

**OHT vs DPN interaction.**
The time effect for each condition is
$$ \delta_{OHT} = \log \mu_{g,OHT,48h} - \log \mu_{g,OHT,24h} = \beta_{g3} + \beta_{g8},$$
$$ \delta_{DPN} = \log \mu_{g,DPN,48h} - \log \mu_{g,DPN,24h} = \beta_{g3} + \beta_{g7},$$
So the interaction effect is
$$\delta_{OHT-DPN} = \beta_{g8} - \beta_{g7}.$$

Let's implement all of these in a contrast matrix.

```{r}
L <- matrix(0, nrow = ncol(fit$coefficients), ncol = 7)
rownames(L) <- colnames(fit$coefficients)
colnames(L) <- c("DPNvsCON24", "DPNvsCON48",
                 "OHTvsCON24", "OHTvsCON48",
                 "DPNvsCONInt", "OHTvsCONInt",
                 "OHTvsDPNInt")
# DPN vs control at 24h
L[2,"DPNvsCON24"] <- 1
# DPN vs control at 48h
L[c(2,8),"DPNvsCON48"] <- 1
# OHT vs control at 24h
L[3,"OHTvsCON24"] <- 1
# OHT vs control at 48h
L[c(3,9),"OHTvsCON48"] <- 1
# DPN control interaction
L[8,"DPNvsCONInt"] <- 1
# OHT control interaction
L[9,"OHTvsCONInt"] <- 1
# OHT DPN interaction
L[c(9,8),"OHTvsDPNInt"] <- c(1, -1)

L
```

And, finally, we can assess each hypothesis using the `glmLRT` function implemented in `edgeR`. We can assess each hypothesis separately by looping over the contrasts.

```{r}
lrtList <- list() #list of results
for(cc in 1:ncol(L)) lrtList[[cc]] <- glmLRT(fit, contrast = L[,cc])

# p-value histograms
pvalList <- lapply(lrtList, function(x) x$table$PValue)
pvalMat <- do.call(cbind,  pvalList)
colnames(pvalMat) <- colnames(L)
par(mfrow=c(3,3))
sapply(1:ncol(pvalMat), function(ii) hist(pvalMat[,ii], 
                                          main = colnames(pvalMat)[ii],
                                          xlab = "p-value"))
```

### Multiple testing

```{r}
# number of DE genes
padjMat <- apply(pvalMat, 2, p.adjust, method="fdr")
colSums(padjMat <= 0.05 )
```

We are finding low numbers of DE genes between treatments at a 5\% FDR level. This was already reflected in the the MDS plots.

### Visualization

Let's visualize some results for the DPN vs control at 48h contrast.

```{r}
library(scales) # for scales::alpha()
deGenes <- p.adjust(lrtList[[2]]$table$PValue, "fdr") <= 0.05

## volcano plot
plot(x = lrtList[[2]]$table$logFC,
     y = -log10(lrtList[[2]]$table$PValue),
     xlab = "log Fold-change",
     ylab = "-log10 P-value",
     pch = 16, col = alpha(deGenes+1, .4), 
     cex=2/3, bty='l')
legend("topright", c("DE", "not DE"),
       col = 2:1, pch=16, bty='n')

## MD-plot
plot(x = lrtList[[2]]$table$logCPM,
     y = lrtList[[2]]$table$logFC,
     xlab = "Average log CPM",
     ylab = "Log fold-change",
     pch = 16, col = alpha(deGenes+1, .4), 
     cex=2/3, bty='l')
legend("topright", c("DE", "not DE"),
       col = 2:1, pch=16, bty='n')
abline(h=0, col="orange", lwd=2, lty=2)

```

 ---

```{r}
# extract all DE genes
deGenes <- rownames(lrtList[[2]]$table)[p.adjust(lrtList[[2]]$table$PValue, "fdr") <= 0.05]
deGenes
# order according to absolute fold-change
orderedDEGenes <- deGenes[order(abs(lrtList[[2]]$table[deGenes, "logFC"]), decreasing = TRUE)]

par(mfrow=c(3,3))
for(kk in 1:9){
  boxplot(log1p(assays(se)$counts[orderedDEGenes[kk],]) ~ interaction(treatment, time))
  boxplot(fit$fitted.values[orderedDEGenes[kk],] ~ interaction(treatment, time))
}
```


## Contrast on the time effect

Based on the MDS plot, we can expect comparatively more DE genes for the time effect. For didactic purposes, here we assess an **average time effect across the three treatments**. The analysis shows how flexible one can be when using contrasts.

Mean of time 24h:
$$\log \mu_{g,24h} = \frac{1}{3} \left\{ \underbrace{(\beta_{g0})}_\text{Control, 24h} + \underbrace{(\beta_{g0} + \beta_{g1})}_\text{DPN, 24h} + \underbrace{(\beta_{g0} + \beta_{g2})}_\text{OHT, 24h} \right\}$$

Mean of time 48h:
$$\log \mu_{g,48h} = \frac{1}{3} \left\{ \underbrace{(\beta_{g0} + \beta_{g3})}_\text{Control, 48h} + \underbrace{(\beta_{g0} + \beta_{g1} + \beta_{g3} + \beta_{g7})}_\text{DPN, 48h} + \underbrace{(\beta_{g0} + \beta_{g2}+ \beta_{g3} + \beta_{g8})}_\text{OHT, 48h} \right\}$$

Difference:
$$ \log \left( \frac{\mu_{g,48h}}{\mu_{g,24h}} \right) = \beta_{g3} + \frac{1}{3}(\beta_{g7} + \beta_{g8}) $$


```{r}
Ltime <- matrix(0, nrow = ncol(fit$coefficients), ncol = 1)
rownames(Ltime) <- colnames(fit$coefficients)
Ltime[c("time48h", "treatmentDPN:time48h", "treatmentOHT:time48h"),1] <- c(1, 1/3, 1/3)

lrtTime <- glmLRT(fit, contrast=Ltime)
hist(lrtTime$table$PValue) # very different p-value distribution

sum(p.adjust(lrtTime$table$PValue, "fdr") <= 0.05) # many DE genes
```


# Alternative parameterizations

While our design matrix here was parameterized as `~ treatment*time + patient` alternative, equivalent parameterizations are also possible.
Below, we demonstrate another parameterization that could work, too, and can be more intuitive. In this parameterization, we estimate a mean for each experimental condition, without an intercept, which can be convenient to think about how to set up contrasts.

```{r}
treatTime <- as.factor(paste0(treatment, time))
table(treatTime)

design2 <- model.matrix(~ 0 + treatTime + patient)

dge2 <- calcNormFactors(se)
dge2 <- estimateDisp(dge2, design2)
plotBCV(dge2)
fit2 <- glmFit(dge2, design2)
head(fit2$coefficients)

## for example: the estimate for the DPN24h vs control 24h is still the same,
## but requires a different combination of parameters
plot(fit$coefficients[,"treatmentDPN"], 
     fit2$coefficients[,"treatTimeDPN24h"] - fit2$coefficients[,"treatTimeControl24h"],
     xlab="Intercept model estimate", ylab="No intercept model estimate")

# Let's implement the DPNvsCON48 contrast
L2 <- matrix(0, nrow = ncol(fit2$coefficients), ncol = 1)
rownames(L2) <- colnames(fit2$coefficients)
L2[c("treatTimeDPN48h", "treatTimeControl48h"),1] <- c(1, -1)
lrt2 <- glmLRT(fit2, contrast=L2[,1])
hist(lrt2$table$PValue)


plot(x=lrt2$table$PValue, y=lrtList[[2]]$table$PValue,
     xlab="No intercept model p-value",
     ylab="Intercept model p-value")
```


# Additional Challenge (Opportunity?): The importance of reproducible analysis

Finally, it is crucial to make your analysis reproducible using tools such as RMarkdown and GitHub. Please sit back and watch this [amazing lecture from Professor Keith Baggerly on "The Importance of Reproducible Research in High-Throughput Biology](https://www.youtube.com/watch?v=8QJfNS7XXwA).

</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("sequencing_rnaseqIntro.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
