<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koen Van den Berge" />

<meta name="date" content="2020-11-16" />

<title>scRNA-seq workflow: Macosko et al.</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SGA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./docs/intro.pdf">1.1 Position</a>
    </li>
    <li>
      <a href="recapGeneralLinearModel.html">1.2 Recap linear model</a>
    </li>
    <li>
      <a href="./docs/martens_proteomics_bioinformatics.pdf">2.1 PDA - identification</a>
    </li>
    <li>
      <a href="pda_quantification_preprocessing.html">2.2 PDA- Quantification - Preprocessing</a>
    </li>
    <li>
      <a href="pda_robustSummarisation_peptideModels.html">2.2 PDA - Wrap-up Preprocessing - Peptide-Level-Models</a>
    </li>
    <li>
      <a href="pda_quantification_inference.html">2.3 PDA - Quantification - Differential Abundance Analysis</a>
    </li>
    <li>
      <a href="pda_blocking_wrapup.html">2.3 PDA - Wrap-up DA - Blocking</a>
    </li>
    <li>
      <a href="technicalDetailsProteomics.html">2.4. PDA - DA Technical Details</a>
    </li>
    <li>
      <a href="./docs/stagewiseTesting.pdf">2.5 PDA - Stage-wise testing</a>
    </li>
    <li>
      <a href="sequencing_intro.html">3.1 SDA - Intro to sequencing</a>
    </li>
    <li>
      <a href="sequencing_countData.html">3.2 SDA - Working with count data</a>
    </li>
    <li>
      <a href="sequencing_rnaseqIntro.html">3.3 SDA - Intro to RNA-seq analysis</a>
    </li>
    <li>
      <a href="sequencing_technicalDE.html">3.4 SDA - Technical topics in DE analysis</a>
    </li>
    <li>
      <a href="https://docs.google.com/presentation/d/1sfbtw52qWgA7TcDIrGB3fWksLM0Mu9O7gBqYgGrt9fM/edit?usp=sharing">4.1 scRNA-seq - General concepts and workflow</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="multipleRegression_KPNA2.html">1 Intro KPNA2</a>
    </li>
    <li>
      <a href="https://www.compomics.com/bioinformatics-for-proteomics/">2.1 Identification</a>
    </li>
    <li>
      <a href="pda_tutorialPreprocessing.html">2.2 DA - Preprocessing</a>
    </li>
    <li>
      <a href="pda_tutorialDesign.html">2.3 DA - Design</a>
    </li>
    <li>
      <a href="sequencing_preprocessing.sh">3.1 Preprocessing</a>
    </li>
    <li>
      <a href="sequencing_lab_oneGene.html">3.2 GLM for one gene</a>
    </li>
    <li>
      <a href="sequencing_hw.html">3.3 Bulk RNA-seq homework</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">scRNA-seq workflow: Macosko et al.</h1>
<h4 class="author">Koen Van den Berge</h4>
<h4 class="date">11/16/2020</h4>

</div>


<div id="import-data" class="section level1">
<h1><span class="header-section-number">1</span> Import data</h1>
<p>The <code>scRNAseq</code> package provides convenient access to several datasets. See the <a href="http://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html">package Bioconductor page</a> for more information.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># install BiocManager package if not installed yet.</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="co"># BiocManager is the package installer for Bioconductor software.</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb1-4"><a href="#cb1-4"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-5"><a href="#cb1-5"></a></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co"># install packages if not yet installed.</span></span>
<span id="cb1-7"><a href="#cb1-7"></a>pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;SingleCellExperiment&quot;</span>, <span class="st">&quot;DropletUtils&quot;</span>, <span class="st">&quot;scRNAseq&quot;</span>, <span class="st">&quot;scater&quot;</span>, <span class="st">&quot;scuttle&quot;</span>, <span class="st">&quot;scran&quot;</span>, <span class="st">&quot;BiocSingular&quot;</span>, <span class="st">&quot;scDblFinder&quot;</span>, <span class="st">&quot;glmpca&quot;</span>, <span class="st">&quot;uwot&quot;</span>)</span>
<span id="cb1-8"><a href="#cb1-8"></a>notInstalled &lt;-<span class="st"> </span>pkgs[<span class="op">!</span>pkgs <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="dv">1</span>]]</span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="cf">if</span>(<span class="kw">length</span>(notInstalled) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>){</span>
<span id="cb1-10"><a href="#cb1-10"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(notInstalled)</span>
<span id="cb1-11"><a href="#cb1-11"></a>}</span></code></pre></div>
<pre><code>## &#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
## &#39;?repositories&#39; for details
## 
## replacement repositories:
##     CRAN: https://cloud.r-project.org</code></pre>
<pre><code>## Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.2 (2021-11-01)</code></pre>
<pre><code>## Installing package(s) &#39;DropletUtils&#39;, &#39;scater&#39;, &#39;scuttle&#39;, &#39;scran&#39;,
##   &#39;BiocSingular&#39;, &#39;scDblFinder&#39;, &#39;glmpca&#39;, &#39;uwot&#39;</code></pre>
<pre><code>## also installing the dependencies &#39;sparseMatrixStats&#39;, &#39;rhdf5filters&#39;, &#39;R.oo&#39;, &#39;R.methodsS3&#39;, &#39;sitmo&#39;, &#39;RcppHNSW&#39;, &#39;beeswarm&#39;, &#39;vipor&#39;, &#39;DelayedMatrixStats&#39;, &#39;HDF5Array&#39;, &#39;rhdf5&#39;, &#39;R.utils&#39;, &#39;dqrng&#39;, &#39;beachmat&#39;, &#39;Rhdf5lib&#39;, &#39;BiocNeighbors&#39;, &#39;ggbeeswarm&#39;, &#39;viridis&#39;, &#39;Rtsne&#39;, &#39;ggrepel&#39;, &#39;statmod&#39;, &#39;bluster&#39;, &#39;metapod&#39;, &#39;ScaledMatrix&#39;, &#39;irlba&#39;, &#39;rsvd&#39;, &#39;xgboost&#39;, &#39;FNN&#39;, &#39;RSpectra&#39;, &#39;RcppAnnoy&#39;, &#39;RcppProgress&#39;</code></pre>
<pre><code>## 
##   There is a binary version available but the source version is later:
##      binary source needs_compilation
## uwot 0.1.10 0.1.11              TRUE
## 
## 
## The downloaded binary packages are in
##  /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//RtmpRq735B/downloaded_packages</code></pre>
<pre><code>## installing the source package &#39;uwot&#39;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># Code below might ask you to create an ExperimentHub directory. </span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co"># Type &#39;yes&#39; and hit Enter, to allow this.</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(scRNAseq))</span>
<span id="cb8-4"><a href="#cb8-4"></a>sce &lt;-<span class="st"> </span><span class="kw">MacoskoRetinaData</span>()</span></code></pre></div>
<pre><code>## snapshotDate(): 2021-10-19</code></pre>
<pre><code>## see ?scRNAseq and browseVignettes(&#39;scRNAseq&#39;) for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<pre><code>## see ?scRNAseq and browseVignettes(&#39;scRNAseq&#39;) for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
</div>
<div id="a-singlecellexperiment-object" class="section level1">
<h1><span class="header-section-number">2</span> A <code>SingleCellExperiment</code> object</h1>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>sce</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 24658 49300 
## metadata(0):
## assays(1): counts
## rownames(24658): KITL TMTC3 ... 1110059M19RIK GM20861
## rowData names(0):
## colnames(49300): r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA ... p1_TAACGCGCTCCT
##   p1_ATTCTTGTTCTT
## colData names(2): cell.id cluster
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
<div id="accessing-data-from-a-singlecellexperiment-object" class="section level2">
<h2><span class="header-section-number">2.1</span> Accessing data from a <code>SingleCellExperiment</code> object</h2>
<p>Please see <a href="http://bioconductor.org/books/release/OSCA/data-infrastructure.html">Figure 4.1 in OSCA</a> for an overview of a <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># Data: assays</span></span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="kw">assays</span>(sce)</span></code></pre></div>
<pre><code>## List of length 1
## names(1): counts</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">assays</span>(sce)<span class="op">$</span>counts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
##               r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA r1_GCGCAACTGCTC r1_GATTGGGAGGCA
## KITL                        .               .               1               .
## TMTC3                       3               .               .               .
## CEP290                      1               3               .               2
## 4930430F08RIK               2               1               2               .
## 1700017N19RIK               .               .               .               .
##               r1_CCTCCTAGTTGG
## KITL                        .
## TMTC3                       2
## CEP290                      1
## 4930430F08RIK               1
## 1700017N19RIK               .</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="co"># Feature metadata: rowData</span></span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="kw">rowData</span>(sce) <span class="co"># empty for now</span></span></code></pre></div>
<pre><code>## DataFrame with 24658 rows and 0 columns</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co"># Cell metadata: colData</span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="kw">colData</span>(sce)</span></code></pre></div>
<pre><code>## DataFrame with 49300 rows and 2 columns
##                         cell.id   cluster
##                     &lt;character&gt; &lt;integer&gt;
## r1_GGCCGCAGTCCG r1_GGCCGCAGTCCG         2
## r1_CTTGTGCGGGAA r1_CTTGTGCGGGAA         2
## r1_GCGCAACTGCTC r1_GCGCAACTGCTC         2
## r1_GATTGGGAGGCA r1_GATTGGGAGGCA         2
## r1_CCTCCTAGTTGG r1_CCTCCTAGTTGG        NA
## ...                         ...       ...
## p1_TCAAAAGCCGGG p1_TCAAAAGCCGGG        24
## p1_ATTAAGTTCCAA p1_ATTAAGTTCCAA        34
## p1_CTGTCTGAGACC p1_CTGTCTGAGACC         2
## p1_TAACGCGCTCCT p1_TAACGCGCTCCT        24
## p1_ATTCTTGTTCTT p1_ATTCTTGTTCTT        24</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co"># Reduced dimensions: reducedDims</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="kw">reducedDims</span>(sce) <span class="co"># empty for now</span></span></code></pre></div>
<pre><code>## List of length 0
## names(0):</code></pre>
</div>
<div id="creating-a-new-singlecellexperiment-object" class="section level2">
<h2><span class="header-section-number">2.2</span> Creating a new <code>SingleCellExperiment</code> object</h2>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>sceNew &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">assays</span>(sce)<span class="op">$</span>counts))</span>
<span id="cb26-2"><a href="#cb26-2"></a>sceNew</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 24658 49300 
## metadata(0):
## assays(1): counts
## rownames(24658): KITL TMTC3 ... 1110059M19RIK GM20861
## rowData names(0):
## colnames(49300): r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA ... p1_TAACGCGCTCCT
##   p1_ATTCTTGTTCTT
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw">rm</span>(sceNew)</span></code></pre></div>
</div>
<div id="storing-metadata-in-a-singlecellexperiment-object" class="section level2">
<h2><span class="header-section-number">2.3</span> Storing (meta)data in a <code>SingleCellExperiment</code> object</h2>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>fakeGeneNames &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sce))</span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="kw">rowData</span>(sce)<span class="op">$</span>fakeName &lt;-<span class="st"> </span>fakeGeneNames</span>
<span id="cb29-3"><a href="#cb29-3"></a><span class="kw">head</span>(<span class="kw">rowData</span>(sce))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 1 column
##                  fakeName
##               &lt;character&gt;
## KITL                gene1
## TMTC3               gene2
## CEP290              gene3
## 4930430F08RIK       gene4
## 1700017N19RIK       gene5
## MGAT4C              gene6</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co"># Remove again by setting to NULL</span></span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="kw">rowData</span>(sce)<span class="op">$</span>fakeName &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb31-3"><a href="#cb31-3"></a></span>
<span id="cb31-4"><a href="#cb31-4"></a><span class="kw">assays</span>(sce)<span class="op">$</span>logCounts &lt;-<span class="st"> </span><span class="kw">log1p</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts)</span>
<span id="cb31-5"><a href="#cb31-5"></a><span class="kw">assays</span>(sce)</span></code></pre></div>
<pre><code>## List of length 2
## names(2): counts logCounts</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="kw">assays</span>(sce)<span class="op">$</span>logCounts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
##               r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA r1_GCGCAACTGCTC r1_GATTGGGAGGCA
## KITL                .               .               0.6931472        .       
## TMTC3               1.3862944       .               .                .       
## CEP290              0.6931472       1.3862944       .                1.098612
## 4930430F08RIK       1.0986123       0.6931472       1.0986123        .       
## 1700017N19RIK       .               .               .                .       
##               r1_CCTCCTAGTTGG
## KITL                .        
## TMTC3               1.0986123
## CEP290              0.6931472
## 4930430F08RIK       0.6931472
## 1700017N19RIK       .</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="kw">assays</span>(sce)<span class="op">$</span>logCounts &lt;-<span class="st"> </span><span class="ot">NULL</span></span></code></pre></div>
</div>
</div>
<div id="filtering-non-informative-genes" class="section level1">
<h1><span class="header-section-number">3</span> Filtering non-informative genes</h1>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span></span>
<span id="cb36-2"><a href="#cb36-2"></a><span class="kw">table</span>(keep)</span></code></pre></div>
<pre><code>## keep
## FALSE  TRUE 
##  6771 17887</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>sce &lt;-<span class="st"> </span>sce[keep,]</span></code></pre></div>
</div>
<div id="quality-control" class="section level1">
<h1><span class="header-section-number">4</span> Quality control</h1>
<div id="calculate-qc-variables" class="section level2">
<h2><span class="header-section-number">4.1</span> Calculate QC variables</h2>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="kw">library</span>(scater)</span></code></pre></div>
<pre><code>## Loading required package: scuttle</code></pre>
<pre><code>## Loading required package: ggplot2</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>is.mito &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^MT-&quot;</span>, <span class="kw">rownames</span>(sce))</span>
<span id="cb42-2"><a href="#cb42-2"></a><span class="kw">sum</span>(is.mito) <span class="co"># 28 mitochondrial genes</span></span></code></pre></div>
<pre><code>## [1] 28</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a>df &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span>is.mito))</span>
<span id="cb44-2"><a href="#cb44-2"></a><span class="co">## add the QC variables to sce object</span></span>
<span id="cb44-3"><a href="#cb44-3"></a><span class="kw">colData</span>(sce) &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">colData</span>(sce), df)</span>
<span id="cb44-4"><a href="#cb44-4"></a><span class="co"># the QC variables have now been added to the colData of our SCE object.</span></span>
<span id="cb44-5"><a href="#cb44-5"></a><span class="kw">colData</span>(sce)</span></code></pre></div>
<pre><code>## DataFrame with 49300 rows and 8 columns
##                         cell.id   cluster       sum  detected subsets_Mito_sum
##                     &lt;character&gt; &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;        &lt;numeric&gt;
## r1_GGCCGCAGTCCG r1_GGCCGCAGTCCG         2     37478      7235              427
## r1_CTTGTGCGGGAA r1_CTTGTGCGGGAA         2     32034      6921              503
## r1_GCGCAACTGCTC r1_GCGCAACTGCTC         2     28140      6390              460
## r1_GATTGGGAGGCA r1_GATTGGGAGGCA         2     20352      5727              326
## r1_CCTCCTAGTTGG r1_CCTCCTAGTTGG        NA     19550      5769              264
## ...                         ...       ...       ...       ...              ...
## p1_TCAAAAGCCGGG p1_TCAAAAGCCGGG        24       817       537               13
## p1_ATTAAGTTCCAA p1_ATTAAGTTCCAA        34       817       574               10
## p1_CTGTCTGAGACC p1_CTGTCTGAGACC         2       816       636               24
## p1_TAACGCGCTCCT p1_TAACGCGCTCCT        24       816       488               27
## p1_ATTCTTGTTCTT p1_ATTCTTGTTCTT        24       816       484               16
##                 subsets_Mito_detected subsets_Mito_percent     total
##                             &lt;integer&gt;            &lt;numeric&gt; &lt;numeric&gt;
## r1_GGCCGCAGTCCG                    14              1.13934     37478
## r1_CTTGTGCGGGAA                    15              1.57021     32034
## r1_GCGCAACTGCTC                    13              1.63468     28140
## r1_GATTGGGAGGCA                    11              1.60181     20352
## r1_CCTCCTAGTTGG                     9              1.35038     19550
## ...                               ...                  ...       ...
## p1_TCAAAAGCCGGG                     4              1.59119       817
## p1_ATTAAGTTCCAA                     5              1.22399       817
## p1_CTGTCTGAGACC                     7              2.94118       816
## p1_TAACGCGCTCCT                     5              3.30882       816
## p1_ATTCTTGTTCTT                     4              1.96078       816</code></pre>
</div>
<div id="eda" class="section level2">
<h2><span class="header-section-number">4.2</span> EDA</h2>
<p>High-quality cells should have many features expressed, and a low contribution of mitochondrial genes. Here, we see that several cells have a very low number of expressed genes, and where most of the molecules are derived from mitochondrial genes. This indicates likely damaged cells, presumably because of loss of cytoplasmic RNA from perforated cells, so we’d want to remove these for the downstream analysis.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a><span class="co"># Number of genes vs library size</span></span>
<span id="cb46-2"><a href="#cb46-2"></a><span class="kw">plotColData</span>(sce, <span class="dt">x =</span> <span class="st">&quot;sum&quot;</span>, <span class="dt">y=</span><span class="st">&quot;detected&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;cluster&quot;</span>) </span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="co"># Mitochondrial genes</span></span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="kw">plotColData</span>(sce, <span class="dt">x =</span> <span class="st">&quot;detected&quot;</span>, <span class="dt">y=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-8-2.png" width="672" /></p>
</div>
<div id="qc-using-adaptive-thresholds" class="section level2">
<h2><span class="header-section-number">4.3</span> QC using adaptive thresholds</h2>
<p>Below, we remove cells that are outlying with respect to</p>
<ol style="list-style-type: decimal">
<li>A low sequencing depth (number of UMIs);</li>
<li>A low number of genes detected;</li>
<li>A high percentage of reads from mitochondrial genes.</li>
</ol>
<p>We remove a total of <span class="math inline">\(3423\)</span> cells, most of which because of an outlyingly high percentage of reads from mitochondrial genes.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a>lowLib &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(df<span class="op">$</span>sum, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)</span>
<span id="cb48-2"><a href="#cb48-2"></a>lowFeatures &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(df<span class="op">$</span>detected, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)</span>
<span id="cb48-3"><a href="#cb48-3"></a>highMito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(df<span class="op">$</span>subsets_Mito_percent, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</span>
<span id="cb48-4"><a href="#cb48-4"></a></span>
<span id="cb48-5"><a href="#cb48-5"></a><span class="kw">table</span>(lowLib)</span></code></pre></div>
<pre><code>## lowLib
## FALSE 
## 49300</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a><span class="kw">table</span>(lowFeatures)</span></code></pre></div>
<pre><code>## lowFeatures
## FALSE  TRUE 
## 49287    13</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1"></a><span class="kw">table</span>(highMito)</span></code></pre></div>
<pre><code>## highMito
## FALSE  TRUE 
## 45890  3410</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1"></a>discardCells &lt;-<span class="st"> </span>(lowLib <span class="op">|</span><span class="st"> </span>lowFeatures <span class="op">|</span><span class="st"> </span>highMito)</span>
<span id="cb54-2"><a href="#cb54-2"></a><span class="kw">table</span>(discardCells)</span></code></pre></div>
<pre><code>## discardCells
## FALSE  TRUE 
## 45877  3423</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1"></a><span class="kw">colData</span>(sce)<span class="op">$</span>discardCells &lt;-<span class="st"> </span>discardCells</span>
<span id="cb56-2"><a href="#cb56-2"></a></span>
<span id="cb56-3"><a href="#cb56-3"></a><span class="co"># visualize cells to be removed</span></span>
<span id="cb56-4"><a href="#cb56-4"></a><span class="kw">plotColData</span>(sce, <span class="dt">x =</span> <span class="st">&quot;detected&quot;</span>, <span class="dt">y=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;discardCells&quot;</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="identifying-and-removing-empty-droplets" class="section level2">
<h2><span class="header-section-number">4.4</span> Identifying and removing empty droplets</h2>
<p>Note that the removal of cells with low sequencing depth using the adaptive threshold procedure above is a way of removing empty droplets. Other approaches are possible, e.g., removing cells by statistical testing using <code>emtpyDrops</code>. This does require us to specify a lower bound on the total number of UMIs, below which all cells are considered to correspond to empty droplets. This lower bound may not be trivial to derive, but the <code>barcodeRanks</code> function can be useful to identify an elbow/knee point.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="kw">library</span>(DropletUtils)</span>
<span id="cb57-2"><a href="#cb57-2"></a>bcrank &lt;-<span class="st"> </span><span class="kw">barcodeRanks</span>(<span class="kw">counts</span>(sce))</span>
<span id="cb57-3"><a href="#cb57-3"></a></span>
<span id="cb57-4"><a href="#cb57-4"></a><span class="co"># Only showing unique points for plotting speed.</span></span>
<span id="cb57-5"><a href="#cb57-5"></a>uniq &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">duplicated</span>(bcrank<span class="op">$</span>rank)</span>
<span id="cb57-6"><a href="#cb57-6"></a><span class="kw">plot</span>(bcrank<span class="op">$</span>rank[uniq], bcrank<span class="op">$</span>total[uniq], <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>,</span>
<span id="cb57-7"><a href="#cb57-7"></a>    <span class="dt">xlab=</span><span class="st">&quot;Rank&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Total UMI count&quot;</span>, <span class="dt">cex.lab=</span><span class="fl">1.2</span>)</span>
<span id="cb57-8"><a href="#cb57-8"></a></span>
<span id="cb57-9"><a href="#cb57-9"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">metadata</span>(bcrank)<span class="op">$</span>inflection, <span class="dt">col=</span><span class="st">&quot;darkgreen&quot;</span>, <span class="dt">lty=</span><span class="dv">2</span>)</span>
<span id="cb57-10"><a href="#cb57-10"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">metadata</span>(bcrank)<span class="op">$</span>knee, <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">lty=</span><span class="dv">2</span>)</span>
<span id="cb57-11"><a href="#cb57-11"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">350</span>, <span class="dt">col=</span><span class="st">&quot;orange&quot;</span>, <span class="dt">lty=</span><span class="dv">2</span>) <span class="co"># picked visually myself</span></span>
<span id="cb57-12"><a href="#cb57-12"></a></span>
<span id="cb57-13"><a href="#cb57-13"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Inflection&quot;</span>, <span class="st">&quot;Knee&quot;</span>, <span class="st">&quot;Empirical knee point&quot;</span>), </span>
<span id="cb57-14"><a href="#cb57-14"></a>        <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&quot;darkgreen&quot;</span>, <span class="st">&quot;dodgerblue&quot;</span>, <span class="st">&quot;orange&quot;</span>), <span class="dt">lty=</span><span class="dv">2</span>, <span class="dt">cex=</span><span class="fl">1.2</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1"></a><span class="kw">set.seed</span>(<span class="dv">100</span>)</span>
<span id="cb58-2"><a href="#cb58-2"></a>limit &lt;-<span class="st"> </span><span class="dv">350</span>   </span>
<span id="cb58-3"><a href="#cb58-3"></a>all.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(<span class="kw">counts</span>(sce), <span class="dt">lower=</span>limit, <span class="dt">test.ambient=</span><span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## Warning in smooth.spline(x[new.keep], y[new.keep], df = df, ...): not using
## invalid df; must have 1 &lt; df &lt;= n := #{unique x} = 9</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1"></a><span class="co"># p-values for cells with total UMI count under the lower bound.</span></span>
<span id="cb60-2"><a href="#cb60-2"></a><span class="kw">hist</span>(all.out<span class="op">$</span>PValue[all.out<span class="op">$</span>Total <span class="op">&lt;=</span><span class="st"> </span>limit <span class="op">&amp;</span><span class="st"> </span>all.out<span class="op">$</span>Total <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>],</span>
<span id="cb60-3"><a href="#cb60-3"></a>    <span class="dt">xlab=</span><span class="st">&quot;P-value&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, <span class="dt">col=</span><span class="st">&quot;grey80&quot;</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-10-2.png" width="672" /></p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1"></a><span class="co"># but note that it would remove a very high number of cells</span></span>
<span id="cb61-2"><a href="#cb61-2"></a><span class="kw">length</span>(<span class="kw">which</span>(all.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.001</span>))</span></code></pre></div>
<pre><code>## [1] 25080</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1"></a><span class="co"># so we stick to the more lenient adaptive filtering strategy</span></span>
<span id="cb63-2"><a href="#cb63-2"></a><span class="co"># remove cells identified using adaptive thresholds</span></span>
<span id="cb63-3"><a href="#cb63-3"></a>sce &lt;-<span class="st"> </span>sce[, <span class="op">!</span><span class="kw">colData</span>(sce)<span class="op">$</span>discardCells]</span></code></pre></div>
</div>
<div id="identifying-and-removing-doublets" class="section level2">
<h2><span class="header-section-number">4.5</span> Identifying and removing doublets</h2>
<p>We will use <a href="https://bioconductor.org/packages/3.14/bioc/html/scDblFinder.html">scDblFinder</a> to detect doublet cells.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1"></a><span class="co">## perform doublet detection</span></span>
<span id="cb64-2"><a href="#cb64-2"></a><span class="kw">library</span>(scDblFinder)</span>
<span id="cb64-3"><a href="#cb64-3"></a><span class="kw">set.seed</span>(<span class="dv">211103</span>)</span>
<span id="cb64-4"><a href="#cb64-4"></a>sampleID &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="kw">strsplit</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>cell.id, <span class="dt">split=</span><span class="st">&quot;_&quot;</span>), <span class="st">&quot;[[&quot;</span>, <span class="dv">1</span>))</span>
<span id="cb64-5"><a href="#cb64-5"></a><span class="kw">table</span>(sampleID)</span></code></pre></div>
<pre><code>## sampleID
##   p1   r1   r2   r3   r4   r5   r6 
## 3942 5953 8414 5319 7015 7487 7747</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1"></a>sce &lt;-<span class="st"> </span><span class="kw">scDblFinder</span>(sce, <span class="dt">returnType=</span><span class="st">&quot;table&quot;</span>,</span>
<span id="cb66-2"><a href="#cb66-2"></a>                 <span class="dt">samples =</span> <span class="kw">factor</span>(sampleID))</span>
<span id="cb66-3"><a href="#cb66-3"></a><span class="kw">table</span>(sce<span class="op">$</span>scDblFinder.class)</span></code></pre></div>
<pre><code>## 
## singlet doublet 
##   41220    4657</code></pre>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a><span class="co">## visualize these scores</span></span>
<span id="cb68-2"><a href="#cb68-2"></a><span class="co">## explore doublet score wrt original cluster labels</span></span>
<span id="cb68-3"><a href="#cb68-3"></a><span class="kw">boxplot</span>(<span class="kw">log1p</span>(sce<span class="op">$</span>scDblFinder.score) <span class="op">~</span><span class="st"> </span><span class="kw">factor</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>cluster, <span class="dt">exclude=</span><span class="ot">NULL</span>))</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1"></a>tab &lt;-<span class="st"> </span><span class="kw">table</span>(sce<span class="op">$</span>scDblFinder.class, sce<span class="op">$</span>cluster, </span>
<span id="cb69-2"><a href="#cb69-2"></a>      <span class="dt">exclude=</span><span class="ot">NULL</span>)</span>
<span id="cb69-3"><a href="#cb69-3"></a>tab</span></code></pre></div>
<pre><code>##          
##               1     2     3     4     5     6     7     8     9    10    11
##   singlet   213   370   253    63    69   177   264   141   305   165   180
##   doublet     2     7    22     9     5    28    53     8    32    24    27
##          
##              12    13    14    15    16    17    18    19    20    21    22
##   singlet   211    47    85    51   196   304    65   111   318   172   201
##   doublet    43     3    23    21    55    67    15    14    64    75    60
##          
##              23    24    25    26    27    28    29    30    31    32    33
##   singlet   216 25026  1604  1932   472   357   464   517   412   266   658
##   doublet    43  1917   169   229   157   124   117   113    90    52   155
##          
##              34    35    36    37    38    39  &lt;NA&gt;
##   singlet  1431    36    42   231    56    64  3475
##   doublet   133     7     3     9     3     2   677</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1"></a><span class="kw">t</span>(<span class="kw">t</span>(tab) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(tab))</span></code></pre></div>
<pre><code>##          
##                     1           2           3           4           5
##   singlet 0.990697674 0.981432361 0.920000000 0.875000000 0.932432432
##   doublet 0.009302326 0.018567639 0.080000000 0.125000000 0.067567568
##          
##                     6           7           8           9          10
##   singlet 0.863414634 0.832807571 0.946308725 0.905044510 0.873015873
##   doublet 0.136585366 0.167192429 0.053691275 0.094955490 0.126984127
##          
##                    11          12          13          14          15
##   singlet 0.869565217 0.830708661 0.940000000 0.787037037 0.708333333
##   doublet 0.130434783 0.169291339 0.060000000 0.212962963 0.291666667
##          
##                    16          17          18          19          20
##   singlet 0.780876494 0.819407008 0.812500000 0.888000000 0.832460733
##   doublet 0.219123506 0.180592992 0.187500000 0.112000000 0.167539267
##          
##                    21          22          23          24          25
##   singlet 0.696356275 0.770114943 0.833976834 0.928849794 0.904681331
##   doublet 0.303643725 0.229885057 0.166023166 0.071150206 0.095318669
##          
##                    26          27          28          29          30
##   singlet 0.894030541 0.750397456 0.742203742 0.798623064 0.820634921
##   doublet 0.105969459 0.249602544 0.257796258 0.201376936 0.179365079
##          
##                    31          32          33          34          35
##   singlet 0.820717131 0.836477987 0.809348093 0.914961637 0.837209302
##   doublet 0.179282869 0.163522013 0.190651907 0.085038363 0.162790698
##          
##                    36          37          38          39        &lt;NA&gt;
##   singlet 0.933333333 0.962500000 0.949152542 0.969696970 0.836946050
##   doublet 0.066666667 0.037500000 0.050847458 0.030303030 0.163053950</code></pre>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1"></a><span class="kw">barplot</span>(<span class="kw">t</span>(<span class="kw">t</span>(tab) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(tab))[<span class="dv">2</span>,],</span>
<span id="cb73-2"><a href="#cb73-2"></a>        <span class="dt">xlab =</span> <span class="st">&quot;Cluster&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Fraction of doublets&quot;</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-11-2.png" width="672" /></p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1"></a><span class="co"># remove doublets</span></span>
<span id="cb74-2"><a href="#cb74-2"></a>sce &lt;-<span class="st"> </span>sce[,<span class="op">!</span>sce<span class="op">$</span>scDblFinder.class <span class="op">==</span><span class="st"> &quot;doublet&quot;</span>]</span></code></pre></div>
</div>
</div>
<div id="normalization" class="section level1">
<h1><span class="header-section-number">5</span> Normalization</h1>
<p>For normalization, the size factors <span class="math inline">\(s_i\)</span> computed here are simply scaled library sizes: <span class="math display">\[ N_i = \sum_g Y_{gi} \]</span> <span class="math display">\[ s_i = N_i / \bar{N}_i \]</span></p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1"></a>sce &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce)</span>
<span id="cb75-2"><a href="#cb75-2"></a></span>
<span id="cb75-3"><a href="#cb75-3"></a><span class="co"># note we also returned log counts: see the additional logcounts assay.</span></span>
<span id="cb75-4"><a href="#cb75-4"></a>sce</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 17887 41220 
## metadata(0):
## assays(2): counts logcounts
## rownames(17887): KITL TMTC3 ... GM16012 GM21464
## rowData names(0):
## colnames(41220): r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA ... p1_TAACGCGCTCCT
##   p1_ATTCTTGTTCTT
## colData names(15): cell.id cluster ... scDblFinder.cxds_score
##   sizeFactor
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1"></a><span class="co"># you can extract size factors using</span></span>
<span id="cb77-2"><a href="#cb77-2"></a>sf &lt;-<span class="st"> </span><span class="kw">librarySizeFactors</span>(sce)</span>
<span id="cb77-3"><a href="#cb77-3"></a><span class="kw">mean</span>(sf) <span class="co"># equal to 1 due to scaling.</span></span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1"></a><span class="kw">plot</span>(<span class="dt">x=</span> <span class="kw">log</span>(<span class="kw">colSums</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts)), </span>
<span id="cb79-2"><a href="#cb79-2"></a>     <span class="dt">y=</span>sf)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
</div>
<div id="feature-selection" class="section level1">
<h1><span class="header-section-number">6</span> Feature selection</h1>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1"></a><span class="kw">library</span>(scran)</span>
<span id="cb80-2"><a href="#cb80-2"></a>dec &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(sce)</span>
<span id="cb80-3"><a href="#cb80-3"></a>fitRetina &lt;-<span class="st"> </span><span class="kw">metadata</span>(dec)</span>
<span id="cb80-4"><a href="#cb80-4"></a><span class="kw">plot</span>(fitRetina<span class="op">$</span>mean, fitRetina<span class="op">$</span>var, </span>
<span id="cb80-5"><a href="#cb80-5"></a>     <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</span>
<span id="cb80-6"><a href="#cb80-6"></a>    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</span>
<span id="cb80-7"><a href="#cb80-7"></a><span class="kw">curve</span>(fitRetina<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1"></a><span class="co"># get 10% highly variable genes</span></span>
<span id="cb81-2"><a href="#cb81-2"></a>hvg &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(dec, <span class="dt">prop=</span><span class="fl">0.1</span>)</span>
<span id="cb81-3"><a href="#cb81-3"></a><span class="kw">head</span>(hvg)</span></code></pre></div>
<pre><code>## [1] &quot;RHO&quot;     &quot;CALM1&quot;   &quot;MEG3&quot;    &quot;GNGT1&quot;   &quot;RPGRIP1&quot; &quot;SAG&quot;</code></pre>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1"></a><span class="co"># plot these </span></span>
<span id="cb83-2"><a href="#cb83-2"></a><span class="kw">plot</span>(fitRetina<span class="op">$</span>mean, fitRetina<span class="op">$</span>var, </span>
<span id="cb83-3"><a href="#cb83-3"></a>     <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;orange&quot;</span>, <span class="st">&quot;darkseagreen3&quot;</span>)[(<span class="kw">names</span>(fitRetina<span class="op">$</span>mean) <span class="op">%in%</span><span class="st"> </span>hvg)<span class="op">+</span><span class="dv">1</span>],</span>
<span id="cb83-4"><a href="#cb83-4"></a>     <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</span>
<span id="cb83-5"><a href="#cb83-5"></a>    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</span>
<span id="cb83-6"><a href="#cb83-6"></a><span class="kw">curve</span>(fitRetina<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</span>
<span id="cb83-7"><a href="#cb83-7"></a><span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, </span>
<span id="cb83-8"><a href="#cb83-8"></a>       <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;Selected&quot;</span>, <span class="st">&quot;Not selected&quot;</span>), </span>
<span id="cb83-9"><a href="#cb83-9"></a>       <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;darkseagreen3&quot;</span>, <span class="st">&quot;orange&quot;</span>),</span>
<span id="cb83-10"><a href="#cb83-10"></a>       <span class="dt">pch =</span> <span class="dv">16</span>,</span>
<span id="cb83-11"><a href="#cb83-11"></a>       <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-13-2.png" width="672" /></p>
</div>
<div id="dimensionality-reduction" class="section level1">
<h1><span class="header-section-number">7</span> Dimensionality reduction</h1>
<p>Note that, below, we color the cells using the known, true cell type label as defined in the metadata, to empirically evaluate the dimensionality reduction. In reality, we don’t know this yet at this stage.</p>
<div id="the-most-basic-dr" class="section level2">
<h2><span class="header-section-number">7.1</span> The most basic DR</h2>
<p>Just by looking at the top two genes based on our feature selection criterion, we can already see some separation according to the cell type!</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1"></a><span class="kw">colData</span>(sce)<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">colData</span>(sce)<span class="op">$</span>cluster)</span>
<span id="cb84-2"><a href="#cb84-2"></a>cl &lt;-<span class="st"> </span><span class="kw">colData</span>(sce)<span class="op">$</span>cluster</span>
<span id="cb84-3"><a href="#cb84-3"></a></span>
<span id="cb84-4"><a href="#cb84-4"></a><span class="kw">par</span>(<span class="dt">bty=</span><span class="st">&#39;l&#39;</span>)</span>
<span id="cb84-5"><a href="#cb84-5"></a><span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">assays</span>(sce)<span class="op">$</span>counts[hvg[<span class="dv">1</span>],],</span>
<span id="cb84-6"><a href="#cb84-6"></a>     <span class="dt">y =</span> <span class="kw">assays</span>(sce)<span class="op">$</span>counts[hvg[<span class="dv">2</span>],],</span>
<span id="cb84-7"><a href="#cb84-7"></a>     <span class="dt">col =</span> <span class="kw">as.numeric</span>(cl),</span>
<span id="cb84-8"><a href="#cb84-8"></a>     <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb84-9"><a href="#cb84-9"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Most informative gene&quot;</span>,</span>
<span id="cb84-10"><a href="#cb84-10"></a>     <span class="dt">ylab =</span> <span class="st">&quot;Second most informative gene&quot;</span>,</span>
<span id="cb84-11"><a href="#cb84-11"></a>     <span class="dt">main =</span> <span class="st">&quot;Cells colored acc to cell type&quot;</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
</div>
<div id="linear-dimensionality-reduction-pca" class="section level2">
<h2><span class="header-section-number">7.2</span> Linear dimensionality reduction: PCA</h2>
<p>We are able to recover quite some structure. However, many cell populations remain obscure, and the plot is overcrowded.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1"></a><span class="kw">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb85-2"><a href="#cb85-2"></a>sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce, <span class="dt">ncomponents=</span><span class="dv">30</span>, <span class="dt">subset_row=</span>hvg)</span>
<span id="cb85-3"><a href="#cb85-3"></a><span class="kw">plotPCA</span>(sce, <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 3475 rows containing missing values (geom_point).</code></pre>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<div id="pca-without-feature-selection" class="section level3">
<h3><span class="header-section-number">7.2.1</span> PCA without feature selection</h3>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1"></a><span class="kw">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb87-2"><a href="#cb87-2"></a>sceNoFS &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce, <span class="dt">ncomponents=</span><span class="dv">30</span>, <span class="dt">subset_row=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sce))</span>
<span id="cb87-3"><a href="#cb87-3"></a><span class="kw">plotPCA</span>(sceNoFS, <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 3475 rows containing missing values (geom_point).</code></pre>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1"></a><span class="kw">rm</span>(sceNoFS)</span></code></pre></div>
</div>
</div>
<div id="a-generalization-of-pca-for-exponential-family-distributions." class="section level2">
<h2><span class="header-section-number">7.3</span> A generalization of PCA for exponential family distributions.</h2>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1"></a><span class="kw">library</span>(glmpca)</span>
<span id="cb90-2"><a href="#cb90-2"></a><span class="kw">set.seed</span>(<span class="dv">211103</span>)</span>
<span id="cb90-3"><a href="#cb90-3"></a>poipca &lt;-<span class="st"> </span><span class="kw">glmpca</span>(<span class="kw">assays</span>(sce)<span class="op">$</span>counts[hvg,],</span>
<span id="cb90-4"><a href="#cb90-4"></a>                 <span class="dt">L=</span><span class="dv">2</span>, <span class="dt">fam=</span><span class="st">&quot;poi&quot;</span>,</span>
<span id="cb90-5"><a href="#cb90-5"></a>                 <span class="dt">minibatch=</span><span class="st">&quot;stochastic&quot;</span>)</span>
<span id="cb90-6"><a href="#cb90-6"></a><span class="kw">reducedDim</span>(sce, <span class="st">&quot;PoiPCA&quot;</span>) &lt;-<span class="st"> </span>poipca<span class="op">$</span>factors</span>
<span id="cb90-7"><a href="#cb90-7"></a><span class="kw">plotReducedDim</span>(sce, </span>
<span id="cb90-8"><a href="#cb90-8"></a>               <span class="dt">dimred=</span><span class="st">&quot;PoiPCA&quot;</span>,</span>
<span id="cb90-9"><a href="#cb90-9"></a>               <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 3475 rows containing missing values (geom_point).</code></pre>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
</div>
<div id="non-linear-dimensionality-reduction-umap" class="section level2">
<h2><span class="header-section-number">7.4</span> Non-linear dimensionality reduction: UMAP</h2>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1"></a>sce &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce, <span class="dt">dimred =</span> <span class="st">&#39;PCA&#39;</span>, <span class="dt">external_neighbors=</span><span class="ot">TRUE</span>)</span>
<span id="cb92-2"><a href="#cb92-2"></a><span class="kw">plotUMAP</span>(sce,</span>
<span id="cb92-3"><a href="#cb92-3"></a>         <span class="dt">colour_by =</span> <span class="st">&quot;cluster&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: Removed 3475 rows containing missing values (geom_point).</code></pre>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
</div>
</div>
<div id="clustering" class="section level1">
<h1><span class="header-section-number">8</span> Clustering</h1>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1"></a><span class="co"># Build a shared nearest-neighbor graph from PCA space</span></span>
<span id="cb94-2"><a href="#cb94-2"></a>g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</span>
<span id="cb94-3"><a href="#cb94-3"></a><span class="co"># Louvain clustering on the SNN graph, and add to sce</span></span>
<span id="cb94-4"><a href="#cb94-4"></a><span class="kw">colData</span>(sce)<span class="op">$</span>label &lt;-<span class="st"> </span><span class="kw">factor</span>(igraph<span class="op">::</span><span class="kw">cluster_louvain</span>(g)<span class="op">$</span>membership)</span>
<span id="cb94-5"><a href="#cb94-5"></a></span>
<span id="cb94-6"><a href="#cb94-6"></a><span class="co"># Visualization.</span></span>
<span id="cb94-7"><a href="#cb94-7"></a><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by=</span><span class="st">&quot;label&quot;</span>)</span></code></pre></div>
<p><img src="singleCell_MacoskoWorkflow_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
</div>

<div id="rmd-source-code">---
title: 'scRNA-seq workflow: Macosko et al.'
author: "Koen Van den Berge"
date: "11/16/2020"
output: 
  html_document:
    toc: true
    toc_float: true
---

# Import data

The `scRNAseq` package provides convenient access to several datasets. See the [package Bioconductor page](http://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html) for more information.

```{r}
# install BiocManager package if not installed yet.
# BiocManager is the package installer for Bioconductor software.
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# install packages if not yet installed.
pkgs <- c("SingleCellExperiment", "DropletUtils", "scRNAseq", "scater", "scuttle", "scran", "BiocSingular", "scDblFinder", "glmpca", "uwot")
notInstalled <- pkgs[!pkgs %in% installed.packages()[,1]]
if(length(notInstalled) > 0){
  BiocManager::install(notInstalled)
}

# Code below might ask you to create an ExperimentHub directory. 
# Type 'yes' and hit Enter, to allow this.
suppressPackageStartupMessages(library(scRNAseq))
sce <- MacoskoRetinaData()
```

# A `SingleCellExperiment` object


```{r}
sce
```

## Accessing data from a `SingleCellExperiment` object

Please see [Figure 4.1 in OSCA](http://bioconductor.org/books/release/OSCA/data-infrastructure.html) for an overview of a `SingleCellExperiment` object.

```{r}
# Data: assays
assays(sce)
assays(sce)$counts[1:5, 1:5]

# Feature metadata: rowData
rowData(sce) # empty for now

# Cell metadata: colData
colData(sce)

# Reduced dimensions: reducedDims
reducedDims(sce) # empty for now
```

## Creating a new `SingleCellExperiment` object

```{r}
sceNew <- SingleCellExperiment(assays = list(counts = assays(sce)$counts))
sceNew

rm(sceNew)
```

## Storing (meta)data in a `SingleCellExperiment` object

```{r}
fakeGeneNames <- paste0("gene", 1:nrow(sce))
rowData(sce)$fakeName <- fakeGeneNames
head(rowData(sce))
# Remove again by setting to NULL
rowData(sce)$fakeName <- NULL

assays(sce)$logCounts <- log1p(assays(sce)$counts)
assays(sce)
assays(sce)$logCounts[1:5, 1:5]
assays(sce)$logCounts <- NULL
```

# Filtering non-informative genes

```{r}
keep <- rowSums(assays(sce)$counts > 0) > 10
table(keep)

sce <- sce[keep,]
```


# Quality control

## Calculate QC variables

```{r}
library(scater)
is.mito <- grepl("^MT-", rownames(sce))
sum(is.mito) # 28 mitochondrial genes
df <- perCellQCMetrics(sce, subsets=list(Mito=is.mito))
## add the QC variables to sce object
colData(sce) <- cbind(colData(sce), df)
# the QC variables have now been added to the colData of our SCE object.
colData(sce)
```

## EDA

High-quality cells should have many features expressed, and a low contribution of mitochondrial genes. Here, we see that several cells have a very low number of expressed genes, and where most of the molecules are derived from mitochondrial genes. This indicates likely damaged cells, presumably because of loss of cytoplasmic RNA from perforated cells, so we'd want to remove these for the downstream analysis.

```{r}
# Number of genes vs library size
plotColData(sce, x = "sum", y="detected", colour_by="cluster") 

# Mitochondrial genes
plotColData(sce, x = "detected", y="subsets_Mito_percent")
```

## QC using adaptive thresholds

Below, we remove cells that are outlying with respect to

 1. A low sequencing depth (number of UMIs);
 2. A low number of genes detected;
 3. A high percentage of reads from mitochondrial genes.
 
We remove a total of $3423$ cells, most of which because of an outlyingly high percentage of reads from mitochondrial genes.

```{r}
lowLib <- isOutlier(df$sum, type="lower", log=TRUE)
lowFeatures <- isOutlier(df$detected, type="lower", log=TRUE)
highMito <- isOutlier(df$subsets_Mito_percent, type="higher")

table(lowLib)
table(lowFeatures)
table(highMito)

discardCells <- (lowLib | lowFeatures | highMito)
table(discardCells)
colData(sce)$discardCells <- discardCells

# visualize cells to be removed
plotColData(sce, x = "detected", y="subsets_Mito_percent", colour_by = "discardCells")

```


## Identifying and removing empty droplets

Note that the removal of cells with low sequencing depth using the adaptive threshold procedure above is a way of removing empty droplets. 
Other approaches are possible, e.g., removing cells by statistical testing using `emtpyDrops`.
This does require us to specify a lower bound on the total number of UMIs, below which all cells are considered to correspond to empty droplets.
This lower bound may not be trivial to derive, but the `barcodeRanks` function can be useful to identify an elbow/knee point.

```{r}
library(DropletUtils)
bcrank <- barcodeRanks(counts(sce))

# Only showing unique points for plotting speed.
uniq <- !duplicated(bcrank$rank)
plot(bcrank$rank[uniq], bcrank$total[uniq], log="xy",
    xlab="Rank", ylab="Total UMI count", cex.lab=1.2)

abline(h=metadata(bcrank)$inflection, col="darkgreen", lty=2)
abline(h=metadata(bcrank)$knee, col="dodgerblue", lty=2)
abline(h=350, col="orange", lty=2) # picked visually myself

legend("topright", legend=c("Inflection", "Knee", "Empirical knee point"), 
        col=c("darkgreen", "dodgerblue", "orange"), lty=2, cex=1.2)

set.seed(100)
limit <- 350   
all.out <- emptyDrops(counts(sce), lower=limit, test.ambient=TRUE)
# p-values for cells with total UMI count under the lower bound.
hist(all.out$PValue[all.out$Total <= limit & all.out$Total > 0],
    xlab="P-value", main="", col="grey80")

# but note that it would remove a very high number of cells
length(which(all.out$FDR <= 0.001))

# so we stick to the more lenient adaptive filtering strategy
# remove cells identified using adaptive thresholds
sce <- sce[, !colData(sce)$discardCells]
```

## Identifying and removing doublets

We will use [scDblFinder](https://bioconductor.org/packages/3.14/bioc/html/scDblFinder.html) to detect doublet cells.



```{r}
## perform doublet detection
library(scDblFinder)
set.seed(211103)
sampleID <- unlist(lapply(strsplit(colData(sce)$cell.id, split="_"), "[[", 1))
table(sampleID)
sce <- scDblFinder(sce, returnType="table",
                 samples = factor(sampleID))
table(sce$scDblFinder.class)


## visualize these scores
## explore doublet score wrt original cluster labels
boxplot(log1p(sce$scDblFinder.score) ~ factor(colData(sce)$cluster, exclude=NULL))

tab <- table(sce$scDblFinder.class, sce$cluster, 
      exclude=NULL)
tab
t(t(tab) / colSums(tab))

barplot(t(t(tab) / colSums(tab))[2,],
        xlab = "Cluster", ylab = "Fraction of doublets")

# remove doublets
sce <- sce[,!sce$scDblFinder.class == "doublet"]
```


# Normalization

For normalization, the size factors $s_i$ computed here are simply scaled library sizes:
\[ N_i = \sum_g Y_{gi} \]
\[ s_i = N_i / \bar{N}_i \]

```{r}
sce <- logNormCounts(sce)

# note we also returned log counts: see the additional logcounts assay.
sce

# you can extract size factors using
sf <- librarySizeFactors(sce)
mean(sf) # equal to 1 due to scaling.
plot(x= log(colSums(assays(sce)$counts)), 
     y=sf)
```

# Feature selection

```{r}
library(scran)
dec <- modelGeneVar(sce)
fitRetina <- metadata(dec)
plot(fitRetina$mean, fitRetina$var, 
     xlab="Mean of log-expression",
    ylab="Variance of log-expression")
curve(fitRetina$trend(x), col="dodgerblue", add=TRUE, lwd=2)

# get 10% highly variable genes
hvg <- getTopHVGs(dec, prop=0.1)
head(hvg)

# plot these 
plot(fitRetina$mean, fitRetina$var, 
     col = c("orange", "darkseagreen3")[(names(fitRetina$mean) %in% hvg)+1],
     xlab="Mean of log-expression",
    ylab="Variance of log-expression")
curve(fitRetina$trend(x), col="dodgerblue", add=TRUE, lwd=2)
legend("topleft", 
       legend = c("Selected", "Not selected"), 
       col = c("darkseagreen3", "orange"),
       pch = 16,
       bty='n')
```

# Dimensionality reduction

Note that, below, we color the cells using the known, true cell type label as defined in the metadata, to empirically evaluate the dimensionality reduction. In reality, we don't know this yet at this stage.

## The most basic DR

Just by looking at the top two genes based on our feature selection criterion, we can already see some separation according to the cell type!

```{r}
colData(sce)$cluster <- as.factor(colData(sce)$cluster)
cl <- colData(sce)$cluster

par(bty='l')
plot(x = assays(sce)$counts[hvg[1],],
     y = assays(sce)$counts[hvg[2],],
     col = as.numeric(cl),
     pch = 16, cex = 1/3,
     xlab = "Most informative gene",
     ylab = "Second most informative gene",
     main = "Cells colored acc to cell type")
```

## Linear dimensionality reduction: PCA

We are able to recover quite some structure. 
However, many cell populations remain obscure, and the plot is overcrowded.

```{r}
set.seed(1234)
sce <- runPCA(sce, ncomponents=30, subset_row=hvg)
plotPCA(sce, colour_by = "cluster")
```

### PCA without feature selection

```{r}
set.seed(1234)
sceNoFS <- runPCA(sce, ncomponents=30, subset_row=1:nrow(sce))
plotPCA(sceNoFS, colour_by = "cluster")
rm(sceNoFS)
```


## A generalization of PCA for exponential family distributions.

```{r, eval=TRUE}
library(glmpca)
set.seed(211103)
poipca <- glmpca(assays(sce)$counts[hvg,],
                 L=2, fam="poi",
                 minibatch="stochastic")
reducedDim(sce, "PoiPCA") <- poipca$factors
plotReducedDim(sce, 
               dimred="PoiPCA",
               colour_by = "cluster")
```



## Non-linear dimensionality reduction: UMAP

```{r}
sce <- runUMAP(sce, dimred = 'PCA', external_neighbors=TRUE)
plotUMAP(sce,
         colour_by = "cluster")
```

# Clustering

```{r}
# Build a shared nearest-neighbor graph from PCA space
g <- buildSNNGraph(sce, use.dimred = 'PCA')
# Louvain clustering on the SNN graph, and add to sce
colData(sce)$label <- factor(igraph::cluster_louvain(g)$membership)

# Visualization.
plotUMAP(sce, colour_by="label")
```
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("singleCell_MacoskoWorkflow.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
